{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Introduction"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We want to see how the pseudo mappers work on our C1 data, there's enough runs that it'd be annoying to generate by hand."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas\n",
    "from pathlib import Path\n",
    "import os\n",
    "import sys\n",
    "from glob import glob\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "LRSC = str(Path('~/proj/long-rna-seq-condor').expanduser())\n",
    "if LRSC not in sys.path:\n",
    "    sys.path.append(LRSC)\n",
    "from woldrnaseq.models import load_experiments, load_library_tables\n",
    "from woldrnaseq.common import find_fastqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.listdir('/woldlab/castor/home/sau'):\n",
    "    raise('Runtime makes sure you have Seans home directory mounted in singularity -B <src>:<dest>')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "c1_remap_root = Path('~/proj/C1_mouse_limb_combined/all_analysis_M21')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calculate kallisto fragment and sd parameters"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From C1_Fluidigm_BioAnalyzer_avgcDNA_and_CV_October14_2020.xlsx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>avg_cDNA</th>\n",
       "      <th>CV%</th>\n",
       "      <th>sd</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>770</td>\n",
       "      <td>78.9</td>\n",
       "      <td>607.530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3</td>\n",
       "      <td>1164</td>\n",
       "      <td>59.6</td>\n",
       "      <td>693.744</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>1214</td>\n",
       "      <td>59.5</td>\n",
       "      <td>722.330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>1132</td>\n",
       "      <td>62.4</td>\n",
       "      <td>706.368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>1144</td>\n",
       "      <td>63.3</td>\n",
       "      <td>724.152</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Run  avg_cDNA   CV%       sd\n",
       "0    3       770  78.9  607.530\n",
       "1    3      1164  59.6  693.744\n",
       "2    3      1214  59.5  722.330\n",
       "3    3      1132  62.4  706.368\n",
       "4    3      1144  63.3  724.152"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sampled = pandas.DataFrame([[3,770,78.9],\n",
    "[3,1164,59.6],\n",
    "[3,1214,59.5],\n",
    "[3,1132,62.4],\n",
    "[3,1144,63.3],\n",
    "[4,2099,75.9],\n",
    "[4,1819,81],\n",
    "[4,1700,92.1],\n",
    "[4,1835,82.7],\n",
    "[4,1699,100],\n",
    "[4,1883,95.6],\n",
    "[4,1955,84.6],\n",
    "[4,1723,93.8],\n",
    "[4,1900,84.7],\n",
    "[4,1919,81.7],\n",
    "[5,1947,81.3],\n",
    "[5,2052,81.2],\n",
    "[5,1812,81.7],\n",
    "[5,1622,88.6],\n",
    "[5,1927,81.6],\n",
    "[5,1846,85.4],\n",
    "[5,1806,85.3],\n",
    "[5,1911,84.2],\n",
    "[6,1856,79.1],\n",
    "[6,1770,76.7],\n",
    "[6,1558,85.7],\n",
    "[6,1115,91.5],\n",
    "[6,1853,78.3],\n",
    "[6,1838,79.9],\n",
    "[6,1794,79.2],\n",
    "[6,1767,80.6],\n",
    "[6,1782,77.6],\n",
    "[6,1791,77.9],\n",
    "[7,1870,71.6],\n",
    "[7,1860,74.4],\n",
    "[7,1900,71.2],\n",
    "[7,1883,72.3],\n",
    "[7,1764,76.6],\n",
    "[7,1885,74.3],\n",
    "[7,1641,82.2],\n",
    "[7,1800,74.6],\n",
    "[7,1882,74.1],\n",
    "[7,1480,87.5],\n",
    "[8,1270,80.4],\n",
    "[8,1808,71.9],\n",
    "[8,1867,70.5],\n",
    "[8,1864,68.3],\n",
    "[8,1902,68.2],\n",
    "[8,1822,76.7],\n",
    "[8,1375,83.7],\n",
    "[8,1932,70.8],\n",
    "[8,1954,71.3],\n",
    "[9,1486,69.3],\n",
    "[9,1475,71.5],\n",
    "[9,1555,69.8],\n",
    "[9,1622,66.3],\n",
    "[9,1496,54.3],\n",
    "[9,1495,65.5],\n",
    "[9,1298,78.1],\n",
    "[9,916,82],\n",
    "[9,1360,79.2],\n",
    "[9,1573,78.3],\n",
    "[10,1814,60.9],\n",
    "[10,1736,62.8],\n",
    "[10,1776,63.9],\n",
    "[10,1798,62.8],\n",
    "[10,1774,62],\n",
    "[10,1795,64.8],\n",
    "[10,1710,65.7],\n",
    "[10,1671,59],\n",
    "[11,1658,62.1],\n",
    "[11,1777,64.1],\n",
    "[11,1726,64.3],\n",
    "[11,1798,62],\n",
    "[11,1711,61.2],\n",
    "[11,1761,61.2],\n",
    "[11,1696,58.3],\n",
    "[11,1863,62.5],\n",
    "[12,1804,61.6],\n",
    "[12,1720,65.6],\n",
    "[12,1657,64.9],\n",
    "[12,1752,63.6],\n",
    "[12,1788,62.9],\n",
    "[12,1549,63.9],\n",
    "[12,1800,61.8],\n",
    "[12,1736,62.4],\n",
    "[12,1441,64.7],\n",
    "[13,1729,63.5],\n",
    "[13,1796,61.4],\n",
    "[13,1758,63.5],\n",
    "[13,1783,64.7],\n",
    "[13,1660,63.5],\n",
    "[13,1841,65.8],\n",
    "[14,1760,63.6],\n",
    "[14,1762,64.5],\n",
    "[14,1912,64.6],\n",
    "[14,1832,62.1],\n",
    "[14,1792,65.4],\n",
    "[14,1755,62.4],\n",
    "[14,1820,62.4],\n",
    "[14,1819,66.9],\n",
    "[14,1280,70.7],\n",
    "[14,1686,65.9],\n",
    "[15,1261,73.6],\n",
    "[15,1768,67.1],\n",
    "[15,1921,63.9],\n",
    "[15,1977,64],\n",
    "[15,1457,85.3],\n",
    "[15,1812,67.6],\n",
    "[15,845,50.4],\n",
    "[15,931,69.5],\n",
    "[16,1760,66.7],\n",
    "[16,1785,66.5],\n",
    "[16,1907,65],\n",
    "[16,1885,67.8],\n",
    "[16,1904,62.6],\n",
    "[17,1753,64.9],\n",
    "[17,1705,67.3],\n",
    "[17,1620,67.8],\n",
    "[17,1941,67.3],\n",
    "[17,1848,68.9],\n",
    "[17,1581,68.7],\n",
    "[17,1914,67.2],], columns=['Run','avg_cDNA','CV%'])\n",
    "sampled['sd'] = sampled['CV%']/100 * sampled['avg_cDNA']\n",
    "sampled.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>avg_cDNA</th>\n",
       "      <th>CV%</th>\n",
       "      <th>sd</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Run</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1084.800</td>\n",
       "      <td>64.7400</td>\n",
       "      <td>690.824800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1853.200</td>\n",
       "      <td>87.2100</td>\n",
       "      <td>1609.615100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1865.375</td>\n",
       "      <td>83.6625</td>\n",
       "      <td>1558.140875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1712.400</td>\n",
       "      <td>80.6500</td>\n",
       "      <td>1372.364900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1796.500</td>\n",
       "      <td>75.8800</td>\n",
       "      <td>1357.001200</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     avg_cDNA      CV%           sd\n",
       "Run                                \n",
       "3    1084.800  64.7400   690.824800\n",
       "4    1853.200  87.2100  1609.615100\n",
       "5    1865.375  83.6625  1558.140875\n",
       "6    1712.400  80.6500  1372.364900\n",
       "7    1796.500  75.8800  1357.001200"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sampled_mean = sampled.groupby('Run').mean()\n",
    "sampled_mean.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load all Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1037, 8)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "libraries = load_library_tables([c1_remap_root / 'libraries-passing.tsv'])\n",
    "libraries.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>replicates</th>\n",
       "      <th>analysis_dir</th>\n",
       "      <th>run</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>experiment</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C1_mouse_e13.5_limb_mesenchyme_mm10_clean_run4</th>\n",
       "      <td>[18251_A1, 18251_A10, 18251_A11, 18251_A12, 18...</td>\n",
       "      <td>~/proj/C1_mouse_limb_combined/all_analysis_M21</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1_mouse_e11.0_limb_mesenchyme_mm10_clean_run5</th>\n",
       "      <td>[18258_A1, 18258_A10, 18258_A11, 18258_A12, 18...</td>\n",
       "      <td>~/proj/C1_mouse_limb_combined/all_analysis_M21</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1_mouse_e11.5_limb_mesenchyme_mm10_clean_run6</th>\n",
       "      <td>[18263_A1, 18263_A10, 18263_A11, 18263_A12, 18...</td>\n",
       "      <td>~/proj/C1_mouse_limb_combined/all_analysis_M21</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1_mouse_e12.5_limb_mesenchyme_mm10_clean_run7</th>\n",
       "      <td>[18270_A1, 18270_A10, 18270_A11, 18270_A12, 18...</td>\n",
       "      <td>~/proj/C1_mouse_limb_combined/all_analysis_M21</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1_mouse_e13.5_limb_mesenchyme_mm10_clean_run8</th>\n",
       "      <td>[18311_A1, 18311_A10, 18311_A11, 18311_A12, 18...</td>\n",
       "      <td>~/proj/C1_mouse_limb_combined/all_analysis_M21</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1_mouse_e10.5_limb_mm10_clean_run1</th>\n",
       "      <td>[18042_A1, 18042_A10, 18042_A11, 18042_A12, 18...</td>\n",
       "      <td>~/proj/C1_mouse_limb_combined/all_analysis_M21</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1_mouse_e10.5_limb_mm10_clean_run2</th>\n",
       "      <td>[17327_A1, 17327_A10, 17327_A11, 17327_A12, 17...</td>\n",
       "      <td>~/proj/C1_mouse_limb_combined/all_analysis_M21</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1_mouse_e10.5_limb_mm10_clean_run3</th>\n",
       "      <td>[18087_F1, 18087_F10, 18087_F11, 18087_F12, 18...</td>\n",
       "      <td>~/proj/C1_mouse_limb_combined/all_analysis_M21</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1_mouse_e14.0_forelimb_run13_December14_2017</th>\n",
       "      <td>[20039_A1, 20039_A10, 20039_A11, 20039_A12, 20...</td>\n",
       "      <td>~/proj/C1_mouse_limb_combined/all_analysis_M21</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1_mouse_e12.5_forelimb_run9_March8_2017_rearray_March152018</th>\n",
       "      <td>[20090_C1, 20090_C10, 20090_C11, 20090_C12, 20...</td>\n",
       "      <td>~/proj/C1_mouse_limb_combined/all_analysis_M21</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1_mouse_e11.5_forelimb_run10_December11_2017</th>\n",
       "      <td>[19906_A1, 19906_A10, 19906_A11, 19906_A12, 19...</td>\n",
       "      <td>~/proj/C1_mouse_limb_combined/all_analysis_M21</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1_mouse_e10.5_forelimb_run15_January13AM_2018</th>\n",
       "      <td>[19912_G10, 19912_G11, 19912_G12, 19912_G7, 19...</td>\n",
       "      <td>~/proj/C1_mouse_limb_combined/all_analysis_M21</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1_mouse_e11.0_forelimb_run16_January13PM_2018</th>\n",
       "      <td>[19914_A1, 19914_A10, 19914_A11, 19914_A12, 19...</td>\n",
       "      <td>~/proj/C1_mouse_limb_combined/all_analysis_M21</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1_mouse_e14.5_forelimb_run17_January16_2018</th>\n",
       "      <td>[19915_B10, 19915_B11, 19915_B12, 19915_B3, 19...</td>\n",
       "      <td>~/proj/C1_mouse_limb_combined/all_analysis_M21</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1_mouse_e12.0_forelimb_run11_December12_2017</th>\n",
       "      <td>[20026_A1, 20026_A10, 20026_A11, 20026_A12, 20...</td>\n",
       "      <td>~/proj/C1_mouse_limb_combined/all_analysis_M21</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1_mouse_e12.0_forelimb_run11_December12_2017</th>\n",
       "      <td>[20031_F1, 20031_F10, 20031_F11, 20031_F12, 20...</td>\n",
       "      <td>~/proj/C1_mouse_limb_combined/all_analysis_M21</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1_mouse_e13.0_forelimb_run12_December13_2017</th>\n",
       "      <td>[20033_A1, 20033_A10, 20033_A11, 20033_A12, 20...</td>\n",
       "      <td>~/proj/C1_mouse_limb_combined/all_analysis_M21</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1_mouse_e15.5_forelimb_run14_December15_2017</th>\n",
       "      <td>[20044_A1, 20044_A10, 20044_A11, 20044_A12, 20...</td>\n",
       "      <td>~/proj/C1_mouse_limb_combined/all_analysis_M21</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                           replicates  \\\n",
       "experiment                                                                                              \n",
       "C1_mouse_e13.5_limb_mesenchyme_mm10_clean_run4      [18251_A1, 18251_A10, 18251_A11, 18251_A12, 18...   \n",
       "C1_mouse_e11.0_limb_mesenchyme_mm10_clean_run5      [18258_A1, 18258_A10, 18258_A11, 18258_A12, 18...   \n",
       "C1_mouse_e11.5_limb_mesenchyme_mm10_clean_run6      [18263_A1, 18263_A10, 18263_A11, 18263_A12, 18...   \n",
       "C1_mouse_e12.5_limb_mesenchyme_mm10_clean_run7      [18270_A1, 18270_A10, 18270_A11, 18270_A12, 18...   \n",
       "C1_mouse_e13.5_limb_mesenchyme_mm10_clean_run8      [18311_A1, 18311_A10, 18311_A11, 18311_A12, 18...   \n",
       "C1_mouse_e10.5_limb_mm10_clean_run1                 [18042_A1, 18042_A10, 18042_A11, 18042_A12, 18...   \n",
       "C1_mouse_e10.5_limb_mm10_clean_run2                 [17327_A1, 17327_A10, 17327_A11, 17327_A12, 17...   \n",
       "C1_mouse_e10.5_limb_mm10_clean_run3                 [18087_F1, 18087_F10, 18087_F11, 18087_F12, 18...   \n",
       "C1_mouse_e14.0_forelimb_run13_December14_2017       [20039_A1, 20039_A10, 20039_A11, 20039_A12, 20...   \n",
       "C1_mouse_e12.5_forelimb_run9_March8_2017_rearra...  [20090_C1, 20090_C10, 20090_C11, 20090_C12, 20...   \n",
       "C1_mouse_e11.5_forelimb_run10_December11_2017       [19906_A1, 19906_A10, 19906_A11, 19906_A12, 19...   \n",
       "C1_mouse_e10.5_forelimb_run15_January13AM_2018      [19912_G10, 19912_G11, 19912_G12, 19912_G7, 19...   \n",
       "C1_mouse_e11.0_forelimb_run16_January13PM_2018      [19914_A1, 19914_A10, 19914_A11, 19914_A12, 19...   \n",
       "C1_mouse_e14.5_forelimb_run17_January16_2018        [19915_B10, 19915_B11, 19915_B12, 19915_B3, 19...   \n",
       "C1_mouse_e12.0_forelimb_run11_December12_2017       [20026_A1, 20026_A10, 20026_A11, 20026_A12, 20...   \n",
       "C1_mouse_e12.0_forelimb_run11_December12_2017       [20031_F1, 20031_F10, 20031_F11, 20031_F12, 20...   \n",
       "C1_mouse_e13.0_forelimb_run12_December13_2017       [20033_A1, 20033_A10, 20033_A11, 20033_A12, 20...   \n",
       "C1_mouse_e15.5_forelimb_run14_December15_2017       [20044_A1, 20044_A10, 20044_A11, 20044_A12, 20...   \n",
       "\n",
       "                                                                                      analysis_dir  \\\n",
       "experiment                                                                                           \n",
       "C1_mouse_e13.5_limb_mesenchyme_mm10_clean_run4      ~/proj/C1_mouse_limb_combined/all_analysis_M21   \n",
       "C1_mouse_e11.0_limb_mesenchyme_mm10_clean_run5      ~/proj/C1_mouse_limb_combined/all_analysis_M21   \n",
       "C1_mouse_e11.5_limb_mesenchyme_mm10_clean_run6      ~/proj/C1_mouse_limb_combined/all_analysis_M21   \n",
       "C1_mouse_e12.5_limb_mesenchyme_mm10_clean_run7      ~/proj/C1_mouse_limb_combined/all_analysis_M21   \n",
       "C1_mouse_e13.5_limb_mesenchyme_mm10_clean_run8      ~/proj/C1_mouse_limb_combined/all_analysis_M21   \n",
       "C1_mouse_e10.5_limb_mm10_clean_run1                 ~/proj/C1_mouse_limb_combined/all_analysis_M21   \n",
       "C1_mouse_e10.5_limb_mm10_clean_run2                 ~/proj/C1_mouse_limb_combined/all_analysis_M21   \n",
       "C1_mouse_e10.5_limb_mm10_clean_run3                 ~/proj/C1_mouse_limb_combined/all_analysis_M21   \n",
       "C1_mouse_e14.0_forelimb_run13_December14_2017       ~/proj/C1_mouse_limb_combined/all_analysis_M21   \n",
       "C1_mouse_e12.5_forelimb_run9_March8_2017_rearra...  ~/proj/C1_mouse_limb_combined/all_analysis_M21   \n",
       "C1_mouse_e11.5_forelimb_run10_December11_2017       ~/proj/C1_mouse_limb_combined/all_analysis_M21   \n",
       "C1_mouse_e10.5_forelimb_run15_January13AM_2018      ~/proj/C1_mouse_limb_combined/all_analysis_M21   \n",
       "C1_mouse_e11.0_forelimb_run16_January13PM_2018      ~/proj/C1_mouse_limb_combined/all_analysis_M21   \n",
       "C1_mouse_e14.5_forelimb_run17_January16_2018        ~/proj/C1_mouse_limb_combined/all_analysis_M21   \n",
       "C1_mouse_e12.0_forelimb_run11_December12_2017       ~/proj/C1_mouse_limb_combined/all_analysis_M21   \n",
       "C1_mouse_e12.0_forelimb_run11_December12_2017       ~/proj/C1_mouse_limb_combined/all_analysis_M21   \n",
       "C1_mouse_e13.0_forelimb_run12_December13_2017       ~/proj/C1_mouse_limb_combined/all_analysis_M21   \n",
       "C1_mouse_e15.5_forelimb_run14_December15_2017       ~/proj/C1_mouse_limb_combined/all_analysis_M21   \n",
       "\n",
       "                                                    run  \n",
       "experiment                                               \n",
       "C1_mouse_e13.5_limb_mesenchyme_mm10_clean_run4        4  \n",
       "C1_mouse_e11.0_limb_mesenchyme_mm10_clean_run5        5  \n",
       "C1_mouse_e11.5_limb_mesenchyme_mm10_clean_run6        6  \n",
       "C1_mouse_e12.5_limb_mesenchyme_mm10_clean_run7        7  \n",
       "C1_mouse_e13.5_limb_mesenchyme_mm10_clean_run8        8  \n",
       "C1_mouse_e10.5_limb_mm10_clean_run1                   1  \n",
       "C1_mouse_e10.5_limb_mm10_clean_run2                   2  \n",
       "C1_mouse_e10.5_limb_mm10_clean_run3                   3  \n",
       "C1_mouse_e14.0_forelimb_run13_December14_2017        13  \n",
       "C1_mouse_e12.5_forelimb_run9_March8_2017_rearra...    9  \n",
       "C1_mouse_e11.5_forelimb_run10_December11_2017        10  \n",
       "C1_mouse_e10.5_forelimb_run15_January13AM_2018       15  \n",
       "C1_mouse_e11.0_forelimb_run16_January13PM_2018       16  \n",
       "C1_mouse_e14.5_forelimb_run17_January16_2018         17  \n",
       "C1_mouse_e12.0_forelimb_run11_December12_2017        11  \n",
       "C1_mouse_e12.0_forelimb_run11_December12_2017        11  \n",
       "C1_mouse_e13.0_forelimb_run12_December13_2017        12  \n",
       "C1_mouse_e15.5_forelimb_run14_December15_2017        14  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "experiments = load_experiments([c1_remap_root / 'experiments-by-run-passing.tsv'])\n",
    "runs = []\n",
    "for experiment, row in experiments.iterrows():\n",
    "    match = re.search('run(?P<run>[\\d]+)', experiment)\n",
    "    runs.append(int(match.group('run')))\n",
    "\n",
    "experiments['run'] = runs\n",
    "experiments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "libraries_with_good_spikes = []\n",
    "library_runs = {}\n",
    "experiments_to_skip = {'C1_mouse_e10.5_limb_mm10_clean_run1', 'C1_mouse_e10.5_limb_mm10_clean_run2'}\n",
    "for experiment_name, row in experiments.iterrows():\n",
    "    if experiment_name not in experiments_to_skip:\n",
    "        libraries_with_good_spikes.extend(row.replicates)\n",
    "        for library_id in row.replicates:\n",
    "            library_runs[library_id] = row.run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "libraries_to_run = libraries.loc[libraries_with_good_spikes]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "library_runs = pandas.Series(library_runs, name='run')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>analysis_dir</th>\n",
       "      <th>genome</th>\n",
       "      <th>annotation</th>\n",
       "      <th>sex</th>\n",
       "      <th>read_1</th>\n",
       "      <th>reference_prefix</th>\n",
       "      <th>analysis_name</th>\n",
       "      <th>stranded</th>\n",
       "      <th>run</th>\n",
       "      <th>avg_cDNA</th>\n",
       "      <th>CV%</th>\n",
       "      <th>sd</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18251_A1</th>\n",
       "      <td>~/proj/C1_mouse_limb_combined/all_analysis_M21...</td>\n",
       "      <td>mm10</td>\n",
       "      <td>M21</td>\n",
       "      <td>male</td>\n",
       "      <td>[/woldlab/castor/home/sau/flowcells/C1_mouse_l...</td>\n",
       "      <td>-</td>\n",
       "      <td>18251_A1</td>\n",
       "      <td>unstranded</td>\n",
       "      <td>4</td>\n",
       "      <td>1853.2</td>\n",
       "      <td>87.21</td>\n",
       "      <td>1609.6151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18251_A10</th>\n",
       "      <td>~/proj/C1_mouse_limb_combined/all_analysis_M21...</td>\n",
       "      <td>mm10</td>\n",
       "      <td>M21</td>\n",
       "      <td>male</td>\n",
       "      <td>[/woldlab/castor/home/sau/flowcells/C1_mouse_l...</td>\n",
       "      <td>-</td>\n",
       "      <td>18251_A10</td>\n",
       "      <td>unstranded</td>\n",
       "      <td>4</td>\n",
       "      <td>1853.2</td>\n",
       "      <td>87.21</td>\n",
       "      <td>1609.6151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18251_A11</th>\n",
       "      <td>~/proj/C1_mouse_limb_combined/all_analysis_M21...</td>\n",
       "      <td>mm10</td>\n",
       "      <td>M21</td>\n",
       "      <td>male</td>\n",
       "      <td>[/woldlab/castor/home/sau/flowcells/C1_mouse_l...</td>\n",
       "      <td>-</td>\n",
       "      <td>18251_A11</td>\n",
       "      <td>unstranded</td>\n",
       "      <td>4</td>\n",
       "      <td>1853.2</td>\n",
       "      <td>87.21</td>\n",
       "      <td>1609.6151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18251_A12</th>\n",
       "      <td>~/proj/C1_mouse_limb_combined/all_analysis_M21...</td>\n",
       "      <td>mm10</td>\n",
       "      <td>M21</td>\n",
       "      <td>male</td>\n",
       "      <td>[/woldlab/castor/home/sau/flowcells/C1_mouse_l...</td>\n",
       "      <td>-</td>\n",
       "      <td>18251_A12</td>\n",
       "      <td>unstranded</td>\n",
       "      <td>4</td>\n",
       "      <td>1853.2</td>\n",
       "      <td>87.21</td>\n",
       "      <td>1609.6151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18251_A2</th>\n",
       "      <td>~/proj/C1_mouse_limb_combined/all_analysis_M21...</td>\n",
       "      <td>mm10</td>\n",
       "      <td>M21</td>\n",
       "      <td>male</td>\n",
       "      <td>[/woldlab/castor/home/sau/flowcells/C1_mouse_l...</td>\n",
       "      <td>-</td>\n",
       "      <td>18251_A2</td>\n",
       "      <td>unstranded</td>\n",
       "      <td>4</td>\n",
       "      <td>1853.2</td>\n",
       "      <td>87.21</td>\n",
       "      <td>1609.6151</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                analysis_dir genome  \\\n",
       "18251_A1   ~/proj/C1_mouse_limb_combined/all_analysis_M21...   mm10   \n",
       "18251_A10  ~/proj/C1_mouse_limb_combined/all_analysis_M21...   mm10   \n",
       "18251_A11  ~/proj/C1_mouse_limb_combined/all_analysis_M21...   mm10   \n",
       "18251_A12  ~/proj/C1_mouse_limb_combined/all_analysis_M21...   mm10   \n",
       "18251_A2   ~/proj/C1_mouse_limb_combined/all_analysis_M21...   mm10   \n",
       "\n",
       "          annotation   sex                                             read_1  \\\n",
       "18251_A1         M21  male  [/woldlab/castor/home/sau/flowcells/C1_mouse_l...   \n",
       "18251_A10        M21  male  [/woldlab/castor/home/sau/flowcells/C1_mouse_l...   \n",
       "18251_A11        M21  male  [/woldlab/castor/home/sau/flowcells/C1_mouse_l...   \n",
       "18251_A12        M21  male  [/woldlab/castor/home/sau/flowcells/C1_mouse_l...   \n",
       "18251_A2         M21  male  [/woldlab/castor/home/sau/flowcells/C1_mouse_l...   \n",
       "\n",
       "          reference_prefix analysis_name    stranded  run  avg_cDNA    CV%  \\\n",
       "18251_A1                 -      18251_A1  unstranded    4    1853.2  87.21   \n",
       "18251_A10                -     18251_A10  unstranded    4    1853.2  87.21   \n",
       "18251_A11                -     18251_A11  unstranded    4    1853.2  87.21   \n",
       "18251_A12                -     18251_A12  unstranded    4    1853.2  87.21   \n",
       "18251_A2                 -      18251_A2  unstranded    4    1853.2  87.21   \n",
       "\n",
       "                  sd  \n",
       "18251_A1   1609.6151  \n",
       "18251_A10  1609.6151  \n",
       "18251_A11  1609.6151  \n",
       "18251_A12  1609.6151  \n",
       "18251_A2   1609.6151  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "libraries_with_cdna = libraries_to_run.merge(library_runs, left_index=True, right_index=True).merge(sampled_mean, left_on='run', right_index=True)\n",
    "libraries_with_cdna.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_fastqs = []\n",
    "for library_id, row in libraries_with_cdna.iterrows():\n",
    "    normalized_row = []\n",
    "    for fastq in row.read_1:\n",
    "        normalized_row.append(fastq.replace('../all_analysis_vdir', str(c1_remap_root.expanduser() / '..' / 'all_analysis_vdir')))\n",
    "    normalized_fastqs.append(normalized_row)\n",
    "libraries_with_cdna['read_1'] = normalized_fastqs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Make sure we can find all the fastqs. There should be no warnings for this call"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "fastqs = list(find_fastqs(libraries_with_cdna, 'read_1'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "c1_pseudo = Path('c1_pseudo')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "c1_kallisto = c1_pseudo / 'kallisto'\n",
    "c1_salmon = c1_pseudo / 'salmon'\n",
    "c1_salmon_decoy = c1_pseudo / 'salmon_decoy'\n",
    "c1_kallisto_minimal = c1_pseudo / 'kallisto_minimal'\n",
    "c1_salmon_minimal = c1_pseudo / 'salmon_minimal'\n",
    "c1_kallisto_minimal_f50  = c1_pseudo / 'kallisto_minimal_f50'\n",
    "c1_kallisto_minimal_f100 = c1_pseudo / 'kallisto_minimal_f100'\n",
    "c1_kallisto_minimal_f200 = c1_pseudo / 'kallisto_minimal_f200'\n",
    "c1_kallisto_minimal_f300 = c1_pseudo / 'kallisto_minimal_f300'\n",
    "c1_kallisto_minimal_f400 = c1_pseudo / 'kallisto_minimal_f400'\n",
    "c1_kallisto_minimal_f600 = c1_pseudo / 'kallisto_minimal_f600'\n",
    "c1_kallisto_minimal_f800 = c1_pseudo / 'kallisto_minimal_f800'\n",
    "c1_kallisto_minimal_long = c1_pseudo / 'kallisto_minimal_long'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_dirs = [\n",
    "    c1_kallisto, \n",
    "    c1_salmon,\n",
    "    c1_salmon_decoy, \n",
    "    c1_kallisto_minimal, \n",
    "    c1_salmon_minimal, \n",
    "    c1_kallisto_minimal_f50,\n",
    "    c1_kallisto_minimal_f100,\n",
    "    c1_kallisto_minimal_f200,\n",
    "    c1_kallisto_minimal_f300,\n",
    "    c1_kallisto_minimal_f400,\n",
    "    c1_kallisto_minimal_f600,\n",
    "    c1_kallisto_minimal_f800,\n",
    "    c1_kallisto_minimal_long,\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "for p in analysis_dirs:\n",
    "    if not p.exists():\n",
    "        p.mkdir()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "for result_dir in analysis_dirs:\n",
    "    log = result_dir / 'logs'\n",
    "    if not log.exists():\n",
    "        log.mkdir()\n",
    "\n",
    "    for library_id in libraries_with_cdna.index:\n",
    "        analysis_dir = result_dir / library_id\n",
    "        if not analysis_dir.exists():\n",
    "            analysis_dir.mkdir()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def kallisto_condor(libraries):\n",
    "    assert 'avg_cDNA' in libraries.columns, \"Didn't receive updated library table\"\n",
    "    header = \"\"\"#!/usr/bin/condor_submit\n",
    "\n",
    "universe=vanilla\n",
    "log=logs/kallisto_c1_quant.log\n",
    "output=logs/kallisto_c1_quant.$(process).out\n",
    "error=logs/kallisto_c1_quant.$(process).out\n",
    "\n",
    "THREADS=16\n",
    "HOME=/woldlab/loxcyc/home/diane\n",
    "ENCODE_GENOME=$(HOME)/proj/encode-202006-jamboree-detrout-rna-sc-pipeline/genome/mm10-M21-male\n",
    "\n",
    "request_cpus=$(THREADS)\n",
    "executable=/woldlab/loxcyc/home/diane/proj/kallisto/build/src/kallisto \n",
    "\"\"\"\n",
    "    gtf = Path('~/proj/encode-202006-jamboree-detrout-rna-sc-pipeline/genome/mm10-M21_minimal-male/mm10-M21_minimal_det-male.gtf').expanduser()\n",
    "    read_1 = dict(find_fastqs(libraries, 'read_1'))\n",
    "    lines = [header]\n",
    "    for library_id, row in libraries.iterrows():\n",
    "        fragment = row['avg_cDNA']\n",
    "        sd = row['sd']\n",
    "        fastqs = ' '.join(read_1[library_id])\n",
    "        # --genomebam --gtf {gtf}\n",
    "        lines.append(f'arguments=\"quant -t $(THREADS) -i $(ENCODE_GENOME)/mm10-M21-male-kallisto-0.46.2.idx -o {library_id} --single --fragment-length={fragment} --sd={sd} {fastqs}\"')\n",
    "        lines.append('queue')\n",
    "            \n",
    "    return '\\n'.join(lines)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(c1_kallisto / 'kallisto-quant.condor', 'wt') as outstream:\n",
    "    outstream.write(kallisto_condor(libraries_with_cdna))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def salmon_condor(libraries):\n",
    "    assert 'avg_cDNA' in libraries.columns, \"Didn't receive updated library table\"\n",
    "    header = \"\"\"#!/usr/bin/condor_submit\n",
    "\n",
    "universe=vanilla\n",
    "log=logs/salmon_c1_quant.log\n",
    "output=logs/salmon_c1_quant.$(process).out\n",
    "error=logs/salmon_c1_quant.$(process).out\n",
    "\n",
    "THREADS=16\n",
    "HOME=/woldlab/loxcyc/home/diane\n",
    "PROJECT=$(HOME)/proj/encode-202006-jamboree-detrout-rna-sc-pipeline\n",
    "SINGULARITY_IMAGE=$(PROJECT)/salmon-container/salmon-unstable.simg \n",
    "ENCODE_GENOME=$(PROJECT)/genome/mm10-M21-male/\n",
    "\n",
    "request_cpus=$(THREADS)\n",
    "executable=/usr/bin/singularity\n",
    "\"\"\"\n",
    "    binds = \"-B /woldlab/castor/home/sau/flowcells:/woldlab/castor/home/sau/flowcells\"\n",
    "    read_1 = dict(find_fastqs(libraries, 'read_1'))\n",
    "    lines = [header]\n",
    "    for library_id, row in libraries.iterrows():\n",
    "        fastqs = ' '.join(read_1[library_id])\n",
    "        lines.append(f'arguments=\"run {binds} $(SINGULARITY_IMAGE) quant -p $(THREADS) -i $(ENCODE_GENOME)/salmon_index -o {library_id} --libType U -r {fastqs}\"')\n",
    "        lines.append('queue')\n",
    "            \n",
    "    return '\\n'.join(lines)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(c1_salmon / 'salmon-quant.condor', 'wt') as outstream:\n",
    "    outstream.write(salmon_condor(libraries_with_cdna))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def salmon_decoy_condor(libraries):\n",
    "    assert 'avg_cDNA' in libraries.columns, \"Didn't receive updated library table\"\n",
    "    header = \"\"\"#!/usr/bin/condor_submit\n",
    "\n",
    "universe=vanilla\n",
    "log=logs/salmon_decoy_c1_quant.log\n",
    "output=logs/salmon_decoy_c1_quant.$(process).out\n",
    "error=logs/salmon_decoy_c1_quant.$(process).out\n",
    "\n",
    "THREADS=16\n",
    "HOME=/woldlab/loxcyc/home/diane\n",
    "PROJECT=$(HOME)/proj/encode-202006-jamboree-detrout-rna-sc-pipeline\n",
    "SINGULARITY_IMAGE=$(PROJECT)/salmon-container/salmon-unstable.simg \n",
    "ENCODE_GENOME=$(PROJECT)/genome/mm10-M21-male/\n",
    "\n",
    "request_cpus=$(THREADS)\n",
    "executable=/usr/bin/singularity\n",
    "\"\"\"\n",
    "    binds = \"-B /woldlab/castor/home/sau/flowcells:/woldlab/castor/home/sau/flowcells\"\n",
    "    read_1 = dict(find_fastqs(libraries, 'read_1'))\n",
    "    lines = [header]\n",
    "    for library_id, row in libraries.iterrows():\n",
    "        fastqs = ' '.join(read_1[library_id])\n",
    "        lines.append(f'arguments=\"run {binds} $(SINGULARITY_IMAGE) quant -p $(THREADS) -i $(ENCODE_GENOME)/salmon_index_decoy -o {library_id} --libType U -r {fastqs}\"')\n",
    "        lines.append('queue')\n",
    "            \n",
    "    return '\\n'.join(lines)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(c1_salmon_decoy / 'salmon-decoy-quant.condor', 'wt') as outstream:\n",
    "    outstream.write(salmon_decoy_condor(libraries_with_cdna))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run with minimal annotation set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def kallisto_minimal_condor(libraries):\n",
    "    assert 'avg_cDNA' in libraries.columns, \"Didn't receive updated library table\"\n",
    "    header = \"\"\"#!/usr/bin/condor_submit\n",
    "\n",
    "universe=vanilla\n",
    "log=logs/kallisto_c1_minimal_quant.log\n",
    "output=logs/kallisto_c1_minimal_quant.$(process).out\n",
    "error=logs/kallisto_c1_minimal_quant.$(process).out\n",
    "\n",
    "THREADS=16\n",
    "HOME=/woldlab/loxcyc/home/diane\n",
    "ENCODE_GENOME=$(HOME)/proj/encode-202006-jamboree-detrout-rna-sc-pipeline/genome/mm10-M21_minimal-male\n",
    "\n",
    "request_cpus=$(THREADS)\n",
    "executable=/woldlab/loxcyc/home/diane/proj/kallisto/build/src/kallisto \n",
    "\"\"\"\n",
    "    gtf = Path('~/proj/encode-202006-jamboree-detrout-rna-sc-pipeline/genome/mm10-M21_minimal-male/mm10-M21_minimal_det-male.gtf').expanduser()\n",
    "    read_1 = dict(find_fastqs(libraries, 'read_1'))\n",
    "    lines = [header]\n",
    "    for library_id, row in libraries.iterrows():\n",
    "        fragment = row['avg_cDNA']\n",
    "        sd = row['sd']\n",
    "        fastqs = ' '.join(read_1[library_id])\n",
    "        # --genomebam --gtf {gtf} \n",
    "        lines.append(f'arguments=\"quant -t $(THREADS) -i $(ENCODE_GENOME)/mm10-M21_minimal-male-kallisto-0.46.2.idx -o {library_id} --single --fragment-length={fragment} --sd={sd} {fastqs}\"')\n",
    "        lines.append('queue')\n",
    "            \n",
    "    return '\\n'.join(lines)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(c1_kallisto_minimal / 'kallisto-minimal-quant.condor', 'wt') as outstream:\n",
    "    outstream.write(kallisto_minimal_condor(libraries_with_cdna))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "def salmon_minimal_condor(libraries):\n",
    "    assert 'avg_cDNA' in libraries.columns, \"Didn't receive updated library table\"\n",
    "    header = \"\"\"#!/usr/bin/condor_submit\n",
    "\n",
    "universe=vanilla\n",
    "log=logs/salmon_c1_minimal_quant.log\n",
    "output=logs/salmon_c1_minimal_quant.$(process).out\n",
    "error=logs/salmon_c1_minimal_quant.$(process).out\n",
    "\n",
    "THREADS=16\n",
    "HOME=/woldlab/loxcyc/home/diane\n",
    "PROJECT=$(HOME)/proj/encode-202006-jamboree-detrout-rna-sc-pipeline\n",
    "SINGULARITY_IMAGE=$(PROJECT)/salmon-container/salmon-unstable.simg \n",
    "ENCODE_GENOME=$(PROJECT)/genome/mm10-M21_minimal-male/\n",
    "\n",
    "request_cpus=$(THREADS)\n",
    "executable=/usr/bin/singularity\n",
    "\"\"\"\n",
    "    binds = \"-B /woldlab/castor/home/sau/flowcells:/woldlab/castor/home/sau/flowcells\"\n",
    "    read_1 = dict(find_fastqs(libraries, 'read_1'))\n",
    "    lines = [header]\n",
    "    for library_id, row in libraries.iterrows():\n",
    "        fastqs = ' '.join(read_1[library_id])\n",
    "        lines.append(f'arguments=\"run {binds} $(SINGULARITY_IMAGE) quant -p $(THREADS) -i $(ENCODE_GENOME)/salmon_index -o {library_id} --libType U -r {fastqs}\"')\n",
    "        lines.append('queue')\n",
    "            \n",
    "    return '\\n'.join(lines)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(c1_salmon_minimal / 'salmon-quant.condor', 'wt') as outstream:\n",
    "    outstream.write(salmon_minimal_condor(libraries_with_cdna))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parameter sweep of fragment length for kallisto minimal "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ever get that feeling you should've been less lazy and stopped copying and pasting?\n",
    "# Then after you refactor everything do you think it was worth it?\n",
    "\n",
    "def kallisto_minimal_fragment_condor(libraries, length):\n",
    "    assert 'avg_cDNA' in libraries.columns, \"Didn't receive updated library table\"\n",
    "    header = f\"\"\"#!/usr/bin/condor_submit\n",
    "\n",
    "universe=vanilla\n",
    "log=logs/kallisto_c1_quant_f{length}.log\n",
    "output=logs/kallisto_c1_quant_f{length}.$(process).out\n",
    "error=logs/kallisto_c1_quant_f{length}.$(process).out\n",
    "\n",
    "THREADS=16\n",
    "HOME=/woldlab/loxcyc/home/diane\n",
    "ENCODE_GENOME=$(HOME)/proj/encode-202006-jamboree-detrout-rna-sc-pipeline/genome/mm10-M21-male\n",
    "\n",
    "request_cpus=$(THREADS)\n",
    "executable=/woldlab/loxcyc/home/diane/proj/kallisto/build/src/kallisto \n",
    "\"\"\"\n",
    "    gtf = Path('~/proj/encode-202006-jamboree-detrout-rna-sc-pipeline/genome/mm10-M21_minimal-male/mm10-M21_minimal_det-male.gtf').expanduser()\n",
    "    read_1 = dict(find_fastqs(libraries, 'read_1'))\n",
    "    lines = [header]\n",
    "    for library_id, row in libraries.iterrows():\n",
    "        sd = length * .2\n",
    "        fastqs = ' '.join(read_1[library_id])\n",
    "        # No genome bams now. --genomebam --gtf {gtf}\n",
    "        lines.append(f'arguments=\"quant -t $(THREADS) -i $(ENCODE_GENOME)/mm10-M21-male-kallisto-0.46.2.idx -o {library_id} --single --fragment-length={length} --sd={sd} {fastqs}\"')\n",
    "        lines.append('queue')\n",
    "        \n",
    "    return \"\\n\".join(lines)\n",
    "\n",
    "def write_kallisto_minimal_fragment_condor(target_dir, libraries, length):\n",
    "    with open(target_dir / f'kallisto-minimal-f{length}-quant.condor', 'wt') as outstream:\n",
    "        outstream.write(kallisto_minimal_fragment_condor(libraries, length))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "length_sweep = [\n",
    "    (c1_kallisto_minimal_f50, 50),\n",
    "    (c1_kallisto_minimal_f100, 100),\n",
    "    (c1_kallisto_minimal_f200, 200),\n",
    "    (c1_kallisto_minimal_f300, 300),\n",
    "    (c1_kallisto_minimal_f400, 400),\n",
    "    (c1_kallisto_minimal_f600, 600),\n",
    "    (c1_kallisto_minimal_f800, 800),\n",
    "]\n",
    "\n",
    "for target_dir, length in length_sweep:\n",
    "    write_kallisto_minimal_fragment_condor(target_dir, libraries_with_cdna, length)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prepare kallisto long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "def kallisto_minimal_long_condor(libraries):\n",
    "    header = \"\"\"#!/usr/bin/condor_submit\n",
    "\n",
    "universe=vanilla\n",
    "log=logs/kallisto_c1_quant_long.log\n",
    "output=logs/kallisto_c1_quant_long.$(process).out\n",
    "error=logs/kallisto_c1_quant_long.$(process).out\n",
    "\n",
    "THREADS=16\n",
    "HOME=/woldlab/loxcyc/home/diane\n",
    "ENCODE_GENOME=$(HOME)/proj/encode-202006-jamboree-detrout-rna-sc-pipeline/genome/mm10-M21-male\n",
    "\n",
    "request_cpus=$(THREADS)\n",
    "executable=/woldlab/loxcyc/home/diane/proj/kallisto-long/build/src/kallisto \n",
    "\"\"\"\n",
    "    gtf = Path('~/proj/encode-202006-jamboree-detrout-rna-sc-pipeline/genome/mm10-M21_minimal-male/mm10-M21_minimal_det-male.gtf').expanduser()\n",
    "    read_1 = dict(find_fastqs(libraries, 'read_1'))\n",
    "    lines = [header]\n",
    "    for library_id, row in libraries.iterrows():\n",
    "        fastqs = ' '.join(read_1[library_id])\n",
    "        # No genome bams now. --genomebam --gtf {gtf}\n",
    "        lines.append(f'arguments=\"quant -t $(THREADS) -i $(ENCODE_GENOME)/mm10-M21-male-kallisto-0.46.2.idx -o {library_id} --long {fastqs}\"')\n",
    "        lines.append('queue')\n",
    "        \n",
    "    return \"\\n\".join(lines)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(c1_kallisto_minimal_long / 'kallisto-minimal-long-quant.condor', 'wt') as outstream:\n",
    "    outstream.write(kallisto_minimal_long_condor(libraries_with_cdna))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prepare standard ENCODE pipeline run with minimal index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['analysis_dir', 'genome', 'annotation', 'sex', 'read_1',\n",
       "       'reference_prefix', 'analysis_name', 'stranded'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "libraries.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "libraries_minimal = libraries.copy().drop('analysis_name', axis=1)\n",
    "libraries_minimal['analysis_dir'] = libraries.index\n",
    "libraries_minimal['read_1'] = libraries_minimal['read_1'].apply(lambda x: ','.join(x))\n",
    "libraries_minimal['annotation'] = 'M21_minimal'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "libraries_minimal.to_csv(c1_pseudo / 'rsem_minimal' / 'libraries.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "experiments.to_csv(c1_pseudo / 'rsem_minimal' / 'experiments.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
