{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "distinguished-slope",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas\n",
    "from pathlib import Path\n",
    "from pprint import pprint\n",
    "from io import BytesIO\n",
    "import re\n",
    "import sys\n",
    "import tarfile\n",
    "from collections import Counter, namedtuple"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "popular-console",
   "metadata": {},
   "outputs": [],
   "source": [
    "EC = str(Path(\"~/proj/encoded_client\").expanduser())\n",
    "if EC not in sys.path:\n",
    "    sys.path.append(EC)\n",
    "from encoded_client.encoded import ENCODED, HTTPError"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "unexpected-error",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "narrow-recycling",
   "metadata": {},
   "outputs": [],
   "source": [
    "known_document_types = {\n",
    "    '/documents/ecb1c489-905b-4afb-8389-ed7fb2e36ff2/': \"Chromium v2\",\n",
    "    '/documents/515a990c-50c4-459a-8036-502feaf5a18c/': \"Chromium v2 technical note\",\n",
    "    '/documents/3408b7a1-6b1b-45d1-a73e-db0556e6abdc/': \"Split-seq analysis protocol\",\n",
    "    '/documents/3cdf327b-df85-4149-9fca-a7541e6f1cb1/': \"Split-seq overview\",\n",
    "    '/documents/c155e58a-13ab-4af9-8591-cee13d639a40/': \"Nuclei Isolation protocol from tissue for 10x multiome\",\n",
    "    '/documents/53fb4382-c54f-4761-89bb-6a01a0dca97d/': \"Chromium GEM v3.1\",\n",
    "    '/documents/4d42b1ef-c9e6-4b31-a443-1266079545d0/': \"Chromium Multiome ATAC + Gene Expression\",\n",
    "    '/documents/d18bad0e-9a86-40f0-8ded-2c74072910e5/': \"Nuclei Isolation protocol from tissue for 10x multiome\",  # probably same as c155e58a-13ab-4af9-8591-cee13d639a40\n",
    "    '/documents/e3110414-4013-425e-94f6-72c549702bde/': \"Pool split protocol\",\n",
    "    '/documents/1bb75b62-ac29-4368-9855-68d410e1963a/': \"GM12878 Cell Culture protocol\",\n",
    "    '/documents/28fef8e0-e171-450d-b66d-e3d199531cf2/': \"Ambion spike in mixture\",\n",
    "    '/documents/a6e06058-5019-488f-9195-8bcd2e9229c1/': \"Laser capture microdisection\",\n",
    "    '/documents/c4b8952c-d9f1-42e7-8ea5-54659d14f46b/': \"Human tissue preservation protocol flash-freezing (version 1)\",\n",
    "    '/documents/c5c4174c-0f33-4793-be9d-f2b6a9357dae/': \"Nuclei Isolation protocol from tissue for 10x multiome\", # probably same as c155e58a-13ab-4af9-8591-cee13d639a40\n",
    "    '/documents/84a66218-cd6a-4959-adc0-91ecc513117f/': \"Chromium Multiome ATAC + Gene Expression\",\n",
    "    '/documents/c6235c97-a978-41b6-8a7e-1e16b983f4ba/': \"inDrop Sample Collection Protocol\",\n",
    "    '/documents/8179222b-9ad7-4274-a07e-820dbcac6aca/': \"inDrop Library Prep\",\n",
    "    '/documents/80953249-5d87-4f2c-84e0-0212c404c8d6/': \"Nuclei Isolation protocol from tissue 10x snRNA\", # probably same as c155e58a-13ab-4af9-8591-cee13d639a40\n",
    "    '/documents/57177fd0-da98-4e8f-95a4-459fbaeea146/': \"Chromium GEM v3.1\", # psame as 53fb4382-c54f-4761-89bb-6a01a0dca97d\n",
    "    '/documents/5f5193b0-0c61-40c6-a286-b1f6d6f0c0da/': \"SmartSeq\",\n",
    "    '/documents/9a7e281f-b901-422b-9f01-378ba4b9e97f/': \"Mouse adrenal gland nuclei isolation protocol for single nucleus RNA-seq\",\n",
    "    '/documents/f1b60b4e-b5d2-4b8d-af0b-06b767fa07c3/': \"Biosample purification protocol\",\n",
    "    '/documents/49fd5b3b-9878-4ec8-92bf-ef6dce9116e3/': \"SMARTer Ultra Low RNA kit\",\n",
    "    \"/documents/f16e44c9-ee9d-4ede-9bb5-886ea451d7a0/\": \"Cortex nuclei isolation protocol\",\n",
    "    \"/documents/eb624ae4-53ea-4de0-ba9e-11b40eb20e2c/\": \"cell isolation protocol\",\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "pressed-sellers",
   "metadata": {},
   "outputs": [],
   "source": [
    "document_to_protocol = {\n",
    "    '/documents/ecb1c489-905b-4afb-8389-ed7fb2e36ff2/': \"Chromium v2\",\n",
    "    #'/documents/515a990c-50c4-459a-8036-502feaf5a18c/': \"Chromium v2\",\n",
    "    '/documents/3408b7a1-6b1b-45d1-a73e-db0556e6abdc/': \"Split-seq\",\n",
    "    '/documents/3cdf327b-df85-4149-9fca-a7541e6f1cb1/': \"Split-seq\",\n",
    "    #'/documents/c155e58a-13ab-4af9-8591-cee13d639a40/': \"Nuclei Isolation protocol from tissue for 10x multiome\",\n",
    "    '/documents/53fb4382-c54f-4761-89bb-6a01a0dca97d/': \"Chromium GEM v3.1\",\n",
    "    '/documents/4d42b1ef-c9e6-4b31-a443-1266079545d0/': \"Chromium Multiome ATAC + Gene Expression\", # same as 53fb4382-c54f-4761-89bb-6a01a0dca97d\n",
    "    #'/documents/d18bad0e-9a86-40f0-8ded-2c74072910e5/': \"Nuclei Isolation protocol from tissue for 10x multiome\",  # probably same as c155e58a-13ab-4af9-8591-cee13d639a40\n",
    "    '/documents/d18bad0e-9a86-40f0-8ded-2c74072910e5/': \"Chromium Multiome ATAC + Gene Expression\",\n",
    "    '/documents/e3110414-4013-425e-94f6-72c549702bde/': \"SmartSeq Pool\",\n",
    "    '/documents/a6e06058-5019-488f-9195-8bcd2e9229c1/': \"SmartSeq\",\n",
    "    #'/documents/c5c4174c-0f33-4793-be9d-f2b6a9357dae/': \"Nuclei Isolation protocol from tissue for 10x multiome\", # probably same as c155e58a-13ab-4af9-8591-cee13d639a40\n",
    "    '/documents/84a66218-cd6a-4959-adc0-91ecc513117f/': \"Chromium Multiome ATAC + Gene Expression\",\n",
    "    '/documents/c6235c97-a978-41b6-8a7e-1e16b983f4ba/': \"inDrop\",\n",
    "    '/documents/8179222b-9ad7-4274-a07e-820dbcac6aca/': \"inDrop\",\n",
    "    #'/documents/80953249-5d87-4f2c-84e0-0212c404c8d6/': \"Nuclei Isolation protocol from tissue 10x snRNA\", # probably same as c155e58a-13ab-4af9-8591-cee13d639a40\n",
    "    '/documents/57177fd0-da98-4e8f-95a4-459fbaeea146/': \"Chromium GEM v3.1\", # psame as 53fb4382-c54f-4761-89bb-6a01a0dca97d\n",
    "    '/documents/5f5193b0-0c61-40c6-a286-b1f6d6f0c0da/': \"SmartSeq\",\n",
    "    '/documents/49fd5b3b-9878-4ec8-92bf-ef6dce9116e3/': \"SmartSeq\",\n",
    "}\n",
    "\n",
    "#From alex https://en-coding.slack.com/archives/C014UJ5F73Q/p1633558205130000\n",
    "#10X 3' protocol v2 and v3 were Forward-stranded, 5'-v2 is Reverse-stranded. I am not aware of any of their protocols that are unstranded.\n",
    "protocol_defaults = {\n",
    "    \"Chromium v2\": {\n",
    "        \"stranded\": \"Forward\",\n",
    "        \"umi_length\": 10,\n",
    "        #\"inclusion_list_url\": \"https://encd-6248-78c147870-qing.demo.encodedcc.org/files/737K-august-2016/@@download/737K-august-2016.txt.gz\"\n",
    "        #\"inclusion_list_url\": \"https://woldlab.caltech.edu/~diane/genome/737K-arc-v1.txt.gz\"\n",
    "        #\"inclusion_list_url\": \"https://encd-6248-24bd77146-qing.demo.encodedcc.org/files/737K-august-2016/@@download/737K-august-2016.txt.gz\"\n",
    "        \"inclusion_accession\": \"737K-august-2016\"\n",
    "    },\n",
    "    \"Chromium GEM v3.1\": {\n",
    "        \"stranded\": \"Forward\",\n",
    "        \"umi_length\": 12,\n",
    "        #\"inclusion_list_url\": \"https://encd-6248-24bd77146-qing.demo.encodedcc.org/files/3M-february-2018/@@download/3M-february-2018.txt.gz\"\n",
    "        \"inclusion_accession\": \"3M-february-2018\"\n",
    "    },\n",
    "    \"Chromium Multiome ATAC + Gene Expression\": {\n",
    "        \"stranded\": \"Forward\",\n",
    "        \"umi_length\": 12,\n",
    "        \"inclusion_accession\": \"737K-arc-v1(GEX)\",\n",
    "        #\"inclusion_list_url\": \"https://encd-6248-78c147870-qing.demo.encodedcc.org/files/737K-arc-v1(GEX)/@@download/737K-arc-v1(GEX).txt.gz\",\n",
    "        #\"inclusion_list_url\": \"https://woldlab.caltech.edu/~diane/genome/737K-arc-v1(GEX).txt.gz\",\n",
    "        #\"inclusion_list_url\": \"https://encd-6248-24bd77146-qing.demo.encodedcc.org/files/737K-arc-v1(GEX)/@@download/737K-arc-v1(GEX).txt.gz\"\n",
    "    },\n",
    "    \"Split-seq\" : {\n",
    "        \"stranded\": \"Forward\",\n",
    "        \"inclusion_list_url\": \"https://woldlab.caltech.edu/~diane/genome/parse-splitseq-v1.tar.gz\",\n",
    "        \"inclusion_accession\": \"encode:parse-biosciences-v1\",\n",
    "    }\n",
    "}\n",
    "\n",
    "default_genome_index = {\n",
    "    #\"/organisms/human/\": \"https://encd-6248-78c147870-qing.demo.encodedcc.org/files/TSTFF202658/@@download/TSTFF202658.tar.gz\"\n",
    "    \"/organisms/human/\": {\n",
    "        #\"genome_index_url\": \"https://woldlab.caltech.edu/~diane/genome/GRCh38-V29-male-2.7.8a.tar.gz\",\n",
    "        \"genome_accession\": \"encode:starsolo-GRCh38-V29-male-index\",\n",
    "        #\"assembly\": \"GRCh38\",\n",
    "        #\"genome_annotation\": \"V29\",\n",
    "    },\n",
    "    \"/organisms/mouse/\": {\n",
    "        #\"genome_index_url\": \"https://woldlab.caltech.edu/~diane/genome/mm10-M21-male-2.7.9a/mm10-M21-male-2.7.9a-star.tar.gz\",\n",
    "        \"genome_accession\": \"encode:starsolo-mm10-M21-male-index\",\n",
    "        #\"assembly\": \"mm10\",\n",
    "        #\"genome_annotation\": \"M21\",\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "seventh-blink",
   "metadata": {},
   "outputs": [],
   "source": [
    "server = ENCODED(\"www.encodeproject.org\")\n",
    "#server = ENCODED(\"test.encodedcc.org\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "appointed-socket",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENCFF930GKT Lab custom GRCh38 V29\n",
      "GeneFull_Ex50pAS_Unique_filtered.tar.gz\n",
      "ENCFF831OBZ Lab custom GRCh38 V29\n",
      "ENCFF820LUU Lab custom GRCh38 V29\n",
      "ENCFF701VIQ Lab custom GRCh38 V29\n",
      "ENCFF220CCK Lab custom GRCh38 V29\n"
     ]
    }
   ],
   "source": [
    "experiment = server.get_json(\"/experiments/ENCSR203THX/\")\n",
    "has_processed = []\n",
    "for f in experiment[\"files\"]:\n",
    "    submitted_file_name = Path(f[\"submitted_file_name\"])\n",
    "    if f.get(\"output_type\") == \"sparse gene count matrix of unique reads\":\n",
    "        print(submitted_file_name)\n",
    "    elif len(f.get(\"analyses\", [])) > 0:\n",
    "        title = f[\"analyses\"][0][\"title\"]\n",
    "        print(f['accession'], title)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "noble-assessment",
   "metadata": {},
   "outputs": [],
   "source": [
    "query = server.search_jsonld(type=\"Experiment\", assay_title=\"scRNA-seq\", limit=\"all\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "immediate-proof",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENCLB779BCG Forward != reverse\n",
      "ENCLB338RBX Forward != unstranded\n",
      "ENCLB774YSO Forward != unstranded\n",
      "ENCLB067EKD Forward != reverse\n",
      "ENCLB043KZW Forward != reverse\n",
      "ENCLB243NMQ Forward != reverse\n",
      "ENCLB113GDH Forward != reverse\n",
      "ENCLB765NUA Forward != unstranded\n",
      "ENCLB273ZWE Forward != unstranded\n",
      "ENCLB555STO Forward != reverse\n",
      "ENCLB950DIT Forward != reverse\n",
      "ENCLB487ULG Forward != reverse\n",
      "ENCLB145HAH Forward != reverse\n",
      "ENCLB457YKG Forward != reverse\n",
      "ENCLB234BIK Forward != unstranded\n",
      "ENCLB527JDU Forward != unstranded\n",
      "ENCLB835IXS Forward != reverse\n",
      "ENCLB419YTE Forward != unstranded\n",
      "ENCLB943WKY Forward != reverse\n",
      "ENCLB508DVG Forward != reverse\n",
      "ENCLB943KZN Forward != reverse\n",
      "ENCLB433LMI Forward != reverse\n",
      "ENCLB863YCR Forward != reverse\n",
      "ENCLB150JOE Forward != reverse\n",
      "ENCLB138XBO Forward != reverse\n",
      "ENCLB155EAA Forward != reverse\n",
      "ENCLB888MGZ Forward != reverse\n",
      "ENCLB213TAB Forward != reverse\n",
      "ENCLB398IAZ Forward != reverse\n",
      "ENCLB513JLQ Forward != unstranded\n",
      "ENCLB025XJD Forward != unstranded\n",
      "ENCLB335ITX Forward != unstranded\n",
      "ENCLB971YWN Forward != unstranded\n",
      "ENCLB866TKE Forward != unstranded\n",
      "ENCLB461YMD Forward != reverse\n",
      "ENCLB761CFQ Forward != reverse\n",
      "ENCLB179BOW Forward != reverse\n",
      "ENCLB414SND Forward != reverse\n",
      "ENCLB812SMT Forward != unstranded\n",
      "ENCLB615BOP Forward != unstranded\n",
      "ENCLB701RZY Forward != reverse\n",
      "ENCLB025GOX Forward != reverse\n",
      "ENCLB514RKT Forward != unstranded\n",
      "ENCLB294FBZ Forward != unstranded\n",
      "ENCLB280ZGL Forward != unstranded\n",
      "ENCLB952MZJ Forward != unstranded\n",
      "ENCLB392SOP Forward != reverse\n",
      "ENCLB801PEP Forward != reverse\n",
      "ENCLB342YHF Forward != unstranded\n",
      "ENCLB338QDU Forward != reverse\n",
      "ENCLB507KDD Forward != reverse\n",
      "ENCLB218CWN Forward != reverse\n",
      "ENCLB793KIA Forward != reverse\n",
      "ENCLB140VOU Forward != reverse\n",
      "ENCLB872TNB Forward != reverse\n",
      "ENCLB001JCO Forward != reverse\n",
      "ENCLB399EQG Forward != reverse\n",
      "ENCLB814PUF Forward != unstranded\n",
      "ENCLB327PYA Forward != reverse\n",
      "ENCLB151JZY Forward != reverse\n",
      "ENCLB061STU Forward != reverse\n",
      "ENCLB297FDO Forward != unstranded\n",
      "ENCLB002DZK Forward != unstranded\n",
      "ENCLB549SMV Forward != unstranded\n",
      "ENCLB611BAN Forward != reverse\n",
      "ENCLB997ETR Forward != reverse\n",
      "ENCLB954OKJ Forward != reverse\n",
      "ENCLB129RYA Forward != reverse\n",
      "ENCLB065ONF Forward != reverse\n",
      "ENCLB196FUF Forward != reverse\n",
      "ENCLB357XZN Forward != reverse\n",
      "ENCLB504FSE Forward != reverse\n",
      "ENCLB038DYV Forward != reverse\n",
      "ENCLB529BXD Forward != reverse\n",
      "ENCLB452QHE Forward != reverse\n",
      "ENCLB366ZFV Forward != reverse\n",
      "ENCLB681VIY Forward != reverse\n",
      "ENCLB986LXE Forward != reverse\n",
      "ENCLB593ESF Forward != reverse\n",
      "ENCLB765JXX Forward != reverse\n",
      "ENCLB455UTN Forward != reverse\n",
      "ENCLB123ZDT Forward != reverse\n",
      "ENCLB039CPV Forward != reverse\n",
      "ENCLB181TPV Forward != unstranded\n",
      "ENCLB907NAX Forward != reverse\n",
      "ENCLB039THH Forward != unstranded\n",
      "ENCLB212WDJ Forward != reverse\n",
      "ENCLB960RBM Forward != unstranded\n",
      "ENCLB814OYP Forward != unstranded\n",
      "ENCLB129LMS Forward != unstranded\n",
      "ENCLB296LDF Forward != unstranded\n",
      "ENCLB592LXS Forward != unstranded\n",
      "ENCLB971CVF Forward != unstranded\n",
      "ENCLB396URL Forward != unstranded\n",
      "ENCLB062TYC Forward != unstranded\n",
      "ENCLB323ARH Forward != reverse\n",
      "ENCLB008XGB Forward != reverse\n",
      "ENCLB900EOL Forward != reverse\n",
      "ENCLB258LQF Forward != reverse\n",
      "ENCLB505NAL Forward != reverse\n",
      "ENCLB295OOS Forward != reverse\n",
      "ENCLB103OKD Forward != reverse\n",
      "ENCLB849AUZ Forward != reverse\n",
      "ENCLB032OPR Forward != reverse\n",
      "ENCLB460PSY Forward != reverse\n",
      "ENCLB845QWH Forward != reverse\n",
      "ENCLB861GQS Forward != reverse\n",
      "ENCLB802MGH Forward != reverse\n",
      "ENCLB000QLP Forward != reverse\n",
      "ENCLB860YRM Forward != reverse\n",
      "ENCLB280OVJ Forward != reverse\n",
      "ENCLB610ELT Forward != reverse\n",
      "ENCLB790XHI Forward != reverse\n",
      "ENCLB063JPD Forward != reverse\n",
      "ENCLB228GCE Forward != reverse\n",
      "ENCLB272QLF Forward != reverse\n",
      "ENCLB909DQG Forward != unstranded\n",
      "ENCLB619VVF Forward != unstranded\n",
      "ENCLB549KAN Forward != unstranded\n",
      "ENCLB674XYV Forward != reverse\n",
      "ENCLB932FNN Forward != reverse\n",
      "ENCLB418IDT Forward != reverse\n",
      "ENCLB846XMI Forward != not set\n",
      "ENCLB673QFY Forward != reverse\n",
      "ENCLB643HGB Forward != unstranded\n",
      "ENCLB578ADC Forward != unstranded\n",
      "ENCLB527WWJ Forward != unstranded\n",
      "ENCLB328JIP Forward != unstranded\n",
      "ENCLB831IHU Forward != unstranded\n",
      "ENCLB836BRZ Forward != unstranded\n",
      "ENCLB638WEJ Forward != unstranded\n",
      "ENCLB540ICC Forward != unstranded\n",
      "ENCLB607FDJ Forward != reverse\n",
      "ENCLB256HGX Forward != unstranded\n",
      "ENCLB753SMP Forward != unstranded\n",
      "ENCLB619PRE Forward != unstranded\n",
      "ENCLB910PXF Forward != unstranded\n",
      "ENCLB232EFI Forward != unstranded\n",
      "ENCLB858UWU Forward != unstranded\n",
      "ENCLB481LCB Forward != unstranded\n",
      "ENCLB613RTM Forward != unstranded\n",
      "ENCLB093BII Forward != unstranded\n",
      "ENCLB776BDA Forward != unstranded\n",
      "ENCLB866MAL Forward != reverse\n",
      "ENCLB337GLO Forward != unstranded\n",
      "ENCLB275OVJ Forward != reverse\n",
      "ENCLB715SBJ Forward != unstranded\n",
      "ENCLB154DIW Forward != unstranded\n",
      "ENCLB704BQZ Forward != unstranded\n",
      "ENCLB463YUO Forward != unstranded\n",
      "ENCLB673NZA Forward != reverse\n",
      "ENCLB970EMO Forward != unstranded\n",
      "ENCLB540USO Forward != unstranded\n",
      "ENCLB765SCG Forward != unstranded\n",
      "ENCLB883MZY Forward != unstranded\n",
      "ENCLB266USG Forward != unstranded\n",
      "ENCLB295AKM Forward != unstranded\n",
      "ENCLB449WLH Forward != unstranded\n",
      "ENCLB595FHR Forward != unstranded\n",
      "ENCLB675MBI Forward != reverse\n",
      "ENCLB621JUD Forward != unstranded\n",
      "ENCLB390NYX Forward != unstranded\n",
      "ENCLB342BHZ Forward != unstranded\n",
      "ENCLB570ZQE Forward != unstranded\n",
      "ENCLB372WFL Forward != unstranded\n",
      "ENCLB958FDX Forward != unstranded\n",
      "ENCLB236NYK Forward != unstranded\n",
      "ENCLB849GLG Forward != unstranded\n",
      "ENCLB819OXS Forward != unstranded\n",
      "ENCLB689QBZ Forward != unstranded\n",
      "ENCLB536VVO Forward != unstranded\n",
      "ENCLB950YDB Forward != unstranded\n",
      "ENCLB808STP Forward != reverse\n",
      "ENCLB363YDA Forward != unstranded\n",
      "ENCLB788WXA Forward != unstranded\n",
      "ENCLB003RYS Forward != unstranded\n",
      "ENCLB590WTC Forward != unstranded\n",
      "ENCLB524WLL Forward != unstranded\n",
      "ENCLB585BUB Forward != unstranded\n",
      "ENCLB003PKT Forward != unstranded\n",
      "ENCLB445HCF Forward != unstranded\n",
      "ENCLB633YKU Forward != unstranded\n",
      "ENCLB320HXY Forward != reverse\n",
      "ENCLB731QGM Forward != unstranded\n",
      "ENCLB260YXF Forward != unstranded\n",
      "ENCLB275WWO Forward != reverse\n",
      "ENCLB039CRC Forward != unstranded\n",
      "ENCLB278RJO Forward != unstranded\n",
      "ENCLB442AMV Forward != reverse\n",
      "ENCLB927HPV Forward != unstranded\n",
      "ENCLB320ZLU Forward != reverse\n",
      "ENCLB793ZEJ Forward != unstranded\n",
      "ENCLB620HWQ Forward != unstranded\n",
      "ENCLB284ZJK Forward != unstranded\n",
      "ENCLB069PRB Forward != reverse\n",
      "ENCLB301CVY Forward != unstranded\n",
      "ENCLB019ZEA Forward != unstranded\n",
      "ENCLB964XJT Forward != unstranded\n",
      "ENCLB828XLK Forward != unstranded\n",
      "ENCLB134RGQ Forward != unstranded\n",
      "ENCLB960KQP Forward != unstranded\n",
      "ENCLB751QGW Forward != unstranded\n",
      "ENCLB639CLJ Forward != unstranded\n",
      "ENCLB577LGZ Forward != unstranded\n",
      "ENCLB996PVL Forward != unstranded\n",
      "ENCLB796MAN Forward != unstranded\n",
      "ENCLB891WAL Forward != unstranded\n",
      "ENCLB797XUM Forward != unstranded\n",
      "ENCLB256KMY Forward != unstranded\n",
      "ENCLB840OSY Forward != unstranded\n",
      "ENCLB984ILH Forward != unstranded\n",
      "ENCLB381HJR Forward != unstranded\n",
      "ENCLB208PDA Forward != unstranded\n",
      "ENCLB966DFA Forward != unstranded\n",
      "ENCLB850RYI Forward != reverse\n",
      "ENCLB818NOG Forward != unstranded\n",
      "ENCLB293BAX Forward != reverse\n",
      "ENCLB549HEK Forward != unstranded\n",
      "ENCLB621EIT Forward != unstranded\n",
      "ENCLB734ODO Forward != unstranded\n",
      "ENCLB845DSU Forward != unstranded\n",
      "ENCLB841LXQ Forward != unstranded\n",
      "ENCLB370RDJ Forward != unstranded\n",
      "ENCLB859SFU Forward != unstranded\n",
      "ENCLB454IQF Forward != unstranded\n",
      "ENCLB089YWN Forward != reverse\n",
      "ENCLB871LRF Forward != unstranded\n",
      "ENCLB807VEG Forward != reverse\n",
      "ENCLB927YFW Forward != unstranded\n",
      "ENCLB773HLI Forward != unstranded\n",
      "ENCLB020CON Forward != unstranded\n",
      "ENCLB242YJU Forward != unstranded\n",
      "ENCLB854VKF Forward != reverse\n",
      "ENCLB961HIF Forward != unstranded\n",
      "ENCLB246EKZ Forward != unstranded\n",
      "ENCLB921TTM Forward != unstranded\n",
      "ENCLB972LZE Forward != unstranded\n",
      "ENCLB678IDN Forward != unstranded\n",
      "ENCLB139QES Forward != reverse\n",
      "ENCLB514RLH Forward != unstranded\n",
      "ENCLB519CDG Forward != unstranded\n",
      "ENCLB607NFB Forward != unstranded\n",
      "ENCLB856JIZ Forward != unstranded\n",
      "ENCLB944QTS Forward != unstranded\n",
      "ENCLB352VYR Forward != unstranded\n",
      "ENCLB669VKT Forward != unstranded\n",
      "ENCLB063XDV Forward != reverse\n",
      "ENCLB223DPT Forward != unstranded\n",
      "ENCLB122RJO Forward != unstranded\n",
      "ENCLB225LKT Forward != unstranded\n",
      "ENCLB578UHU Forward != unstranded\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENCLB063DMS Forward != unstranded\n",
      "ENCLB561FUO Forward != unstranded\n",
      "ENCLB807HIR Forward != reverse\n",
      "ENCLB474FGN Forward != unstranded\n",
      "ENCLB048AFA Forward != unstranded\n",
      "ENCLB219PGK Forward != unstranded\n",
      "ENCLB349KYC Forward != reverse\n",
      "ENCLB850KUN Forward != unstranded\n",
      "ENCLB866KKG Forward != unstranded\n",
      "ENCLB112QYE Forward != unstranded\n",
      "ENCLB903KKG Forward != unstranded\n",
      "ENCLB965KXU Forward != unstranded\n",
      "ENCLB012QYK Forward != reverse\n",
      "ENCLB635BYD Forward != unstranded\n",
      "ENCLB054SXC Forward != reverse\n",
      "ENCLB823WAF Forward != reverse\n",
      "ENCLB878KRT Forward != unstranded\n",
      "ENCLB497BTJ Forward != unstranded\n",
      "ENCLB464YXA Forward != unstranded\n",
      "ENCLB475AQQ Forward != unstranded\n",
      "ENCLB713GHH Forward != reverse\n",
      "ENCLB268GZV Forward != unstranded\n",
      "ENCLB261TVA Forward != unstranded\n",
      "ENCLB414YSS Forward != unstranded\n",
      "ENCLB401UMP Forward != reverse\n",
      "ENCLB401DAZ Forward != reverse\n",
      "ENCLB851FFF Forward != reverse\n",
      "ENCLB780ICU Forward != unstranded\n",
      "ENCLB615ZSX Forward != unstranded\n",
      "ENCLB568RWC Forward != reverse\n",
      "ENCLB766ZBL Forward != reverse\n",
      "ENCLB767RCV Forward != unstranded\n",
      "ENCLB995TFV Forward != reverse\n",
      "ENCLB339NZV Forward != unstranded\n",
      "ENCLB398WHZ Forward != unstranded\n",
      "ENCLB335FQZ Forward != unstranded\n",
      "ENCLB669IFQ Forward != unstranded\n",
      "ENCLB919ROH Forward != unstranded\n",
      "ENCLB511RUE Forward != unstranded\n",
      "ENCLB608YNH Forward != reverse\n",
      "ENCLB871PHY Forward != reverse\n",
      "ENCLB794POB Forward != unstranded\n",
      "ENCLB201GIH Forward != unstranded\n",
      "ENCLB785BIJ Forward != unstranded\n",
      "ENCLB585ZUX Forward != unstranded\n",
      "ENCLB903CLN Forward != unstranded\n",
      "ENCLB847VJF Forward != reverse\n",
      "ENCLB835OER Forward != unstranded\n",
      "ENCLB791LRQ Forward != unstranded\n",
      "ENCLB779UHQ Forward != unstranded\n",
      "ENCLB549NJM Forward != unstranded\n",
      "ENCLB713ETA Forward != reverse\n",
      "ENCLB215OBG Forward != unstranded\n",
      "ENCLB371QTK Forward != unstranded\n",
      "ENCLB357LOI Forward != unstranded\n",
      "ENCLB132HUB Forward != unstranded\n",
      "ENCLB134MUP Forward != unstranded\n",
      "ENCLB879UGT Forward != unstranded\n",
      "ENCLB626ZWO Forward != unstranded\n",
      "ENCLB043WMO Forward != unstranded\n",
      "ENCLB590ZKY Forward != unstranded\n",
      "ENCLB840KBQ Forward != unstranded\n",
      "ENCLB983BCX Forward != reverse\n",
      "ENCLB432ZEQ Forward != reverse\n",
      "ENCLB687SMG Forward != unstranded\n",
      "ENCLB663TUU Forward != reverse\n",
      "ENCLB533FKA Forward != unstranded\n",
      "ENCLB525NMC Forward != unstranded\n",
      "ENCLB893EDU Forward != unstranded\n",
      "ENCLB225KMP Forward != reverse\n",
      "ENCLB994DPH Forward != unstranded\n",
      "ENCLB804GBP Forward != unstranded\n",
      "ENCLB994FPH Forward != unstranded\n",
      "ENCLB391MDC Forward != unstranded\n",
      "ENCLB080DFC Forward != unstranded\n",
      "ENCLB640NHB Forward != unstranded\n",
      "ENCLB004ADO Forward != unstranded\n",
      "ENCLB852FRE Forward != unstranded\n",
      "ENCLB578XRO Forward != unstranded\n",
      "ENCLB440WWI Forward != unstranded\n",
      "ENCLB479LPQ Forward != unstranded\n",
      "ENCLB618UUX Forward != unstranded\n",
      "ENCLB495DSX Forward != reverse\n",
      "ENCLB676AMA Forward != reverse\n",
      "ENCLB487OLR Forward != unstranded\n",
      "ENCLB704KJI Forward != unstranded\n",
      "ENCLB929UNM Forward != unstranded\n",
      "ENCLB570ZDP Forward != unstranded\n",
      "ENCLB502XLE Forward != unstranded\n",
      "ENCLB474TVI Forward != unstranded\n",
      "ENCLB706UWR Forward != unstranded\n",
      "ENCLB873RKB Forward != unstranded\n",
      "ENCLB204EKU Forward != unstranded\n",
      "ENCLB093OVC Forward != unstranded\n",
      "ENCLB676XJX Forward != unstranded\n",
      "ENCLB465HEU Forward != unstranded\n",
      "ENCLB826MTF Forward != unstranded\n",
      "ENCLB786EOH Forward != reverse\n",
      "ENCLB820YPQ Forward != unstranded\n",
      "ENCLB066LUV Forward != unstranded\n",
      "ENCLB064HGO Forward != unstranded\n",
      "ENCLB532ZTA Forward != unstranded\n",
      "ENCLB794PMK Forward != reverse\n",
      "ENCLB860NKI Forward != unstranded\n",
      "ENCLB415QIR Forward != unstranded\n",
      "ENCLB107BBQ Forward != unstranded\n",
      "ENCLB706KFB Forward != unstranded\n",
      "ENCLB940XLD Forward != unstranded\n",
      "ENCLB367HZZ Forward != unstranded\n",
      "ENCLB514JPK Forward != unstranded\n",
      "ENCLB057SXW Forward != unstranded\n",
      "ENCLB925PUE Forward != unstranded\n",
      "ENCLB571LOU Forward != unstranded\n",
      "ENCLB446NDM Forward != unstranded\n",
      "ENCLB033ZJD Forward != unstranded\n",
      "ENCLB918XPI Forward != unstranded\n",
      "ENCLB186WYP Forward != unstranded\n",
      "ENCLB497GHS Forward != unstranded\n",
      "ENCLB956KBS Forward != unstranded\n",
      "ENCLB395VZG Forward != unstranded\n",
      "ENCLB486GRF Forward != unstranded\n",
      "ENCLB352HCM Forward != unstranded\n",
      "ENCLB236QCO Forward != unstranded\n",
      "ENCLB270IKS Forward != unstranded\n",
      "ENCLB410GHL Forward != unstranded\n",
      "ENCLB104XFX Forward != reverse\n",
      "ENCLB822ZBL Forward != unstranded\n",
      "ENCLB747VGK Forward != unstranded\n",
      "ENCLB560GBC Forward != reverse\n",
      "ENCLB311ZJJ Forward != unstranded\n",
      "ENCLB655KFD Forward != unstranded\n",
      "ENCLB561OFB Forward != unstranded\n",
      "ENCLB534VTN Forward != unstranded\n",
      "ENCLB959OHU Forward != unstranded\n",
      "ENCLB526RVS Forward != reverse\n",
      "ENCLB041UXR Forward != unstranded\n",
      "ENCLB538QVH Forward != unstranded\n",
      "ENCLB324TXL Forward != unstranded\n",
      "ENCLB549UXG Forward != unstranded\n",
      "ENCLB842EQT Forward != unstranded\n",
      "ENCLB583JZE Forward != unstranded\n",
      "ENCLB836HUJ Forward != unstranded\n",
      "ENCLB428UEV Forward != unstranded\n",
      "ENCLB229VWB Forward != unstranded\n",
      "ENCLB540OBF Forward != reverse\n",
      "ENCLB566WWJ Forward != reverse\n",
      "ENCLB042NUM Forward != unstranded\n",
      "ENCLB935EGG Forward != unstranded\n",
      "ENCLB938HOJ Forward != unstranded\n",
      "matching_strand 0\n"
     ]
    }
   ],
   "source": [
    "documents_seen = Counter()\n",
    "results = []\n",
    "needed_metadata = []\n",
    "experiment_cache = {}\n",
    "matching_strand = 0\n",
    "for row in query[\"@graph\"]:\n",
    "    if row[\"status\"] in (\"in progress\", \"released\"):\n",
    "        experiment = server.get_json(row[\"@id\"])\n",
    "        documents = set(experiment.get(\"documents\", []))\n",
    "        construction_method = set()\n",
    "        for replicate in experiment['replicates']:\n",
    "            library = replicate['library']\n",
    "            library_barcode = library.get(\"barcode_details\")\n",
    "            for d in library.get(\"documents\", []):\n",
    "                documents.add(d)\n",
    "            construction_platform = library.get('construction_platform', None)\n",
    "            if construction_platform is not None:                \n",
    "                construction_method.add(construction_platform[\"term_name\"])\n",
    "            construction_platform = library.get(\"construction_method\", None)\n",
    "            if construction_platform is not None:\n",
    "                construction_method.add(construction_platform[0])\n",
    "\n",
    "            biosample = library[\"biosample\"]\n",
    "            for d in biosample.get(\"documents\", []):\n",
    "                documents.add(d)\n",
    "            subcellular_fraction = biosample.get(\"subcellular_fraction_term_name\", \"\")\n",
    "                \n",
    "        has_processed = False\n",
    "        for f in experiment[\"files\"]:\n",
    "            submitted_file_name = Path(f[\"submitted_file_name\"])            \n",
    "            if f.get(\"output_type\") == \"sparse gene count matrix of unique reads\":\n",
    "                has_processed = True\n",
    "            # split-seq doesn't have the filtered files\n",
    "            elif f.get(\"output_type\") == \"unfiltered sparse gene count matrix of unique reads\":\n",
    "                has_processed = True\n",
    "            # split-seq uses bulk files\n",
    "            elif f.get(\"output_type\") == \"gene quantification\":\n",
    "                has_processed = True\n",
    "\n",
    "        sc_type = set()\n",
    "\n",
    "        for d in documents:\n",
    "            assert d in known_document_types\n",
    "            if d in document_to_protocol:\n",
    "                sc_type.add(document_to_protocol[d])\n",
    "        if len(experiment[\"documents\"]) > 0:\n",
    "            documents_seen[\",\".join(experiment[\"documents\"])] += 1\n",
    "          \n",
    "        if len(sc_type) == 1:\n",
    "            sc_type = sc_type.pop()\n",
    "        elif len(sc_type) < 1:\n",
    "            sc_type = \"Unknown type: {}\".format(experiment[\"accession\"], documents)\n",
    "        elif len(sc_type) > 1:\n",
    "            sc_type = \"Mixed types: {}\".format(experiment[\"accession\"], sc_type)\n",
    "        \n",
    "        strand = protocol_defaults.get(sc_type, {}).get(\"stranded\", None)\n",
    "        \n",
    "        if strand is not None:\n",
    "            strand_specificity = library.get(\"strand_specificity\", \"not set\").lower()\n",
    "            if strand.lower() != strand_specificity:\n",
    "                print(\"{} {} != {}\".format(library[\"accession\"], strand, strand_specificity))\n",
    "            else:\n",
    "                matching_strand += 1\n",
    "        \n",
    "        if (sc_type.startswith(\"Chromium\") or sc_type.startswith(\"Split-seq\")):\n",
    "            could_process = True\n",
    "        else:\n",
    "            could_process = False\n",
    "        \n",
    "        experiment_cache[row[\"accession\"]] = experiment\n",
    "        results.append({\n",
    "            \"experiment\": row[\"accession\"], \n",
    "            \"library\": library[\"accession\"],\n",
    "            \"status\": row[\"status\"],\n",
    "            \"could_process\": \"Y\" if could_process else \"N\",\n",
    "            \"lab\": row[\"lab\"][\"title\"].split(',')[1].strip(), \n",
    "            \"description\": experiment.get(\"description\", \"\"),\n",
    "            \"documents\": documents,\n",
    "            \"document_len\": len(documents),\n",
    "            \"construction_method\": \",\".join(construction_method),\n",
    "            #\"library_barcode\": library_barcode,\n",
    "            #\"library_description\": library.get(\"description\", \"\"),\n",
    "            \"biosample_ontology\": biosample[\"biosample_ontology\"][\"@id\"],\n",
    "            \"biosample_term\": biosample[\"biosample_ontology\"][\"term_name\"],\n",
    "            \"sc_type\": sc_type,\n",
    "            \"organism\": biosample[\"organism\"][\"name\"],\n",
    "            \"subcellular_fraction\": subcellular_fraction,\n",
    "            \"has_processed\": \"Y\" if has_processed else \"N\",\n",
    "        })\n",
    "        \n",
    "print(\"matching_strand\", matching_strand)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "referenced-plant",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Experiment   Status       Lab        Ready Done  Organism Cellular sc_type              bioterm\n",
      "ENCSR133IBO  in progress  Stanford   Y     N     human             Chromium Multiome AT liver\n",
      "ENCSR192UEL  in progress  Stanford   Y     N     human             Chromium Multiome AT liver\n",
      "ENCSR291IHH  in progress  Stanford   Y     N     human             Chromium Multiome AT liver\n",
      "ENCSR292BJF  in progress  Stanford   Y     N     human             Chromium Multiome AT liver\n",
      "ENCSR474DKO  in progress  Stanford   Y     N     human             Chromium Multiome AT liver\n",
      "ENCSR523HMC  in progress  Stanford   Y     N     human             Chromium Multiome AT liver\n",
      "ENCSR807BYQ  in progress  Stanford   Y     N     human             Chromium Multiome AT liver\n",
      "ENCSR955TRT  in progress  Stanford   Y     N     human             Chromium Multiome AT liver\n",
      "ENCSR000AIZ  released     Caltech    N     N     human             SmartSeq Pool        GM12878\n",
      "ENCSR000AJG  released     Caltech    N     N     human             SmartSeq Pool        GM12878\n",
      "ENCSR137RHZ  released     Caltech    N     N     human             SmartSeq             pyramidal neuron\n",
      "ENCSR156HHE  released     Caltech    Y     Y     mouse             Chromium v2          forelimb\n",
      "ENCSR157NLF  released     Caltech    N     N     human             SmartSeq             Purkinje cell\n",
      "ENCSR236EGS  released     Caltech    N     N     mouse             SmartSeq             cerebral cortex, layer 5\n",
      "ENCSR268AXN  released     Caltech    N     N     mouse             SmartSeq             pyramidal neuron\n",
      "ENCSR306UHC  released     Caltech    Y     Y     mouse             Chromium v2          forelimb\n",
      "ENCSR313IUO  released     Caltech    N     N     human             SmartSeq             cerebellar granule cell\n",
      "ENCSR337QTQ  released     Caltech    Y     Y     mouse             Chromium v2          forelimb\n",
      "ENCSR415JJO  released     Caltech    N     N     human             SmartSeq             cerebellar granule cell\n",
      "ENCSR416FHN  released     Caltech    Y     Y     mouse             Chromium v2          autopod region\n",
      "ENCSR423XRG  released     Caltech    N     N     human             SmartSeq             pyramidal neuron\n",
      "ENCSR476JWN  released     Caltech    Y     Y     mouse             Chromium v2          forelimb\n",
      "ENCSR642XVO  released     Caltech    Y     Y     mouse             Chromium v2          forelimb\n",
      "ENCSR698TMC  released     Caltech    Y     Y     mouse             Chromium v2          forelimb zygopod\n",
      "ENCSR745DNK  released     Caltech    Y     Y     mouse             Chromium v2          forelimb\n",
      "ENCSR758TAD  released     Caltech    N     N     mouse             SmartSeq             Purkinje cell\n",
      "ENCSR787QXE  released     Caltech    Y     Y     mouse             Chromium v2          forelimb\n",
      "ENCSR792LNZ  released     Caltech    Y     Y     mouse             Chromium v2          forelimb\n",
      "ENCSR805QYO  released     Caltech    N     N     mouse             SmartSeq             Purkinje cell\n",
      "ENCSR814JMM  released     Caltech    N     N     mouse             SmartSeq             cerebellar granule layer\n",
      "ENCSR822EWO  released     Caltech    N     N     mouse             SmartSeq             cerebellar granule layer\n",
      "ENCSR847RRR  released     Caltech    Y     Y     mouse             Chromium v2          forelimb stylopod\n",
      "ENCSR874BOF  released     Caltech    Y     Y     mouse             Chromium v2          forelimb\n",
      "ENCSR002SMQ  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR005TCL  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT psoas muscle\n",
      "ENCSR008CVR  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR011EDH  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT fallopian tube\n",
      "ENCSR012APQ  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT heart left ventricle\n",
      "ENCSR021MHR  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT gastrocnemius\n",
      "ENCSR021QIA  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT placenta\n",
      "ENCSR021YYL  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT adrenal gland\n",
      "ENCSR028MBY  released     Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    mucosa of descending colon\n",
      "ENCSR035BDI  released     Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    adrenal gland\n",
      "ENCSR056QLB  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR060SEA  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT bile duct\n",
      "ENCSR060ZXZ  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT heart right ventricle\n",
      "ENCSR067BOK  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR075HVN  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT layer of hippocampus\n",
      "ENCSR076ZLE  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR079LVG  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT left ventricle myocardium superior\n",
      "ENCSR084XKX  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR085XEW  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR087YKG  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT right cardiac atrium\n",
      "ENCSR093GXF  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR125XUV  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT gastrocnemius\n",
      "ENCSR132MEW  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT left cerebral cortex\n",
      "ENCSR133XJX  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT left cerebral cortex\n",
      "ENCSR138JCM  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR142RRV  released     Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    upper lobe of right lung\n",
      "ENCSR151ONW  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT Right ventricle myocardium inferior\n",
      "ENCSR157FDD  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR157YJO  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT adrenal gland\n",
      "ENCSR163GJZ  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT gastrocnemius\n",
      "ENCSR175TRJ  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR176WWW  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR188GDL  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT gastrocnemius\n",
      "ENCSR190TRK  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR201WJO  released     Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    mucosa of gallbladder\n",
      "ENCSR203THX  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT psoas muscle\n",
      "ENCSR203YOV  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR204RHR  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR224OUG  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT adrenal gland\n",
      "ENCSR224RWH  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT left cerebral cortex\n",
      "ENCSR225OII  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT lung\n",
      "ENCSR231FNL  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR237HWJ  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR258SZI  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT heart right ventricle\n",
      "ENCSR259VOY  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT heart left ventricle\n",
      "ENCSR261BXB  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT pancreas\n",
      "ENCSR261LPM  released     Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    right lobe of liver\n",
      "ENCSR273JWD  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR281NBH  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT pancreas\n",
      "ENCSR281RIH  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT psoas muscle\n",
      "ENCSR289ANH  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT left colon\n",
      "ENCSR297ZTQ  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT ovary\n",
      "ENCSR298MPV  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT ovary\n",
      "ENCSR328GTN  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR333SLJ  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT heart\n",
      "ENCSR337HKS  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT natural killer cell\n",
      "ENCSR345CVL  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR349AHE  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR352DXB  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT heart left ventricle\n",
      "ENCSR356VJZ  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT adrenal gland\n",
      "ENCSR362HUP  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT ovary\n",
      "ENCSR362YDM  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT adrenal gland\n",
      "ENCSR367YUX  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT right lobe of liver\n",
      "ENCSR369UFT  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT left lobe of liver\n",
      "ENCSR398PIW  released     Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    right lobe of liver\n",
      "ENCSR398YBK  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR402PHR  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT adrenal gland\n",
      "ENCSR405YKM  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT heart left ventricle\n",
      "ENCSR415HEP  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT heart\n",
      "ENCSR417FCM  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT heart\n",
      "ENCSR421YXX  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT left cerebral cortex\n",
      "ENCSR426PDU  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT heart right ventricle\n",
      "ENCSR438FRT  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT gastrocnemius\n",
      "ENCSR439ZVQ  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT heart left ventricle\n",
      "ENCSR445LCA  released     Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    pancreas\n",
      "ENCSR455MGH  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR460MGC  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT mucosa of descending colon\n",
      "ENCSR462HCD  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT heart\n",
      "ENCSR463EEP  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT left cardiac atrium\n",
      "ENCSR467GGT  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT left cerebral cortex\n",
      "ENCSR471YTA  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT psoas muscle\n",
      "ENCSR472RRP  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT pancreas\n",
      "ENCSR478FQR  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT pancreas\n",
      "ENCSR478PUL  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT gastrocnemius\n",
      "ENCSR481QQR  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT heart left ventricle\n",
      "ENCSR484YSK  released     Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    right lobe of liver\n",
      "ENCSR485GOL  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR488UUT  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR489URW  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR522HNR  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT right cardiac atrium\n",
      "ENCSR527MGL  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT left cardiac atrium\n",
      "ENCSR533IGQ  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT heart\n",
      "ENCSR539XHZ  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT layer of hippocampus\n",
      "ENCSR540DHJ  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR548HED  released     Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    kidney\n",
      "ENCSR568GGW  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT heart\n",
      "ENCSR587SPO  released     Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    left colon\n",
      "ENCSR602SGU  released     Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    right lobe of liver\n",
      "ENCSR620AJZ  released     Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    adrenal gland\n",
      "ENCSR624WDN  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT left cerebral cortex\n",
      "ENCSR630LZS  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR632EIK  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT left cerebral cortex\n",
      "ENCSR641MWO  released     Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    psoas muscle\n",
      "ENCSR642GFM  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT fallopian tube\n",
      "ENCSR644VVE  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT heart\n",
      "ENCSR654MFX  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR656VWP  released     Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    colonic mucosa\n",
      "ENCSR662VTU  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT upper lobe of left lung\n",
      "ENCSR665HPR  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT heart\n",
      "ENCSR671FDY  released     Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    lower lobe of right lung\n",
      "ENCSR681GPY  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR683ONV  released     Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    lower lobe of left lung\n",
      "ENCSR684KYI  released     Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    pancreas\n",
      "ENCSR700UVN  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT left cardiac atrium\n",
      "ENCSR707ECV  released     Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    colonic mucosa\n",
      "ENCSR707LLI  released     Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    mucosa of descending colon\n",
      "ENCSR716VLR  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT posterior vena cava\n",
      "ENCSR720TUP  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT left cerebral cortex\n",
      "ENCSR724KET  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT adrenal gland\n",
      "ENCSR726IPC  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT adrenal gland\n",
      "ENCSR727OYO  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT heart left ventricle\n",
      "ENCSR742NWO  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT layer of hippocampus\n",
      "ENCSR749GDE  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT adrenal gland\n",
      "ENCSR751BHQ  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT heart left ventricle\n",
      "ENCSR753YOZ  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR753YPW  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT lower lobe of left lung\n",
      "ENCSR755CIF  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT heart left ventricle\n",
      "ENCSR762LML  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR762LZV  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT uterus\n",
      "ENCSR763BII  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR769KWM  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT bile duct\n",
      "ENCSR771MOE  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT gastrocnemius\n",
      "ENCSR777RUZ  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR788SNY  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR789PLK  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT placenta\n",
      "ENCSR801DHT  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT heart left ventricle\n",
      "ENCSR813PWQ  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT Right ventricle myocardium superior\n",
      "ENCSR814LMX  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR835JFN  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT layer of hippocampus\n",
      "ENCSR835ZXV  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT gastrocnemius\n",
      "ENCSR851JBE  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR876EUH  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT ovary\n",
      "ENCSR899GYX  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR906MRL  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR908CQZ  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT adrenal gland\n",
      "ENCSR919ENI  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR938CZJ  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT ovary\n",
      "ENCSR943HBF  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT right cardiac atrium\n",
      "ENCSR944XPC  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT layer of hippocampus\n",
      "ENCSR962JKS  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR964JZN  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT ureter\n",
      "ENCSR966DDY  released     Stanford   Y     Y     human    nucleus  Chromium Multiome AT left lung\n",
      "ENCSR968RPR  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT layer of hippocampus\n",
      "ENCSR980OCK  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR987SOK  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT layer of hippocampus\n",
      "ENCSR988AMY  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT adrenal gland\n",
      "ENCSR991LHO  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR994VEG  released     Stanford   Y     Y     human             Chromium Multiome AT heart left ventricle\n",
      "ENCSR995ZKH  released     Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT layer of hippocampus\n",
      "ENCSR019BZL  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR026VDA  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR035ZCY  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR036OQE  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR037LFG  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR041KFB  released     UCI        Y     Y     mouse    nucleus  Split-seq            heart\n",
      "ENCSR047KXM  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR049KDI  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR079CKL  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR079GGS  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR081SXM  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR083EPD  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR084RQP  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR092RYV  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR093CLX  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR094YHS  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR097PDV  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR108ZIJ  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR109DAN  released     UCI        Y     Y     mouse    nucleus  Split-seq            myotube\n",
      "ENCSR111SCI  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR118FCJ  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR122DEZ  released     UCI        Y     Y     mouse    nucleus  Split-seq            heart\n",
      "ENCSR125HRM  released     UCI        Y     Y     mouse    nucleus  Split-seq            heart\n",
      "ENCSR127LUZ  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR132KMR  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR134FSZ  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR136FOI  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR137UJM  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR138QIF  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR141EJG  released     UCI        Y     Y     mouse    nucleus  Split-seq            heart\n",
      "ENCSR145HUM  released     UCI        Y     Y     mouse    nucleus  Split-seq            heart\n",
      "ENCSR147TEP  released     UCI        Y     Y     mouse    nucleus  Split-seq            heart\n",
      "ENCSR148VQR  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR150KKS  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR160AKB  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR160FFZ  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR160VMU  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR161BNJ  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR165FLF  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR165VMQ  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR167HUO  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR174ZVM  released     UCI        Y     Y     mouse    nucleus  Split-seq            heart\n",
      "ENCSR178PKG  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR182SMX  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR187HKH  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR196TWX  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR201SNQ  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR203XJO  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR216UTT  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR220GBV  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR231VJO  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR232VMC  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR237OCP  released     UCI        Y     Y     mouse    nucleus  Split-seq            heart\n",
      "ENCSR247PQY  released     UCI        Y     Y     mouse    nucleus  Split-seq            heart\n",
      "ENCSR247QPJ  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR248MFG  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR250SCW  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR253YLA  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR254SLA  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR257BOU  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR257VQO  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR263VNQ  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR266RTS  released     UCI        Y     Y     mouse    nucleus  Split-seq            heart\n",
      "ENCSR277UKO  released     UCI        Y     Y     mouse    nucleus  Split-seq            heart\n",
      "ENCSR279GBC  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR291RIX  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR302VFA  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR303EIR  released     UCI        Y     Y     mouse    nucleus  Split-seq            heart\n",
      "ENCSR305SPL  released     UCI        Y     Y     mouse    nucleus  Split-seq            heart\n",
      "ENCSR305VTL  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR308DHB  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR312AOC  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR320EFC  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR328JOY  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR329JYG  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR337DKH  released     UCI        Y     Y     mouse    nucleus  Split-seq            heart\n",
      "ENCSR355CTS  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR358NKP  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR358SBF  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR386PPF  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR388EIN  released     UCI        Y     Y     mouse             Split-seq            C2C12\n",
      "ENCSR396XOO  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR399EUZ  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR399TQO  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR402GQZ  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR403MNL  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR409OGG  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR419OCT  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR422JWZ  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR425UJQ  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR435CMT  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR440PMT  released     UCI        Y     Y     mouse    nucleus  Split-seq            heart\n",
      "ENCSR442EUE  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR442NQZ  released     UCI        Y     Y     mouse    nucleus  Split-seq            heart\n",
      "ENCSR450IJU  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR452BKB  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR452OYH  released     UCI        Y     Y     mouse    nucleus  Split-seq            C2C12\n",
      "ENCSR454ZHR  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR455YDL  released     UCI        Y     Y     mouse    nucleus  Split-seq            heart\n",
      "ENCSR459ESG  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR462NDO  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR467JRI  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR472OOH  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR480RWU  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR483SSM  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR484MVN  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR488XVC  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR490RCS  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR496OYJ  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR498SXJ  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR499ZYZ  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR503QZJ  released     UCI        Y     Y     mouse             Split-seq            C2C12\n",
      "ENCSR514XCG  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR526LFP  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR527LIA  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR531SRE  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR544IJY  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR544MGC  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR545URU  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR548JRH  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR551OZP  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR552XCF  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR554FGW  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR559AAG  released     UCI        Y     Y     mouse    nucleus  Split-seq            heart\n",
      "ENCSR567YKE  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR569LJC  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR571QJI  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR572SMD  released     UCI        Y     Y     mouse    nucleus  Split-seq            myotube\n",
      "ENCSR573UKV  released     UCI        Y     Y     mouse    nucleus  Split-seq            heart\n",
      "ENCSR575TNG  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR576UHY  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR577OFZ  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR580UNE  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR581NLU  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR581ZDJ  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR591UDP  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR594LDC  released     UCI        Y     Y     mouse    nucleus  Split-seq            heart\n",
      "ENCSR611QJH  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR613VKV  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR617CNE  released     UCI        Y     Y     mouse    nucleus  Split-seq            heart\n",
      "ENCSR627PLH  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR627TLB  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR627XNK  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR632SJM  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR642QUS  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR644SME  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR646KEB  released     UCI        Y     Y     mouse    nucleus  Split-seq            heart\n",
      "ENCSR650QSN  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR650VIQ  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR653HPZ  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR659DFF  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR662TPD  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR666LZG  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR667UMR  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR688APT  released     UCI        Y     Y     mouse    nucleus  Split-seq            heart\n",
      "ENCSR692LZA  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR693KAV  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR694GJL  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR696NAL  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR697WWY  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR711CNO  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR711LRI  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR725IGH  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR728NFJ  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR729BRE  released     UCI        Y     Y     mouse    nucleus  Split-seq            heart\n",
      "ENCSR730JFZ  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR732MDJ  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR736JKX  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR743OKJ  released     UCI        Y     Y     mouse    nucleus  Split-seq            heart\n",
      "ENCSR763LIT  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR764CYB  released     UCI        Y     Y     mouse    nucleus  Split-seq            heart\n",
      "ENCSR767FKE  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR773CKJ  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR776XMP  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR790DPF  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR793HMA  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR794RPU  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR800QGK  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR809WJL  released     UCI        Y     Y     mouse    nucleus  Split-seq            heart\n",
      "ENCSR816PAX  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR819UHV  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR823CLM  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR823HRL  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR824CFI  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR826EHD  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR827AOS  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR839BHY  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR846QIZ  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR855URG  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR859XOI  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR860LQG  released     UCI        Y     Y     mouse    nucleus  Split-seq            heart\n",
      "ENCSR868XSG  released     UCI        Y     Y     mouse    nucleus  Split-seq            heart\n",
      "ENCSR869HKZ  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR877NAX  released     UCI        Y     Y     mouse    nucleus  Split-seq            C2C12\n",
      "ENCSR883XZK  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR888TQC  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR899EKD  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR906KCJ  released     UCI        Y     Y     mouse    nucleus  Split-seq            heart\n",
      "ENCSR908PNN  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR918LWU  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR921SBQ  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR926QCE  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR927BSJ  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR928UPG  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR930FUM  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR938YKJ  released     UCI        Y     Y     mouse    nucleus  Split-seq            heart\n",
      "ENCSR940VCU  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR943REW  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR944KUB  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR950SIX  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR961HBK  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR964OQW  released     UCI        Y     Y     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR966QJD  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR967YET  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR973WCI  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR974ODT  released     UCI        Y     Y     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR975YAQ  released     UCI        Y     Y     mouse    nucleus  Split-seq            heart\n",
      "ENCSR979MRY  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR982GYD  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR983TAK  released     UCI        Y     Y     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR998YMP  released     UCI        Y     Y     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR562QQA  released     UMass      N     N     mouse             inDrop               dendritic cell\n",
      "ENCSR568FZM  released     UMass      N     N     mouse             inDrop               dendritic cell\n",
      "ENCSR597YMV  released     UMass      N     N     mouse             inDrop               dendritic cell\n"
     ]
    }
   ],
   "source": [
    "#report_format = \"{accession:12} {status:12} {lab:10} {document_len:3} {has_processed:4} {construction_method:35.35} {sc_type:20}\"\n",
    "#  {document_len:3}\n",
    "report_format = \"{experiment:12} {status:12} {lab:10} {could_process:5} {has_processed:5} {organism:8.8} {subcellular_fraction:8.8} {sc_type:20.20} {biosample_term}\"\n",
    "print(report_format.format(**{\n",
    "    \"experiment\": \"Experiment\",\n",
    "    \"description\": \"Description\",\n",
    "    \"status\": \"Status\", \n",
    "    \"lab\": \"Lab\", \n",
    "    \"document_len\": \"Doc\", \n",
    "    #\"construction\",\n",
    "    \"has_processed\": \"Done\",\n",
    "    \"could_process\": \"Ready\",\n",
    "    \"organism\": \"Organism\",\n",
    "    #\"construction_method\": \"Const. Method\",\n",
    "    \"subcellular_fraction\": \"Cellular\",\n",
    "    #\"biosample_ontology\": \"biosample ontology\",\n",
    "    \"biosample_term\": \"bioterm\",\n",
    "    \"sc_type\": \"sc_type\",\n",
    "}))\n",
    "for row in sorted(results, key=lambda x: (x[\"status\"], x[\"lab\"], x[\"experiment\"])):\n",
    "    #print(report_format.format(row[0], row[1], row[2], len(row[3]), \",\".join(row[4])[:20], \",\".join(row[5])))\n",
    "    print(report_format.format(**row))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "corporate-sussex",
   "metadata": {},
   "outputs": [],
   "source": [
    "pandas.DataFrame(results).sort_values([\"status\", \"lab\", \"experiment\"]).to_excel(\"scrna-experiments.xlsx\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "twelve-league",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unprocessed 0 of 416\n"
     ]
    }
   ],
   "source": [
    "print(\"Unprocessed {} of {}\".format(sum([not x[\"has_processed\"] for x in results]), len(results)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "attached-state",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unprocessed 0 of 416\n"
     ]
    }
   ],
   "source": [
    "print(\"Unprocessed {} of {}\".format(sum([(not x[\"has_processed\"] and x[\"sc_type\"].startswith(\"Chromium\")) for x in results]), len(results)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "oriental-harris",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Could process 167 of 416\n"
     ]
    }
   ],
   "source": [
    "print(\"Could process {} of {}\".format(sum([(x[\"could_process\"] and x[\"sc_type\"].startswith(\"Chromium\") and x[\"lab\"].startswith(\"Stanford\"))  for x in results]), len(results)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "colored-thesaurus",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unprocessed 0 of 416\n"
     ]
    }
   ],
   "source": [
    "print(\"Unprocessed {} of {}\".format(sum([(not x[\"has_processed\"] and x[\"sc_type\"].startswith(\"Split-seq\")) for x in results]), len(results)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "numerous-integral",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unprocessed 12 of 416\n"
     ]
    }
   ],
   "source": [
    "print(\"Unprocessed {} of {}\".format(sum([(x[\"sc_type\"].startswith(\"Chromium v2\") and x[\"lab\"] == \"Caltech\") for x in results]), len(results)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "confident-greeting",
   "metadata": {},
   "outputs": [],
   "source": [
    "mouse_limb_experiments = [x[\"experiment\"] for x in results if (x[\"sc_type\"].startswith(\"Chromium v2\") and x[\"lab\"] == \"Caltech\" and x[\"has_processed\"] == \"N\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "handled-garden",
   "metadata": {},
   "outputs": [],
   "source": [
    "for exp in mouse_limb_experiments:\n",
    "    print(\"https://www.encodeproject.org/experiments/{}/\".format(exp))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "conservative-prime",
   "metadata": {},
   "outputs": [],
   "source": [
    "# First I figured out what documents were present, then I made my mapping then \n",
    "# I produced the above report, but I lost generating what documents existed.\n",
    "if 0:\n",
    "    documents_seen = Counter()\n",
    "    results = []\n",
    "    needed_metadata = []\n",
    "    for row in query[\"@graph\"]:\n",
    "        if row[\"status\"] in (\"in progress\", \"released\"):\n",
    "\n",
    "            experiment = server.get_json(row[\"@id\"])\n",
    "            documents = set(experiment.get(\"documents\", []))\n",
    "            construction_method = set()\n",
    "            for replicate in experiment['replicates']:\n",
    "                library = replicate['library']\n",
    "                library_barcode = library.get(\"barcode_details\")\n",
    "                for d in library.get(\"documents\", []):\n",
    "                    documents.add(d)\n",
    "                construction_platform = library.get('construction_platform', None)\n",
    "                if construction_platform is not None:                \n",
    "                    construction_method.add(construction_platform[\"term_name\"])\n",
    "                construction_platform = library.get(\"construction_method\", None)\n",
    "                if construction_platform is not None:\n",
    "                    construction_method.add(construction_platform[0])\n",
    "\n",
    "                biosample = library[\"biosample\"]\n",
    "                for d in biosample.get(\"documents\", []):\n",
    "                    documents.add(d)\n",
    "                subcellular_fraction = biosample.get(\"subcellular_fraction_term_name\")\n",
    "\n",
    "            has_processed = False\n",
    "            for f in experiment[\"files\"]:\n",
    "                if f[\"file_type\"] == \"fastq\":\n",
    "                    submitted_file_name = Path(f[\"submitted_file_name\"])\n",
    "                    print(experiment[\"accession\"], f[\"accession\"], f.get(\"paired_end\"), f.get(\"paired_with\"), submitted_file_name.name)\n",
    "                    #for detail in f.get(\"flowcell_details\", []):                    \n",
    "                        #for barcode_attribute in [\"barcode_in_read\", \"barcode\", \"barcode_position\"]:\n",
    "                        #    value = detail.get(barcode_attribute)\n",
    "                        #    if value is not None:\n",
    "                        #       print(row['accession'], barcode_attribute, value)\n",
    "                else:\n",
    "                    has_processed = True\n",
    "\n",
    "            sc_type = set()\n",
    "            for d in documents:\n",
    "                assert d in known_document_types\n",
    "                if d in document_to_protocol:\n",
    "                    sc_type.add(document_to_protocol[d])\n",
    "            documents_seen[\",\".join(experiment[\"documents\"])] += 1\n",
    "\n",
    "            results.append({\n",
    "                \"accession\": row[\"accession\"], \n",
    "                \"status\": row[\"status\"], \n",
    "                \"lab\": row[\"lab\"][\"title\"].split(',')[1], \n",
    "                \"description\": experiment.get(\"description\", \"\"),\n",
    "                \"documents\": documents,\n",
    "                \"document_len\": len(documents),\n",
    "                \"construction_method\": \",\".join(construction_method),\n",
    "                \"library_barcode\": library_barcode, \n",
    "                \"sc_type\": \",\".join(sc_type),\n",
    "                \"subcellular_fraction\": subcellular_fraction,\n",
    "                \"has_processed\": has_processed,\n",
    "            })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "breathing-numbers",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'www.encodeproject.org'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "server.server"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "alike-horror",
   "metadata": {},
   "outputs": [],
   "source": [
    "Filereps = namedtuple(\"Filereps\", [\"rep\", \"date_created\", \"paired\", \"accession\", \"paired_with\", \"submitted_name\"])\n",
    "illumina_parts = re.compile(\"_L(?P<lane>[\\d]{3})_R(?P<read>[\\d])_(?P<chunk>[\\d]{3}).fastq.gz\")\n",
    "\n",
    "def create_experiment_level_config(sc_type, experiment):\n",
    "    accession = experiment[\"accession\"]\n",
    "    description = experiment.get(\"description\", \"\")\n",
    "    default_config = {\n",
    "        \"experiment_accession\": accession,\n",
    "        \"description\": description,\n",
    "        \"lab\": \"/labs/barbara-wold/\",\n",
    "        \"award\": \"UM1HG009443\",\n",
    "        #\"encode_portal_host\": \"encd-6248-78c147870-qing.demo.encodedcc.org\",\n",
    "        #\"encode_portal_host\": \"encd-6248-24bd77146-qing.demo.encodedcc.org\",\n",
    "        \"encode_portal_host\": server.server,\n",
    "        #\"star_container\": \"https://woldlab.caltech.edu/~diane/containers/star-bash-dev_EoI-head.sif\",\n",
    "        \"star_container\": \"https://woldlab.caltech.edu/~diane/containers/star-bash-2.7.10a.sif\",\n",
    "    \n",
    "        # these should really scale some\n",
    "        \"mem_mb\": 65536,\n",
    "        \"disk_mb\": 51200,\n",
    "\n",
    "        \"automatic_submission\": True\n",
    "    }\n",
    "    if sc_type not in protocol_defaults:\n",
    "        print(\"Can't process {} {}: {}\".format(accession, description, sc_type))\n",
    "        return\n",
    "    default_config.update(protocol_defaults[sc_type])\n",
    "    \n",
    "    fractions = set()\n",
    "    config = {}\n",
    "    for rep in experiment[\"replicates\"]:\n",
    "        library_id = rep[\"library\"][\"@id\"]\n",
    "        library_accession = rep[\"library\"][\"accession\"]\n",
    "        config.setdefault(library_id, {}).update(default_config)\n",
    "        config[library_id][\"library_accession\"] = library_accession\n",
    "        biosample = rep[\"library\"][\"biosample\"]\n",
    "        organism_id = biosample['organism']['@id']\n",
    "        if organism_id in default_genome_index:\n",
    "            config[library_id].update(default_genome_index[organism_id])\n",
    "        else:\n",
    "            print(\"Unable to process {} {} until {} index\".format(accession, description, organism_id))\n",
    "        subcellular_fraction = biosample.get(\"subcellular_fraction_term_name\", \"\")\n",
    "        fractions.add(subcellular_fraction)\n",
    "        if subcellular_fraction == \"nucleus\":\n",
    "            config[library_id][\"include_intron\"] = True\n",
    "        else:\n",
    "            config[library_id][\"include_intron\"] = False\n",
    "\n",
    "    assert(len(fractions)) == 1\n",
    "    return config\n",
    "\n",
    "\n",
    "def add_missing_paired_with_tag(experiment):\n",
    "    # the split-seq experiments don't used-paired with, but there's only 2 fastqs so maybe fine?\n",
    "    \n",
    "    overrides = {\n",
    "        # ENCSR552XCF\n",
    "        #\"/files/ENCFF504RHF/\": \"/files/ENCFF273WGE/\",\n",
    "        #\"/files/ENCFF261RDF/\": \"/files/ENCFF238TYU/\",\n",
    "        # ENCSR399EUZ\n",
    "        \"/files/ENCFF015GGR/\": \"/files/ENCFF587CBE/\",\n",
    "        \"/files/ENCFF491GMR/\": \"/files/ENCFF333SWR/\",\n",
    "        # ENCSR019BZL\n",
    "        \"/files/ENCFF287LTE/\": \"/files/ENCFF925EHZ/\",\n",
    "        \"/files/ENCFF946RCV/\": \"/files/ENCFF554GZW/\",\n",
    "        # ENCSR026VDA\n",
    "        \"/files/ENCFF446ACV/\": \"/files/ENCFF470AMS/\",\n",
    "        \"/files/ENCFF306NMW/\": \"/files/ENCFF199VWC/\",\n",
    "        # ENCSR035ZCY\n",
    "        \"/files/ENCFF826DZV/\": \"/files/ENCFF792QPO/\",\n",
    "        \"/files/ENCFF125AIB/\": \"/files/ENCFF222BPO/\",\n",
    "        # ENCSR036OQE\n",
    "        \"/files/ENCFF567SUU/\": \"/files/ENCFF367TWB/\",\n",
    "        \"/files/ENCFF907MHB/\": \"/files/ENCFF532DDQ/\",\n",
    "        # ENCSR037LFG\n",
    "        # \"/files//\": \"/files//\",\n",
    "        # \"/files//\": \"/files//\",\n",
    "        # \"/files//\": \"/files//\",\n",
    "        # \"/files//\": \"/files//\",\n",
    "        # \"/files//\": \"/files//\",\n",
    "        # \"/files//\": \"/files//\",\n",
    "        # \"/files//\": \"/files//\",\n",
    "        # \"/files//\": \"/files//\",\n",
    "        # \"/files//\": \"/files//\",\n",
    "        # \"/files//\": \"/files//\",\n",
    "    }\n",
    "    # add reverse\n",
    "    overrides.update({overrides[k]: k for k in overrides})\n",
    "    \n",
    "    fastqs = []\n",
    "    has_paired_with = True\n",
    "    for f in experiment[\"files\"]:\n",
    "        if f[\"file_format\"] == \"fastq\":\n",
    "            if \"paired_with\" not in f:\n",
    "                if f[\"@id\"] in overrides:\n",
    "                    f[\"paired_with\"] = overrides[f[\"@id\"]]\n",
    "                else:\n",
    "                    has_paired_with = False\n",
    "            fastqs.append(f)\n",
    "    \n",
    "    if not has_paired_with:\n",
    "        # First group fastqs by rep.\n",
    "        reps = {}\n",
    "        for f in fastqs:\n",
    "            reps.setdefault(f[\"replicate\"][\"uuid\"], []).append(f)\n",
    "            \n",
    "        for rep in reps:\n",
    "            fastqs = reps[rep]\n",
    "            if len(fastqs) == 2:\n",
    "                # we can fix it up\n",
    "                fastqs[0][\"paired_with\"] = fastqs[1][\"@id\"]\n",
    "                fastqs[1][\"paired_with\"] = fastqs[0][\"@id\"]\n",
    "            else:\n",
    "                print(\"Unable to fixup paired_with tag on {} has {} fastqs\".format(experiment[\"accession\"], len(fastqs)))\n",
    "\n",
    "def generate_config(sc_type, experiment, verbose=False):\n",
    "    if sc_type in (\"SmartSeq\", \"SmartSeq Pool\", \"inDrop\"):\n",
    "        return\n",
    "\n",
    "    paired_indexing = {\n",
    "        \"1\": \"indexed reads\",\n",
    "        \"2\": \"reads\",\n",
    "    }\n",
    "    \n",
    "    accession = experiment[\"accession\"]    \n",
    "    config = create_experiment_level_config(sc_type, experiment)\n",
    "    \n",
    "    add_missing_paired_with_tag(experiment)\n",
    "        \n",
    "    files = []\n",
    "    paired_with = {}\n",
    "    read1 = {}\n",
    "    read2 = {}\n",
    "    for f in experiment[\"files\"]:\n",
    "        if f['file_format'] == 'fastq':\n",
    "            # f[\"submitted_file_name\"],\n",
    "            paired_accession = f[\"paired_with\"][len(\"/files/\"):-1]\n",
    "            paired_with[f[\"accession\"]] = paired_accession\n",
    "            \n",
    "            submitted_file_name = Path(f[\"submitted_file_name\"])\n",
    "            lanes = []\n",
    "            for detail in f.get(\"flowcell_details\", []):\n",
    "                if \"lane\" in detail and len(detail[\"lane\"]) > 0:\n",
    "                    lanes.append(detail[\"lane\"])\n",
    "            if len(lanes) == 0:\n",
    "                match = illumina_parts.search(submitted_file_name.name)\n",
    "                if match:\n",
    "                    lanes.append(str(int(match.group(\"lane\"))))\n",
    "         \n",
    "            file_library_id = f['replicate'][\"library\"]\n",
    "            if \"paired_end\" in f:\n",
    "                read = paired_indexing[f[\"paired_end\"]]\n",
    "            elif f.get(\"output_type\") in ('index reads', 'reads'):\n",
    "                read = f[\"output_type\"]\n",
    "                if len(lanes) == 0:\n",
    "                    lanes = [\"1\"]\n",
    "            else:\n",
    "                raise ValueError(\"Unable to figure out which is the index read\")\n",
    "\n",
    "            #lanes = \",\".join(lanes)\n",
    "            #if len(lanes) == 0:\n",
    "            #    print(\"In {} Unable to determine lane with {}\".format(\n",
    "            #        accession, f[\"accession\"]))\n",
    "                \n",
    "            if verbose:\n",
    "                print(f[\"accession\"], Path(submitted_file_name).name, f[\"date_created\"])\n",
    "                \n",
    "            fr = Filereps(file_library_id, f[\"date_created\"], read, f['accession'], paired_with[f[\"accession\"]], Path(submitted_file_name).name)\n",
    "            #print(fr)\n",
    "            files.append(fr)\n",
    "\n",
    "    files = sorted(files)\n",
    "    if verbose:\n",
    "        for f in files:\n",
    "            print(f)\n",
    "\n",
    "    paired_files = {}\n",
    "    for f in files:\n",
    "        paired_files[f.paired_with] = f\n",
    "        \n",
    "\n",
    "    for fr in files:\n",
    "        if verbose:\n",
    "            print(\"paired\", fr.paired, fr.accession)\n",
    "        if fr.paired in (\"indexed reads\", \"index reads\"):\n",
    "            config[fr.rep].setdefault(\"read1\", []).append(fr.accession)\n",
    "            config[fr.rep].setdefault(\"read2\", []).append(paired_files[fr.accession].accession)\n",
    "            \n",
    "\n",
    "    for library_id in config:\n",
    "        yield config[library_id]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "streaming-buffalo",
   "metadata": {},
   "outputs": [],
   "source": [
    "config = list(generate_config(\"Split-seq\", experiment_cache[\"ENCSR552XCF\"], verbose=False))\n",
    "#print(config)\n",
    "assert config[0][\"read1\"] == [\"ENCFF273WGE\"]\n",
    "assert config[0][\"read2\"] == [\"ENCFF504RHF\"]\n",
    "assert config[0][\"inclusion_accession\"] == \"encode:parse-biosciences-v1\"\n",
    "assert config[1][\"read1\"] == [\"ENCFF238TYU\"]\n",
    "assert config[1][\"read2\"] == [\"ENCFF261RDF\"]\n",
    "assert config[1][\"inclusion_accession\"] == \"encode:parse-biosciences-v1\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "ideal-approach",
   "metadata": {},
   "outputs": [],
   "source": [
    "# new one\n",
    "config = list(generate_config(\"Split-seq\", experiment_cache[\"ENCSR503QZJ\"], verbose=False))\n",
    "assert config[0][\"read1\"] == [\"ENCFF910URO\"]\n",
    "assert config[0][\"read2\"] == [\"ENCFF955PGO\"]\n",
    "assert config[0][\"inclusion_accession\"] == \"encode:parse-biosciences-v1\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "interracial-building",
   "metadata": {},
   "outputs": [],
   "source": [
    "config = list(generate_config(\"Chromium Multiome ATAC + Gene Expression\", experiment_cache[\"ENCSR203YOV\"], verbose=False))\n",
    "\n",
    "#print(config[0])\n",
    "assert config[0][\"read1\"] == [\"ENCFF053PFO\", \"ENCFF740BAJ\", \"ENCFF865XHP\", \"ENCFF694BUP\"]\n",
    "assert config[0][\"read2\"] == [\"ENCFF325QXA\", \"ENCFF681BML\", \"ENCFF974AZJ\", \"ENCFF040AHY\"]\n",
    "assert config[0][\"inclusion_accession\"] == '737K-arc-v1(GEX)'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "retired-steam",
   "metadata": {},
   "outputs": [],
   "source": [
    "config = list(generate_config(\"Chromium v2\", experiment_cache[\"ENCSR787QXE\"], verbose=False))\n",
    "#print(config[0])\n",
    "\n",
    "assert config[0][\"read1\"] == [\"ENCFF817GKI\", \"ENCFF947WRU\", \"ENCFF321GGP\", \"ENCFF951GHL\"]\n",
    "assert config[0][\"read2\"] == [\"ENCFF820VIA\", \"ENCFF093KXG\", \"ENCFF665DKW\", \"ENCFF665SYG\"]\n",
    "assert config[0][\"inclusion_accession\"] == '737K-august-2016'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "proprietary-validity",
   "metadata": {},
   "outputs": [],
   "source": [
    "config = list(generate_config(\"Chromium Multiome ATAC + Gene Expression\", experiment_cache[\"ENCSR005TCL\"]))\n",
    "#print(config[0])\n",
    "assert len(config) == 1\n",
    "assert config[0][\"read1\"] == [\"ENCFF121UIQ\", \"ENCFF663ZRK\"]\n",
    "assert config[0][\"read2\"] == [\"ENCFF194KTM\", \"ENCFF364QFJ\"]\n",
    "assert config[0][\"inclusion_accession\"] == \"737K-arc-v1(GEX)\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "empty-major",
   "metadata": {},
   "outputs": [],
   "source": [
    "config = list(generate_config(\"Chromium Multiome ATAC + Gene Expression\", experiment_cache[\"ENCSR337HKS\"]))\n",
    "#print(config[0])\n",
    "assert len(config) == 1\n",
    "assert config[0]['read1'] == [\"ENCFF298GMK\", \"ENCFF931UDZ\"]\n",
    "assert config[0][\"read2\"] == [\"ENCFF878VQP\", \"ENCFF323AMX\"]\n",
    "assert config[0][\"inclusion_accession\"] == \"737K-arc-v1(GEX)\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "difficult-optimization",
   "metadata": {},
   "outputs": [],
   "source": [
    "#list(generate_config(\"Split-seq\", experiment_cache['ENCSR003FZY']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "forty-hands",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_configs = {}\n",
    "for row in sorted(results, key=lambda x: (x[\"status\"], x[\"lab\"], x[\"experiment\"])):\n",
    "    for config in generate_config(row[\"sc_type\"], experiment_cache[row['experiment']]):\n",
    "        library = config['library_accession']\n",
    "        if config is not None:\n",
    "            all_configs[library] = config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "virtual-basket",
   "metadata": {},
   "outputs": [],
   "source": [
    "import yaml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "alternative-harrison",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "automatic_submission: true\n",
      "award: UM1HG009443\n",
      "description: snRNA on human adrenal gland. Please note that this experiment is part\n",
      "  of 10x multiome and has a corresponding scATAC experiment (ENCSR693GAD)\n",
      "disk_mb: 51200\n",
      "encode_portal_host: www.encodeproject.org\n",
      "experiment_accession: ENCSR724KET\n",
      "genome_accession: encode:starsolo-GRCh38-V29-male-index\n",
      "include_intron: true\n",
      "inclusion_accession: 737K-arc-v1(GEX)\n",
      "lab: /labs/barbara-wold/\n",
      "library_accession: ENCLB002DZK\n",
      "mem_mb: 65536\n",
      "read1:\n",
      "- ENCFF150FBF\n",
      "- ENCFF385IAW\n",
      "read2:\n",
      "- ENCFF351VBS\n",
      "- ENCFF503CCI\n",
      "star_container: https://woldlab.caltech.edu/~diane/containers/star-bash-2.7.10a.sif\n",
      "stranded: Forward\n",
      "umi_length: 12\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(yaml.dump(all_configs[\"ENCLB002DZK\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "cellular-windsor",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "automatic_submission: true\n",
      "award: UM1HG009443\n",
      "description: snRNA on human psoas muscle. Please note that this experiment is part\n",
      "  of 10x multiome and has a correspoding scATAC experiment (ENCSR332XEW).\n",
      "disk_mb: 51200\n",
      "encode_portal_host: www.encodeproject.org\n",
      "experiment_accession: ENCSR005TCL\n",
      "genome_accession: encode:starsolo-GRCh38-V29-male-index\n",
      "include_intron: true\n",
      "inclusion_accession: 737K-arc-v1(GEX)\n",
      "lab: /labs/barbara-wold/\n",
      "library_accession: ENCLB549KAN\n",
      "mem_mb: 65536\n",
      "read1:\n",
      "- ENCFF121UIQ\n",
      "- ENCFF663ZRK\n",
      "read2:\n",
      "- ENCFF194KTM\n",
      "- ENCFF364QFJ\n",
      "star_container: https://woldlab.caltech.edu/~diane/containers/star-bash-2.7.10a.sif\n",
      "stranded: Forward\n",
      "umi_length: 12\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(yaml.dump(all_configs[\"ENCLB549KAN\"]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "proof-arbor",
   "metadata": {},
   "source": [
    "Ali said there's one multiome sample he'd like it processed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "occupied-athens",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "automatic_submission: true\n",
      "award: UM1HG009443\n",
      "description: snRNA on human right ventricle myocardium inferior. Please note that\n",
      "  this experiment is part of 10x multiome and has a corresponding scATAC experiment\n",
      "  (ENCSR628POB)\n",
      "disk_mb: 51200\n",
      "encode_portal_host: www.encodeproject.org\n",
      "experiment_accession: ENCSR151ONW\n",
      "genome_accession: encode:starsolo-GRCh38-V29-male-index\n",
      "include_intron: true\n",
      "inclusion_accession: 737K-arc-v1(GEX)\n",
      "lab: /labs/barbara-wold/\n",
      "library_accession: ENCLB297FDO\n",
      "mem_mb: 65536\n",
      "read1:\n",
      "- ENCFF424PGM\n",
      "- ENCFF989IIG\n",
      "read2:\n",
      "- ENCFF906BLI\n",
      "- ENCFF859KJN\n",
      "star_container: https://woldlab.caltech.edu/~diane/containers/star-bash-2.7.10a.sif\n",
      "stranded: Forward\n",
      "umi_length: 12\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(yaml.dump(all_configs[\"ENCLB297FDO\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "affiliated-indianapolis",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "automatic_submission: true\n",
      "award: UM1HG009443\n",
      "description: 10X Genomics Chromium 3' single-cell RNA-Seq (v2 kit) of C57Bl6 wild-type\n",
      "  15.0 day embryonic mouse proximal forelimb\n",
      "disk_mb: 51200\n",
      "encode_portal_host: www.encodeproject.org\n",
      "experiment_accession: ENCSR847RRR\n",
      "genome_accession: encode:starsolo-mm10-M21-male-index\n",
      "include_intron: false\n",
      "inclusion_accession: 737K-august-2016\n",
      "lab: /labs/barbara-wold/\n",
      "library_accession: ENCLB938HOJ\n",
      "mem_mb: 65536\n",
      "read1:\n",
      "- ENCFF471WVX\n",
      "read2:\n",
      "- ENCFF555EDW\n",
      "star_container: https://woldlab.caltech.edu/~diane/containers/star-bash-2.7.10a.sif\n",
      "stranded: Forward\n",
      "umi_length: 10\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(yaml.dump(all_configs[\"ENCLB938HOJ\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "operating-bahamas",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "automatic_submission: true\n",
      "award: UM1HG009443\n",
      "description: Please note that this experiment is part of 10x multiome and has a corresponding\n",
      "  scATAC experiment (ENCSR777GVP).\n",
      "disk_mb: 51200\n",
      "encode_portal_host: www.encodeproject.org\n",
      "experiment_accession: ENCSR700UVN\n",
      "genome_accession: encode:starsolo-GRCh38-V29-male-index\n",
      "include_intron: true\n",
      "inclusion_accession: 737K-arc-v1(GEX)\n",
      "lab: /labs/barbara-wold/\n",
      "library_accession: ENCLB527WWJ\n",
      "mem_mb: 65536\n",
      "read1:\n",
      "- ENCFF660WUU\n",
      "- ENCFF381PIJ\n",
      "read2:\n",
      "- ENCFF563YLA\n",
      "- ENCFF365DGX\n",
      "star_container: https://woldlab.caltech.edu/~diane/containers/star-bash-2.7.10a.sif\n",
      "stranded: Forward\n",
      "umi_length: 12\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(yaml.dump(all_configs[\"ENCLB527WWJ\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "useful-nitrogen",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "automatic_submission: true\n",
      "award: UM1HG009443\n",
      "description: Please note that this experiment is part of 10x multiome and has a corresponding\n",
      "  scATAC experiment (ENCSR668XPI).\n",
      "disk_mb: 51200\n",
      "encode_portal_host: www.encodeproject.org\n",
      "experiment_accession: ENCSR012APQ\n",
      "genome_accession: encode:starsolo-GRCh38-V29-male-index\n",
      "include_intron: true\n",
      "inclusion_accession: 737K-arc-v1(GEX)\n",
      "lab: /labs/barbara-wold/\n",
      "library_accession: ENCLB062TYC\n",
      "mem_mb: 65536\n",
      "read1:\n",
      "- ENCFF955XWD\n",
      "- ENCFF295OZA\n",
      "read2:\n",
      "- ENCFF439KKH\n",
      "- ENCFF088UOS\n",
      "star_container: https://woldlab.caltech.edu/~diane/containers/star-bash-2.7.10a.sif\n",
      "stranded: Forward\n",
      "umi_length: 12\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(yaml.dump(all_configs[\"ENCLB062TYC\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "baking-michael",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "automatic_submission: true\n",
      "award: UM1HG009443\n",
      "description: ''\n",
      "disk_mb: 51200\n",
      "encode_portal_host: www.encodeproject.org\n",
      "experiment_accession: ENCSR075HVN\n",
      "genome_accession: encode:starsolo-mm10-M21-male-index\n",
      "include_intron: true\n",
      "inclusion_accession: 737K-arc-v1(GEX)\n",
      "lab: /labs/barbara-wold/\n",
      "library_accession: ENCLB349KYC\n",
      "mem_mb: 65536\n",
      "read1:\n",
      "- ENCFF153JKN\n",
      "- ENCFF459HYO\n",
      "read2:\n",
      "- ENCFF537XAH\n",
      "- ENCFF958QIC\n",
      "star_container: https://woldlab.caltech.edu/~diane/containers/star-bash-2.7.10a.sif\n",
      "stranded: Forward\n",
      "umi_length: 12\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(yaml.dump(all_configs[\"ENCLB349KYC\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "downtown-glossary",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "automatic_submission: true\n",
      "award: UM1HG009443\n",
      "description: 10X Genomics Chromium 3' single-cell RNA-Seq (v2 kit) of C57Bl6 wild-type\n",
      "  13.0 day embryonic mouse forelimb\n",
      "disk_mb: 51200\n",
      "encode_portal_host: www.encodeproject.org\n",
      "experiment_accession: ENCSR156HHE\n",
      "genome_accession: encode:starsolo-mm10-M21-male-index\n",
      "include_intron: false\n",
      "inclusion_accession: 737K-august-2016\n",
      "lab: /labs/barbara-wold/\n",
      "library_accession: ENCLB497BTJ\n",
      "mem_mb: 65536\n",
      "read1:\n",
      "- ENCFF151FBB\n",
      "read2:\n",
      "- ENCFF233PFV\n",
      "star_container: https://woldlab.caltech.edu/~diane/containers/star-bash-2.7.10a.sif\n",
      "stranded: Forward\n",
      "umi_length: 10\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(yaml.dump(all_configs[\"ENCLB497BTJ\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "american-activation",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "automatic_submission: true\n",
      "award: UM1HG009443\n",
      "description: snRNA on human pancreas\n",
      "disk_mb: 51200\n",
      "encode_portal_host: www.encodeproject.org\n",
      "experiment_accession: ENCSR684KYI\n",
      "genome_accession: encode:starsolo-GRCh38-V29-male-index\n",
      "include_intron: true\n",
      "inclusion_accession: 3M-february-2018\n",
      "lab: /labs/barbara-wold/\n",
      "library_accession: ENCLB549SMV\n",
      "mem_mb: 65536\n",
      "read1:\n",
      "- ENCFF649AKD\n",
      "- ENCFF846ATA\n",
      "read2:\n",
      "- ENCFF039MXO\n",
      "- ENCFF760KEL\n",
      "star_container: https://woldlab.caltech.edu/~diane/containers/star-bash-2.7.10a.sif\n",
      "stranded: Forward\n",
      "umi_length: 12\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(yaml.dump(all_configs[\"ENCLB549SMV\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "underlying-audio",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "automatic_submission: true\n",
      "award: UM1HG009443\n",
      "description: 10X Genomics Chromium 3' single-cell RNA-Seq (v2 kit) of C57Bl6 wild-type\n",
      "  13.0 day embryonic mouse forelimb\n",
      "disk_mb: 51200\n",
      "encode_portal_host: www.encodeproject.org\n",
      "experiment_accession: ENCSR156HHE\n",
      "genome_accession: encode:starsolo-mm10-M21-male-index\n",
      "include_intron: false\n",
      "inclusion_accession: 737K-august-2016\n",
      "lab: /labs/barbara-wold/\n",
      "library_accession: ENCLB497BTJ\n",
      "mem_mb: 65536\n",
      "read1:\n",
      "- ENCFF151FBB\n",
      "read2:\n",
      "- ENCFF233PFV\n",
      "star_container: https://woldlab.caltech.edu/~diane/containers/star-bash-2.7.10a.sif\n",
      "stranded: Forward\n",
      "umi_length: 10\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(yaml.dump(all_configs[\"ENCLB497BTJ\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "caroline-rwanda",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "intermediate-producer",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gene_model(config):\n",
    "    return \"GeneFull_Ex50pAS\" if config['include_intron'] else \"Gene\"\n",
    "\n",
    "\n",
    "def prepare_bulk_subworkflow(all_configs, libraries, submit=False):\n",
    "    batch_config = []\n",
    "    bulk_rule_input = []\n",
    "    \n",
    "    for library in libraries:\n",
    "        config = all_configs[library]\n",
    "        #config[\"automatic_submission\"] = False\n",
    "        submit_host = config[\"encode_portal_host\"]\n",
    "        library_dir = Path(library)\n",
    "        workflow_name = \"job_{}\".format(library)\n",
    "\n",
    "        current = \"\"\"subworkflow {workflow_name}:\n",
    "    workdir:\n",
    "        \"{workdir}\"\n",
    "    snakefile:\n",
    "        \"{snakefile}\"\n",
    "    configfile:\n",
    "        \"{configfile}\"\n",
    "\"\"\".format(\n",
    "        workflow_name=workflow_name,\n",
    "        workdir=str(library_dir),\n",
    "        snakefile=str(Path(\"/home/diane/woldlab-rna-seq/workflow/process-encode-10x.snakefile\")),\n",
    "        configfile=str(library_dir / \"config.yaml\"))\n",
    "        batch_config.append(current)\n",
    "        \n",
    "        posted_csv = \"posted.{}.csv\".format(submit_host)\n",
    "        metadata_csv = \"metadata.{}.csv\".format(submit_host)\n",
    "        target_rule = posted_csv if config[\"automatic_submission\"] else metadata_csv\n",
    "        bulk_rule_input.append(\"{}(\\\"{}\\\")\".format(workflow_name, target_rule))\n",
    "        final_out_rule = \"Log.final.out.{}.qc-upload\".format(submit_host)\n",
    "        bulk_rule_input.append(\"{}(\\\"{}\\\")\".format(workflow_name, final_out_rule))\n",
    "        summary_rule = \"Solo.out/{}/Summary.csv.{}.qc-upload\".format(get_gene_model(config), submit_host)\n",
    "        bulk_rule_input.append(\"{}(\\\"{}\\\")\".format(workflow_name, summary_rule))\n",
    "        pct_count_mt_rule = \"pct_count_mt.{}_EM_filtered.png.{}.qc-upload\".format(get_gene_model(config), submit_host)\n",
    "        bulk_rule_input.append(\"{}(\\\"{}\\\")\".format(workflow_name, pct_count_mt_rule))\n",
    "    \n",
    "    formatted_rules = []\n",
    "    for rule in bulk_rule_input:\n",
    "        formatted_rules.append(\"{prefix}{rule},\\n\".format(prefix=\" \" * 8, rule=rule))\n",
    "    batch_config.append(\"\"\"rule world:\n",
    "    input:\n",
    "{rules}\n",
    "\"\"\".format(rules=\"\".join(formatted_rules)))\n",
    "    return \"\\n\".join(batch_config)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "impressive-surgery",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "subworkflow job_ENCLB002DZK:\n",
      "    workdir:\n",
      "        \"ENCLB002DZK\"\n",
      "    snakefile:\n",
      "        \"/home/diane/woldlab-rna-seq/workflow/process-encode-10x.snakefile\"\n",
      "    configfile:\n",
      "        \"ENCLB002DZK/config.yaml\"\n",
      "\n",
      "subworkflow job_ENCLB549KAN:\n",
      "    workdir:\n",
      "        \"ENCLB549KAN\"\n",
      "    snakefile:\n",
      "        \"/home/diane/woldlab-rna-seq/workflow/process-encode-10x.snakefile\"\n",
      "    configfile:\n",
      "        \"ENCLB549KAN/config.yaml\"\n",
      "\n",
      "rule world:\n",
      "    input:\n",
      "        job_ENCLB002DZK(\"posted.www.encodeproject.org.csv\"),\n",
      "        job_ENCLB002DZK(\"Log.final.out.www.encodeproject.org.qc-upload\"),\n",
      "        job_ENCLB002DZK(\"Solo.out/GeneFull_Ex50pAS/Summary.csv.www.encodeproject.org.qc-upload\"),\n",
      "        job_ENCLB002DZK(\"pct_count_mt.GeneFull_Ex50pAS_EM_filtered.png.www.encodeproject.org.qc-upload\"),\n",
      "        job_ENCLB549KAN(\"posted.www.encodeproject.org.csv\"),\n",
      "        job_ENCLB549KAN(\"Log.final.out.www.encodeproject.org.qc-upload\"),\n",
      "        job_ENCLB549KAN(\"Solo.out/GeneFull_Ex50pAS/Summary.csv.www.encodeproject.org.qc-upload\"),\n",
      "        job_ENCLB549KAN(\"pct_count_mt.GeneFull_Ex50pAS_EM_filtered.png.www.encodeproject.org.qc-upload\"),\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(prepare_bulk_subworkflow(all_configs, [\"ENCLB002DZK\", \"ENCLB549KAN\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "equal-martial",
   "metadata": {},
   "outputs": [],
   "source": [
    "def prepare_bulk_subworkflow_run(destination, all_configs, libraries):\n",
    "    with tarfile.open(destination, \"w:gz\") as archive:\n",
    "        for library in libraries:\n",
    "            library_dir = Path(library)\n",
    "            config_data = yaml.dump(all_configs[library]).encode(\"utf-8\")\n",
    "            config_stream = BytesIO(config_data)\n",
    "            config_filename = library_dir / \"config.yaml\"\n",
    "            config_info = tarfile.TarInfo(str(config_filename))\n",
    "            config_info.size = len(config_data)\n",
    "            \n",
    "            archive.addfile(config_info, fileobj=config_stream)\n",
    "            \n",
    "        world_data = prepare_bulk_subworkflow(all_configs, libraries).encode(\"utf-8\")\n",
    "        world_stream = BytesIO(world_data)\n",
    "        world_info = tarfile.TarInfo(\"Snakefile\")\n",
    "        world_info.size = len(world_data)\n",
    "        archive.addfile(world_info, world_stream)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "incredible-notification",
   "metadata": {},
   "outputs": [],
   "source": [
    "prepare_bulk_subworkflow_run(\"/dev/shm/all.tar.gz\", all_configs, [\"ENCLB002DZK\", \"ENCLB549KAN\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "editorial-residence",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "constant-robertson",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "every-warning",
   "metadata": {},
   "outputs": [],
   "source": [
    "def prepare_bulk_sbatch_10x(all_configs, libraries, submit=False):\n",
    "    commands = [\"#!/bin/bash\", \"\"]\n",
    "    \n",
    "    for library in libraries:\n",
    "        config = all_configs[library]\n",
    "        library_dir = Path(library)\n",
    "\n",
    "        commands.append(\"pushd {}\".format(library_dir))\n",
    "        commands.append(\"sbatch --time 1-0 --mem 64GB --cpus-per-task 24 -- snakemake --use-singularity --snakefile ~/woldlab-rna-seq/workflow/process-encode-10x.snakefile --cluster \\\"sbatch --time 1-0 --mem {resources.mem_mb} --cpus-per-task {threads}\\\" --jobs 40 --cores 40\")\n",
    "        commands.append(\"popd\")\n",
    "    \n",
    "    return \"\\n\".join(commands)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "biological-potato",
   "metadata": {},
   "outputs": [],
   "source": [
    "def prepare_bulk_sbatch_10x_run(destination, all_configs, libraries):\n",
    "    valid_libraries = []\n",
    "    with tarfile.open(destination, \"w:gz\") as archive:\n",
    "        for library in libraries:\n",
    "            config = all_configs[library]\n",
    "            if len(config.get(\"read1\", [])) == 0 or len(config.get(\"read2\",[])) == 0:\n",
    "                print(\n",
    "                    \"WARNING: {}/{} has no fastqs, skipping\".format(\n",
    "                        config[\"experiment_accession\"], config[\"library_accession\"]))\n",
    "                continue\n",
    "            library_dir = Path(library)\n",
    "            config_data = yaml.dump(all_configs[library]).encode(\"utf-8\")\n",
    "            config_stream = BytesIO(config_data)\n",
    "            config_filename = library_dir / \"config.yaml\"\n",
    "            config_info = tarfile.TarInfo(str(config_filename))\n",
    "            config_info.size = len(config_data)\n",
    "            \n",
    "            archive.addfile(config_info, fileobj=config_stream)\n",
    "            valid_libraries.append(library)\n",
    "            \n",
    "        world_data = prepare_bulk_sbatch_10x(all_configs, valid_libraries).encode(\"utf-8\")\n",
    "        world_stream = BytesIO(world_data)\n",
    "        world_info = tarfile.TarInfo(\"run.sh\")\n",
    "        world_info.size = len(world_data)\n",
    "        archive.addfile(world_info, world_stream)\n",
    "\n",
    "    return valid_libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "valid-glory",
   "metadata": {},
   "outputs": [],
   "source": [
    "def prepare_bulk_sbatch_split(all_configs, libraries, submit=False):\n",
    "    commands = [\"#!/bin/bash\", \"\"]\n",
    "    \n",
    "    for library in libraries:\n",
    "        config = all_configs[library]\n",
    "        library_dir = Path(library)\n",
    "\n",
    "        commands.append(\"pushd {}\".format(library_dir))\n",
    "        commands.append(\"sbatch --time 1-0 --mem 64GB --cpus-per-task 24 -- snakemake --use-singularity --snakefile ~/woldlab-rna-seq/workflow/process-encode-splitseq.snakefile --cluster \\\"sbatch --time 1-0 --mem {resources.mem_mb} --cpus-per-task {threads}\\\" --jobs 40 --cores 40\")\n",
    "        commands.append(\"popd\")\n",
    "    \n",
    "    return \"\\n\".join(commands)\n",
    "\n",
    "def prepare_bulk_sbatch_split_run(destination, all_configs, libraries):\n",
    "    with tarfile.open(destination, \"w:gz\") as archive:\n",
    "        for library in libraries:\n",
    "            library_dir = Path(library)\n",
    "            config_data = yaml.dump(all_configs[library]).encode(\"utf-8\")\n",
    "            config_stream = BytesIO(config_data)\n",
    "            config_filename = library_dir / \"config.yaml\"\n",
    "            config_info = tarfile.TarInfo(str(config_filename))\n",
    "            config_info.size = len(config_data)\n",
    "            \n",
    "            archive.addfile(config_info, fileobj=config_stream)\n",
    "            \n",
    "        world_data = prepare_bulk_sbatch_split(all_configs, libraries).encode(\"utf-8\")\n",
    "        world_stream = BytesIO(world_data)\n",
    "        world_info = tarfile.TarInfo(\"run.sh\")\n",
    "        world_info.size = len(world_data)\n",
    "        archive.addfile(world_info, world_stream)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "covered-wildlife",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = pandas.DataFrame(results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "original-filter",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Preparing 27 jobs\n"
     ]
    }
   ],
   "source": [
    "experiments_to_run_2_6 = [\"ENCSR005TCL\",\"ENCSR011EDH\",\"ENCSR012APQ\",\"ENCSR021QIA\",\"ENCSR021YYL\"]\n",
    "\n",
    "experiments_to_run_7_21 = [\n",
    "\"ENCSR028MBY\",\n",
    "\"ENCSR035BDI\",\n",
    "\"ENCSR060SEA\",\n",
    "\"ENCSR060ZXZ\",\n",
    "\"ENCSR075HVN\",\n",
    "\"ENCSR079LVG\",\n",
    "\"ENCSR087YKG\",\n",
    "\"ENCSR132MEW\",\n",
    "\"ENCSR133XJX\",\n",
    "\"ENCSR142RRV\",\n",
    "\"ENCSR151ONW\",\n",
    "\"ENCSR157YJO\",\n",
    "\"ENCSR201WJO\",\n",
    "\"ENCSR203THX\",\n",
    "\"ENCSR224OUG\",\n",
    "]\n",
    "\n",
    "#experiments_to_run = set([\"ENCSR402PHR\",\"ENCSR157YJO\",\"ENCSR021YYL\",\"ENCSR988AMY\",\"ENCSR908CQZ\",\"ENCSR749GDE\",\"ENCSR224OUG\",\"ENCSR356VJZ\", \"ENCSR021YYL\"])\n",
    "experiments_to_run_22_51 = [\n",
    "\"ENCSR224RWH\",\n",
    "\"ENCSR225OII\",\n",
    "\"ENCSR258SZI\",\n",
    "\"ENCSR259VOY\",\n",
    "\"ENCSR261BXB\",\n",
    "\"ENCSR261LPM\",\n",
    "\"ENCSR281NBH\",\n",
    "\"ENCSR281RIH\",\n",
    "\"ENCSR289ANH\",\n",
    "\"ENCSR297ZTQ\",\n",
    "\"ENCSR298MPV\",\n",
    "\"ENCSR333SLJ\",\n",
    "\"ENCSR337HKS\",\n",
    "\"ENCSR352DXB\",\n",
    "\"ENCSR356VJZ\",\n",
    "\"ENCSR362HUP\",\n",
    "\"ENCSR362YDM\",\n",
    "\"ENCSR367YUX\",\n",
    "\"ENCSR369UFT\",\n",
    "\"ENCSR398PIW\",\n",
    "\"ENCSR402PHR\",\n",
    "\"ENCSR405YKM\",\n",
    "\"ENCSR415HEP\",\n",
    "\"ENCSR417FCM\",\n",
    "\"ENCSR421YXX\",\n",
    "\"ENCSR426PDU\",\n",
    "\"ENCSR439ZVQ\",\n",
    "\"ENCSR445LCA\",\n",
    "\"ENCSR460MGC\",\n",
    "\"ENCSR462HCD\",\n",
    "]\n",
    "\n",
    "experiments_to_run_52_81 = [\n",
    "\"ENCSR463EEP\",\n",
    "\"ENCSR467GGT\",\n",
    "\"ENCSR471YTA\",\n",
    "\"ENCSR472RRP\",\n",
    "\"ENCSR478FQR\",\n",
    "\"ENCSR481QQR\",\n",
    "\"ENCSR484YSK\",\n",
    "\"ENCSR522HNR\",\n",
    "\"ENCSR527MGL\",\n",
    "\"ENCSR533IGQ\",\n",
    "\"ENCSR539XHZ\",\n",
    "\"ENCSR548HED\",\n",
    "\"ENCSR568GGW\",\n",
    "\"ENCSR587SPO\",\n",
    "\"ENCSR602SGU\",\n",
    "\"ENCSR620AJZ\",\n",
    "\"ENCSR624WDN\",\n",
    "\"ENCSR632EIK\",\n",
    "\"ENCSR641MWO\",\n",
    "\"ENCSR642GFM\",\n",
    "\"ENCSR644VVE\",\n",
    "\"ENCSR656VWP\",\n",
    "\"ENCSR662VTU\",\n",
    "\"ENCSR665HPR\",\n",
    "\"ENCSR671FDY\",\n",
    "\"ENCSR683ONV\",\n",
    "\"ENCSR684KYI\",\n",
    "\"ENCSR700UVN\",\n",
    "\"ENCSR707ECV\",\n",
    "\"ENCSR707LLI\",\n",
    "]\n",
    "\n",
    "experiments_to_run_82_107 = [\"ENCSR716VLR\",\n",
    "\"ENCSR720TUP\",\n",
    "\"ENCSR724KET\",\n",
    "\"ENCSR726IPC\",\n",
    "\"ENCSR727OYO\",\n",
    "\"ENCSR742NWO\",\n",
    "\"ENCSR749GDE\",\n",
    "\"ENCSR751BHQ\",\n",
    "\"ENCSR753YPW\",\n",
    "\"ENCSR755CIF\",\n",
    "\"ENCSR762LZV\",\n",
    "\"ENCSR769KWM\",\n",
    "\"ENCSR789PLK\",\n",
    "\"ENCSR801DHT\",\n",
    "\"ENCSR813PWQ\",\n",
    "\"ENCSR835JFN\",\n",
    "\"ENCSR876EUH\",\n",
    "\"ENCSR908CQZ\",\n",
    "\"ENCSR938CZJ\",\n",
    "\"ENCSR943HBF\",\n",
    "\"ENCSR944XPC\",\n",
    "\"ENCSR964JZN\",\n",
    "\"ENCSR966DDY\",\n",
    "\"ENCSR968RPR\",\n",
    "\"ENCSR987SOK\",\n",
    "\"ENCSR988AMY\",\n",
    "\"ENCSR995ZKH\",\n",
    "]\n",
    "\n",
    "experiments_to_run = experiments_to_run_82_107\n",
    "print(\"Preparing {} jobs\".format(len(experiments_to_run)))\n",
    "ready_metadata = metadata[metadata[\"experiment\"].isin(experiments_to_run)]\n",
    "\n",
    "#prepare_bulk_sbatch_10x_run(\"/dev/shm/test-run-82-108.tar.gz\", all_configs, ready_metadata[\"library\"].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "second-windsor",
   "metadata": {},
   "outputs": [],
   "source": [
    "#prepare_bulk_10x_run(\"/dev/shm/adrenal_multiome.tar.gz\", all_configs, ready_metadata[\"library\"].values)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "royal-heavy",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://www.encodeproject.org/experiments/ENCSR755CIF/\n",
      "https://www.encodeproject.org/experiments/ENCSR751BHQ/\n",
      "https://www.encodeproject.org/experiments/ENCSR995ZKH/\n",
      "https://www.encodeproject.org/experiments/ENCSR726IPC/\n",
      "https://www.encodeproject.org/experiments/ENCSR876EUH/\n",
      "https://www.encodeproject.org/experiments/ENCSR769KWM/\n",
      "https://www.encodeproject.org/experiments/ENCSR789PLK/\n",
      "https://www.encodeproject.org/experiments/ENCSR813PWQ/\n",
      "https://www.encodeproject.org/experiments/ENCSR966DDY/\n",
      "https://www.encodeproject.org/experiments/ENCSR724KET/\n",
      "https://www.encodeproject.org/experiments/ENCSR716VLR/\n",
      "https://www.encodeproject.org/experiments/ENCSR964JZN/\n",
      "https://www.encodeproject.org/experiments/ENCSR762LZV/\n",
      "https://www.encodeproject.org/experiments/ENCSR943HBF/\n",
      "https://www.encodeproject.org/experiments/ENCSR753YPW/\n",
      "https://www.encodeproject.org/experiments/ENCSR727OYO/\n",
      "https://www.encodeproject.org/experiments/ENCSR801DHT/\n",
      "https://www.encodeproject.org/experiments/ENCSR938CZJ/\n",
      "https://www.encodeproject.org/experiments/ENCSR749GDE/\n",
      "https://www.encodeproject.org/experiments/ENCSR968RPR/\n",
      "https://www.encodeproject.org/experiments/ENCSR835JFN/\n",
      "https://www.encodeproject.org/experiments/ENCSR908CQZ/\n",
      "https://www.encodeproject.org/experiments/ENCSR944XPC/\n",
      "https://www.encodeproject.org/experiments/ENCSR988AMY/\n",
      "https://www.encodeproject.org/experiments/ENCSR742NWO/\n",
      "https://www.encodeproject.org/experiments/ENCSR720TUP/\n",
      "https://www.encodeproject.org/experiments/ENCSR987SOK/\n"
     ]
    }
   ],
   "source": [
    "for i, row in ready_metadata.iterrows():\n",
    "    config = all_configs[row.library]\n",
    "    print(\"https://{}/experiments/{}/\".format(config[\"encode_portal_host\"], row.experiment))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "limited-meter",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ENCLB779BCG',\n",
       " 'ENCLB338RBX',\n",
       " 'ENCLB774YSO',\n",
       " 'ENCLB067EKD',\n",
       " 'ENCLB043KZW',\n",
       " 'ENCLB243NMQ',\n",
       " 'ENCLB113GDH',\n",
       " 'ENCLB765NUA',\n",
       " 'ENCLB555STO',\n",
       " 'ENCLB950DIT',\n",
       " 'ENCLB487ULG',\n",
       " 'ENCLB145HAH',\n",
       " 'ENCLB457YKG',\n",
       " 'ENCLB234BIK',\n",
       " 'ENCLB835IXS',\n",
       " 'ENCLB943WKY',\n",
       " 'ENCLB508DVG',\n",
       " 'ENCLB943KZN',\n",
       " 'ENCLB433LMI',\n",
       " 'ENCLB863YCR',\n",
       " 'ENCLB150JOE',\n",
       " 'ENCLB138XBO',\n",
       " 'ENCLB155EAA',\n",
       " 'ENCLB888MGZ',\n",
       " 'ENCLB213TAB',\n",
       " 'ENCLB398IAZ',\n",
       " 'ENCLB513JLQ',\n",
       " 'ENCLB025XJD',\n",
       " 'ENCLB335ITX',\n",
       " 'ENCLB461YMD',\n",
       " 'ENCLB761CFQ',\n",
       " 'ENCLB179BOW',\n",
       " 'ENCLB414SND',\n",
       " 'ENCLB812SMT',\n",
       " 'ENCLB615BOP',\n",
       " 'ENCLB701RZY',\n",
       " 'ENCLB025GOX',\n",
       " 'ENCLB280ZGL',\n",
       " 'ENCLB392SOP',\n",
       " 'ENCLB801PEP',\n",
       " 'ENCLB342YHF',\n",
       " 'ENCLB338QDU',\n",
       " 'ENCLB507KDD',\n",
       " 'ENCLB218CWN',\n",
       " 'ENCLB793KIA',\n",
       " 'ENCLB140VOU',\n",
       " 'ENCLB872TNB',\n",
       " 'ENCLB001JCO',\n",
       " 'ENCLB399EQG',\n",
       " 'ENCLB814PUF',\n",
       " 'ENCLB327PYA',\n",
       " 'ENCLB151JZY',\n",
       " 'ENCLB061STU',\n",
       " 'ENCLB297FDO',\n",
       " 'ENCLB002DZK',\n",
       " 'ENCLB549SMV',\n",
       " 'ENCLB611BAN',\n",
       " 'ENCLB997ETR',\n",
       " 'ENCLB954OKJ',\n",
       " 'ENCLB129RYA',\n",
       " 'ENCLB065ONF',\n",
       " 'ENCLB196FUF',\n",
       " 'ENCLB357XZN',\n",
       " 'ENCLB504FSE',\n",
       " 'ENCLB038DYV',\n",
       " 'ENCLB529BXD',\n",
       " 'ENCLB452QHE',\n",
       " 'ENCLB366ZFV',\n",
       " 'ENCLB681VIY',\n",
       " 'ENCLB986LXE',\n",
       " 'ENCLB593ESF',\n",
       " 'ENCLB765JXX',\n",
       " 'ENCLB455UTN',\n",
       " 'ENCLB123ZDT',\n",
       " 'ENCLB039CPV',\n",
       " 'ENCLB181TPV',\n",
       " 'ENCLB907NAX',\n",
       " 'ENCLB039THH',\n",
       " 'ENCLB212WDJ',\n",
       " 'ENCLB960RBM',\n",
       " 'ENCLB814OYP',\n",
       " 'ENCLB296LDF',\n",
       " 'ENCLB592LXS',\n",
       " 'ENCLB971CVF',\n",
       " 'ENCLB396URL',\n",
       " 'ENCLB062TYC',\n",
       " 'ENCLB323ARH',\n",
       " 'ENCLB008XGB',\n",
       " 'ENCLB900EOL',\n",
       " 'ENCLB258LQF',\n",
       " 'ENCLB505NAL',\n",
       " 'ENCLB295OOS',\n",
       " 'ENCLB103OKD',\n",
       " 'ENCLB849AUZ',\n",
       " 'ENCLB032OPR',\n",
       " 'ENCLB460PSY',\n",
       " 'ENCLB845QWH',\n",
       " 'ENCLB861GQS',\n",
       " 'ENCLB802MGH',\n",
       " 'ENCLB000QLP',\n",
       " 'ENCLB860YRM',\n",
       " 'ENCLB280OVJ',\n",
       " 'ENCLB610ELT',\n",
       " 'ENCLB790XHI',\n",
       " 'ENCLB063JPD',\n",
       " 'ENCLB228GCE',\n",
       " 'ENCLB272QLF',\n",
       " 'ENCLB909DQG',\n",
       " 'ENCLB619VVF',\n",
       " 'ENCLB549KAN',\n",
       " 'ENCLB674XYV',\n",
       " 'ENCLB932FNN',\n",
       " 'ENCLB418IDT',\n",
       " 'ENCLB846XMI',\n",
       " 'ENCLB673QFY',\n",
       " 'ENCLB643HGB',\n",
       " 'ENCLB578ADC',\n",
       " 'ENCLB527WWJ',\n",
       " 'ENCLB328JIP',\n",
       " 'ENCLB831IHU',\n",
       " 'ENCLB607FDJ',\n",
       " 'ENCLB866MAL',\n",
       " 'ENCLB275OVJ',\n",
       " 'ENCLB673NZA',\n",
       " 'ENCLB675MBI',\n",
       " 'ENCLB808STP',\n",
       " 'ENCLB320HXY',\n",
       " 'ENCLB275WWO',\n",
       " 'ENCLB442AMV',\n",
       " 'ENCLB320ZLU',\n",
       " 'ENCLB069PRB',\n",
       " 'ENCLB850RYI',\n",
       " 'ENCLB293BAX',\n",
       " 'ENCLB089YWN',\n",
       " 'ENCLB807VEG',\n",
       " 'ENCLB854VKF',\n",
       " 'ENCLB139QES',\n",
       " 'ENCLB063XDV',\n",
       " 'ENCLB807HIR',\n",
       " 'ENCLB349KYC',\n",
       " 'ENCLB012QYK',\n",
       " 'ENCLB054SXC',\n",
       " 'ENCLB823WAF',\n",
       " 'ENCLB713GHH',\n",
       " 'ENCLB401UMP',\n",
       " 'ENCLB401DAZ',\n",
       " 'ENCLB851FFF',\n",
       " 'ENCLB568RWC',\n",
       " 'ENCLB766ZBL',\n",
       " 'ENCLB995TFV',\n",
       " 'ENCLB608YNH',\n",
       " 'ENCLB871PHY',\n",
       " 'ENCLB847VJF',\n",
       " 'ENCLB713ETA',\n",
       " 'ENCLB983BCX',\n",
       " 'ENCLB432ZEQ',\n",
       " 'ENCLB663TUU',\n",
       " 'ENCLB225KMP',\n",
       " 'ENCLB495DSX',\n",
       " 'ENCLB676AMA',\n",
       " 'ENCLB786EOH',\n",
       " 'ENCLB794PMK',\n",
       " 'ENCLB104XFX',\n",
       " 'ENCLB560GBC',\n",
       " 'ENCLB526RVS',\n",
       " 'ENCLB540OBF',\n",
       " 'ENCLB566WWJ']"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prepare_bulk_sbatch_10x_run(\"/dev/shm/run-1-108.tar.gz\", all_configs, metadata[metadata[\"lab\"] == \"Stanford\"][\"library\"].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "drawn-orleans",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ENCLB338RBX',\n",
       " 'ENCLB067EKD',\n",
       " 'ENCLB835IXS',\n",
       " 'ENCLB025XJD',\n",
       " 'ENCLB179BOW',\n",
       " 'ENCLB812SMT',\n",
       " 'ENCLB701RZY',\n",
       " 'ENCLB342YHF',\n",
       " 'ENCLB061STU',\n",
       " 'ENCLB002DZK',\n",
       " 'ENCLB954OKJ',\n",
       " 'ENCLB196FUF',\n",
       " 'ENCLB455UTN',\n",
       " 'ENCLB123ZDT',\n",
       " 'ENCLB039THH',\n",
       " 'ENCLB960RBM',\n",
       " 'ENCLB296LDF',\n",
       " 'ENCLB418IDT',\n",
       " 'ENCLB442AMV',\n",
       " 'ENCLB850RYI',\n",
       " 'ENCLB089YWN',\n",
       " 'ENCLB139QES',\n",
       " 'ENCLB012QYK',\n",
       " 'ENCLB568RWC',\n",
       " 'ENCLB225KMP',\n",
       " 'ENCLB676AMA',\n",
       " 'ENCLB104XFX']"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prepare_bulk_sbatch_10x_run(\"/dev/shm/test-run-82-108.tar.gz\", all_configs, ready_metadata[\"library\"].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "gothic-corner",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prepare_bulk_sbatch_10x_run(\"/dev/shm/mouse-limb.tar.gz\", all_configs, metadata[metadata[\"experiment\"].isin(mouse_limb_experiments)][\"library\"].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "impressed-typing",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_dcc_accession(server, accession):\n",
    "    metadata = None\n",
    "    try:\n",
    "        metadata = server.get_json(accession)\n",
    "    except HTTPError as e:\n",
    "        if e.response.status_code == 404:\n",
    "            print(\"Accession {} not found, discovery will fail\".format(accession))\n",
    "        else:\n",
    "            print(\"Other error {} for {}\".format(e.response.status_code, accession))\n",
    "            raise e\n",
    "\n",
    "    return metadata\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "mounted-richmond",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uuid</th>\n",
       "      <th>accession</th>\n",
       "      <th>dataset</th>\n",
       "      <th>file_format</th>\n",
       "      <th>output_type</th>\n",
       "      <th>assembly</th>\n",
       "      <th>genome_annotation</th>\n",
       "      <th>derived_from:array</th>\n",
       "      <th>md5sum</th>\n",
       "      <th>file_size:integer</th>\n",
       "      <th>submitted_file_name</th>\n",
       "      <th>award</th>\n",
       "      <th>lab</th>\n",
       "      <th>step_run</th>\n",
       "      <th>aliases:array</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR224OUG</td>\n",
       "      <td>bam</td>\n",
       "      <td>alignments</td>\n",
       "      <td>mm10</td>\n",
       "      <td>M21</td>\n",
       "      <td>encode:starsolo-mm10-M21-male-index,/files/ENC...</td>\n",
       "      <td>38bf053481a36b565a45e2e4fedb2bbf</td>\n",
       "      <td>14365291492</td>\n",
       "      <td>Aligned.sortedByCoord.out.bam</td>\n",
       "      <td>UM1HG009443</td>\n",
       "      <td>/labs/barbara-wold/</td>\n",
       "      <td>barbara-wold:starsolo-10x-alignment-step-run</td>\n",
       "      <td>barbara-wold:ENCLB607FDJ_alignment_2022-01-19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR224OUG</td>\n",
       "      <td>tar</td>\n",
       "      <td>sparse gene count matrix of unique reads</td>\n",
       "      <td>mm10</td>\n",
       "      <td>M21</td>\n",
       "      <td>737K-arc-v1(GEX),barbara-wold:ENCLB607FDJ_alig...</td>\n",
       "      <td>d4214b016b19730e185ce11dd5396514</td>\n",
       "      <td>39391130</td>\n",
       "      <td>GeneFull_Ex50pAS_Unique_filtered.tar.gz</td>\n",
       "      <td>UM1HG009443</td>\n",
       "      <td>/labs/barbara-wold/</td>\n",
       "      <td>barbara-wold:starsolo-10x-quantification-step-run</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR224OUG</td>\n",
       "      <td>tar</td>\n",
       "      <td>sparse gene count matrix of all reads</td>\n",
       "      <td>mm10</td>\n",
       "      <td>M21</td>\n",
       "      <td>737K-arc-v1(GEX),barbara-wold:ENCLB607FDJ_alig...</td>\n",
       "      <td>68d9aa21aae8e891f85f880cd5bd8cab</td>\n",
       "      <td>46821115</td>\n",
       "      <td>GeneFull_Ex50pAS_EM_filtered.tar.gz</td>\n",
       "      <td>UM1HG009443</td>\n",
       "      <td>/labs/barbara-wold/</td>\n",
       "      <td>barbara-wold:starsolo-10x-quantification-step-run</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR224OUG</td>\n",
       "      <td>tar</td>\n",
       "      <td>unfiltered sparse gene count matrix of unique ...</td>\n",
       "      <td>mm10</td>\n",
       "      <td>M21</td>\n",
       "      <td>737K-arc-v1(GEX),barbara-wold:ENCLB607FDJ_alig...</td>\n",
       "      <td>920b78efae954ffb082faf2aaa921ba5</td>\n",
       "      <td>59389166</td>\n",
       "      <td>GeneFull_Ex50pAS_Unique_raw.tar.gz</td>\n",
       "      <td>UM1HG009443</td>\n",
       "      <td>/labs/barbara-wold/</td>\n",
       "      <td>barbara-wold:starsolo-10x-quantification-step-run</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR224OUG</td>\n",
       "      <td>tar</td>\n",
       "      <td>unfiltered sparse gene count matrix of all reads</td>\n",
       "      <td>mm10</td>\n",
       "      <td>M21</td>\n",
       "      <td>737K-arc-v1(GEX),barbara-wold:ENCLB607FDJ_alig...</td>\n",
       "      <td>7d6019e2816f28f19eeecd47a3442608</td>\n",
       "      <td>76437934</td>\n",
       "      <td>GeneFull_Ex50pAS_EM_raw.tar.gz</td>\n",
       "      <td>UM1HG009443</td>\n",
       "      <td>/labs/barbara-wold/</td>\n",
       "      <td>barbara-wold:starsolo-10x-quantification-step-run</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENCSR224OUG</td>\n",
       "      <td>tar</td>\n",
       "      <td>unfiltered sparse splice junction count matrix...</td>\n",
       "      <td>mm10</td>\n",
       "      <td>M21</td>\n",
       "      <td>737K-arc-v1(GEX),barbara-wold:ENCLB607FDJ_alig...</td>\n",
       "      <td>2dd5935279637f61abf89e5778881a77</td>\n",
       "      <td>14019963</td>\n",
       "      <td>SJ_Unique_raw.tar.gz</td>\n",
       "      <td>UM1HG009443</td>\n",
       "      <td>/labs/barbara-wold/</td>\n",
       "      <td>barbara-wold:starsolo-10x-quantification-step-run</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   uuid  accession      dataset file_format  \\\n",
       "0   NaN        NaN  ENCSR224OUG         bam   \n",
       "1   NaN        NaN  ENCSR224OUG         tar   \n",
       "2   NaN        NaN  ENCSR224OUG         tar   \n",
       "3   NaN        NaN  ENCSR224OUG         tar   \n",
       "4   NaN        NaN  ENCSR224OUG         tar   \n",
       "5   NaN        NaN  ENCSR224OUG         tar   \n",
       "\n",
       "                                         output_type assembly  \\\n",
       "0                                         alignments     mm10   \n",
       "1           sparse gene count matrix of unique reads     mm10   \n",
       "2              sparse gene count matrix of all reads     mm10   \n",
       "3  unfiltered sparse gene count matrix of unique ...     mm10   \n",
       "4   unfiltered sparse gene count matrix of all reads     mm10   \n",
       "5  unfiltered sparse splice junction count matrix...     mm10   \n",
       "\n",
       "  genome_annotation                                 derived_from:array  \\\n",
       "0               M21  encode:starsolo-mm10-M21-male-index,/files/ENC...   \n",
       "1               M21  737K-arc-v1(GEX),barbara-wold:ENCLB607FDJ_alig...   \n",
       "2               M21  737K-arc-v1(GEX),barbara-wold:ENCLB607FDJ_alig...   \n",
       "3               M21  737K-arc-v1(GEX),barbara-wold:ENCLB607FDJ_alig...   \n",
       "4               M21  737K-arc-v1(GEX),barbara-wold:ENCLB607FDJ_alig...   \n",
       "5               M21  737K-arc-v1(GEX),barbara-wold:ENCLB607FDJ_alig...   \n",
       "\n",
       "                             md5sum  file_size:integer  \\\n",
       "0  38bf053481a36b565a45e2e4fedb2bbf        14365291492   \n",
       "1  d4214b016b19730e185ce11dd5396514           39391130   \n",
       "2  68d9aa21aae8e891f85f880cd5bd8cab           46821115   \n",
       "3  920b78efae954ffb082faf2aaa921ba5           59389166   \n",
       "4  7d6019e2816f28f19eeecd47a3442608           76437934   \n",
       "5  2dd5935279637f61abf89e5778881a77           14019963   \n",
       "\n",
       "                       submitted_file_name        award                  lab  \\\n",
       "0            Aligned.sortedByCoord.out.bam  UM1HG009443  /labs/barbara-wold/   \n",
       "1  GeneFull_Ex50pAS_Unique_filtered.tar.gz  UM1HG009443  /labs/barbara-wold/   \n",
       "2      GeneFull_Ex50pAS_EM_filtered.tar.gz  UM1HG009443  /labs/barbara-wold/   \n",
       "3       GeneFull_Ex50pAS_Unique_raw.tar.gz  UM1HG009443  /labs/barbara-wold/   \n",
       "4           GeneFull_Ex50pAS_EM_raw.tar.gz  UM1HG009443  /labs/barbara-wold/   \n",
       "5                     SJ_Unique_raw.tar.gz  UM1HG009443  /labs/barbara-wold/   \n",
       "\n",
       "                                            step_run  \\\n",
       "0       barbara-wold:starsolo-10x-alignment-step-run   \n",
       "1  barbara-wold:starsolo-10x-quantification-step-run   \n",
       "2  barbara-wold:starsolo-10x-quantification-step-run   \n",
       "3  barbara-wold:starsolo-10x-quantification-step-run   \n",
       "4  barbara-wold:starsolo-10x-quantification-step-run   \n",
       "5  barbara-wold:starsolo-10x-quantification-step-run   \n",
       "\n",
       "                                   aliases:array  \n",
       "0  barbara-wold:ENCLB607FDJ_alignment_2022-01-19  \n",
       "1                                            NaN  \n",
       "2                                            NaN  \n",
       "3                                            NaN  \n",
       "4                                            NaN  \n",
       "5                                            NaN  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from io import StringIO\n",
    "submitted = pandas.read_csv(StringIO(\"\"\"uuid,accession,dataset,file_format,output_type,assembly,genome_annotation,derived_from:array,md5sum,file_size:integer,submitted_file_name,award,lab,step_run,aliases:array\n",
    ",,ENCSR224OUG,bam,alignments,mm10,M21,\"encode:starsolo-mm10-M21-male-index,/files/ENCFF494DXJ/,/files/ENCFF586ZOY/,/files/ENCFF201AFX/,/files/ENCFF468KNY/\",38bf053481a36b565a45e2e4fedb2bbf,14365291492,Aligned.sortedByCoord.out.bam,UM1HG009443,/labs/barbara-wold/,barbara-wold:starsolo-10x-alignment-step-run,barbara-wold:ENCLB607FDJ_alignment_2022-01-19\n",
    ",,ENCSR224OUG,tar,sparse gene count matrix of unique reads,mm10,M21,\"737K-arc-v1(GEX),barbara-wold:ENCLB607FDJ_alignment_2022-01-19\",d4214b016b19730e185ce11dd5396514,39391130,GeneFull_Ex50pAS_Unique_filtered.tar.gz,UM1HG009443,/labs/barbara-wold/,barbara-wold:starsolo-10x-quantification-step-run,\n",
    ",,ENCSR224OUG,tar,sparse gene count matrix of all reads,mm10,M21,\"737K-arc-v1(GEX),barbara-wold:ENCLB607FDJ_alignment_2022-01-19\",68d9aa21aae8e891f85f880cd5bd8cab,46821115,GeneFull_Ex50pAS_EM_filtered.tar.gz,UM1HG009443,/labs/barbara-wold/,barbara-wold:starsolo-10x-quantification-step-run,\n",
    ",,ENCSR224OUG,tar,unfiltered sparse gene count matrix of unique reads,mm10,M21,\"737K-arc-v1(GEX),barbara-wold:ENCLB607FDJ_alignment_2022-01-19\",920b78efae954ffb082faf2aaa921ba5,59389166,GeneFull_Ex50pAS_Unique_raw.tar.gz,UM1HG009443,/labs/barbara-wold/,barbara-wold:starsolo-10x-quantification-step-run,\n",
    ",,ENCSR224OUG,tar,unfiltered sparse gene count matrix of all reads,mm10,M21,\"737K-arc-v1(GEX),barbara-wold:ENCLB607FDJ_alignment_2022-01-19\",7d6019e2816f28f19eeecd47a3442608,76437934,GeneFull_Ex50pAS_EM_raw.tar.gz,UM1HG009443,/labs/barbara-wold/,barbara-wold:starsolo-10x-quantification-step-run,\n",
    ",,ENCSR224OUG,tar,unfiltered sparse splice junction count matrix of unique reads,mm10,M21,\"737K-arc-v1(GEX),barbara-wold:ENCLB607FDJ_alignment_2022-01-19\",2dd5935279637f61abf89e5778881a77,14019963,SJ_Unique_raw.tar.gz,UM1HG009443,/labs/barbara-wold/,barbara-wold:starsolo-10x-quantification-step-run,\n",
    "\"\"\"))\n",
    "submitted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "practical-avenue",
   "metadata": {},
   "outputs": [],
   "source": [
    "problems = set(\"\"\"Problems with ENCSR203THX /libraries/ENCLB592LXS/\n",
    "Problems with ENCSR203THX /libraries/ENCLB592LXS/\n",
    "Problems with ENCSR203THX /libraries/ENCLB592LXS/\n",
    "Problems with ENCSR203THX /libraries/ENCLB592LXS/\n",
    "Problems with ENCSR298MPV /libraries/ENCLB814OYP/\n",
    "Problems with ENCSR298MPV /libraries/ENCLB814OYP/\n",
    "Problems with ENCSR298MPV /libraries/ENCLB814OYP/\n",
    "Problems with ENCSR298MPV /libraries/ENCLB814OYP/\n",
    "Problems with ENCSR337HKS /libraries/ENCLB831IHU/\n",
    "Problems with ENCSR337HKS /libraries/ENCLB831IHU/\n",
    "Problems with ENCSR337HKS /libraries/ENCLB831IHU/\n",
    "Problems with ENCSR337HKS /libraries/ENCLB831IHU/\n",
    "Problems with ENCSR405YKM /libraries/ENCLB181TPV/\n",
    "Problems with ENCSR405YKM /libraries/ENCLB181TPV/\n",
    "Problems with ENCSR405YKM /libraries/ENCLB181TPV/\n",
    "Problems with ENCSR405YKM /libraries/ENCLB181TPV/\n",
    "Problems with ENCSR463EEP /libraries/ENCLB619VVF/\n",
    "Problems with ENCSR463EEP /libraries/ENCLB619VVF/\n",
    "Problems with ENCSR463EEP /libraries/ENCLB619VVF/\n",
    "Problems with ENCSR463EEP /libraries/ENCLB619VVF/\n",
    "Problems with ENCSR527MGL /libraries/ENCLB909DQG/\n",
    "Problems with ENCSR527MGL /libraries/ENCLB909DQG/\n",
    "Problems with ENCSR527MGL /libraries/ENCLB909DQG/\n",
    "Problems with ENCSR527MGL /libraries/ENCLB909DQG/\n",
    "Problems with ENCSR662VTU /libraries/ENCLB615BOP/\n",
    "Problems with ENCSR662VTU /libraries/ENCLB615BOP/\n",
    "Problems with ENCSR662VTU /libraries/ENCLB615BOP/\n",
    "Problems with ENCSR662VTU /libraries/ENCLB615BOP/\n",
    "Problems with ENCSR700UVN /libraries/ENCLB527WWJ/\n",
    "Problems with ENCSR700UVN /libraries/ENCLB527WWJ/\n",
    "Problems with ENCSR700UVN /libraries/ENCLB527WWJ/\n",
    "Problems with ENCSR700UVN /libraries/ENCLB527WWJ/\n",
    "Problems with ENCSR769KWM /libraries/ENCLB812SMT/\n",
    "Problems with ENCSR769KWM /libraries/ENCLB812SMT/\n",
    "Problems with ENCSR769KWM /libraries/ENCLB812SMT/\n",
    "Problems with ENCSR769KWM /libraries/ENCLB812SMT/\n",
    "Problems with ENCSR003FZY /libraries/ENCLB194KKC/\n",
    "Problems with ENCSR003FZY /libraries/ENCLB194KKC/\n",
    "Problems with ENCSR003FZY /libraries/ENCLB194KKC/\n",
    "Problems with ENCSR003FZY /libraries/ENCLB194KKC/\n",
    "Problems with ENCSR024CHH /libraries/ENCLB491BDA/\n",
    "Problems with ENCSR024CHH /libraries/ENCLB491BDA/\n",
    "Problems with ENCSR029GVP /libraries/ENCLB631JOU/\n",
    "Problems with ENCSR029GVP /libraries/ENCLB631JOU/\n",
    "Problems with ENCSR029GVP /libraries/ENCLB631JOU/\n",
    "Problems with ENCSR029GVP /libraries/ENCLB631JOU/\n",
    "Problems with ENCSR029OIF /libraries/ENCLB048POP/\n",
    "Problems with ENCSR029OIF /libraries/ENCLB048POP/\n",
    "Problems with ENCSR029OIF /libraries/ENCLB048POP/\n",
    "Problems with ENCSR029OIF /libraries/ENCLB048POP/\n",
    "Problems with ENCSR035SWS /libraries/ENCLB315XSJ/\n",
    "Problems with ENCSR035SWS /libraries/ENCLB315XSJ/\n",
    "Problems with ENCSR035SWS /libraries/ENCLB315XSJ/\n",
    "Problems with ENCSR035SWS /libraries/ENCLB315XSJ/\n",
    "Problems with ENCSR043HKT /libraries/ENCLB771NKO/\n",
    "Problems with ENCSR043HKT /libraries/ENCLB771NKO/\n",
    "Problems with ENCSR043HKT /libraries/ENCLB771NKO/\n",
    "Problems with ENCSR043HKT /libraries/ENCLB771NKO/\n",
    "Problems with ENCSR044GSN /libraries/ENCLB102VRS/\n",
    "Problems with ENCSR044GSN /libraries/ENCLB102VRS/\n",
    "Problems with ENCSR044GSN /libraries/ENCLB102VRS/\n",
    "Problems with ENCSR044GSN /libraries/ENCLB102VRS/\n",
    "Problems with ENCSR062ICV /libraries/ENCLB440ABL/\n",
    "Problems with ENCSR062ICV /libraries/ENCLB440ABL/\n",
    "Problems with ENCSR062ICV /libraries/ENCLB440ABL/\n",
    "Problems with ENCSR062ICV /libraries/ENCLB440ABL/\n",
    "Problems with ENCSR062WSS /libraries/ENCLB046ZBE/\n",
    "Problems with ENCSR062WSS /libraries/ENCLB046ZBE/\n",
    "Problems with ENCSR062WSS /libraries/ENCLB046ZBE/\n",
    "Problems with ENCSR062WSS /libraries/ENCLB046ZBE/\n",
    "Problems with ENCSR080VLK /libraries/ENCLB400DWX/\n",
    "Problems with ENCSR080VLK /libraries/ENCLB400DWX/\n",
    "Problems with ENCSR081SRK /libraries/ENCLB798BJP/\n",
    "Problems with ENCSR081SRK /libraries/ENCLB798BJP/\n",
    "Problems with ENCSR081SRK /libraries/ENCLB798BJP/\n",
    "Problems with ENCSR081SRK /libraries/ENCLB798BJP/\n",
    "Problems with ENCSR087LWL /libraries/ENCLB892ZAW/\n",
    "Problems with ENCSR087LWL /libraries/ENCLB892ZAW/\n",
    "Problems with ENCSR087LWL /libraries/ENCLB892ZAW/\n",
    "Problems with ENCSR087LWL /libraries/ENCLB892ZAW/\n",
    "Problems with ENCSR090OFY /libraries/ENCLB061KPS/\n",
    "Problems with ENCSR090OFY /libraries/ENCLB061KPS/\n",
    "Problems with ENCSR090OFY /libraries/ENCLB061KPS/\n",
    "Problems with ENCSR090OFY /libraries/ENCLB061KPS/\n",
    "Problems with ENCSR093ZTB /libraries/ENCLB988YCJ/\n",
    "Problems with ENCSR093ZTB /libraries/ENCLB988YCJ/\n",
    "Problems with ENCSR093ZTB /libraries/ENCLB988YCJ/\n",
    "Problems with ENCSR093ZTB /libraries/ENCLB988YCJ/\n",
    "Problems with ENCSR096PUV /libraries/ENCLB169FGN/\n",
    "Problems with ENCSR096PUV /libraries/ENCLB169FGN/\n",
    "Problems with ENCSR096PUV /libraries/ENCLB169FGN/\n",
    "Problems with ENCSR096PUV /libraries/ENCLB169FGN/\n",
    "Problems with ENCSR105JHX /libraries/ENCLB792DPM/\n",
    "Problems with ENCSR105JHX /libraries/ENCLB792DPM/\n",
    "Problems with ENCSR105JHX /libraries/ENCLB792DPM/\n",
    "Problems with ENCSR105JHX /libraries/ENCLB792DPM/\n",
    "Problems with ENCSR115JCE /libraries/ENCLB156LUM/\n",
    "Problems with ENCSR115JCE /libraries/ENCLB156LUM/\n",
    "Problems with ENCSR115JCE /libraries/ENCLB156LUM/\n",
    "Problems with ENCSR115JCE /libraries/ENCLB156LUM/\n",
    "Problems with ENCSR119HVP /libraries/ENCLB372GFV/\n",
    "Problems with ENCSR119HVP /libraries/ENCLB372GFV/\n",
    "Problems with ENCSR119HVP /libraries/ENCLB372GFV/\n",
    "Problems with ENCSR119HVP /libraries/ENCLB372GFV/\n",
    "Problems with ENCSR120UHU /libraries/ENCLB343QQX/\n",
    "Problems with ENCSR120UHU /libraries/ENCLB343QQX/\n",
    "Problems with ENCSR120UHU /libraries/ENCLB343QQX/\n",
    "Problems with ENCSR120UHU /libraries/ENCLB343QQX/\n",
    "Problems with ENCSR126YAO /libraries/ENCLB848YMF/\n",
    "Problems with ENCSR126YAO /libraries/ENCLB848YMF/\n",
    "Problems with ENCSR126YAO /libraries/ENCLB848YMF/\n",
    "Problems with ENCSR126YAO /libraries/ENCLB848YMF/\n",
    "Problems with ENCSR135ELG /libraries/ENCLB722NFY/\n",
    "Problems with ENCSR135ELG /libraries/ENCLB722NFY/\n",
    "Problems with ENCSR145DBY /libraries/ENCLB230VGI/\n",
    "Problems with ENCSR145DBY /libraries/ENCLB230VGI/\n",
    "Problems with ENCSR146AFJ /libraries/ENCLB323HIM/\n",
    "Problems with ENCSR146AFJ /libraries/ENCLB323HIM/\n",
    "Problems with ENCSR146AFJ /libraries/ENCLB323HIM/\n",
    "Problems with ENCSR146AFJ /libraries/ENCLB323HIM/\n",
    "Problems with ENCSR147UYS /libraries/ENCLB564RVH/\n",
    "Problems with ENCSR147UYS /libraries/ENCLB564RVH/\n",
    "Problems with ENCSR147UYS /libraries/ENCLB564RVH/\n",
    "Problems with ENCSR147UYS /libraries/ENCLB564RVH/\n",
    "Problems with ENCSR156BQX /libraries/ENCLB894TFF/\n",
    "Problems with ENCSR156BQX /libraries/ENCLB894TFF/\n",
    "Problems with ENCSR156BQX /libraries/ENCLB894TFF/\n",
    "Problems with ENCSR156BQX /libraries/ENCLB894TFF/\n",
    "Problems with ENCSR162LWP /libraries/ENCLB672PSR/\n",
    "Problems with ENCSR162LWP /libraries/ENCLB672PSR/\n",
    "Problems with ENCSR162LWP /libraries/ENCLB672PSR/\n",
    "Problems with ENCSR162LWP /libraries/ENCLB672PSR/\n",
    "Problems with ENCSR164REI /libraries/ENCLB249TPV/\n",
    "Problems with ENCSR164REI /libraries/ENCLB249TPV/\n",
    "Problems with ENCSR164REI /libraries/ENCLB249TPV/\n",
    "Problems with ENCSR164REI /libraries/ENCLB249TPV/\n",
    "Problems with ENCSR165CAC /libraries/ENCLB323CCI/\n",
    "Problems with ENCSR165CAC /libraries/ENCLB323CCI/\n",
    "Problems with ENCSR169QXA /libraries/ENCLB213LKC/\n",
    "Problems with ENCSR169QXA /libraries/ENCLB213LKC/\n",
    "Problems with ENCSR169QXA /libraries/ENCLB213LKC/\n",
    "Problems with ENCSR169QXA /libraries/ENCLB213LKC/\n",
    "Problems with ENCSR171BEG /libraries/ENCLB143IPZ/\n",
    "Problems with ENCSR171BEG /libraries/ENCLB143IPZ/\n",
    "Problems with ENCSR171BEG /libraries/ENCLB143IPZ/\n",
    "Problems with ENCSR171BEG /libraries/ENCLB143IPZ/\n",
    "Problems with ENCSR172XWR /libraries/ENCLB820VCW/\n",
    "Problems with ENCSR172XWR /libraries/ENCLB820VCW/\n",
    "Problems with ENCSR172XWR /libraries/ENCLB820VCW/\n",
    "Problems with ENCSR172XWR /libraries/ENCLB820VCW/\n",
    "Problems with ENCSR178WRO /libraries/ENCLB829GDB/\n",
    "Problems with ENCSR178WRO /libraries/ENCLB829GDB/\n",
    "Problems with ENCSR186OUD /libraries/ENCLB072RWF/\n",
    "Problems with ENCSR186OUD /libraries/ENCLB072RWF/\n",
    "Problems with ENCSR186OUD /libraries/ENCLB072RWF/\n",
    "Problems with ENCSR186OUD /libraries/ENCLB072RWF/\n",
    "Problems with ENCSR186TDH /libraries/ENCLB716MVI/\n",
    "Problems with ENCSR186TDH /libraries/ENCLB716MVI/\n",
    "Problems with ENCSR186TDH /libraries/ENCLB716MVI/\n",
    "Problems with ENCSR186TDH /libraries/ENCLB716MVI/\n",
    "Problems with ENCSR188BZZ /libraries/ENCLB603YRZ/\n",
    "Problems with ENCSR188BZZ /libraries/ENCLB603YRZ/\n",
    "Problems with ENCSR188BZZ /libraries/ENCLB603YRZ/\n",
    "Problems with ENCSR188BZZ /libraries/ENCLB603YRZ/\n",
    "Problems with ENCSR191YNY /libraries/ENCLB266ZRP/\n",
    "Problems with ENCSR191YNY /libraries/ENCLB266ZRP/\n",
    "Problems with ENCSR192TYH /libraries/ENCLB925LWZ/\n",
    "Problems with ENCSR192TYH /libraries/ENCLB925LWZ/\n",
    "Problems with ENCSR192TYH /libraries/ENCLB925LWZ/\n",
    "Problems with ENCSR192TYH /libraries/ENCLB925LWZ/\n",
    "Problems with ENCSR193PLQ /libraries/ENCLB082IOS/\n",
    "Problems with ENCSR193PLQ /libraries/ENCLB082IOS/\n",
    "Problems with ENCSR193PLQ /libraries/ENCLB082IOS/\n",
    "Problems with ENCSR193PLQ /libraries/ENCLB082IOS/\n",
    "Problems with ENCSR204FUI /libraries/ENCLB072YJA/\n",
    "Problems with ENCSR204FUI /libraries/ENCLB072YJA/\n",
    "Problems with ENCSR204FUI /libraries/ENCLB072YJA/\n",
    "Problems with ENCSR204FUI /libraries/ENCLB072YJA/\n",
    "Problems with ENCSR226SBC /libraries/ENCLB889CHC/\n",
    "Problems with ENCSR226SBC /libraries/ENCLB889CHC/\n",
    "Problems with ENCSR226SBC /libraries/ENCLB889CHC/\n",
    "Problems with ENCSR226SBC /libraries/ENCLB889CHC/\n",
    "Problems with ENCSR227CPT /libraries/ENCLB793OLP/\n",
    "Problems with ENCSR227CPT /libraries/ENCLB793OLP/\n",
    "Problems with ENCSR227CPT /libraries/ENCLB793OLP/\n",
    "Problems with ENCSR227CPT /libraries/ENCLB793OLP/\n",
    "Problems with ENCSR236MGX /libraries/ENCLB692PHT/\n",
    "Problems with ENCSR236MGX /libraries/ENCLB692PHT/\n",
    "Problems with ENCSR236MGX /libraries/ENCLB692PHT/\n",
    "Problems with ENCSR236MGX /libraries/ENCLB692PHT/\n",
    "Problems with ENCSR237EZS /libraries/ENCLB010VHV/\n",
    "Problems with ENCSR237EZS /libraries/ENCLB010VHV/\n",
    "Problems with ENCSR237EZS /libraries/ENCLB010VHV/\n",
    "Problems with ENCSR237EZS /libraries/ENCLB010VHV/\n",
    "Problems with ENCSR240CYU /libraries/ENCLB601XFB/\n",
    "Problems with ENCSR240CYU /libraries/ENCLB601XFB/\n",
    "Problems with ENCSR240CYU /libraries/ENCLB601XFB/\n",
    "Problems with ENCSR240CYU /libraries/ENCLB601XFB/\n",
    "Problems with ENCSR244FME /libraries/ENCLB590EJB/\n",
    "Problems with ENCSR244FME /libraries/ENCLB590EJB/\n",
    "Problems with ENCSR244FME /libraries/ENCLB590EJB/\n",
    "Problems with ENCSR244FME /libraries/ENCLB590EJB/\n",
    "Problems with ENCSR248IOH /libraries/ENCLB570NOI/\n",
    "Problems with ENCSR248IOH /libraries/ENCLB570NOI/\n",
    "Problems with ENCSR248IOH /libraries/ENCLB570NOI/\n",
    "Problems with ENCSR248IOH /libraries/ENCLB570NOI/\n",
    "Problems with ENCSR250ZVV /libraries/ENCLB274ZBT/\n",
    "Problems with ENCSR250ZVV /libraries/ENCLB274ZBT/\n",
    "Problems with ENCSR250ZVV /libraries/ENCLB274ZBT/\n",
    "Problems with ENCSR250ZVV /libraries/ENCLB274ZBT/\n",
    "Problems with ENCSR266HBP /libraries/ENCLB931EHJ/\n",
    "Problems with ENCSR266HBP /libraries/ENCLB931EHJ/\n",
    "Problems with ENCSR266HBP /libraries/ENCLB931EHJ/\n",
    "Problems with ENCSR266HBP /libraries/ENCLB931EHJ/\n",
    "Problems with ENCSR268TST /libraries/ENCLB548SZL/\n",
    "Problems with ENCSR268TST /libraries/ENCLB548SZL/\n",
    "Problems with ENCSR284OLV /libraries/ENCLB164ZHY/\n",
    "Problems with ENCSR284OLV /libraries/ENCLB164ZHY/\n",
    "Problems with ENCSR284OLV /libraries/ENCLB164ZHY/\n",
    "Problems with ENCSR284OLV /libraries/ENCLB164ZHY/\n",
    "Problems with ENCSR287VRB /libraries/ENCLB544SEN/\n",
    "Problems with ENCSR287VRB /libraries/ENCLB544SEN/\n",
    "Problems with ENCSR287VRB /libraries/ENCLB544SEN/\n",
    "Problems with ENCSR287VRB /libraries/ENCLB544SEN/\n",
    "Problems with ENCSR299DGH /libraries/ENCLB773PPF/\n",
    "Problems with ENCSR299DGH /libraries/ENCLB773PPF/\n",
    "Problems with ENCSR299DGH /libraries/ENCLB773PPF/\n",
    "Problems with ENCSR299DGH /libraries/ENCLB773PPF/\n",
    "Problems with ENCSR306ZET /libraries/ENCLB018ZYA/\n",
    "Problems with ENCSR306ZET /libraries/ENCLB018ZYA/\n",
    "Problems with ENCSR306ZET /libraries/ENCLB018ZYA/\n",
    "Problems with ENCSR306ZET /libraries/ENCLB018ZYA/\n",
    "Problems with ENCSR311TRI /libraries/ENCLB321YXH/\n",
    "Problems with ENCSR311TRI /libraries/ENCLB321YXH/\n",
    "Problems with ENCSR311TRI /libraries/ENCLB321YXH/\n",
    "Problems with ENCSR311TRI /libraries/ENCLB321YXH/\n",
    "Problems with ENCSR321BZA /libraries/ENCLB715WPG/\n",
    "Problems with ENCSR321BZA /libraries/ENCLB715WPG/\n",
    "Problems with ENCSR322KUB /libraries/ENCLB908NNT/\n",
    "Problems with ENCSR322KUB /libraries/ENCLB908NNT/\n",
    "Problems with ENCSR322KUB /libraries/ENCLB908NNT/\n",
    "Problems with ENCSR322KUB /libraries/ENCLB908NNT/\n",
    "Problems with ENCSR327PSD /libraries/ENCLB919KOR/\n",
    "Problems with ENCSR327PSD /libraries/ENCLB919KOR/\n",
    "Problems with ENCSR331DYV /libraries/ENCLB905KRE/\n",
    "Problems with ENCSR331DYV /libraries/ENCLB905KRE/\n",
    "Problems with ENCSR331DYV /libraries/ENCLB905KRE/\n",
    "Problems with ENCSR331DYV /libraries/ENCLB905KRE/\n",
    "Problems with ENCSR331KXG /libraries/ENCLB661DWA/\n",
    "Problems with ENCSR331KXG /libraries/ENCLB661DWA/\n",
    "Problems with ENCSR334YXW /libraries/ENCLB457VVB/\n",
    "Problems with ENCSR334YXW /libraries/ENCLB457VVB/\n",
    "Problems with ENCSR345BZU /libraries/ENCLB211BVF/\n",
    "Problems with ENCSR345BZU /libraries/ENCLB211BVF/\n",
    "Problems with ENCSR356NFK /libraries/ENCLB797XAC/\n",
    "Problems with ENCSR356NFK /libraries/ENCLB797XAC/\n",
    "Problems with ENCSR356NFK /libraries/ENCLB797XAC/\n",
    "Problems with ENCSR356NFK /libraries/ENCLB797XAC/\n",
    "Problems with ENCSR356PVK /libraries/ENCLB604HIX/\n",
    "Problems with ENCSR356PVK /libraries/ENCLB604HIX/\n",
    "Problems with ENCSR356PVK /libraries/ENCLB604HIX/\n",
    "Problems with ENCSR356PVK /libraries/ENCLB604HIX/\n",
    "Problems with ENCSR358XDG /libraries/ENCLB239BXF/\n",
    "Problems with ENCSR358XDG /libraries/ENCLB239BXF/\n",
    "Problems with ENCSR358XDG /libraries/ENCLB239BXF/\n",
    "Problems with ENCSR358XDG /libraries/ENCLB239BXF/\n",
    "Problems with ENCSR366MPW /libraries/ENCLB129SKD/\n",
    "Problems with ENCSR366MPW /libraries/ENCLB129SKD/\n",
    "Problems with ENCSR366MPW /libraries/ENCLB129SKD/\n",
    "Problems with ENCSR366MPW /libraries/ENCLB129SKD/\n",
    "Problems with ENCSR382ZGH /libraries/ENCLB988KOO/\n",
    "Problems with ENCSR382ZGH /libraries/ENCLB988KOO/\n",
    "Problems with ENCSR382ZGH /libraries/ENCLB988KOO/\n",
    "Problems with ENCSR382ZGH /libraries/ENCLB988KOO/\n",
    "Problems with ENCSR388RND /libraries/ENCLB943OZA/\n",
    "Problems with ENCSR388RND /libraries/ENCLB943OZA/\n",
    "Problems with ENCSR388RND /libraries/ENCLB943OZA/\n",
    "Problems with ENCSR388RND /libraries/ENCLB943OZA/\n",
    "Problems with ENCSR389TZP /libraries/ENCLB306NVG/\n",
    "Problems with ENCSR389TZP /libraries/ENCLB306NVG/\n",
    "Problems with ENCSR389TZP /libraries/ENCLB306NVG/\n",
    "Problems with ENCSR389TZP /libraries/ENCLB306NVG/\n",
    "Problems with ENCSR390GEG /libraries/ENCLB690LBQ/\n",
    "Problems with ENCSR390GEG /libraries/ENCLB690LBQ/\n",
    "Problems with ENCSR390GEG /libraries/ENCLB690LBQ/\n",
    "Problems with ENCSR390GEG /libraries/ENCLB690LBQ/\n",
    "Problems with ENCSR400PBH /libraries/ENCLB010FKE/\n",
    "Problems with ENCSR400PBH /libraries/ENCLB010FKE/\n",
    "Problems with ENCSR400PBH /libraries/ENCLB010FKE/\n",
    "Problems with ENCSR400PBH /libraries/ENCLB010FKE/\n",
    "Problems with ENCSR405CPX /libraries/ENCLB614YSB/\n",
    "Problems with ENCSR405CPX /libraries/ENCLB614YSB/\n",
    "Problems with ENCSR405CPX /libraries/ENCLB614YSB/\n",
    "Problems with ENCSR405CPX /libraries/ENCLB614YSB/\n",
    "Problems with ENCSR405RFP /libraries/ENCLB261MTS/\n",
    "Problems with ENCSR405RFP /libraries/ENCLB261MTS/\n",
    "Problems with ENCSR405RFP /libraries/ENCLB261MTS/\n",
    "Problems with ENCSR405RFP /libraries/ENCLB261MTS/\n",
    "Problems with ENCSR408PZM /libraries/ENCLB336ZNO/\n",
    "Problems with ENCSR408PZM /libraries/ENCLB336ZNO/\n",
    "Problems with ENCSR408PZM /libraries/ENCLB336ZNO/\n",
    "Problems with ENCSR408PZM /libraries/ENCLB336ZNO/\n",
    "Problems with ENCSR408ZUR /libraries/ENCLB382NBT/\n",
    "Problems with ENCSR408ZUR /libraries/ENCLB382NBT/\n",
    "Problems with ENCSR408ZUR /libraries/ENCLB382NBT/\n",
    "Problems with ENCSR408ZUR /libraries/ENCLB382NBT/\n",
    "Problems with ENCSR415CYS /libraries/ENCLB867BPN/\n",
    "Problems with ENCSR415CYS /libraries/ENCLB867BPN/\n",
    "Problems with ENCSR415CYS /libraries/ENCLB867BPN/\n",
    "Problems with ENCSR415CYS /libraries/ENCLB867BPN/\n",
    "Problems with ENCSR421VUT /libraries/ENCLB673SUG/\n",
    "Problems with ENCSR421VUT /libraries/ENCLB673SUG/\n",
    "Problems with ENCSR421VUT /libraries/ENCLB673SUG/\n",
    "Problems with ENCSR421VUT /libraries/ENCLB673SUG/\n",
    "Problems with ENCSR423AXI /libraries/ENCLB489WXX/\n",
    "Problems with ENCSR423AXI /libraries/ENCLB489WXX/\n",
    "Problems with ENCSR423AXI /libraries/ENCLB489WXX/\n",
    "Problems with ENCSR423AXI /libraries/ENCLB489WXX/\n",
    "Problems with ENCSR423CNV /libraries/ENCLB622RLX/\n",
    "Problems with ENCSR423CNV /libraries/ENCLB622RLX/\n",
    "Problems with ENCSR423CNV /libraries/ENCLB622RLX/\n",
    "Problems with ENCSR423CNV /libraries/ENCLB622RLX/\n",
    "Problems with ENCSR423LFS /libraries/ENCLB516LRK/\n",
    "Problems with ENCSR423LFS /libraries/ENCLB516LRK/\n",
    "Problems with ENCSR423LFS /libraries/ENCLB516LRK/\n",
    "Problems with ENCSR423LFS /libraries/ENCLB516LRK/\n",
    "Problems with ENCSR424RYS /libraries/ENCLB228MHF/\n",
    "Problems with ENCSR424RYS /libraries/ENCLB228MHF/\n",
    "Problems with ENCSR424RYS /libraries/ENCLB228MHF/\n",
    "Problems with ENCSR424RYS /libraries/ENCLB228MHF/\n",
    "Problems with ENCSR426MLQ /libraries/ENCLB368QMP/\n",
    "Problems with ENCSR426MLQ /libraries/ENCLB368QMP/\n",
    "Problems with ENCSR427OIY /libraries/ENCLB319YUS/\n",
    "Problems with ENCSR427OIY /libraries/ENCLB319YUS/\n",
    "Problems with ENCSR427OIY /libraries/ENCLB319YUS/\n",
    "Problems with ENCSR427OIY /libraries/ENCLB319YUS/\n",
    "Problems with ENCSR429ZKT /libraries/ENCLB523JWN/\n",
    "Problems with ENCSR429ZKT /libraries/ENCLB523JWN/\n",
    "Problems with ENCSR429ZKT /libraries/ENCLB523JWN/\n",
    "Problems with ENCSR429ZKT /libraries/ENCLB523JWN/\n",
    "Problems with ENCSR441YBL /libraries/ENCLB903BTS/\n",
    "Problems with ENCSR441YBL /libraries/ENCLB903BTS/\n",
    "Problems with ENCSR441YBL /libraries/ENCLB903BTS/\n",
    "Problems with ENCSR441YBL /libraries/ENCLB903BTS/\n",
    "Problems with ENCSR462YAG /libraries/ENCLB954TLZ/\n",
    "Problems with ENCSR462YAG /libraries/ENCLB954TLZ/\n",
    "Problems with ENCSR462YAG /libraries/ENCLB954TLZ/\n",
    "Problems with ENCSR462YAG /libraries/ENCLB954TLZ/\n",
    "Problems with ENCSR470GMO /libraries/ENCLB592ZMI/\n",
    "Problems with ENCSR470GMO /libraries/ENCLB592ZMI/\n",
    "Problems with ENCSR470GMO /libraries/ENCLB592ZMI/\n",
    "Problems with ENCSR470GMO /libraries/ENCLB592ZMI/\n",
    "Problems with ENCSR477EIC /libraries/ENCLB983QCI/\n",
    "Problems with ENCSR477EIC /libraries/ENCLB983QCI/\n",
    "Problems with ENCSR488RZS /libraries/ENCLB989CJP/\n",
    "Problems with ENCSR488RZS /libraries/ENCLB989CJP/\n",
    "Problems with ENCSR488RZS /libraries/ENCLB989CJP/\n",
    "Problems with ENCSR488RZS /libraries/ENCLB989CJP/\n",
    "Problems with ENCSR489KPP /libraries/ENCLB399LDV/\n",
    "Problems with ENCSR489KPP /libraries/ENCLB399LDV/\n",
    "Problems with ENCSR489KPP /libraries/ENCLB399LDV/\n",
    "Problems with ENCSR489KPP /libraries/ENCLB399LDV/\n",
    "Problems with ENCSR493ZQU /libraries/ENCLB862YMG/\n",
    "Problems with ENCSR493ZQU /libraries/ENCLB862YMG/\n",
    "Problems with ENCSR493ZQU /libraries/ENCLB862YMG/\n",
    "Problems with ENCSR493ZQU /libraries/ENCLB862YMG/\n",
    "Problems with ENCSR498KBD /libraries/ENCLB188WUM/\n",
    "Problems with ENCSR498KBD /libraries/ENCLB188WUM/\n",
    "Problems with ENCSR510HAO /libraries/ENCLB805MAL/\n",
    "Problems with ENCSR510HAO /libraries/ENCLB805MAL/\n",
    "Problems with ENCSR510HAO /libraries/ENCLB805MAL/\n",
    "Problems with ENCSR510HAO /libraries/ENCLB805MAL/\n",
    "Problems with ENCSR511XOL /libraries/ENCLB460OEF/\n",
    "Problems with ENCSR511XOL /libraries/ENCLB460OEF/\n",
    "Problems with ENCSR511XOL /libraries/ENCLB460OEF/\n",
    "Problems with ENCSR511XOL /libraries/ENCLB460OEF/\n",
    "Problems with ENCSR519HYQ /libraries/ENCLB369KJQ/\n",
    "Problems with ENCSR519HYQ /libraries/ENCLB369KJQ/\n",
    "Problems with ENCSR519HYQ /libraries/ENCLB369KJQ/\n",
    "Problems with ENCSR519HYQ /libraries/ENCLB369KJQ/\n",
    "Problems with ENCSR526WJT /libraries/ENCLB656UZY/\n",
    "Problems with ENCSR526WJT /libraries/ENCLB656UZY/\n",
    "Problems with ENCSR526WJT /libraries/ENCLB656UZY/\n",
    "Problems with ENCSR526WJT /libraries/ENCLB656UZY/\n",
    "Problems with ENCSR536EAF /libraries/ENCLB891WFR/\n",
    "Problems with ENCSR536EAF /libraries/ENCLB891WFR/\n",
    "Problems with ENCSR536EAF /libraries/ENCLB891WFR/\n",
    "Problems with ENCSR536EAF /libraries/ENCLB891WFR/\n",
    "Problems with ENCSR542LRP /libraries/ENCLB700LWU/\n",
    "Problems with ENCSR542LRP /libraries/ENCLB700LWU/\n",
    "Problems with ENCSR555ATU /libraries/ENCLB908IUA/\n",
    "Problems with ENCSR555ATU /libraries/ENCLB908IUA/\n",
    "Problems with ENCSR560WNI /libraries/ENCLB354YXN/\n",
    "Problems with ENCSR560WNI /libraries/ENCLB354YXN/\n",
    "Problems with ENCSR560WNI /libraries/ENCLB354YXN/\n",
    "Problems with ENCSR560WNI /libraries/ENCLB354YXN/\n",
    "Problems with ENCSR561PWI /libraries/ENCLB785FXH/\n",
    "Problems with ENCSR561PWI /libraries/ENCLB785FXH/\n",
    "Problems with ENCSR563WPL /libraries/ENCLB782BWD/\n",
    "Problems with ENCSR563WPL /libraries/ENCLB782BWD/\n",
    "Problems with ENCSR563WPL /libraries/ENCLB782BWD/\n",
    "Problems with ENCSR563WPL /libraries/ENCLB782BWD/\n",
    "Problems with ENCSR566XHU /libraries/ENCLB416NMR/\n",
    "Problems with ENCSR566XHU /libraries/ENCLB416NMR/\n",
    "Problems with ENCSR576RBX /libraries/ENCLB902RKO/\n",
    "Problems with ENCSR576RBX /libraries/ENCLB902RKO/\n",
    "Problems with ENCSR586LMA /libraries/ENCLB332MGL/\n",
    "Problems with ENCSR586LMA /libraries/ENCLB332MGL/\n",
    "Problems with ENCSR586LMA /libraries/ENCLB332MGL/\n",
    "Problems with ENCSR586LMA /libraries/ENCLB332MGL/\n",
    "Problems with ENCSR595LVS /libraries/ENCLB678GYJ/\n",
    "Problems with ENCSR595LVS /libraries/ENCLB678GYJ/\n",
    "Problems with ENCSR595LVS /libraries/ENCLB678GYJ/\n",
    "Problems with ENCSR595LVS /libraries/ENCLB678GYJ/\n",
    "Problems with ENCSR601CQM /libraries/ENCLB524IGV/\n",
    "Problems with ENCSR601CQM /libraries/ENCLB524IGV/\n",
    "Problems with ENCSR601CQM /libraries/ENCLB524IGV/\n",
    "Problems with ENCSR601CQM /libraries/ENCLB524IGV/\n",
    "Problems with ENCSR635GFL /libraries/ENCLB589IWN/\n",
    "Problems with ENCSR635GFL /libraries/ENCLB589IWN/\n",
    "Problems with ENCSR635GFL /libraries/ENCLB589IWN/\n",
    "Problems with ENCSR635GFL /libraries/ENCLB589IWN/\n",
    "Problems with ENCSR635GRA /libraries/ENCLB250XVZ/\n",
    "Problems with ENCSR635GRA /libraries/ENCLB250XVZ/\n",
    "Problems with ENCSR635GRA /libraries/ENCLB250XVZ/\n",
    "Problems with ENCSR635GRA /libraries/ENCLB250XVZ/\n",
    "Problems with ENCSR637AZA /libraries/ENCLB581NBT/\n",
    "Problems with ENCSR637AZA /libraries/ENCLB581NBT/\n",
    "Problems with ENCSR637AZA /libraries/ENCLB581NBT/\n",
    "Problems with ENCSR637AZA /libraries/ENCLB581NBT/\n",
    "Problems with ENCSR642TKQ /libraries/ENCLB765ZDM/\n",
    "Problems with ENCSR642TKQ /libraries/ENCLB765ZDM/\n",
    "Problems with ENCSR644EUQ /libraries/ENCLB146JLP/\n",
    "Problems with ENCSR644EUQ /libraries/ENCLB146JLP/\n",
    "Problems with ENCSR644YWO /libraries/ENCLB517TXV/\n",
    "Problems with ENCSR644YWO /libraries/ENCLB517TXV/\n",
    "Problems with ENCSR644YWO /libraries/ENCLB517TXV/\n",
    "Problems with ENCSR644YWO /libraries/ENCLB517TXV/\n",
    "Problems with ENCSR651QGH /libraries/ENCLB308YIM/\n",
    "Problems with ENCSR651QGH /libraries/ENCLB308YIM/\n",
    "Problems with ENCSR651QGH /libraries/ENCLB308YIM/\n",
    "Problems with ENCSR651QGH /libraries/ENCLB308YIM/\n",
    "Problems with ENCSR671RJG /libraries/ENCLB885OKL/\n",
    "Problems with ENCSR671RJG /libraries/ENCLB885OKL/\n",
    "Problems with ENCSR671RJG /libraries/ENCLB885OKL/\n",
    "Problems with ENCSR671RJG /libraries/ENCLB885OKL/\n",
    "Problems with ENCSR692OZP /libraries/ENCLB194LYE/\n",
    "Problems with ENCSR692OZP /libraries/ENCLB194LYE/\n",
    "Problems with ENCSR692OZP /libraries/ENCLB194LYE/\n",
    "Problems with ENCSR692OZP /libraries/ENCLB194LYE/\n",
    "Problems with ENCSR692ZWB /libraries/ENCLB598TEK/\n",
    "Problems with ENCSR692ZWB /libraries/ENCLB598TEK/\n",
    "Problems with ENCSR692ZWB /libraries/ENCLB598TEK/\n",
    "Problems with ENCSR692ZWB /libraries/ENCLB598TEK/\n",
    "Problems with ENCSR694DQP /libraries/ENCLB170MZP/\n",
    "Problems with ENCSR694DQP /libraries/ENCLB170MZP/\n",
    "Problems with ENCSR694DQP /libraries/ENCLB170MZP/\n",
    "Problems with ENCSR694DQP /libraries/ENCLB170MZP/\n",
    "Problems with ENCSR732MDS /libraries/ENCLB701VGG/\n",
    "Problems with ENCSR732MDS /libraries/ENCLB701VGG/\n",
    "Problems with ENCSR732MDS /libraries/ENCLB701VGG/\n",
    "Problems with ENCSR732MDS /libraries/ENCLB701VGG/\n",
    "Problems with ENCSR745UMG /libraries/ENCLB709WAG/\n",
    "Problems with ENCSR745UMG /libraries/ENCLB709WAG/\n",
    "Problems with ENCSR773FPF /libraries/ENCLB536TLF/\n",
    "Problems with ENCSR773FPF /libraries/ENCLB536TLF/\n",
    "Problems with ENCSR773FPF /libraries/ENCLB536TLF/\n",
    "Problems with ENCSR773FPF /libraries/ENCLB536TLF/\n",
    "Problems with ENCSR774KTU /libraries/ENCLB727KBM/\n",
    "Problems with ENCSR774KTU /libraries/ENCLB727KBM/\n",
    "Problems with ENCSR774KTU /libraries/ENCLB727KBM/\n",
    "Problems with ENCSR774KTU /libraries/ENCLB727KBM/\n",
    "Problems with ENCSR799GLH /libraries/ENCLB899ORD/\n",
    "Problems with ENCSR799GLH /libraries/ENCLB899ORD/\n",
    "Problems with ENCSR799GLH /libraries/ENCLB899ORD/\n",
    "Problems with ENCSR799GLH /libraries/ENCLB899ORD/\n",
    "Problems with ENCSR801MJO /libraries/ENCLB449BXX/\n",
    "Problems with ENCSR801MJO /libraries/ENCLB449BXX/\n",
    "Problems with ENCSR801MJO /libraries/ENCLB449BXX/\n",
    "Problems with ENCSR801MJO /libraries/ENCLB449BXX/\n",
    "Problems with ENCSR813ULV /libraries/ENCLB963QFG/\n",
    "Problems with ENCSR813ULV /libraries/ENCLB963QFG/\n",
    "Problems with ENCSR813ULV /libraries/ENCLB963QFG/\n",
    "Problems with ENCSR813ULV /libraries/ENCLB963QFG/\n",
    "Problems with ENCSR827YHY /libraries/ENCLB389BSV/\n",
    "Problems with ENCSR827YHY /libraries/ENCLB389BSV/\n",
    "Problems with ENCSR827YHY /libraries/ENCLB389BSV/\n",
    "Problems with ENCSR827YHY /libraries/ENCLB389BSV/\n",
    "Problems with ENCSR828TLP /libraries/ENCLB611DGC/\n",
    "Problems with ENCSR828TLP /libraries/ENCLB611DGC/\n",
    "Problems with ENCSR828TLP /libraries/ENCLB611DGC/\n",
    "Problems with ENCSR828TLP /libraries/ENCLB611DGC/\n",
    "Problems with ENCSR830MSX /libraries/ENCLB231PCO/\n",
    "Problems with ENCSR830MSX /libraries/ENCLB231PCO/\n",
    "Problems with ENCSR830MSX /libraries/ENCLB231PCO/\n",
    "Problems with ENCSR830MSX /libraries/ENCLB231PCO/\n",
    "Problems with ENCSR834SCG /libraries/ENCLB849DVZ/\n",
    "Problems with ENCSR834SCG /libraries/ENCLB849DVZ/\n",
    "Problems with ENCSR834SCG /libraries/ENCLB849DVZ/\n",
    "Problems with ENCSR834SCG /libraries/ENCLB849DVZ/\n",
    "Problems with ENCSR847STS /libraries/ENCLB735WRJ/\n",
    "Problems with ENCSR847STS /libraries/ENCLB735WRJ/\n",
    "Problems with ENCSR847STS /libraries/ENCLB735WRJ/\n",
    "Problems with ENCSR847STS /libraries/ENCLB735WRJ/\n",
    "Problems with ENCSR851RNW /libraries/ENCLB349QMF/\n",
    "Problems with ENCSR851RNW /libraries/ENCLB349QMF/\n",
    "Problems with ENCSR851RNW /libraries/ENCLB349QMF/\n",
    "Problems with ENCSR851RNW /libraries/ENCLB349QMF/\n",
    "Problems with ENCSR852NQO /libraries/ENCLB299EAG/\n",
    "Problems with ENCSR852NQO /libraries/ENCLB299EAG/\n",
    "Problems with ENCSR852NQO /libraries/ENCLB299EAG/\n",
    "Problems with ENCSR852NQO /libraries/ENCLB299EAG/\n",
    "Problems with ENCSR855ZYP /libraries/ENCLB193HSH/\n",
    "Problems with ENCSR855ZYP /libraries/ENCLB193HSH/\n",
    "Problems with ENCSR855ZYP /libraries/ENCLB193HSH/\n",
    "Problems with ENCSR855ZYP /libraries/ENCLB193HSH/\n",
    "Problems with ENCSR877CFF /libraries/ENCLB227FXC/\n",
    "Problems with ENCSR877CFF /libraries/ENCLB227FXC/\n",
    "Problems with ENCSR877CFF /libraries/ENCLB227FXC/\n",
    "Problems with ENCSR877CFF /libraries/ENCLB227FXC/\n",
    "Problems with ENCSR877UAI /libraries/ENCLB747XAC/\n",
    "Problems with ENCSR877UAI /libraries/ENCLB747XAC/\n",
    "Problems with ENCSR877UAI /libraries/ENCLB747XAC/\n",
    "Problems with ENCSR877UAI /libraries/ENCLB747XAC/\n",
    "Problems with ENCSR880DYE /libraries/ENCLB977UQY/\n",
    "Problems with ENCSR880DYE /libraries/ENCLB977UQY/\n",
    "Problems with ENCSR880DYE /libraries/ENCLB977UQY/\n",
    "Problems with ENCSR880DYE /libraries/ENCLB977UQY/\n",
    "Problems with ENCSR880MVS /libraries/ENCLB586GGX/\n",
    "Problems with ENCSR880MVS /libraries/ENCLB586GGX/\n",
    "Problems with ENCSR880MVS /libraries/ENCLB586GGX/\n",
    "Problems with ENCSR880MVS /libraries/ENCLB586GGX/\n",
    "Problems with ENCSR888NNP /libraries/ENCLB639GHE/\n",
    "Problems with ENCSR888NNP /libraries/ENCLB639GHE/\n",
    "Problems with ENCSR888NNP /libraries/ENCLB639GHE/\n",
    "Problems with ENCSR888NNP /libraries/ENCLB639GHE/\n",
    "Problems with ENCSR894JVL /libraries/ENCLB868TVO/\n",
    "Problems with ENCSR894JVL /libraries/ENCLB868TVO/\n",
    "Problems with ENCSR894JVL /libraries/ENCLB868TVO/\n",
    "Problems with ENCSR894JVL /libraries/ENCLB868TVO/\n",
    "Problems with ENCSR905YQE /libraries/ENCLB822SVY/\n",
    "Problems with ENCSR905YQE /libraries/ENCLB822SVY/\n",
    "Problems with ENCSR905YQE /libraries/ENCLB822SVY/\n",
    "Problems with ENCSR905YQE /libraries/ENCLB822SVY/\n",
    "Problems with ENCSR912IHV /libraries/ENCLB849GNT/\n",
    "Problems with ENCSR912IHV /libraries/ENCLB849GNT/\n",
    "Problems with ENCSR912IHV /libraries/ENCLB849GNT/\n",
    "Problems with ENCSR912IHV /libraries/ENCLB849GNT/\n",
    "Problems with ENCSR915UNZ /libraries/ENCLB151ZAG/\n",
    "Problems with ENCSR915UNZ /libraries/ENCLB151ZAG/\n",
    "Problems with ENCSR915UNZ /libraries/ENCLB151ZAG/\n",
    "Problems with ENCSR915UNZ /libraries/ENCLB151ZAG/\n",
    "Problems with ENCSR923NSW /libraries/ENCLB143DFU/\n",
    "Problems with ENCSR923NSW /libraries/ENCLB143DFU/\n",
    "Problems with ENCSR923NSW /libraries/ENCLB143DFU/\n",
    "Problems with ENCSR923NSW /libraries/ENCLB143DFU/\n",
    "Problems with ENCSR929PEE /libraries/ENCLB959AIF/\n",
    "Problems with ENCSR929PEE /libraries/ENCLB959AIF/\n",
    "Problems with ENCSR932DTM /libraries/ENCLB905AUU/\n",
    "Problems with ENCSR932DTM /libraries/ENCLB905AUU/\n",
    "Problems with ENCSR932DTM /libraries/ENCLB905AUU/\n",
    "Problems with ENCSR932DTM /libraries/ENCLB905AUU/\n",
    "Problems with ENCSR948XEM /libraries/ENCLB315FQV/\n",
    "Problems with ENCSR948XEM /libraries/ENCLB315FQV/\n",
    "Problems with ENCSR962DRE /libraries/ENCLB697JVD/\n",
    "Problems with ENCSR962DRE /libraries/ENCLB697JVD/\n",
    "Problems with ENCSR962DRE /libraries/ENCLB697JVD/\n",
    "Problems with ENCSR962DRE /libraries/ENCLB697JVD/\n",
    "Problems with ENCSR969VQQ /libraries/ENCLB170FEN/\n",
    "Problems with ENCSR969VQQ /libraries/ENCLB170FEN/\n",
    "Problems with ENCSR972KJU /libraries/ENCLB762GOP/\n",
    "Problems with ENCSR972KJU /libraries/ENCLB762GOP/\n",
    "Problems with ENCSR972KJU /libraries/ENCLB762GOP/\n",
    "Problems with ENCSR972KJU /libraries/ENCLB762GOP/\n",
    "Problems with ENCSR973ZJO /libraries/ENCLB374JXM/\n",
    "Problems with ENCSR973ZJO /libraries/ENCLB374JXM/\n",
    "Problems with ENCSR973ZJO /libraries/ENCLB374JXM/\n",
    "Problems with ENCSR973ZJO /libraries/ENCLB374JXM/\n",
    "Problems with ENCSR974VTR /libraries/ENCLB415PDV/\n",
    "Problems with ENCSR974VTR /libraries/ENCLB415PDV/\n",
    "Problems with ENCSR974VTR /libraries/ENCLB415PDV/\n",
    "Problems with ENCSR974VTR /libraries/ENCLB415PDV/\n",
    "Problems with ENCSR976FIL /libraries/ENCLB124EIT/\n",
    "Problems with ENCSR976FIL /libraries/ENCLB124EIT/\n",
    "Problems with ENCSR976FIL /libraries/ENCLB124EIT/\n",
    "Problems with ENCSR976FIL /libraries/ENCLB124EIT/\n",
    "Problems with ENCSR983WDT /libraries/ENCLB708YOS/\n",
    "Problems with ENCSR983WDT /libraries/ENCLB708YOS/\n",
    "Problems with ENCSR983WDT /libraries/ENCLB708YOS/\n",
    "Problems with ENCSR983WDT /libraries/ENCLB708YOS/\n",
    "Problems with ENCSR985FZQ /libraries/ENCLB292IDD/\n",
    "Problems with ENCSR985FZQ /libraries/ENCLB292IDD/\n",
    "Problems with ENCSR985FZQ /libraries/ENCLB292IDD/\n",
    "Problems with ENCSR985FZQ /libraries/ENCLB292IDD/\n",
    "Problems with ENCSR986QTH /libraries/ENCLB571XKG/\n",
    "Problems with ENCSR986QTH /libraries/ENCLB571XKG/\n",
    "Problems with ENCSR109DAN /libraries/ENCLB527JDU/\n",
    "Problems with ENCSR109DAN /libraries/ENCLB527JDU/\n",
    "Problems with ENCSR388EIN /libraries/ENCLB952MZJ/\n",
    "Problems with ENCSR388EIN /libraries/ENCLB952MZJ/\n",
    "Problems with ENCSR452OYH /libraries/ENCLB294FBZ/\n",
    "Problems with ENCSR452OYH /libraries/ENCLB294FBZ/\n",
    "Problems with ENCSR503QZJ /libraries/ENCLB129LMS/\n",
    "Problems with ENCSR503QZJ /libraries/ENCLB129LMS/\n",
    "Problems with ENCSR572SMD /libraries/ENCLB273ZWE/\n",
    "Problems with ENCSR572SMD /libraries/ENCLB273ZWE/\n",
    "Problems with ENCSR877NAX /libraries/ENCLB514RKT/\n",
    "Problems with ENCSR877NAX /libraries/ENCLB514RKT/\"\"\".split(\"\\n\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "collective-bundle",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "169"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(problems)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "patient-jason",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ENCSR311TRI',\n",
       " 'ENCSR426MLQ',\n",
       " 'ENCSR172XWR',\n",
       " 'ENCSR062WSS',\n",
       " 'ENCSR390GEG',\n",
       " 'ENCSR192TYH',\n",
       " 'ENCSR560WNI',\n",
       " 'ENCSR932DTM',\n",
       " 'ENCSR830MSX',\n",
       " 'ENCSR135ELG',\n",
       " 'ENCSR266HBP',\n",
       " 'ENCSR421VUT',\n",
       " 'ENCSR204FUI',\n",
       " 'ENCSR236MGX',\n",
       " 'ENCSR035SWS',\n",
       " 'ENCSR644EUQ',\n",
       " 'ENCSR694DQP',\n",
       " 'ENCSR477EIC',\n",
       " 'ENCSR193PLQ',\n",
       " 'ENCSR165CAC',\n",
       " 'ENCSR334YXW',\n",
       " 'ENCSR356NFK',\n",
       " 'ENCSR595LVS',\n",
       " 'ENCSR635GRA',\n",
       " 'ENCSR813ULV',\n",
       " 'ENCSR093ZTB',\n",
       " 'ENCSR503QZJ',\n",
       " 'ENCSR178WRO',\n",
       " 'ENCSR322KUB',\n",
       " 'ENCSR327PSD',\n",
       " 'ENCSR358XDG',\n",
       " 'ENCSR635GFL',\n",
       " 'ENCSR519HYQ',\n",
       " 'ENCSR915UNZ',\n",
       " 'ENCSR888NNP',\n",
       " 'ENCSR226SBC',\n",
       " 'ENCSR146AFJ',\n",
       " 'ENCSR164REI',\n",
       " 'ENCSR389TZP',\n",
       " 'ENCSR566XHU',\n",
       " 'ENCSR877CFF',\n",
       " 'ENCSR306ZET',\n",
       " 'ENCSR969VQQ',\n",
       " 'ENCSR250ZVV',\n",
       " 'ENCSR912IHV',\n",
       " 'ENCSR126YAO',\n",
       " 'ENCSR405CPX',\n",
       " 'ENCSR489KPP',\n",
       " 'ENCSR186OUD',\n",
       " 'ENCSR452OYH',\n",
       " 'ENCSR096PUV',\n",
       " 'ENCSR488RZS',\n",
       " 'ENCSR356PVK',\n",
       " 'ENCSR298MPV',\n",
       " 'ENCSR542LRP',\n",
       " 'ENCSR923NSW',\n",
       " 'ENCSR769KWM',\n",
       " 'ENCSR852NQO',\n",
       " 'ENCSR120UHU',\n",
       " 'ENCSR572SMD',\n",
       " 'ENCSR929PEE',\n",
       " 'ENCSR894JVL',\n",
       " 'ENCSR774KTU',\n",
       " 'ENCSR080VLK',\n",
       " 'ENCSR423CNV',\n",
       " 'ENCSR003FZY',\n",
       " 'ENCSR331DYV',\n",
       " 'ENCSR692ZWB',\n",
       " 'ENCSR115JCE',\n",
       " 'ENCSR171BEG',\n",
       " 'ENCSR240CYU',\n",
       " 'ENCSR388RND',\n",
       " 'ENCSR188BZZ',\n",
       " 'ENCSR976FIL',\n",
       " 'ENCSR510HAO',\n",
       " 'ENCSR284OLV',\n",
       " 'ENCSR191YNY',\n",
       " 'ENCSR644YWO',\n",
       " 'ENCSR400PBH',\n",
       " 'ENCSR642TKQ',\n",
       " 'ENCSR268TST',\n",
       " 'ENCSR799GLH',\n",
       " 'ENCSR244FME',\n",
       " 'ENCSR555ATU',\n",
       " 'ENCSR470GMO',\n",
       " 'ENCSR081SRK',\n",
       " 'ENCSR287VRB',\n",
       " 'ENCSR855ZYP',\n",
       " 'ENCSR586LMA',\n",
       " 'ENCSR526WJT',\n",
       " 'ENCSR162LWP',\n",
       " 'ENCSR877NAX',\n",
       " 'ENCSR156BQX',\n",
       " 'ENCSR441YBL',\n",
       " 'ENCSR498KBD',\n",
       " 'ENCSR827YHY',\n",
       " 'ENCSR563WPL',\n",
       " 'ENCSR877UAI',\n",
       " 'ENCSR145DBY',\n",
       " 'ENCSR169QXA',\n",
       " 'ENCSR024CHH',\n",
       " 'ENCSR985FZQ',\n",
       " 'ENCSR405YKM',\n",
       " 'ENCSR972KJU',\n",
       " 'ENCSR248IOH',\n",
       " 'ENCSR299DGH',\n",
       " 'ENCSR345BZU',\n",
       " 'ENCSR405RFP',\n",
       " 'ENCSR423LFS',\n",
       " 'ENCSR237EZS',\n",
       " 'ENCSR062ICV',\n",
       " 'ENCSR880MVS',\n",
       " 'ENCSR983WDT',\n",
       " 'ENCSR637AZA',\n",
       " 'ENCSR692OZP',\n",
       " 'ENCSR366MPW',\n",
       " 'ENCSR044GSN',\n",
       " 'ENCSR408ZUR',\n",
       " 'ENCSR671RJG',\n",
       " 'ENCSR773FPF',\n",
       " 'ENCSR321BZA',\n",
       " 'ENCSR601CQM',\n",
       " 'ENCSR527MGL',\n",
       " 'ENCSR105JHX',\n",
       " 'ENCSR043HKT',\n",
       " 'ENCSR109DAN',\n",
       " 'ENCSR029GVP',\n",
       " 'ENCSR745UMG',\n",
       " 'ENCSR880DYE',\n",
       " 'ENCSR463EEP',\n",
       " 'ENCSR948XEM',\n",
       " 'ENCSR511XOL',\n",
       " 'ENCSR203THX',\n",
       " 'ENCSR801MJO',\n",
       " 'ENCSR662VTU',\n",
       " 'ENCSR462YAG',\n",
       " 'ENCSR962DRE',\n",
       " 'ENCSR651QGH',\n",
       " 'ENCSR408PZM',\n",
       " 'ENCSR090OFY',\n",
       " 'ENCSR147UYS',\n",
       " 'ENCSR561PWI',\n",
       " 'ENCSR331KXG',\n",
       " 'ENCSR493ZQU',\n",
       " 'ENCSR427OIY',\n",
       " 'ENCSR227CPT',\n",
       " 'ENCSR186TDH',\n",
       " 'ENCSR029OIF',\n",
       " 'ENCSR429ZKT',\n",
       " 'ENCSR415CYS',\n",
       " 'ENCSR700UVN',\n",
       " 'ENCSR536EAF',\n",
       " 'ENCSR905YQE',\n",
       " 'ENCSR847STS',\n",
       " 'ENCSR732MDS',\n",
       " 'ENCSR974VTR',\n",
       " 'ENCSR423AXI',\n",
       " 'ENCSR424RYS',\n",
       " 'ENCSR986QTH',\n",
       " 'ENCSR851RNW',\n",
       " 'ENCSR388EIN',\n",
       " 'ENCSR973ZJO',\n",
       " 'ENCSR119HVP',\n",
       " 'ENCSR337HKS',\n",
       " 'ENCSR087LWL',\n",
       " 'ENCSR834SCG',\n",
       " 'ENCSR576RBX',\n",
       " 'ENCSR828TLP',\n",
       " 'ENCSR382ZGH']"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "problem_experiments = []\n",
    "for row in problems:\n",
    "    problem_experiments.append(row.split()[2])\n",
    "problem_experiments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "mysterious-extra",
   "metadata": {},
   "outputs": [],
   "source": [
    "processed = set()\n",
    "processed.update(experiments_to_run_2_6)\n",
    "processed.update(experiments_to_run_7_21)\n",
    "processed.update(experiments_to_run_22_51)\n",
    "processed.update(experiments_to_run_52_81)\n",
    "processed.update(experiments_to_run_82_107)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "frequent-dominican",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ENCSR203THX',\n",
       " 'ENCSR298MPV',\n",
       " 'ENCSR337HKS',\n",
       " 'ENCSR405YKM',\n",
       " 'ENCSR463EEP',\n",
       " 'ENCSR527MGL',\n",
       " 'ENCSR662VTU',\n",
       " 'ENCSR700UVN',\n",
       " 'ENCSR769KWM'}"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "processed.intersection(problem_experiments)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "sapphire-ability",
   "metadata": {},
   "source": [
    "# UCI Submission"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "flexible-massachusetts",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aliases</th>\n",
       "      <th>biosample_ontology</th>\n",
       "      <th>description</th>\n",
       "      <th>assay_term_name</th>\n",
       "      <th>lab</th>\n",
       "      <th>award</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ali-mortazavi:experiment_illumina_sr_B6Cast_ad...</td>\n",
       "      <td>/biosample-types/tissue_UBERON_0002369/</td>\n",
       "      <td>Illumina short-read B6/Cast F1 adrenal P8mo ma...</td>\n",
       "      <td>single-cell RNA sequencing assay</td>\n",
       "      <td>ali-mortazavi</td>\n",
       "      <td>UM1HG009443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ali-mortazavi:experiment_illumina_sr_5xFADCast...</td>\n",
       "      <td>/biosample-types/tissue_UBERON_0000948/</td>\n",
       "      <td>Illumina short-read 5xFAD/Cast F1 heart P8mo f...</td>\n",
       "      <td>single-cell RNA sequencing assay</td>\n",
       "      <td>ali-mortazavi</td>\n",
       "      <td>UM1HG009443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ali-mortazavi:experiment_illumina_sr_5xFADCast...</td>\n",
       "      <td>/biosample-types/tissue_UBERON_0001388/</td>\n",
       "      <td>Illumina short-read 5xFAD/Cast F1 gastrocnemiu...</td>\n",
       "      <td>single-cell RNA sequencing assay</td>\n",
       "      <td>ali-mortazavi</td>\n",
       "      <td>UM1HG009443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ali-mortazavi:experiment_illumina_sr_B6Cast_ga...</td>\n",
       "      <td>/biosample-types/tissue_UBERON_0001388/</td>\n",
       "      <td>Illumina short-read B6/Cast F1 gastrocnemius P...</td>\n",
       "      <td>single-cell RNA sequencing assay</td>\n",
       "      <td>ali-mortazavi</td>\n",
       "      <td>UM1HG009443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ali-mortazavi:experiment_illumina_sr_B6Cast_ad...</td>\n",
       "      <td>/biosample-types/tissue_UBERON_0002369/</td>\n",
       "      <td>Illumina short-read B6/Cast F1 adrenal P8mo fe...</td>\n",
       "      <td>single-cell RNA sequencing assay</td>\n",
       "      <td>ali-mortazavi</td>\n",
       "      <td>UM1HG009443</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             aliases  \\\n",
       "0  ali-mortazavi:experiment_illumina_sr_B6Cast_ad...   \n",
       "1  ali-mortazavi:experiment_illumina_sr_5xFADCast...   \n",
       "2  ali-mortazavi:experiment_illumina_sr_5xFADCast...   \n",
       "3  ali-mortazavi:experiment_illumina_sr_B6Cast_ga...   \n",
       "4  ali-mortazavi:experiment_illumina_sr_B6Cast_ad...   \n",
       "\n",
       "                        biosample_ontology  \\\n",
       "0  /biosample-types/tissue_UBERON_0002369/   \n",
       "1  /biosample-types/tissue_UBERON_0000948/   \n",
       "2  /biosample-types/tissue_UBERON_0001388/   \n",
       "3  /biosample-types/tissue_UBERON_0001388/   \n",
       "4  /biosample-types/tissue_UBERON_0002369/   \n",
       "\n",
       "                                         description  \\\n",
       "0  Illumina short-read B6/Cast F1 adrenal P8mo ma...   \n",
       "1  Illumina short-read 5xFAD/Cast F1 heart P8mo f...   \n",
       "2  Illumina short-read 5xFAD/Cast F1 gastrocnemiu...   \n",
       "3  Illumina short-read B6/Cast F1 gastrocnemius P...   \n",
       "4  Illumina short-read B6/Cast F1 adrenal P8mo fe...   \n",
       "\n",
       "                    assay_term_name            lab        award  \n",
       "0  single-cell RNA sequencing assay  ali-mortazavi  UM1HG009443  \n",
       "1  single-cell RNA sequencing assay  ali-mortazavi  UM1HG009443  \n",
       "2  single-cell RNA sequencing assay  ali-mortazavi  UM1HG009443  \n",
       "3  single-cell RNA sequencing assay  ali-mortazavi  UM1HG009443  \n",
       "4  single-cell RNA sequencing assay  ali-mortazavi  UM1HG009443  "
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "experiment_tsv = StringIO(\"\"\"aliases\tbiosample_ontology\tdescription\tassay_term_name\tlab\taward\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P8mo_male_8mo_12k\t/biosample-types/tissue_UBERON_0002369/\tIllumina short-read B6/Cast F1 adrenal P8mo male 8mo 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_5xFADCast_heart_P8mo_female_8mo_12k\t/biosample-types/tissue_UBERON_0000948/\tIllumina short-read 5xFAD/Cast F1 heart P8mo female 8mo 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_5xFADCast_gastrocnemius_P8mo_female_8mo_12k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read 5xFAD/Cast F1 gastrocnemius P8mo female 8mo 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P8mo_male_8mo_12k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P8mo male 8mo 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P8mo_female_8mo_12k\t/biosample-types/tissue_UBERON_0002369/\tIllumina short-read B6/Cast F1 adrenal P8mo female 8mo 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_5xFADCast_adrenal_P8mo_female_8mo_12k\t/biosample-types/tissue_UBERON_0002369/\tIllumina short-read 5xFAD/Cast F1 adrenal P8mo female 8mo 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_5xFADCast_adrenal_P8mo_male_8mo_12k\t/biosample-types/tissue_UBERON_0002369/\tIllumina short-read 5xFAD/Cast F1 adrenal P8mo male 8mo 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_heart_P8mo_female_8mo_12k\t/biosample-types/tissue_UBERON_0000948/\tIllumina short-read B6/Cast F1 heart P8mo female 8mo 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_5xFADCast_gastrocnemius_P8mo_male_8mo_12k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read 5xFAD/Cast F1 gastrocnemius P8mo male 8mo 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_heart_P8mo_male_8mo_12k\t/biosample-types/tissue_UBERON_0000948/\tIllumina short-read B6/Cast F1 heart P8mo male 8mo 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P8mo_female_8mo_12k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P8mo female 8mo 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_5xFADCast_heart_P8mo_male_8mo_12k\t/biosample-types/tissue_UBERON_0000948/\tIllumina short-read 5xFAD/Cast F1 heart P8mo male 8mo 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_cortex_P8mo_male_8mo_12k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read B6/Cast F1 cortex P8mo male 8mo 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_5xFADCast_cortex_P8mo_female_8mo_12k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read 5xFAD/Cast F1 cortex P8mo female 8mo 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_cortex_P8mo_female_8mo_12k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read B6/Cast F1 cortex P8mo female 8mo 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_5xFADCast_cortex_P8mo_male_8mo_12k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read 5xFAD/Cast F1 cortex P8mo male 8mo 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_5xFADCast_hippocampus_P8mo_female_8mo_12k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read 5xFAD/Cast F1 hippocampus P8mo female 8mo 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_5xFADCast_hippocampus_P8mo_male_8mo_12k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read 5xFAD/Cast F1 hippocampus P8mo male 8mo 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P8mo_male_8mo_12k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P8mo male 8mo 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P8mo_female_8mo_12k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P8mo female 8mo 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_cortex_P8mo_male_8mo_2k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read B6/Cast F1 cortex P8mo male 8mo 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_5xFADCast_cortex_P8mo_female_8mo_2k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read 5xFAD/Cast F1 cortex P8mo female 8mo 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_cortex_P8mo_female_8mo_2k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read B6/Cast F1 cortex P8mo female 8mo 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_5xFADCast_cortex_P8mo_male_8mo_2k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read 5xFAD/Cast F1 cortex P8mo male 8mo 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_5xFADCast_hippocampus_P8mo_female_8mo_2k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read 5xFAD/Cast F1 hippocampus P8mo female 8mo 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_5xFADCast_hippocampus_P8mo_male_8mo_2k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read 5xFAD/Cast F1 hippocampus P8mo male 8mo 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P8mo_male_8mo_2k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P8mo male 8mo 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P8mo_female_8mo_2k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P8mo female 8mo 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P18-20mo_male_normal_12k\t/biosample-types/tissue_UBERON_0002369/\tIllumina short-read B6/Cast F1 adrenal P18-20mo male normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P36_male_normal_12k\t/biosample-types/tissue_UBERON_0002369/\tIllumina short-read B6/Cast F1 adrenal P36 male normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P4_male_normal_12k\t/biosample-types/tissue_UBERON_0002369/\tIllumina short-read B6/Cast F1 adrenal P4 male normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P18-20mo_female_normal_12k\t/biosample-types/tissue_UBERON_0002369/\tIllumina short-read B6/Cast F1 adrenal P18-20mo female normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P14_male_normal_12k\t/biosample-types/tissue_UBERON_0002369/\tIllumina short-read B6/Cast F1 adrenal P14 male normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P2mo_male_normal_12k\t/biosample-types/tissue_UBERON_0002369/\tIllumina short-read B6/Cast F1 adrenal P2mo male normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P36_female_normal_12k\t/biosample-types/tissue_UBERON_0002369/\tIllumina short-read B6/Cast F1 adrenal P36 female normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P4_female_normal_12k\t/biosample-types/tissue_UBERON_0002369/\tIllumina short-read B6/Cast F1 adrenal P4 female normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P2mo_female_normal_12k\t/biosample-types/tissue_UBERON_0002369/\tIllumina short-read B6/Cast F1 adrenal P2mo female normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P10_female_normal_12k\t/biosample-types/tissue_UBERON_0002369/\tIllumina short-read B6/Cast F1 adrenal P10 female normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P25_female_normal_12k\t/biosample-types/tissue_UBERON_0002369/\tIllumina short-read B6/Cast F1 adrenal P25 female normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P10_male_normal_12k\t/biosample-types/tissue_UBERON_0002369/\tIllumina short-read B6/Cast F1 adrenal P10 male normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P25_male_normal_12k\t/biosample-types/tissue_UBERON_0002369/\tIllumina short-read B6/Cast F1 adrenal P25 male normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P14_female_normal_12k\t/biosample-types/tissue_UBERON_0002369/\tIllumina short-read B6/Cast F1 adrenal P14 female normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_cortex_P2mo_female_normal_12k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read B6/Cast F1 cortex P2mo female normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_cortex_P25_female_normal_12k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read B6/Cast F1 cortex P25 female normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_cortex_P36_female_normal_12k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read B6/Cast F1 cortex P36 female normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_cortex_P14_male_normal_12k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read B6/Cast F1 cortex P14 male normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_cortex_P2mo_male_normal_12k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read B6/Cast F1 cortex P2mo male normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_cortex_P10_male_normal_12k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read B6/Cast F1 cortex P10 male normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_cortex_P4_male_normal_12k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read B6/Cast F1 cortex P4 male normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_cortex_P4_female_normal_12k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read B6/Cast F1 cortex P4 female normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_cortex_P25_male_normal_12k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read B6/Cast F1 cortex P25 male normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_cortex_P18-20mo_female_normal_12k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read B6/Cast F1 cortex P18-20mo female normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_cortex_P10_female_normal_12k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read B6/Cast F1 cortex P10 female normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_cortex_P14_female_normal_12k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read B6/Cast F1 cortex P14 female normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_cortex_P36_male_normal_12k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read B6/Cast F1 cortex P36 male normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_cortex_P18-20mo_male_normal_12k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read B6/Cast F1 cortex P18-20mo male normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P25_male_normal_12k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P25 male normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P36_male_normal_12k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P36 male normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P4_female_normal_12k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P4 female normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P18-20mo_female_normal_12k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P18-20mo female normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P25_female_normal_12k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P25 female normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P18-20mo_male_normal_12k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P18-20mo male normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P4_male_normal_12k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P4 male normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P2mo_male_normal_12k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P2mo male normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P10_female_normal_12k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P10 female normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P10_male_normal_12k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P10 male normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P14_male_normal_12k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P14 male normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P2mo_female_normal_12k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P2mo female normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P14_female_normal_12k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P14 female normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P36_female_normal_12k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P36 female normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_heart_P10_male_normal_12k\t/biosample-types/tissue_UBERON_0000948/\tIllumina short-read B6/Cast F1 heart P10 male normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_heart_P10_female_normal_12k\t/biosample-types/tissue_UBERON_0000948/\tIllumina short-read B6/Cast F1 heart P10 female normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_heart_P36_female_normal_12k\t/biosample-types/tissue_UBERON_0000948/\tIllumina short-read B6/Cast F1 heart P36 female normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_heart_P14_male_normal_12k\t/biosample-types/tissue_UBERON_0000948/\tIllumina short-read B6/Cast F1 heart P14 male normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_heart_P4_female_normal_12k\t/biosample-types/tissue_UBERON_0000948/\tIllumina short-read B6/Cast F1 heart P4 female normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_heart_P2mo_female_normal_12k\t/biosample-types/tissue_UBERON_0000948/\tIllumina short-read B6/Cast F1 heart P2mo female normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_heart_P2mo_male_normal_12k\t/biosample-types/tissue_UBERON_0000948/\tIllumina short-read B6/Cast F1 heart P2mo male normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_heart_P4_male_normal_12k\t/biosample-types/tissue_UBERON_0000948/\tIllumina short-read B6/Cast F1 heart P4 male normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_heart_P25_male_normal_12k\t/biosample-types/tissue_UBERON_0000948/\tIllumina short-read B6/Cast F1 heart P25 male normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_heart_P14_female_normal_12k\t/biosample-types/tissue_UBERON_0000948/\tIllumina short-read B6/Cast F1 heart P14 female normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_heart_P18-20mo_male_normal_12k\t/biosample-types/tissue_UBERON_0000948/\tIllumina short-read B6/Cast F1 heart P18-20mo male normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_heart_P25_female_normal_12k\t/biosample-types/tissue_UBERON_0000948/\tIllumina short-read B6/Cast F1 heart P25 female normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_heart_P18-20mo_female_normal_12k\t/biosample-types/tissue_UBERON_0000948/\tIllumina short-read B6/Cast F1 heart P18-20mo female normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_heart_P36_male_normal_12k\t/biosample-types/tissue_UBERON_0000948/\tIllumina short-read B6/Cast F1 heart P36 male normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P2mo_male_normal_12k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P2mo male normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P25_male_normal_12k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P25 male normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P18-20mo_male_normal_12k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P18-20mo male normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P36_female_normal_12k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P36 female normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P2mo_female_normal_12k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P2mo female normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P10_male_normal_12k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P10 male normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P10_female_normal_12k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P10 female normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P25_female_normal_12k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P25 female normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P14_female_normal_12k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P14 female normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P14_male_normal_12k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P14 male normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P36_male_normal_12k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P36 male normal 12k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P25_male_normal_15k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P25 male normal 15k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P36_male_normal_15k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P36 male normal 15k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P4_female_normal_15k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P4 female normal 15k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P18-20mo_female_normal_15k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P18-20mo female normal 15k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P25_female_normal_15k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P25 female normal 15k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P18-20mo_male_normal_15k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P18-20mo male normal 15k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P4_male_normal_15k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P4 male normal 15k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P2mo_male_normal_15k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P2mo male normal 15k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P10_female_normal_15k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P10 female normal 15k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P10_male_normal_15k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P10 male normal 15k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P14_male_normal_15k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P14 male normal 15k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P2mo_female_normal_15k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P2mo female normal 15k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P14_female_normal_15k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P14 female normal 15k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P36_female_normal_15k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P36 female normal 15k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P25_male_normal_1k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P25 male normal 1k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P36_male_normal_1k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P36 male normal 1k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P4_female_normal_1k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P4 female normal 1k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P18-20mo_female_normal_1k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P18-20mo female normal 1k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P25_female_normal_1k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P25 female normal 1k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P18-20mo_male_normal_1k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P18-20mo male normal 1k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P4_male_normal_1k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P4 male normal 1k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P2mo_male_normal_1k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P2mo male normal 1k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P10_female_normal_1k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P10 female normal 1k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P10_male_normal_1k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P10 male normal 1k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P14_male_normal_1k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P14 male normal 1k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P2mo_female_normal_1k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P2mo female normal 1k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P14_female_normal_1k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P14 female normal 1k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P36_female_normal_1k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P36 female normal 1k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P18-20mo_male_normal_2k\t/biosample-types/tissue_UBERON_0002369/\tIllumina short-read B6/Cast F1 adrenal P18-20mo male normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P36_male_normal_2k\t/biosample-types/tissue_UBERON_0002369/\tIllumina short-read B6/Cast F1 adrenal P36 male normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P4_male_normal_2k\t/biosample-types/tissue_UBERON_0002369/\tIllumina short-read B6/Cast F1 adrenal P4 male normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P18-20mo_female_normal_2k\t/biosample-types/tissue_UBERON_0002369/\tIllumina short-read B6/Cast F1 adrenal P18-20mo female normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P14_male_normal_2k\t/biosample-types/tissue_UBERON_0002369/\tIllumina short-read B6/Cast F1 adrenal P14 male normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P2mo_male_normal_2k\t/biosample-types/tissue_UBERON_0002369/\tIllumina short-read B6/Cast F1 adrenal P2mo male normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P36_female_normal_2k\t/biosample-types/tissue_UBERON_0002369/\tIllumina short-read B6/Cast F1 adrenal P36 female normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P4_female_normal_2k\t/biosample-types/tissue_UBERON_0002369/\tIllumina short-read B6/Cast F1 adrenal P4 female normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P2mo_female_normal_2k\t/biosample-types/tissue_UBERON_0002369/\tIllumina short-read B6/Cast F1 adrenal P2mo female normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P10_female_normal_2k\t/biosample-types/tissue_UBERON_0002369/\tIllumina short-read B6/Cast F1 adrenal P10 female normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P25_female_normal_2k\t/biosample-types/tissue_UBERON_0002369/\tIllumina short-read B6/Cast F1 adrenal P25 female normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P10_male_normal_2k\t/biosample-types/tissue_UBERON_0002369/\tIllumina short-read B6/Cast F1 adrenal P10 male normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P25_male_normal_2k\t/biosample-types/tissue_UBERON_0002369/\tIllumina short-read B6/Cast F1 adrenal P25 male normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P14_female_normal_2k\t/biosample-types/tissue_UBERON_0002369/\tIllumina short-read B6/Cast F1 adrenal P14 female normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_cortex_P2mo_female_normal_2k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read B6/Cast F1 cortex P2mo female normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_cortex_P25_female_normal_2k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read B6/Cast F1 cortex P25 female normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_cortex_P36_female_normal_2k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read B6/Cast F1 cortex P36 female normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_cortex_P14_male_normal_2k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read B6/Cast F1 cortex P14 male normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_cortex_P2mo_male_normal_2k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read B6/Cast F1 cortex P2mo male normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_cortex_P10_male_normal_2k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read B6/Cast F1 cortex P10 male normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_cortex_P4_male_normal_2k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read B6/Cast F1 cortex P4 male normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_cortex_P4_female_normal_2k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read B6/Cast F1 cortex P4 female normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_cortex_P25_male_normal_2k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read B6/Cast F1 cortex P25 male normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_cortex_P18-20mo_female_normal_2k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read B6/Cast F1 cortex P18-20mo female normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_cortex_P10_female_normal_2k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read B6/Cast F1 cortex P10 female normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_cortex_P14_female_normal_2k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read B6/Cast F1 cortex P14 female normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_cortex_P36_male_normal_2k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read B6/Cast F1 cortex P36 male normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_cortex_P18-20mo_male_normal_2k\t/biosample-types/tissue_NTR_0000646/\tIllumina short-read B6/Cast F1 cortex P18-20mo male normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P25_male_normal_2k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P25 male normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P36_male_normal_2k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P36 male normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P4_female_normal_2k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P4 female normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P18-20mo_female_normal_2k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P18-20mo female normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P25_female_normal_2k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P25 female normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P18-20mo_male_normal_2k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P18-20mo male normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P4_male_normal_2k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P4 male normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P2mo_male_normal_2k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P2mo male normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P10_female_normal_2k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P10 female normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P10_male_normal_2k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P10 male normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P14_male_normal_2k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P14 male normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P2mo_female_normal_2k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P2mo female normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P14_female_normal_2k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P14 female normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_gastrocnemius_P36_female_normal_2k\t/biosample-types/tissue_UBERON_0001388/\tIllumina short-read B6/Cast F1 gastrocnemius P36 female normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_heart_P10_male_normal_2k\t/biosample-types/tissue_UBERON_0000948/\tIllumina short-read B6/Cast F1 heart P10 male normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_heart_P10_female_normal_2k\t/biosample-types/tissue_UBERON_0000948/\tIllumina short-read B6/Cast F1 heart P10 female normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_heart_P36_female_normal_2k\t/biosample-types/tissue_UBERON_0000948/\tIllumina short-read B6/Cast F1 heart P36 female normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_heart_P14_male_normal_2k\t/biosample-types/tissue_UBERON_0000948/\tIllumina short-read B6/Cast F1 heart P14 male normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_heart_P4_female_normal_2k\t/biosample-types/tissue_UBERON_0000948/\tIllumina short-read B6/Cast F1 heart P4 female normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_heart_P2mo_female_normal_2k\t/biosample-types/tissue_UBERON_0000948/\tIllumina short-read B6/Cast F1 heart P2mo female normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_heart_P2mo_male_normal_2k\t/biosample-types/tissue_UBERON_0000948/\tIllumina short-read B6/Cast F1 heart P2mo male normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_heart_P4_male_normal_2k\t/biosample-types/tissue_UBERON_0000948/\tIllumina short-read B6/Cast F1 heart P4 male normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_heart_P25_male_normal_2k\t/biosample-types/tissue_UBERON_0000948/\tIllumina short-read B6/Cast F1 heart P25 male normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_heart_P14_female_normal_2k\t/biosample-types/tissue_UBERON_0000948/\tIllumina short-read B6/Cast F1 heart P14 female normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_heart_P18-20mo_male_normal_2k\t/biosample-types/tissue_UBERON_0000948/\tIllumina short-read B6/Cast F1 heart P18-20mo male normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_heart_P25_female_normal_2k\t/biosample-types/tissue_UBERON_0000948/\tIllumina short-read B6/Cast F1 heart P25 female normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_heart_P18-20mo_female_normal_2k\t/biosample-types/tissue_UBERON_0000948/\tIllumina short-read B6/Cast F1 heart P18-20mo female normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_heart_P36_male_normal_2k\t/biosample-types/tissue_UBERON_0000948/\tIllumina short-read B6/Cast F1 heart P36 male normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P2mo_male_normal_2k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P2mo male normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P25_male_normal_2k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P25 male normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P18-20mo_male_normal_2k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P18-20mo male normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P36_female_normal_2k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P36 female normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P2mo_female_normal_2k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P2mo female normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P10_male_normal_2k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P10 male normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P10_female_normal_2k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P10 female normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P25_female_normal_2k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P25 female normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P14_female_normal_2k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P14 female normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P14_male_normal_2k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P14 male normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P36_male_normal_2k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P36 male normal 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P2mo_male_ont_match_2k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P2mo male ont_match 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P25_male_ont_match_2k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P25 male ont_match 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P18-20mo_male_ont_match_2k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P18-20mo male ont_match 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P36_female_ont_match_2k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P36 female ont_match 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P2mo_female_ont_match_2k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P2mo female ont_match 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P10_male_ont_match_2k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P10 male ont_match 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P10_female_ont_match_2k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P10 female ont_match 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P25_female_ont_match_2k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P25 female ont_match 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P14_female_ont_match_2k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P14 female ont_match 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P14_male_ont_match_2k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P14 male ont_match 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_hippocampus_P36_male_ont_match_2k\t/biosample-types/tissue_UBERON_0002305/\tIllumina short-read B6/Cast F1 hippocampus P36 male ont_match 2k\tsingle-cell RNA sequencing assay\tali-mortazavi\tUM1HG009443\n",
    "\"\"\")\n",
    "\n",
    "uci_splitseq = pandas.read_csv(experiment_tsv, sep=\"\\t\")\n",
    "uci_splitseq.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "roman-middle",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "402\n"
     ]
    }
   ],
   "source": [
    "uci_libraries = []\n",
    "for i, row in uci_splitseq.iterrows():\n",
    "    experiment = server.get_json(row[\"aliases\"])\n",
    "    for replicate in experiment['replicates']:\n",
    "        uci_libraries.append(replicate[\"library\"][\"accession\"])\n",
    "print(len(uci_libraries))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "formed-diagram",
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(yaml.dump(all_configs[uci_libraries[3]]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "norman-persian",
   "metadata": {},
   "outputs": [],
   "source": [
    "#prepare_bulk_sbatch_split_run(\"/dev/shm/splitseq-0-49.tar.gz\", all_configs, uci_libraries[0:50])\n",
    "#prepare_bulk_sbatch_split_run(\"/dev/shm/splitseq-50-99.tar.gz\", all_configs, uci_libraries[50:100])\n",
    "#prepare_bulk_sbatch_split_run(\"/dev/shm/splitseq-100-199.tar.gz\", all_configs, uci_libraries[100:200])\n",
    "#prepare_bulk_sbatch_split_run(\"/dev/shm/splitseq-200-299.tar.gz\", all_configs, uci_libraries[200:300])\n",
    "#prepare_bulk_sbatch_split_run(\"/dev/shm/splitseq-300-end.tar.gz\", all_configs, uci_libraries[300:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "facial-vatican",
   "metadata": {},
   "outputs": [],
   "source": [
    "#for lib_id in uci_libraries[350::2]:\n",
    "#    print(\"https://www.encodeproject.org/experiments/{}/\".format(all_configs[lib_id][\"experiment_accession\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "vietnamese-formula",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "numeric-lender",
   "metadata": {},
   "source": [
    "Process some later gastroc samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "separate-pickup",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://www.encodeproject.org/experiment/ENCSR021MHR/ ENCLB673NZA\n",
      "https://www.encodeproject.org/experiment/ENCSR771MOE/ ENCLB275WWO\n",
      "https://www.encodeproject.org/experiment/ENCSR478PUL/ ENCLB320ZLU\n",
      "https://www.encodeproject.org/experiment/ENCSR438FRT/ ENCLB854VKF\n",
      "https://www.encodeproject.org/experiment/ENCSR835ZXV/ ENCLB995TFV\n",
      "https://www.encodeproject.org/experiment/ENCSR163GJZ/ ENCLB786EOH\n",
      "https://www.encodeproject.org/experiment/ENCSR125XUV/ ENCLB794PMK\n",
      "https://www.encodeproject.org/experiment/ENCSR188GDL/ ENCLB526RVS\n"
     ]
    }
   ],
   "source": [
    "to_run_feb_9 = []\n",
    "graph = server.get_json(\"https://www.encodeproject.org/search/?type=Experiment&assay_title=scRNA-seq&replicates.library.biosample.donor.organism.scientific_name=Mus+musculus&biosample_ontology.term_name=gastrocnemius&lab.title=Michael+Snyder,+Stanford\")\n",
    "for row in graph[\"@graph\"]:\n",
    "    experiment = server.get_json(row[\"accession\"])\n",
    "    for replicate in experiment[\"replicates\"]:\n",
    "        library = replicate[\"library\"]\n",
    "        print(\"https://www.encodeproject.org/experiment/{}/\".format(experiment[\"accession\"]), library[\"accession\"])\n",
    "        to_run_feb_9.append(library[\"accession\"])\n",
    "\n",
    "#prepare_bulk_sbatch_10x_run(\"/dev/shm/gastroc-multiome-feb9.tar.gz\", all_configs, to_run_feb_9)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "experimental-stopping",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "chronic-summary",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28\n"
     ]
    }
   ],
   "source": [
    "experiment_feb17_tsv = StringIO(\"\"\"aliases\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P18-20mo_male_ont_match_2k\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P36_male_ont_match_2k\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P4_male_ont_match_2k\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P18-20mo_female_ont_match_2k\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P14_male_ont_match_2k\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P2mo_male_ont_match_2k\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P36_female_ont_match_2k\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P4_female_ont_match_2k\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P2mo_female_ont_match_2k\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P10_female_ont_match_2k\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P25_female_ont_match_2k\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P10_male_ont_match_2k\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P25_male_ont_match_2k\n",
    "ali-mortazavi:experiment_illumina_sr_B6Cast_adrenal_P14_female_ont_match_2k\n",
    "\"\"\")\n",
    "uci_splitseq_feb17 = pandas.read_csv(experiment_feb17_tsv, sep=\"\\t\")\n",
    "uci_splitseq_feb17.head()\n",
    "\n",
    "uci_libraries_feb17 = []\n",
    "for i, row in uci_splitseq_feb17.iterrows():\n",
    "    experiment = server.get_json(row[\"aliases\"])\n",
    "    for replicate in experiment['replicates']:\n",
    "        uci_libraries_feb17.append(replicate[\"library\"][\"accession\"])\n",
    "print(len(uci_libraries_feb17))\n",
    "\n",
    "#prepare_bulk_sbatch_split_run(\"/dev/shm/splitseq-feb17.tar.gz\", all_configs, uci_libraries_feb17)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "answering-paragraph",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ENCLB765UWE',\n",
       " 'ENCLB107BBQ',\n",
       " 'ENCLB680GUP',\n",
       " 'ENCLB856JIZ',\n",
       " 'ENCLB924MNV',\n",
       " 'ENCLB454IQF',\n",
       " 'ENCLB048VRI',\n",
       " 'ENCLB866KKG',\n",
       " 'ENCLB109THY',\n",
       " 'ENCLB689QBZ',\n",
       " 'ENCLB126DIM',\n",
       " 'ENCLB419YTE',\n",
       " 'ENCLB691GNG',\n",
       " 'ENCLB585ZUX',\n",
       " 'ENCLB517ACD',\n",
       " 'ENCLB093BII',\n",
       " 'ENCLB851NLO',\n",
       " 'ENCLB381HJR',\n",
       " 'ENCLB741YUG',\n",
       " 'ENCLB339NZV',\n",
       " 'ENCLB679MGP',\n",
       " 'ENCLB903CLN',\n",
       " 'ENCLB096XEI',\n",
       " 'ENCLB842EQT',\n",
       " 'ENCLB668FHK',\n",
       " 'ENCLB944QTS',\n",
       " 'ENCLB469KQZ',\n",
       " 'ENCLB765SCG']"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uci_libraries_feb17"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "premium-accessory",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "amber-helen",
   "metadata": {},
   "source": [
    "# Another batch from Annika Jun 21 2022"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "subsequent-elizabeth",
   "metadata": {},
   "source": [
    "Annika gave me another batch of files to process:\n",
    "\n",
    "I hope youre doing well. \n",
    "We just uploaded more Multiome experiments (all adult heart biosamples) to the ENCODE portal.\n",
    "Would you be able to process the snRNA part of the experiments? \n",
    "Please note that they are all single nuclear RNA-seq, not single cell RNA. \n",
    "\n",
    "The accession numbers for the snRNA-seq experiments are as follows: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "clinical-corruption",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "37"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annika_experiments_20210621 = \"\"\"ENCSR455MGH\n",
    "ENCSR175TRJ\n",
    "ENCSR076ZLE\n",
    "ENCSR763BII\n",
    "ENCSR002SMQ\n",
    "ENCSR753YOZ\n",
    "ENCSR485GOL\n",
    "ENCSR762LML\n",
    "ENCSR190TRK\n",
    "ENCSR008CVR\n",
    "ENCSR540DHJ\n",
    "ENCSR056QLB\n",
    "ENCSR489URW\n",
    "ENCSR204RHR\n",
    "ENCSR328GTN\n",
    "ENCSR488UUT\n",
    "ENCSR788SNY\n",
    "ENCSR906MRL\n",
    "ENCSR994VEG\n",
    "ENCSR084XKX\n",
    "ENCSR345CVL\n",
    "ENCSR349AHE\n",
    "ENCSR962JKS\n",
    "ENCSR681GPY\n",
    "ENCSR273JWD\n",
    "ENCSR630LZS\n",
    "ENCSR654MFX\n",
    "ENCSR157FDD\n",
    "ENCSR067BOK\n",
    "ENCSR814LMX\n",
    "ENCSR176WWW\n",
    "ENCSR398YBK\n",
    "ENCSR231FNL\n",
    "ENCSR980OCK\n",
    "ENCSR991LHO\n",
    "ENCSR899GYX\n",
    "ENCSR203YOV\n",
    "\"\"\".split()\n",
    "len(annika_experiments_20210621)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "documented-manual",
   "metadata": {},
   "source": [
    "So first question do they match what is currently marked as needing to be processed.?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "amber-thursday",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "in_progress = [x[\"experiment\"] for x in results if x[\"status\"] == \"in progress\" and x[\"has_processed\"] == \"N\"]\n",
    "len(in_progress)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "mighty-czech",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(in_progress).intersection(annika_experiments_20210621)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "steady-invalid",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "libraries_to_process 8\n",
      "Processed 8\n",
      "tar: /dev/shm/annika_heart_20220621.tgz: Cannot open: No such file or directory\r\n",
      "tar: Error is not recoverable: exiting now\r\n"
     ]
    }
   ],
   "source": [
    "libraries_to_process = [x[\"library\"] for x in results if (x[\"status\"] == \"in progress\") and x[\"experiment\"] in in_progress]\n",
    "print(\"libraries_to_process\", len(libraries_to_process))\n",
    "\n",
    "for lib in libraries_to_process:\n",
    "    all_configs[lib][\"include_intron\"] = True\n",
    "\n",
    "if len(libraries_to_process) > 0:\n",
    "    processed = prepare_bulk_sbatch_10x_run(\"/dev/shm/annika_heart_20220810.tgz\", all_configs, libraries_to_process)\n",
    "    print(\"Processed\", len(processed))\n",
    "    !tar tavf /dev/shm/annika_heart_20220621.tgz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "furnished-burton",
   "metadata": {},
   "source": [
    "# Group of lab processed that need to be reprocessed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "military-balloon",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ENCLB273ZWE',\n",
       " 'ENCLB527JDU',\n",
       " 'ENCLB514RKT',\n",
       " 'ENCLB294FBZ',\n",
       " 'ENCLB952MZJ',\n",
       " 'ENCLB129LMS']"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "liz_reprocess_experiments = [\"ENCSR503QZJ\", \"ENCSR109DAN\", \"ENCSR572SMD\", \"ENCSR388EIN\", \"ENCSR452OYH\", \"ENCSR877NAX\"]\n",
    "liz_reprocess_libraries = [x[\"library\"] for x in results if x[\"experiment\"] in liz_reprocess_experiments]\n",
    "liz_reprocess_libraries\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "pressed-egypt",
   "metadata": {},
   "outputs": [],
   "source": [
    "prepare_bulk_sbatch_split_run(\"/dev/shm/liz_reprocess.tgz\", all_configs, liz_reprocess_libraries)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "sharp-internship",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'experiment_accession': 'ENCSR109DAN',\n",
       " 'description': 'short-read shallow Split-seq single-nucleus myotube 9,000 cell libs',\n",
       " 'lab': '/labs/barbara-wold/',\n",
       " 'award': 'UM1HG009443',\n",
       " 'encode_portal_host': 'www.encodeproject.org',\n",
       " 'star_container': 'https://woldlab.caltech.edu/~diane/containers/star-bash-2.7.10a.sif',\n",
       " 'mem_mb': 65536,\n",
       " 'disk_mb': 51200,\n",
       " 'automatic_submission': True,\n",
       " 'stranded': 'Forward',\n",
       " 'inclusion_list_url': 'https://woldlab.caltech.edu/~diane/genome/parse-splitseq-v1.tar.gz',\n",
       " 'inclusion_accession': 'encode:parse-biosciences-v1',\n",
       " 'library_accession': 'ENCLB527JDU',\n",
       " 'genome_accession': 'encode:starsolo-mm10-M21-male-index',\n",
       " 'include_intron': True,\n",
       " 'read1': ['ENCFF288PHA'],\n",
       " 'read2': ['ENCFF093WFS']}"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_configs[\"ENCLB527JDU\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "entire-pharmaceutical",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "french-runner",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "arbitrary-firmware",
   "metadata": {},
   "source": [
    "# 8 more multiome heart ventrical samples from Snyder group."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "thick-county",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "libraries_to_process 8\n",
      "Processed 8\n",
      "-rw-r--r-- 0/0             525 1969-12-31 16:00 ENCLB054SXC/config.yaml\r\n",
      "-rw-r--r-- 0/0             525 1969-12-31 16:00 ENCLB823WAF/config.yaml\r\n",
      "-rw-r--r-- 0/0             525 1969-12-31 16:00 ENCLB713GHH/config.yaml\r\n",
      "-rw-r--r-- 0/0             525 1969-12-31 16:00 ENCLB401UMP/config.yaml\r\n",
      "-rw-r--r-- 0/0             525 1969-12-31 16:00 ENCLB401DAZ/config.yaml\r\n",
      "-rw-r--r-- 0/0             525 1969-12-31 16:00 ENCLB851FFF/config.yaml\r\n",
      "-rw-r--r-- 0/0             525 1969-12-31 16:00 ENCLB540OBF/config.yaml\r\n",
      "-rw-r--r-- 0/0             525 1969-12-31 16:00 ENCLB566WWJ/config.yaml\r\n",
      "-rw-r--r-- 0/0            2188 1969-12-31 16:00 run.sh\r\n"
     ]
    }
   ],
   "source": [
    "query = \"https://www.encodeproject.org/search/?type=Experiment&accession=ENCSR777RUZ&accession=ENCSR093GXF&accession=ENCSR919ENI&accession=ENCSR138JCM&accession=ENCSR085XEW&accession=ENCSR237HWJ&accession=ENCSR851JBE&accession=ENCSR451TSH\"\n",
    "sept22_accessions = [\n",
    "    \"ENCSR777RUZ\",\n",
    "    \"ENCSR093GXF\",\n",
    "    \"ENCSR919ENI\",\n",
    "    \"ENCSR138JCM\",\n",
    "    \"ENCSR085XEW\",\n",
    "    \"ENCSR237HWJ\",\n",
    "    \"ENCSR851JBE\",\n",
    "    \"ENCSR451TSH\",\n",
    "]\n",
    "sept22_experiments = []\n",
    "sept22_libraries = [x[\"library\"] for x in results if x[\"experiment\"] in sept22_experiments]\n",
    "sept22_libraries\n",
    "print(len(sept22_libraries))\n",
    "\n",
    "sept22_libraries_to_process = [x[\"library\"] for x in results if (x[\"status\"] == \"in progress\") and x[\"experiment\"] in in_progress]\n",
    "\n",
    "print(\"libraries_to_process\", len(sept22_libraries_to_process))\n",
    "\n",
    "for lib in sept22_libraries_to_process:\n",
    "    all_configs[lib][\"include_intron\"] = True\n",
    "\n",
    "if len(sept22_libraries_to_process) > 0:\n",
    "    processed = prepare_bulk_sbatch_10x_run(\"/dev/shm/snyder_heart_20220822.tgz\", all_configs, sept22_libraries_to_process)\n",
    "    print(\"Processed\", len(processed))\n",
    "    !tar tavf /dev/shm/snyder_heart_20220822.tgz\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "working-billy",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ENCSR955TRT',\n",
       " 'ENCSR292BJF',\n",
       " 'ENCSR523HMC',\n",
       " 'ENCSR807BYQ',\n",
       " 'ENCSR133IBO',\n",
       " 'ENCSR474DKO',\n",
       " 'ENCSR192UEL',\n",
       " 'ENCSR291IHH']"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "jan10_accessions = []\n",
    "query = \"https://www.encodeproject.org/search/?type=Experiment&lab.title=Michael+Snyder%2C+Stanford&status=in+progress&assay_title=scRNA-seq\"\n",
    "graph = server.get_json(query)\n",
    "\n",
    "for row in graph[\"@graph\"]:\n",
    "    if 'Experiment' in row[\"@type\"]:\n",
    "        jan10_accessions.append(row[\"accession\"])\n",
    "\n",
    "jan10_accessions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "suited-console",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8\n",
      "libraries_to_process 8\n",
      "Processed 8\n",
      "-rw-r--r-- 0/0             525 1969-12-31 16:00 ENCLB054SXC/config.yaml\r\n",
      "-rw-r--r-- 0/0             525 1969-12-31 16:00 ENCLB823WAF/config.yaml\r\n",
      "-rw-r--r-- 0/0             525 1969-12-31 16:00 ENCLB713GHH/config.yaml\r\n",
      "-rw-r--r-- 0/0             525 1969-12-31 16:00 ENCLB401UMP/config.yaml\r\n",
      "-rw-r--r-- 0/0             525 1969-12-31 16:00 ENCLB401DAZ/config.yaml\r\n",
      "-rw-r--r-- 0/0             525 1969-12-31 16:00 ENCLB851FFF/config.yaml\r\n",
      "-rw-r--r-- 0/0             525 1969-12-31 16:00 ENCLB540OBF/config.yaml\r\n",
      "-rw-r--r-- 0/0             525 1969-12-31 16:00 ENCLB566WWJ/config.yaml\r\n",
      "-rw-r--r-- 0/0            2188 1969-12-31 16:00 run.sh\r\n"
     ]
    }
   ],
   "source": [
    "jan10_accessions = []\n",
    "query = \"https://www.encodeproject.org/search/?type=Experiment&lab.title=Michael+Snyder%2C+Stanford&status=in+progress&assay_title=scRNA-seq\"\n",
    "graph = server.get_json(query)\n",
    "\n",
    "for row in graph[\"@graph\"]:\n",
    "    if 'Experiment' in row[\"@type\"]:\n",
    "        jan10_accessions.append(row[\"accession\"])\n",
    "\n",
    "jan10_accessions\n",
    "\n",
    "\n",
    "jan10_experiments = []\n",
    "jan10_libraries = [x[\"library\"] for x in results if x[\"experiment\"] in jan10_accessions]\n",
    "jan10_libraries\n",
    "print(len(jan10_libraries))\n",
    "\n",
    "jan10_libraries_to_process = [x[\"library\"] for x in results if (x[\"status\"] == \"in progress\") and x[\"experiment\"] in in_progress]\n",
    "\n",
    "print(\"libraries_to_process\", len(jan10_libraries_to_process))\n",
    "\n",
    "for lib in jan10_libraries_to_process:\n",
    "    all_configs[lib][\"include_intron\"] = True\n",
    "\n",
    "if len(jan10_libraries_to_process) > 0:\n",
    "    processed = prepare_bulk_sbatch_10x_run(\"/dev/shm/snyder_liver_20230110.tgz\", all_configs, jan10_libraries_to_process)\n",
    "    print(\"Processed\", len(processed))\n",
    "    !tar tavf /dev/shm/snyder_liver_20230110.tgz\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "developing-assumption",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
