{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "geographic-parks",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas\n",
    "from pathlib import Path\n",
    "from pprint import pprint\n",
    "from io import BytesIO\n",
    "import re\n",
    "import sys\n",
    "import tarfile\n",
    "from collections import Counter, namedtuple"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "sudden-mitchell",
   "metadata": {},
   "outputs": [],
   "source": [
    "EC = str(Path(\"~/proj/encoded_client\").expanduser())\n",
    "if EC not in sys.path:\n",
    "    sys.path.append(EC)\n",
    "from encoded_client.encoded import ENCODED, HTTPError"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "funny-definition",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "expanded-granny",
   "metadata": {},
   "outputs": [],
   "source": [
    "known_document_types = {\n",
    "    '/documents/ecb1c489-905b-4afb-8389-ed7fb2e36ff2/': \"Chromium v2\",\n",
    "    '/documents/515a990c-50c4-459a-8036-502feaf5a18c/': \"Chromium v2 technical note\",\n",
    "    '/documents/3408b7a1-6b1b-45d1-a73e-db0556e6abdc/': \"Split-seq analysis protocol\",\n",
    "    '/documents/3cdf327b-df85-4149-9fca-a7541e6f1cb1/': \"Split-seq overview\",\n",
    "    '/documents/c155e58a-13ab-4af9-8591-cee13d639a40/': \"Nuclei Isolation protocol from tissue for 10x multiome\",\n",
    "    '/documents/53fb4382-c54f-4761-89bb-6a01a0dca97d/': \"Chromium GEM v3.1\",\n",
    "    '/documents/4d42b1ef-c9e6-4b31-a443-1266079545d0/': \"Chromium Multiome ATAC + Gene Expression\",\n",
    "    '/documents/d18bad0e-9a86-40f0-8ded-2c74072910e5/': \"Nuclei Isolation protocol from tissue for 10x multiome\",  # probably same as c155e58a-13ab-4af9-8591-cee13d639a40\n",
    "    '/documents/e3110414-4013-425e-94f6-72c549702bde/': \"Pool split protocol\",\n",
    "    '/documents/1bb75b62-ac29-4368-9855-68d410e1963a/': \"GM12878 Cell Culture protocol\",\n",
    "    '/documents/28fef8e0-e171-450d-b66d-e3d199531cf2/': \"Ambion spike in mixture\",\n",
    "    '/documents/a6e06058-5019-488f-9195-8bcd2e9229c1/': \"Laser capture microdisection\",\n",
    "    '/documents/c4b8952c-d9f1-42e7-8ea5-54659d14f46b/': \"Human tissue preservation protocol flash-freezing (version 1)\",\n",
    "    '/documents/c5c4174c-0f33-4793-be9d-f2b6a9357dae/': \"Nuclei Isolation protocol from tissue for 10x multiome\", # probably same as c155e58a-13ab-4af9-8591-cee13d639a40\n",
    "    '/documents/84a66218-cd6a-4959-adc0-91ecc513117f/': \"Chromium Multiome ATAC + Gene Expression\",\n",
    "    '/documents/c6235c97-a978-41b6-8a7e-1e16b983f4ba/': \"inDrop Sample Collection Protocol\",\n",
    "    '/documents/8179222b-9ad7-4274-a07e-820dbcac6aca/': \"inDrop Library Prep\",\n",
    "    '/documents/80953249-5d87-4f2c-84e0-0212c404c8d6/': \"Nuclei Isolation protocol from tissue 10x snRNA\", # probably same as c155e58a-13ab-4af9-8591-cee13d639a40\n",
    "    '/documents/57177fd0-da98-4e8f-95a4-459fbaeea146/': \"Chromium GEM v3.1\", # psame as 53fb4382-c54f-4761-89bb-6a01a0dca97d\n",
    "    '/documents/5f5193b0-0c61-40c6-a286-b1f6d6f0c0da/': \"SmartSeq\",\n",
    "    '/documents/9a7e281f-b901-422b-9f01-378ba4b9e97f/': \"Mouse adrenal gland nuclei isolation protocol for single nucleus RNA-seq\",\n",
    "    '/documents/f1b60b4e-b5d2-4b8d-af0b-06b767fa07c3/': \"Biosample purification protocol\",\n",
    "    '/documents/49fd5b3b-9878-4ec8-92bf-ef6dce9116e3/': \"SMARTer Ultra Low RNA kit\",\n",
    "    \"/documents/f16e44c9-ee9d-4ede-9bb5-886ea451d7a0/\": \"Cortex nuclei isolation protocol\",\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "proud-rapid",
   "metadata": {},
   "outputs": [],
   "source": [
    "document_to_protocol = {\n",
    "    '/documents/ecb1c489-905b-4afb-8389-ed7fb2e36ff2/': \"Chromium v2\",\n",
    "    #'/documents/515a990c-50c4-459a-8036-502feaf5a18c/': \"Chromium v2\",\n",
    "    '/documents/3408b7a1-6b1b-45d1-a73e-db0556e6abdc/': \"Split-seq\",\n",
    "    '/documents/3cdf327b-df85-4149-9fca-a7541e6f1cb1/': \"Split-seq\",\n",
    "    #'/documents/c155e58a-13ab-4af9-8591-cee13d639a40/': \"Nuclei Isolation protocol from tissue for 10x multiome\",\n",
    "    '/documents/53fb4382-c54f-4761-89bb-6a01a0dca97d/': \"Chromium GEM v3.1\",\n",
    "    '/documents/4d42b1ef-c9e6-4b31-a443-1266079545d0/': \"Chromium Multiome ATAC + Gene Expression\", # same as 53fb4382-c54f-4761-89bb-6a01a0dca97d\n",
    "    #'/documents/d18bad0e-9a86-40f0-8ded-2c74072910e5/': \"Nuclei Isolation protocol from tissue for 10x multiome\",  # probably same as c155e58a-13ab-4af9-8591-cee13d639a40\n",
    "    '/documents/d18bad0e-9a86-40f0-8ded-2c74072910e5/': \"Chromium Multiome ATAC + Gene Expression\",\n",
    "    '/documents/e3110414-4013-425e-94f6-72c549702bde/': \"SmartSeq Pool\",\n",
    "    '/documents/a6e06058-5019-488f-9195-8bcd2e9229c1/': \"SmartSeq\",\n",
    "    #'/documents/c5c4174c-0f33-4793-be9d-f2b6a9357dae/': \"Nuclei Isolation protocol from tissue for 10x multiome\", # probably same as c155e58a-13ab-4af9-8591-cee13d639a40\n",
    "    '/documents/84a66218-cd6a-4959-adc0-91ecc513117f/': \"Chromium Multiome ATAC + Gene Expression\",\n",
    "    '/documents/c6235c97-a978-41b6-8a7e-1e16b983f4ba/': \"inDrop\",\n",
    "    '/documents/8179222b-9ad7-4274-a07e-820dbcac6aca/': \"inDrop\",\n",
    "    #'/documents/80953249-5d87-4f2c-84e0-0212c404c8d6/': \"Nuclei Isolation protocol from tissue 10x snRNA\", # probably same as c155e58a-13ab-4af9-8591-cee13d639a40\n",
    "    '/documents/57177fd0-da98-4e8f-95a4-459fbaeea146/': \"Chromium GEM v3.1\", # psame as 53fb4382-c54f-4761-89bb-6a01a0dca97d\n",
    "    '/documents/5f5193b0-0c61-40c6-a286-b1f6d6f0c0da/': \"SmartSeq\",\n",
    "    '/documents/49fd5b3b-9878-4ec8-92bf-ef6dce9116e3/': \"SmartSeq\",\n",
    "}\n",
    "\n",
    "#From alex https://en-coding.slack.com/archives/C014UJ5F73Q/p1633558205130000\n",
    "#10X 3' protocol v2 and v3 were Forward-stranded, 5'-v2 is Reverse-stranded. I am not aware of any of their protocols that are unstranded.\n",
    "protocol_defaults = {\n",
    "    \"Chromium v2\": {\n",
    "        \"stranded\": \"Forward\",\n",
    "        \"umi_length\": 10,\n",
    "        #\"inclusion_list_url\": \"https://encd-6248-78c147870-qing.demo.encodedcc.org/files/737K-august-2016/@@download/737K-august-2016.txt.gz\"\n",
    "        #\"inclusion_list_url\": \"https://woldlab.caltech.edu/~diane/genome/737K-arc-v1.txt.gz\"\n",
    "        #\"inclusion_list_url\": \"https://encd-6248-24bd77146-qing.demo.encodedcc.org/files/737K-august-2016/@@download/737K-august-2016.txt.gz\"\n",
    "        \"inclusion_accession\": \"737K-august-2016\"\n",
    "    },\n",
    "    \"Chromium GEM v3.1\": {\n",
    "        \"stranded\": \"Forward\",\n",
    "        \"umi_length\": 12,\n",
    "        #\"inclusion_list_url\": \"https://encd-6248-24bd77146-qing.demo.encodedcc.org/files/3M-february-2018/@@download/3M-february-2018.txt.gz\"\n",
    "        \"inclusion_accession\": \"3M-february-2018\"\n",
    "    },\n",
    "    \"Chromium Multiome ATAC + Gene Expression\": {\n",
    "        \"stranded\": \"Forward\",\n",
    "        \"umi_length\": 12,\n",
    "        \"inclusion_accession\": \"737K-arc-v1(GEX)\",\n",
    "        #\"inclusion_list_url\": \"https://encd-6248-78c147870-qing.demo.encodedcc.org/files/737K-arc-v1(GEX)/@@download/737K-arc-v1(GEX).txt.gz\",\n",
    "        #\"inclusion_list_url\": \"https://woldlab.caltech.edu/~diane/genome/737K-arc-v1(GEX).txt.gz\",\n",
    "        #\"inclusion_list_url\": \"https://encd-6248-24bd77146-qing.demo.encodedcc.org/files/737K-arc-v1(GEX)/@@download/737K-arc-v1(GEX).txt.gz\"\n",
    "    },\n",
    "    \"Split-seq\" : {\n",
    "        \"stranded\": \"Forward\",\n",
    "        \"inclusion_list_url\": \"https://woldlab.caltech.edu/~diane/genome/parse-splitseq-v1.tar.gz\",\n",
    "    }\n",
    "}\n",
    "\n",
    "default_genome_index = {\n",
    "    #\"/organisms/human/\": \"https://encd-6248-78c147870-qing.demo.encodedcc.org/files/TSTFF202658/@@download/TSTFF202658.tar.gz\"\n",
    "    \"/organisms/human/\": {\n",
    "        #\"genome_index_url\": \"https://woldlab.caltech.edu/~diane/genome/GRCh38-V29-male-2.7.8a.tar.gz\",\n",
    "        \"genome_accession\": \"encode:starsolo-GRCh38-V29-male-index\",\n",
    "        #\"assembly\": \"GRCh38\",\n",
    "        #\"genome_annotation\": \"V29\",\n",
    "    },\n",
    "    \"/organisms/mouse/\": {\n",
    "        #\"genome_index_url\": \"https://woldlab.caltech.edu/~diane/genome/mm10-M21-male-2.7.9a/mm10-M21-male-2.7.9a-star.tar.gz\",\n",
    "        \"genome_accession\": \"encode:starsolo-mm10-M21-male-index\",\n",
    "        #\"assembly\": \"mm10\",\n",
    "        #\"genome_annotation\": \"M21\",\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "defensive-combine",
   "metadata": {},
   "outputs": [],
   "source": [
    "server = ENCODED(\"www.encodeproject.org\")\n",
    "#server = ENCODED(\"test.encodedcc.org\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "brilliant-northwest",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GeneFull_Ex50pAS_Unique_filtered.tar.gz\n"
     ]
    }
   ],
   "source": [
    "experiment = server.get_json(\"/experiments/ENCSR203THX/\")\n",
    "has_processed = []\n",
    "for f in experiment[\"files\"]:\n",
    "    submitted_file_name = Path(f[\"submitted_file_name\"])\n",
    "    if f.get(\"output_type\") == \"sparse gene count matrix of unique reads\":\n",
    "        print(submitted_file_name)\n",
    "    elif len(f.get(\"analyses\", [])) > 0:\n",
    "        title = f[\"analyses\"][0][\"title\"]\n",
    "        print(f['accession'], title)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ruled-native",
   "metadata": {},
   "outputs": [],
   "source": [
    "query = server.search_jsonld(type=\"Experiment\", assay_title=\"scRNA-seq\", limit=\"all\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "former-whole",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENCLB523JWN Forward != unstranded\n",
      "ENCLB072RWF Forward != unstranded\n",
      "ENCLB885OKL Forward != unstranded\n",
      "ENCLB849GNT Forward != unstranded\n",
      "ENCLB581NBT Forward != unstranded\n",
      "ENCLB524IGV Forward != unstranded\n",
      "ENCLB010FKE Forward != unstranded\n",
      "ENCLB919KOR Forward != unstranded\n",
      "ENCLB989CJP Forward != unstranded\n",
      "ENCLB010VHV Forward != unstranded\n",
      "ENCLB716MVI Forward != unstranded\n",
      "ENCLB773PPF Forward != unstranded\n",
      "ENCLB902RKO Forward != unstranded\n",
      "ENCLB536TLF Forward != unstranded\n",
      "ENCLB374JXM Forward != unstranded\n",
      "ENCLB343QQX Forward != unstranded\n",
      "ENCLB129SKD Forward != unstranded\n",
      "ENCLB709WAG Forward != unstranded\n",
      "ENCLB692PHT Forward != unstranded\n",
      "ENCLB102VRS Forward != unstranded\n",
      "ENCLB169FGN Forward != unstranded\n",
      "ENCLB988KOO Forward != unstranded\n",
      "ENCLB848YMF Forward != unstranded\n",
      "ENCLB332MGL Forward != unstranded\n",
      "ENCLB323CCI Forward != unstranded\n",
      "ENCLB678GYJ Forward != unstranded\n",
      "ENCLB382NBT Forward != unstranded\n",
      "ENCLB592ZMI Forward != unstranded\n",
      "ENCLB194LYE Forward != unstranded\n",
      "ENCLB735WRJ Forward != unstranded\n",
      "ENCLB354YXN Forward != unstranded\n",
      "ENCLB018ZYA Forward != unstranded\n",
      "ENCLB782BWD Forward != unstranded\n",
      "ENCLB164ZHY Forward != unstranded\n",
      "ENCLB194KKC Forward != unstranded\n",
      "ENCLB954TLZ Forward != unstranded\n",
      "ENCLB306NVG Forward != unstranded\n",
      "ENCLB266ZRP Forward != unstranded\n",
      "ENCLB959AIF Forward != unstranded\n",
      "ENCLB516LRK Forward != unstranded\n",
      "ENCLB061KPS Forward != unstranded\n",
      "ENCLB765ZDM Forward != unstranded\n",
      "ENCLB440ABL Forward != unstranded\n",
      "ENCLB700LWU Forward != unstranded\n",
      "ENCLB336ZNO Forward != unstranded\n",
      "ENCLB747XAC Forward != unstranded\n",
      "ENCLB564RVH Forward != unstranded\n",
      "ENCLB571XKG Forward != unstranded\n",
      "ENCLB822SVY Forward != unstranded\n",
      "ENCLB349QMF Forward != unstranded\n",
      "ENCLB321YXH Forward != unstranded\n",
      "ENCLB611DGC Forward != unstranded\n",
      "ENCLB785FXH Forward != unstranded\n",
      "ENCLB862YMG Forward != unstranded\n",
      "ENCLB227FXC Forward != unstranded\n",
      "ENCLB820VCW Forward != unstranded\n",
      "ENCLB570NOI Forward != unstranded\n",
      "ENCLB143DFU Forward != unstranded\n",
      "ENCLB905AUU Forward != unstranded\n",
      "ENCLB415PDV Forward != unstranded\n",
      "ENCLB722NFY Forward != unstranded\n",
      "ENCLB867BPN Forward != unstranded\n",
      "ENCLB622RLX Forward != unstranded\n",
      "ENCLB072YJA Forward != unstranded\n",
      "ENCLB894TFF Forward != unstranded\n",
      "ENCLB315FQV Forward != unstranded\n",
      "ENCLB727KBM Forward != unstranded\n",
      "ENCLB614YSB Forward != unstranded\n",
      "ENCLB604HIX Forward != unstranded\n",
      "ENCLB797XAC Forward != unstranded\n",
      "ENCLB631JOU Forward != unstranded\n",
      "ENCLB292IDD Forward != unstranded\n",
      "ENCLB715WPG Forward != unstranded\n",
      "ENCLB211BVF Forward != unstranded\n",
      "ENCLB315XSJ Forward != unstranded\n",
      "ENCLB389BSV Forward != unstranded\n",
      "ENCLB771NKO Forward != unstranded\n",
      "ENCLB046ZBE Forward != unstranded\n",
      "ENCLB369KJQ Forward != unstranded\n",
      "ENCLB261MTS Forward != unstranded\n",
      "ENCLB460OEF Forward != unstranded\n",
      "ENCLB792DPM Forward != unstranded\n",
      "ENCLB250XVZ Forward != unstranded\n",
      "ENCLB943OZA Forward != unstranded\n",
      "ENCLB701VGG Forward != unstranded\n",
      "ENCLB889CHC Forward != unstranded\n",
      "ENCLB988YCJ Forward != unstranded\n",
      "ENCLB249TPV Forward != unstranded\n",
      "ENCLB146JLP Forward != unstranded\n",
      "ENCLB193HSH Forward != unstranded\n",
      "ENCLB143IPZ Forward != unstranded\n",
      "ENCLB491BDA Forward != unstranded\n",
      "ENCLB908IUA Forward != unstranded\n",
      "ENCLB697JVD Forward != unstranded\n",
      "ENCLB793OLP Forward != unstranded\n",
      "ENCLB925LWZ Forward != unstranded\n",
      "ENCLB905KRE Forward != unstranded\n",
      "ENCLB082IOS Forward != unstranded\n",
      "ENCLB517TXV Forward != unstranded\n",
      "ENCLB170FEN Forward != unstranded\n",
      "ENCLB274ZBT Forward != unstranded\n",
      "ENCLB983QCI Forward != unstranded\n",
      "ENCLB400DWX Forward != unstranded\n",
      "ENCLB656UZY Forward != unstranded\n",
      "ENCLB690LBQ Forward != unstranded\n",
      "ENCLB457VVB Forward != unstranded\n",
      "ENCLB601XFB Forward != unstranded\n",
      "ENCLB188WUM Forward != unstranded\n",
      "ENCLB368QMP Forward != unstranded\n",
      "ENCLB903BTS Forward != unstranded\n",
      "ENCLB673SUG Forward != unstranded\n",
      "ENCLB891WFR Forward != unstranded\n",
      "ENCLB416NMR Forward != unstranded\n",
      "ENCLB308YIM Forward != unstranded\n",
      "ENCLB963QFG Forward != unstranded\n",
      "ENCLB589IWN Forward != unstranded\n",
      "ENCLB708YOS Forward != unstranded\n",
      "ENCLB548SZL Forward != unstranded\n",
      "ENCLB829GDB Forward != unstranded\n",
      "ENCLB230VGI Forward != unstranded\n",
      "ENCLB603YRZ Forward != unstranded\n",
      "ENCLB048POP Forward != unstranded\n",
      "ENCLB805MAL Forward != unstranded\n",
      "ENCLB762GOP Forward != unstranded\n",
      "ENCLB156LUM Forward != unstranded\n",
      "ENCLB868TVO Forward != unstranded\n",
      "ENCLB399LDV Forward != unstranded\n",
      "ENCLB213LKC Forward != unstranded\n",
      "ENCLB598TEK Forward != unstranded\n",
      "ENCLB231PCO Forward != unstranded\n",
      "ENCLB228MHF Forward != unstranded\n",
      "ENCLB372GFV Forward != unstranded\n",
      "ENCLB489WXX Forward != unstranded\n",
      "ENCLB590EJB Forward != unstranded\n",
      "ENCLB672PSR Forward != unstranded\n",
      "ENCLB661DWA Forward != unstranded\n",
      "ENCLB319YUS Forward != unstranded\n",
      "ENCLB124EIT Forward != unstranded\n",
      "ENCLB586GGX Forward != unstranded\n",
      "ENCLB239BXF Forward != unstranded\n",
      "ENCLB899ORD Forward != unstranded\n",
      "ENCLB798BJP Forward != unstranded\n",
      "ENCLB639GHE Forward != unstranded\n",
      "ENCLB261TVA Forward != unstranded\n",
      "ENCLB935EGG Forward != unstranded\n",
      "ENCLB428UEV Forward != unstranded\n",
      "ENCLB229VWB Forward != unstranded\n",
      "ENCLB583JZE Forward != unstranded\n",
      "ENCLB836HUJ Forward != unstranded\n",
      "ENCLB414YSS Forward != unstranded\n",
      "ENCLB938HOJ Forward != unstranded\n",
      "ENCLB635BYD Forward != unstranded\n",
      "ENCLB497BTJ Forward != unstranded\n",
      "ENCLB042NUM Forward != unstranded\n",
      "ENCLB268GZV Forward != unstranded\n",
      "ENCLB931EHJ Forward != unstranded\n",
      "ENCLB323HIM Forward != unstranded\n",
      "ENCLB063XDV Forward != reverse\n",
      "ENCLB273ZWE Forward != unstranded\n",
      "ENCLB527JDU Forward != unstranded\n",
      "ENCLB544SEN Forward != unstranded\n",
      "ENCLB129RYA Forward != reverse\n",
      "ENCLB299EAG Forward != unstranded\n",
      "ENCLB663TUU Forward != reverse\n",
      "ENCLB297FDO Forward != unstranded\n",
      "ENCLB761CFQ Forward != reverse\n",
      "ENCLB549KAN Forward != unstranded\n",
      "ENCLB179BOW Forward != reverse\n",
      "ENCLB342YHF Forward != unstranded\n",
      "ENCLB225KMP Forward != reverse\n",
      "ENCLB065ONF Forward != reverse\n",
      "ENCLB766ZBL Forward != reverse\n",
      "ENCLB243NMQ Forward != reverse\n",
      "ENCLB338RBX Forward != unstranded\n",
      "ENCLB234BIK Forward != unstranded\n",
      "ENCLB275OVJ Forward != reverse\n",
      "ENCLB568RWC Forward != reverse\n",
      "ENCLB043KZW Forward != reverse\n",
      "ENCLB847VJF Forward != reverse\n",
      "ENCLB129LMS Forward != unstranded\n",
      "ENCLB892ZAW Forward != unstranded\n",
      "ENCLB977UQY Forward != unstranded\n",
      "ENCLB954OKJ Forward != reverse\n",
      "ENCLB849DVZ Forward != unstranded\n",
      "ENCLB560GBC Forward != reverse\n",
      "ENCLB113GDH Forward != reverse\n",
      "ENCLB069PRB Forward != reverse\n",
      "ENCLB513JLQ Forward != unstranded\n",
      "ENCLB814PUF Forward != unstranded\n",
      "ENCLB104XFX Forward != reverse\n",
      "ENCLB002DZK Forward != unstranded\n",
      "ENCLB067EKD Forward != reverse\n",
      "ENCLB765NUA Forward != unstranded\n",
      "ENCLB151ZAG Forward != unstranded\n",
      "ENCLB952MZJ Forward != unstranded\n",
      "ENCLB514RKT Forward != unstranded\n",
      "ENCLB294FBZ Forward != unstranded\n",
      "ENCLB293BAX Forward != reverse\n",
      "ENCLB607FDJ Forward != reverse\n",
      "ENCLB349KYC Forward != reverse\n",
      "ENCLB971CVF Forward != unstranded\n",
      "ENCLB765JXX Forward != reverse\n",
      "ENCLB675MBI Forward != reverse\n",
      "ENCLB062TYC Forward != unstranded\n",
      "ENCLB061STU Forward != reverse\n",
      "ENCLB335ITX Forward != unstranded\n",
      "ENCLB089YWN Forward != reverse\n",
      "ENCLB943WKY Forward != reverse\n",
      "ENCLB280ZGL Forward != unstranded\n",
      "ENCLB449BXX Forward != unstranded\n",
      "ENCLB038DYV Forward != reverse\n",
      "ENCLB170MZP Forward != unstranded\n",
      "ENCLB908NNT Forward != unstranded\n",
      "ENCLB392SOP Forward != reverse\n",
      "ENCLB983BCX Forward != reverse\n",
      "ENCLB399EQG Forward != reverse\n",
      "ENCLB025GOX Forward != reverse\n",
      "ENCLB000QLP Forward != reverse\n",
      "ENCLB801PEP Forward != reverse\n",
      "ENCLB673QFY Forward != reverse\n",
      "ENCLB681VIY Forward != reverse\n",
      "ENCLB802MGH Forward != reverse\n",
      "ENCLB442AMV Forward != reverse\n",
      "ENCLB835IXS Forward != reverse\n",
      "ENCLB997ETR Forward != reverse\n",
      "ENCLB508DVG Forward != reverse\n",
      "ENCLB807VEG Forward != reverse\n",
      "ENCLB012QYK Forward != reverse\n",
      "ENCLB455UTN Forward != reverse\n",
      "ENCLB808STP Forward != reverse\n",
      "ENCLB396URL Forward != unstranded\n",
      "ENCLB123ZDT Forward != reverse\n",
      "ENCLB701RZY Forward != reverse\n",
      "ENCLB850RYI Forward != reverse\n",
      "ENCLB320HXY Forward != reverse\n",
      "ENCLB296LDF Forward != unstranded\n",
      "ENCLB196FUF Forward != reverse\n",
      "ENCLB139QES Forward != reverse\n",
      "ENCLB846XMI Forward != not set\n",
      "ENCLB228GCE Forward != reverse\n",
      "ENCLB039THH Forward != unstranded\n",
      "ENCLB960RBM Forward != unstranded\n",
      "ENCLB774YSO Forward != unstranded\n",
      "ENCLB986LXE Forward != reverse\n",
      "ENCLB212WDJ Forward != reverse\n",
      "ENCLB943KZN Forward != reverse\n",
      "ENCLB504FSE Forward != reverse\n",
      "ENCLB549SMV Forward != unstranded\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENCLB357XZN Forward != reverse\n",
      "ENCLB611BAN Forward != reverse\n",
      "ENCLB025XJD Forward != unstranded\n",
      "ENCLB151JZY Forward != reverse\n",
      "ENCLB807HIR Forward != reverse\n",
      "ENCLB866MAL Forward != reverse\n",
      "ENCLB814OYP Forward != unstranded\n",
      "ENCLB592LXS Forward != unstranded\n",
      "ENCLB615BOP Forward != unstranded\n",
      "ENCLB812SMT Forward != unstranded\n",
      "ENCLB608YNH Forward != reverse\n",
      "ENCLB932FNN Forward != reverse\n",
      "ENCLB181TPV Forward != unstranded\n",
      "ENCLB495DSX Forward != reverse\n",
      "ENCLB327PYA Forward != reverse\n",
      "ENCLB063JPD Forward != reverse\n",
      "ENCLB871PHY Forward != reverse\n",
      "ENCLB328JIP Forward != unstranded\n",
      "ENCLB713ETA Forward != reverse\n",
      "ENCLB676AMA Forward != reverse\n",
      "ENCLB418IDT Forward != reverse\n",
      "ENCLB272QLF Forward != reverse\n",
      "ENCLB578ADC Forward != unstranded\n",
      "ENCLB103OKD Forward != reverse\n",
      "ENCLB432ZEQ Forward != reverse\n",
      "ENCLB643HGB Forward != unstranded\n",
      "ENCLB674XYV Forward != reverse\n",
      "ENCLB909DQG Forward != unstranded\n",
      "ENCLB831IHU Forward != unstranded\n",
      "ENCLB619VVF Forward != unstranded\n",
      "ENCLB527WWJ Forward != unstranded\n",
      "matching_strand 0\n"
     ]
    }
   ],
   "source": [
    "documents_seen = Counter()\n",
    "results = []\n",
    "needed_metadata = []\n",
    "experiment_cache = {}\n",
    "matching_strand = 0\n",
    "for row in query[\"@graph\"]:\n",
    "    if row[\"status\"] in (\"in progress\", \"released\"):\n",
    "        experiment = server.get_json(row[\"@id\"])\n",
    "        documents = set(experiment.get(\"documents\", []))\n",
    "        construction_method = set()\n",
    "        for replicate in experiment['replicates']:\n",
    "            library = replicate['library']\n",
    "            library_barcode = library.get(\"barcode_details\")\n",
    "            for d in library.get(\"documents\", []):\n",
    "                documents.add(d)\n",
    "            construction_platform = library.get('construction_platform', None)\n",
    "            if construction_platform is not None:                \n",
    "                construction_method.add(construction_platform[\"term_name\"])\n",
    "            construction_platform = library.get(\"construction_method\", None)\n",
    "            if construction_platform is not None:\n",
    "                construction_method.add(construction_platform[0])\n",
    "\n",
    "            biosample = library[\"biosample\"]\n",
    "            for d in biosample.get(\"documents\", []):\n",
    "                documents.add(d)\n",
    "            subcellular_fraction = biosample.get(\"subcellular_fraction_term_name\", \"\")\n",
    "                \n",
    "        has_processed = False\n",
    "        for f in experiment[\"files\"]:\n",
    "            submitted_file_name = Path(f[\"submitted_file_name\"])            \n",
    "            if f.get(\"output_type\") == \"sparse gene count matrix of unique reads\":\n",
    "                has_processed = True\n",
    "\n",
    "        sc_type = set()\n",
    "\n",
    "        for d in documents:\n",
    "            assert d in known_document_types\n",
    "            if d in document_to_protocol:\n",
    "                sc_type.add(document_to_protocol[d])\n",
    "        if len(experiment[\"documents\"]) > 0:\n",
    "            documents_seen[\",\".join(experiment[\"documents\"])] += 1\n",
    "          \n",
    "        if len(sc_type) == 1:\n",
    "            sc_type = sc_type.pop()\n",
    "        elif len(sc_type) < 1:\n",
    "            sc_type = \"Unknown type: {}\".format(experiment[\"accession\"], documents)\n",
    "        elif len(sc_type) > 1:\n",
    "            sc_type = \"Mixed types: {}\".format(experiment[\"accession\"], sc_type)\n",
    "        \n",
    "        strand = protocol_defaults.get(sc_type, {}).get(\"stranded\", None)\n",
    "        \n",
    "        if strand is not None:\n",
    "            strand_specificity = library.get(\"strand_specificity\", \"not set\").lower()\n",
    "            if strand.lower() != strand_specificity:\n",
    "                print(\"{} {} != {}\".format(library[\"accession\"], strand, strand_specificity))\n",
    "            else:\n",
    "                matching_strand += 1\n",
    "        \n",
    "        if (sc_type.startswith(\"Chromium\") or sc_type.startswith(\"Split-seq\")):\n",
    "            could_process = True\n",
    "        else:\n",
    "            could_process = False\n",
    "        \n",
    "        experiment_cache[row[\"accession\"]] = experiment\n",
    "        results.append({\n",
    "            \"experiment\": row[\"accession\"], \n",
    "            \"library\": library[\"accession\"],\n",
    "            \"status\": row[\"status\"],\n",
    "            \"could_process\": \"Y\" if could_process else \"N\",\n",
    "            \"lab\": row[\"lab\"][\"title\"].split(',')[1].strip(), \n",
    "            \"description\": experiment.get(\"description\", \"\"),\n",
    "            \"documents\": documents,\n",
    "            \"document_len\": len(documents),\n",
    "            \"construction_method\": \",\".join(construction_method),\n",
    "            #\"library_barcode\": library_barcode,\n",
    "            #\"library_description\": library.get(\"description\", \"\"),\n",
    "            \"biosample_ontology\": biosample[\"biosample_ontology\"][\"@id\"],\n",
    "            \"biosample_term\": biosample[\"biosample_ontology\"][\"term_name\"],\n",
    "            \"sc_type\": sc_type,\n",
    "            \"organism\": biosample[\"organism\"][\"name\"],\n",
    "            \"subcellular_fraction\": subcellular_fraction,\n",
    "            \"has_processed\": \"Y\" if has_processed else \"N\",\n",
    "        })\n",
    "        \n",
    "print(\"matching_strand\", matching_strand)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "distinct-surge",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Experiment   Status       Lab        Ready Done  Organism Cellular sc_type              bioterm\n",
      "ENCSR005TCL  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT psoas muscle\n",
      "ENCSR011EDH  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT fallopian tube\n",
      "ENCSR012APQ  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT heart left ventricle\n",
      "ENCSR021QIA  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT placenta\n",
      "ENCSR021YYL  in progress  Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT adrenal gland\n",
      "ENCSR028MBY  in progress  Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    mucosa of descending colon\n",
      "ENCSR035BDI  in progress  Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    adrenal gland\n",
      "ENCSR060SEA  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT bile duct\n",
      "ENCSR060ZXZ  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT heart right ventricle\n",
      "ENCSR075HVN  in progress  Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT layer of hippocampus\n",
      "ENCSR079LVG  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT left ventricle myocardium superior\n",
      "ENCSR087YKG  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT right cardiac atrium\n",
      "ENCSR132MEW  in progress  Stanford   Y     Y     mouse             Chromium Multiome AT left cerebral cortex\n",
      "ENCSR133XJX  in progress  Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT cerebral cortex\n",
      "ENCSR142RRV  in progress  Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    upper lobe of right lung\n",
      "ENCSR151ONW  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT Right ventricle myocardium inferior\n",
      "ENCSR157YJO  in progress  Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT adrenal gland\n",
      "ENCSR201WJO  in progress  Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    mucosa of gallbladder\n",
      "ENCSR203THX  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT psoas muscle\n",
      "ENCSR224OUG  in progress  Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT adrenal gland\n",
      "ENCSR224RWH  in progress  Stanford   Y     Y     mouse             Chromium Multiome AT left cerebral cortex\n",
      "ENCSR225OII  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT lung\n",
      "ENCSR258SZI  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT heart right ventricle\n",
      "ENCSR259VOY  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT heart left ventricle\n",
      "ENCSR261BXB  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT pancreas\n",
      "ENCSR261LPM  in progress  Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    right lobe of liver\n",
      "ENCSR281NBH  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT pancreas\n",
      "ENCSR281RIH  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT psoas muscle\n",
      "ENCSR289ANH  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT left colon\n",
      "ENCSR297ZTQ  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT ovary\n",
      "ENCSR298MPV  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT ovary\n",
      "ENCSR333SLJ  in progress  Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT heart\n",
      "ENCSR337HKS  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT natural killer cell\n",
      "ENCSR352DXB  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT heart left ventricle\n",
      "ENCSR356VJZ  in progress  Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT adrenal gland\n",
      "ENCSR362HUP  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT ovary\n",
      "ENCSR362YDM  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT adrenal gland\n",
      "ENCSR367YUX  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT right lobe of liver\n",
      "ENCSR369UFT  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT left lobe of liver\n",
      "ENCSR398PIW  in progress  Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    right lobe of liver\n",
      "ENCSR402PHR  in progress  Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT adrenal gland\n",
      "ENCSR405YKM  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT heart left ventricle\n",
      "ENCSR415HEP  in progress  Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT heart\n",
      "ENCSR417FCM  in progress  Stanford   Y     Y     mouse             Chromium Multiome AT heart\n",
      "ENCSR421YXX  in progress  Stanford   Y     Y     mouse             Chromium Multiome AT left cerebral cortex\n",
      "ENCSR426PDU  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT heart right ventricle\n",
      "ENCSR439ZVQ  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT heart left ventricle\n",
      "ENCSR445LCA  in progress  Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    pancreas\n",
      "ENCSR460MGC  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT mucosa of descending colon\n",
      "ENCSR462HCD  in progress  Stanford   Y     Y     mouse             Chromium Multiome AT heart\n",
      "ENCSR463EEP  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT left cardiac atrium\n",
      "ENCSR467GGT  in progress  Stanford   Y     Y     mouse             Chromium Multiome AT left cerebral cortex\n",
      "ENCSR471YTA  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT psoas muscle\n",
      "ENCSR472RRP  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT pancreas\n",
      "ENCSR478FQR  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT pancreas\n",
      "ENCSR481QQR  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT heart left ventricle\n",
      "ENCSR484YSK  in progress  Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    right lobe of liver\n",
      "ENCSR522HNR  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT right cardiac atrium\n",
      "ENCSR527MGL  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT left cardiac atrium\n",
      "ENCSR533IGQ  in progress  Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT heart\n",
      "ENCSR539XHZ  in progress  Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT layer of hippocampus\n",
      "ENCSR548HED  in progress  Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    kidney\n",
      "ENCSR568GGW  in progress  Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT heart\n",
      "ENCSR587SPO  in progress  Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    left colon\n",
      "ENCSR602SGU  in progress  Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    right lobe of liver\n",
      "ENCSR620AJZ  in progress  Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    adrenal gland\n",
      "ENCSR624WDN  in progress  Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT cerebral cortex\n",
      "ENCSR632EIK  in progress  Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT cerebral cortex\n",
      "ENCSR641MWO  in progress  Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    psoas muscle\n",
      "ENCSR642GFM  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT fallopian tube\n",
      "ENCSR644VVE  in progress  Stanford   Y     Y     mouse             Chromium Multiome AT heart\n",
      "ENCSR656VWP  in progress  Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    colonic mucosa\n",
      "ENCSR662VTU  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT upper lobe of left lung\n",
      "ENCSR665HPR  in progress  Stanford   Y     Y     mouse             Chromium Multiome AT heart\n",
      "ENCSR671FDY  in progress  Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    lower lobe of right lung\n",
      "ENCSR683ONV  in progress  Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    lower lobe of left lung\n",
      "ENCSR684KYI  in progress  Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    pancreas\n",
      "ENCSR700UVN  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT left cardiac atrium\n",
      "ENCSR707ECV  in progress  Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    colonic mucosa\n",
      "ENCSR707LLI  in progress  Stanford   Y     Y     human    nucleus  Chromium GEM v3.1    mucosa of descending colon\n",
      "ENCSR716VLR  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT posterior vena cava\n",
      "ENCSR720TUP  in progress  Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT cerebral cortex\n",
      "ENCSR724KET  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT adrenal gland\n",
      "ENCSR726IPC  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT adrenal gland\n",
      "ENCSR727OYO  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT heart left ventricle\n",
      "ENCSR742NWO  in progress  Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT layer of hippocampus\n",
      "ENCSR749GDE  in progress  Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT adrenal gland\n",
      "ENCSR751BHQ  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT heart left ventricle\n",
      "ENCSR753YPW  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT lower lobe of left lung\n",
      "ENCSR755CIF  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT heart left ventricle\n",
      "ENCSR762LZV  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT uterus\n",
      "ENCSR769KWM  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT bile duct\n",
      "ENCSR789PLK  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT placenta\n",
      "ENCSR801DHT  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT heart left ventricle\n",
      "ENCSR813PWQ  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT Right ventricle myocardium superior\n",
      "ENCSR835JFN  in progress  Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT layer of hippocampus\n",
      "ENCSR876EUH  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT ovary\n",
      "ENCSR908CQZ  in progress  Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT adrenal gland\n",
      "ENCSR938CZJ  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT ovary\n",
      "ENCSR943HBF  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT right cardiac atrium\n",
      "ENCSR944XPC  in progress  Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT layer of hippocampus\n",
      "ENCSR964JZN  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT ureter\n",
      "ENCSR966DDY  in progress  Stanford   Y     Y     human    nucleus  Chromium Multiome AT left lung\n",
      "ENCSR968RPR  in progress  Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT layer of hippocampus\n",
      "ENCSR987SOK  in progress  Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT layer of hippocampus\n",
      "ENCSR988AMY  in progress  Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT adrenal gland\n",
      "ENCSR995ZKH  in progress  Stanford   Y     Y     mouse    nucleus  Chromium Multiome AT layer of hippocampus\n",
      "ENCSR002GJC  in progress  UCI        N     N     mouse    nucleus  Unknown type: ENCSR0 gastrocnemius\n",
      "ENCSR003FZY  in progress  UCI        Y     N     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR024CHH  in progress  UCI        Y     N     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR029GVP  in progress  UCI        Y     N     mouse    nucleus  Split-seq            heart\n",
      "ENCSR029OIF  in progress  UCI        Y     N     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR032BAX  in progress  UCI        N     N     mouse    nucleus  Unknown type: ENCSR0 adrenal gland\n",
      "ENCSR035SWS  in progress  UCI        Y     N     mouse    nucleus  Split-seq            heart\n",
      "ENCSR043HKT  in progress  UCI        Y     N     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR044GSN  in progress  UCI        Y     N     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR054JGY  in progress  UCI        N     N     mouse    nucleus  Unknown type: ENCSR0 gastrocnemius\n",
      "ENCSR062ICV  in progress  UCI        Y     N     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR062WSS  in progress  UCI        Y     N     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR080VLK  in progress  UCI        Y     N     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR081SRK  in progress  UCI        Y     N     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR087LWL  in progress  UCI        Y     N     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR090OFY  in progress  UCI        Y     N     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR093ZTB  in progress  UCI        Y     N     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR096PUV  in progress  UCI        Y     N     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR105JHX  in progress  UCI        Y     N     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR115JCE  in progress  UCI        Y     N     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR119HVP  in progress  UCI        Y     N     mouse    nucleus  Split-seq            heart\n",
      "ENCSR120UHU  in progress  UCI        Y     N     mouse    nucleus  Split-seq            heart\n",
      "ENCSR126YAO  in progress  UCI        Y     N     mouse    nucleus  Split-seq            heart\n",
      "ENCSR135ELG  in progress  UCI        Y     N     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR145DBY  in progress  UCI        Y     N     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR146AFJ  in progress  UCI        Y     N     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR147UYS  in progress  UCI        Y     N     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR156BQX  in progress  UCI        Y     N     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR162LWP  in progress  UCI        Y     N     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR164REI  in progress  UCI        Y     N     mouse    nucleus  Split-seq            heart\n",
      "ENCSR165CAC  in progress  UCI        Y     N     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR169QXA  in progress  UCI        Y     N     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR170HAO  in progress  UCI        N     N     mouse    nucleus  Unknown type: ENCSR1 gastrocnemius\n",
      "ENCSR171BEG  in progress  UCI        Y     N     mouse    nucleus  Split-seq            heart\n",
      "ENCSR172XWR  in progress  UCI        Y     N     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR178WRO  in progress  UCI        Y     N     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR186OUD  in progress  UCI        Y     N     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR186TDH  in progress  UCI        Y     N     mouse    nucleus  Split-seq            heart\n",
      "ENCSR188BZZ  in progress  UCI        Y     N     mouse    nucleus  Split-seq            heart\n",
      "ENCSR191YNY  in progress  UCI        Y     N     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR192TYH  in progress  UCI        Y     N     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR193PLQ  in progress  UCI        Y     N     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR204FUI  in progress  UCI        Y     N     mouse    nucleus  Split-seq            heart\n",
      "ENCSR226SBC  in progress  UCI        Y     N     mouse    nucleus  Split-seq            heart\n",
      "ENCSR227CPT  in progress  UCI        Y     N     mouse    nucleus  Split-seq            heart\n",
      "ENCSR236MGX  in progress  UCI        Y     N     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR237EZS  in progress  UCI        Y     N     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR240CYU  in progress  UCI        Y     N     mouse    nucleus  Split-seq            heart\n",
      "ENCSR244FME  in progress  UCI        Y     N     mouse    nucleus  Split-seq            heart\n",
      "ENCSR248IOH  in progress  UCI        Y     N     mouse    nucleus  Split-seq            heart\n",
      "ENCSR250ZVV  in progress  UCI        Y     N     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR266HBP  in progress  UCI        Y     N     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR268TST  in progress  UCI        Y     N     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR284OLV  in progress  UCI        Y     N     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR287VRB  in progress  UCI        Y     N     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR299DGH  in progress  UCI        Y     N     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR306ZET  in progress  UCI        Y     N     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR311TRI  in progress  UCI        Y     N     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR321BZA  in progress  UCI        Y     N     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR322KUB  in progress  UCI        Y     N     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR327PSD  in progress  UCI        Y     N     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR331DYV  in progress  UCI        Y     N     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR331KXG  in progress  UCI        Y     N     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR334YXW  in progress  UCI        Y     N     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR345BZU  in progress  UCI        Y     N     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR356NFK  in progress  UCI        Y     N     mouse    nucleus  Split-seq            heart\n",
      "ENCSR356PVK  in progress  UCI        Y     N     mouse    nucleus  Split-seq            heart\n",
      "ENCSR358XDG  in progress  UCI        Y     N     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR366MPW  in progress  UCI        Y     N     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR382ZGH  in progress  UCI        Y     N     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR388RND  in progress  UCI        Y     N     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR389TZP  in progress  UCI        Y     N     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR390GEG  in progress  UCI        Y     N     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR400PBH  in progress  UCI        Y     N     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR405CPX  in progress  UCI        Y     N     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR405RFP  in progress  UCI        Y     N     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR408PZM  in progress  UCI        Y     N     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR408ZUR  in progress  UCI        Y     N     mouse    nucleus  Split-seq            heart\n",
      "ENCSR415CYS  in progress  UCI        Y     N     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR421VUT  in progress  UCI        Y     N     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR423AXI  in progress  UCI        Y     N     mouse    nucleus  Split-seq            heart\n",
      "ENCSR423CNV  in progress  UCI        Y     N     mouse    nucleus  Split-seq            heart\n",
      "ENCSR423LFS  in progress  UCI        Y     N     mouse    nucleus  Split-seq            heart\n",
      "ENCSR424RYS  in progress  UCI        Y     N     mouse    nucleus  Split-seq            heart\n",
      "ENCSR426MLQ  in progress  UCI        Y     N     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR427OIY  in progress  UCI        Y     N     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR429ZKT  in progress  UCI        Y     N     mouse    nucleus  Split-seq            heart\n",
      "ENCSR431TUC  in progress  UCI        N     N     mouse    nucleus  Unknown type: ENCSR4 adrenal gland\n",
      "ENCSR441YBL  in progress  UCI        Y     N     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR462YAG  in progress  UCI        Y     N     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR470GMO  in progress  UCI        Y     N     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR477EIC  in progress  UCI        Y     N     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR488RZS  in progress  UCI        Y     N     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR489KPP  in progress  UCI        Y     N     mouse    nucleus  Split-seq            heart\n",
      "ENCSR493ZQU  in progress  UCI        Y     N     mouse    nucleus  Split-seq            heart\n",
      "ENCSR495FGW  in progress  UCI        N     N     mouse    nucleus  Unknown type: ENCSR4 gastrocnemius\n",
      "ENCSR498KBD  in progress  UCI        Y     N     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR510HAO  in progress  UCI        Y     N     mouse    nucleus  Split-seq            heart\n",
      "ENCSR511XOL  in progress  UCI        Y     N     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR519HYQ  in progress  UCI        Y     N     mouse    nucleus  Split-seq            heart\n",
      "ENCSR526WJT  in progress  UCI        Y     N     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR536EAF  in progress  UCI        Y     N     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR542LRP  in progress  UCI        Y     N     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR555ATU  in progress  UCI        Y     N     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR560WNI  in progress  UCI        Y     N     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR561PWI  in progress  UCI        Y     N     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR563WPL  in progress  UCI        Y     N     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR566XHU  in progress  UCI        Y     N     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR576RBX  in progress  UCI        Y     N     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR586LMA  in progress  UCI        Y     N     mouse    nucleus  Split-seq            heart\n",
      "ENCSR595LVS  in progress  UCI        Y     N     mouse    nucleus  Split-seq            heart\n",
      "ENCSR601CQM  in progress  UCI        Y     N     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR635GFL  in progress  UCI        Y     N     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR635GRA  in progress  UCI        Y     N     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR637AZA  in progress  UCI        Y     N     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR642TKQ  in progress  UCI        Y     N     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR644EUQ  in progress  UCI        Y     N     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR644YWO  in progress  UCI        Y     N     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR651QGH  in progress  UCI        Y     N     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR671RJG  in progress  UCI        Y     N     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR692OZP  in progress  UCI        Y     N     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR692ZWB  in progress  UCI        Y     N     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR694DQP  in progress  UCI        Y     N     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR732MDS  in progress  UCI        Y     N     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR745UMG  in progress  UCI        Y     N     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR773FPF  in progress  UCI        Y     N     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR774KTU  in progress  UCI        Y     N     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR799GLH  in progress  UCI        Y     N     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR801MJO  in progress  UCI        Y     N     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR813ULV  in progress  UCI        Y     N     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR827YHY  in progress  UCI        Y     N     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR828TLP  in progress  UCI        Y     N     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR830MSX  in progress  UCI        Y     N     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR834SCG  in progress  UCI        Y     N     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR847STS  in progress  UCI        Y     N     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR851RNW  in progress  UCI        Y     N     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR852NQO  in progress  UCI        Y     N     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR855ZYP  in progress  UCI        Y     N     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR877CFF  in progress  UCI        Y     N     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR877UAI  in progress  UCI        Y     N     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR880DYE  in progress  UCI        Y     N     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR880MVS  in progress  UCI        Y     N     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR888NNP  in progress  UCI        Y     N     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR894JVL  in progress  UCI        Y     N     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR905YQE  in progress  UCI        Y     N     mouse    nucleus  Split-seq            left cerebral cortex\n",
      "ENCSR910HZB  in progress  UCI        N     N     mouse    nucleus  Unknown type: ENCSR9 layer of hippocampus\n",
      "ENCSR912IHV  in progress  UCI        Y     N     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR915UNZ  in progress  UCI        Y     N     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR923NSW  in progress  UCI        Y     N     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR929PEE  in progress  UCI        Y     N     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR932DTM  in progress  UCI        Y     N     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR948INJ  in progress  UCI        N     N     mouse    nucleus  Unknown type: ENCSR9 gastrocnemius\n",
      "ENCSR948XEM  in progress  UCI        Y     N     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR962DRE  in progress  UCI        Y     N     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR969VQQ  in progress  UCI        Y     N     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR972KJU  in progress  UCI        Y     N     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR973ZJO  in progress  UCI        Y     N     mouse    nucleus  Split-seq            heart\n",
      "ENCSR974VTR  in progress  UCI        Y     N     mouse    nucleus  Split-seq            heart\n",
      "ENCSR976FIL  in progress  UCI        Y     N     mouse    nucleus  Split-seq            adrenal gland\n",
      "ENCSR983WDT  in progress  UCI        Y     N     mouse    nucleus  Split-seq            layer of hippocampus\n",
      "ENCSR985FZQ  in progress  UCI        Y     N     mouse    nucleus  Split-seq            heart\n",
      "ENCSR986QTH  in progress  UCI        Y     N     mouse    nucleus  Split-seq            gastrocnemius\n",
      "ENCSR000AIZ  released     Caltech    N     N     human             SmartSeq Pool        GM12878\n",
      "ENCSR000AJG  released     Caltech    N     N     human             SmartSeq Pool        GM12878\n",
      "ENCSR137RHZ  released     Caltech    N     N     human             SmartSeq             pyramidal neuron\n",
      "ENCSR156HHE  released     Caltech    Y     N     mouse             Chromium v2          forelimb\n",
      "ENCSR157NLF  released     Caltech    N     N     human             SmartSeq             Purkinje cell\n",
      "ENCSR236EGS  released     Caltech    N     N     mouse             SmartSeq             cerebral cortex, layer 5\n",
      "ENCSR268AXN  released     Caltech    N     N     mouse             SmartSeq             pyramidal neuron\n",
      "ENCSR306UHC  released     Caltech    Y     N     mouse             Chromium v2          forelimb\n",
      "ENCSR313IUO  released     Caltech    N     N     human             SmartSeq             cerebellar granule cell\n",
      "ENCSR337QTQ  released     Caltech    Y     N     mouse             Chromium v2          forelimb\n",
      "ENCSR415JJO  released     Caltech    N     N     human             SmartSeq             cerebellar granule cell\n",
      "ENCSR416FHN  released     Caltech    Y     N     mouse             Chromium v2          autopod region\n",
      "ENCSR423XRG  released     Caltech    N     N     human             SmartSeq             pyramidal neuron\n",
      "ENCSR476JWN  released     Caltech    Y     N     mouse             Chromium v2          forelimb\n",
      "ENCSR642XVO  released     Caltech    Y     N     mouse             Chromium v2          forelimb\n",
      "ENCSR698TMC  released     Caltech    Y     N     mouse             Chromium v2          forelimb zygopod\n",
      "ENCSR745DNK  released     Caltech    Y     N     mouse             Chromium v2          forelimb\n",
      "ENCSR758TAD  released     Caltech    N     N     mouse             SmartSeq             Purkinje cell\n",
      "ENCSR787QXE  released     Caltech    Y     N     mouse             Chromium v2          forelimb\n",
      "ENCSR792LNZ  released     Caltech    Y     N     mouse             Chromium v2          forelimb\n",
      "ENCSR805QYO  released     Caltech    N     N     mouse             SmartSeq             Purkinje cell\n",
      "ENCSR814JMM  released     Caltech    N     N     mouse             SmartSeq             cerebellar granule layer\n",
      "ENCSR822EWO  released     Caltech    N     N     mouse             SmartSeq             cerebellar granule layer\n",
      "ENCSR847RRR  released     Caltech    Y     N     mouse             Chromium v2          forelimb stylopod\n",
      "ENCSR874BOF  released     Caltech    Y     N     mouse             Chromium v2          forelimb\n",
      "ENCSR109DAN  released     UCI        Y     N     mouse    nucleus  Split-seq            myotube\n",
      "ENCSR388EIN  released     UCI        Y     N     mouse             Split-seq            C2C12\n",
      "ENCSR452OYH  released     UCI        Y     N     mouse    nucleus  Split-seq            C2C12\n",
      "ENCSR503QZJ  released     UCI        Y     N     mouse             Split-seq            C2C12\n",
      "ENCSR572SMD  released     UCI        Y     N     mouse    nucleus  Split-seq            myotube\n",
      "ENCSR877NAX  released     UCI        Y     N     mouse    nucleus  Split-seq            C2C12\n",
      "ENCSR562QQA  released     UMass      N     N     mouse             inDrop               dendritic cell\n",
      "ENCSR568FZM  released     UMass      N     N     mouse             inDrop               dendritic cell\n",
      "ENCSR597YMV  released     UMass      N     N     mouse             inDrop               dendritic cell\n"
     ]
    }
   ],
   "source": [
    "#report_format = \"{accession:12} {status:12} {lab:10} {document_len:3} {has_processed:4} {construction_method:35.35} {sc_type:20}\"\n",
    "#  {document_len:3}\n",
    "report_format = \"{experiment:12} {status:12} {lab:10} {could_process:5} {has_processed:5} {organism:8.8} {subcellular_fraction:8.8} {sc_type:20.20} {biosample_term}\"\n",
    "print(report_format.format(**{\n",
    "    \"experiment\": \"Experiment\",\n",
    "    \"description\": \"Description\",\n",
    "    \"status\": \"Status\", \n",
    "    \"lab\": \"Lab\", \n",
    "    \"document_len\": \"Doc\", \n",
    "    #\"construction\",\n",
    "    \"has_processed\": \"Done\",\n",
    "    \"could_process\": \"Ready\",\n",
    "    \"organism\": \"Organism\",\n",
    "    #\"construction_method\": \"Const. Method\",\n",
    "    \"subcellular_fraction\": \"Cellular\",\n",
    "    #\"biosample_ontology\": \"biosample ontology\",\n",
    "    \"biosample_term\": \"bioterm\",\n",
    "    \"sc_type\": \"sc_type\",\n",
    "}))\n",
    "for row in sorted(results, key=lambda x: (x[\"status\"], x[\"lab\"], x[\"experiment\"])):\n",
    "    #print(report_format.format(row[0], row[1], row[2], len(row[3]), \",\".join(row[4])[:20], \",\".join(row[5])))\n",
    "    print(report_format.format(**row))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "latin-seventh",
   "metadata": {},
   "outputs": [],
   "source": [
    "pandas.DataFrame(results).sort_values([\"status\", \"lab\", \"experiment\"]).to_excel(\"scrna-experiments.xlsx\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "least-christmas",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unprocessed 0 of 303\n"
     ]
    }
   ],
   "source": [
    "print(\"Unprocessed {} of {}\".format(sum([not x[\"has_processed\"] for x in results]), len(results)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "specific-settle",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unprocessed 0 of 303\n"
     ]
    }
   ],
   "source": [
    "print(\"Unprocessed {} of {}\".format(sum([(not x[\"has_processed\"] and x[\"sc_type\"].startswith(\"Chromium\")) for x in results]), len(results)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "desperate-visiting",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unprocessed 107 of 303\n"
     ]
    }
   ],
   "source": [
    "print(\"Unprocessed {} of {}\".format(sum([(x[\"could_process\"] and x[\"sc_type\"].startswith(\"Chromium\") and x[\"lab\"].startswith(\"Stanford\"))  for x in results]), len(results)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "driving-teddy",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unprocessed 0 of 303\n"
     ]
    }
   ],
   "source": [
    "print(\"Unprocessed {} of {}\".format(sum([(not x[\"has_processed\"] and x[\"sc_type\"].startswith(\"Split-seq\")) for x in results]), len(results)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "suspended-facial",
   "metadata": {},
   "outputs": [],
   "source": [
    "# First I figured out what documents were present, then I made my mapping then \n",
    "# I produced the above report, but I lost generating what documents existed.\n",
    "if 0:\n",
    "    documents_seen = Counter()\n",
    "    results = []\n",
    "    needed_metadata = []\n",
    "    for row in query[\"@graph\"]:\n",
    "        if row[\"status\"] in (\"in progress\", \"released\"):\n",
    "\n",
    "            experiment = server.get_json(row[\"@id\"])\n",
    "            documents = set(experiment.get(\"documents\", []))\n",
    "            construction_method = set()\n",
    "            for replicate in experiment['replicates']:\n",
    "                library = replicate['library']\n",
    "                library_barcode = library.get(\"barcode_details\")\n",
    "                for d in library.get(\"documents\", []):\n",
    "                    documents.add(d)\n",
    "                construction_platform = library.get('construction_platform', None)\n",
    "                if construction_platform is not None:                \n",
    "                    construction_method.add(construction_platform[\"term_name\"])\n",
    "                construction_platform = library.get(\"construction_method\", None)\n",
    "                if construction_platform is not None:\n",
    "                    construction_method.add(construction_platform[0])\n",
    "\n",
    "                biosample = library[\"biosample\"]\n",
    "                for d in biosample.get(\"documents\", []):\n",
    "                    documents.add(d)\n",
    "                subcellular_fraction = biosample.get(\"subcellular_fraction_term_name\")\n",
    "\n",
    "            has_processed = False\n",
    "            for f in experiment[\"files\"]:\n",
    "                if f[\"file_type\"] == \"fastq\":\n",
    "                    submitted_file_name = Path(f[\"submitted_file_name\"])\n",
    "                    print(experiment[\"accession\"], f[\"accession\"], f.get(\"paired_end\"), f.get(\"paired_with\"), submitted_file_name.name)\n",
    "                    #for detail in f.get(\"flowcell_details\", []):                    \n",
    "                        #for barcode_attribute in [\"barcode_in_read\", \"barcode\", \"barcode_position\"]:\n",
    "                        #    value = detail.get(barcode_attribute)\n",
    "                        #    if value is not None:\n",
    "                        #       print(row['accession'], barcode_attribute, value)\n",
    "                else:\n",
    "                    has_processed = True\n",
    "\n",
    "            sc_type = set()\n",
    "            for d in documents:\n",
    "                assert d in known_document_types\n",
    "                if d in document_to_protocol:\n",
    "                    sc_type.add(document_to_protocol[d])\n",
    "            documents_seen[\",\".join(experiment[\"documents\"])] += 1\n",
    "\n",
    "            results.append({\n",
    "                \"accession\": row[\"accession\"], \n",
    "                \"status\": row[\"status\"], \n",
    "                \"lab\": row[\"lab\"][\"title\"].split(',')[1], \n",
    "                \"description\": experiment.get(\"description\", \"\"),\n",
    "                \"documents\": documents,\n",
    "                \"document_len\": len(documents),\n",
    "                \"construction_method\": \",\".join(construction_method),\n",
    "                \"library_barcode\": library_barcode, \n",
    "                \"sc_type\": \",\".join(sc_type),\n",
    "                \"subcellular_fraction\": subcellular_fraction,\n",
    "                \"has_processed\": has_processed,\n",
    "            })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "aquatic-cuisine",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'www.encodeproject.org'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "server.server"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "binding-aquarium",
   "metadata": {},
   "outputs": [],
   "source": [
    "Filereps = namedtuple(\"Filereps\", [\"rep\", \"lane\", \"paired\", \"accession\"])\n",
    "illumina_parts = re.compile(\"_L(?P<lane>[\\d]{3})_R(?P<read>[\\d])_(?P<chunk>[\\d]{3}).fastq.gz\")\n",
    "\n",
    "experiment = server.get_json(\"ENCSR724KET\")\n",
    "\n",
    "def generate_config(sc_type, experiment):\n",
    "    paired_indexing = {\n",
    "        \"1\": \"indexed reads\",\n",
    "        \"2\": \"reads\",\n",
    "    }\n",
    "    \n",
    "    accession = experiment[\"accession\"]\n",
    "    description = experiment.get(\"description\", \"\")\n",
    "    default_config = {\n",
    "        \"experiment_accession\": accession,\n",
    "        \"description\": description,\n",
    "        \"lab\": \"/labs/barbara-wold/\",\n",
    "        \"award\": \"UM1HG009443\",\n",
    "        #\"encode_portal_host\": \"encd-6248-78c147870-qing.demo.encodedcc.org\",\n",
    "        #\"encode_portal_host\": \"encd-6248-24bd77146-qing.demo.encodedcc.org\",\n",
    "        \"encode_portal_host\": server.server,\n",
    "        #\"star_container\": \"https://woldlab.caltech.edu/~diane/containers/star-bash-dev_EoI-head.sif\",\n",
    "        \"star_container\": \"https://woldlab.caltech.edu/~diane/containers/star-bash-2.7.10a.sif\",\n",
    "    \n",
    "        # these should really scale some\n",
    "        \"mem_mb\": 65536,\n",
    "        \"disk_mb\": 51200,\n",
    "\n",
    "        \"automatic_submission\": True\n",
    "    }\n",
    "    if sc_type not in protocol_defaults:\n",
    "        print(\"Can't process {} {}: {}\".format(accession, description, sc_type))\n",
    "        return\n",
    "    default_config.update(protocol_defaults[sc_type])\n",
    "\n",
    "    fractions = set()\n",
    "    config = {}\n",
    "    for rep in experiment[\"replicates\"]:\n",
    "        library_id = rep[\"library\"][\"@id\"]\n",
    "        config.setdefault(library_id, {}).update(default_config)\n",
    "        config[library_id][\"library_accession\"] = rep[\"library\"][\"accession\"]\n",
    "        biosample = rep[\"library\"][\"biosample\"]\n",
    "        organism_id = biosample['organism']['@id']\n",
    "        if organism_id in default_genome_index:\n",
    "            config[library_id].update(default_genome_index[organism_id])\n",
    "        else:\n",
    "            print(\"Unable to process {} {} until {} index\".format(accession, description, organism_id))\n",
    "        subcellular_fraction = biosample.get(\"subcellular_fraction_term_name\", \"\")\n",
    "        fractions.add(subcellular_fraction)\n",
    "        if subcellular_fraction == \"nucleus\":\n",
    "            config[library_id][\"include_intron\"] = True\n",
    "        else:\n",
    "            config[library_id][\"include_intron\"] = False\n",
    "\n",
    "    assert(len(fractions)) == 1\n",
    "    files = []\n",
    "    for f in experiment[\"files\"]:\n",
    "        if f['file_format'] == 'fastq':\n",
    "            submitted_file_name = Path(f[\"submitted_file_name\"])\n",
    "            lanes = []\n",
    "            for detail in f.get(\"flowcell_details\", []):\n",
    "                if \"lane\" in detail and len(detail[\"lane\"]) > 0:\n",
    "                    lanes.append(detail[\"lane\"])\n",
    "            if len(lanes) == 0:\n",
    "                match = illumina_parts.search(submitted_file_name.name)\n",
    "                if match:\n",
    "                    lanes.append(str(int(match.group(\"lane\"))))\n",
    "            lanes = \",\".join(lanes)\n",
    "            if len(lanes) == 0:\n",
    "                print(\"Problems with\", accession, library_id)\n",
    "                \n",
    "            file_library_id = f['replicate'][\"library\"]\n",
    "            if \"paired_end\" in f:\n",
    "                read = paired_indexing[f[\"paired_end\"]]\n",
    "            elif f.get(\"output_type\") in ('index reads', 'reads'):\n",
    "                read = f[\"output_type\"]\n",
    "            else:\n",
    "                raise ValueError(\"Unable to figure out which is the index read\")\n",
    "            \n",
    "            fr = Filereps(file_library_id, lanes, read, f['accession'])\n",
    "            #print(fr)\n",
    "            #print(submitted_file_name)\n",
    "            files.append(fr)\n",
    "\n",
    "    files = sorted(files)\n",
    "    for i, (read1, read2) in enumerate(zip(files[::2], files[1::2])):\n",
    "        config[read1.rep].setdefault(\"read1\", []).append(read1.accession)\n",
    "        config[read1.rep].setdefault(\"read2\", []).append(read2.accession)\n",
    "\n",
    "    for library_id in config:\n",
    "        yield config[library_id]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "expressed-contents",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'experiment_accession': 'ENCSR005TCL',\n",
       "  'description': 'snRNA on human psoas muscle. Please note that this experiment is part of 10x multiome and has a correspoding scATAC experiment (ENCSR332XEW).',\n",
       "  'lab': '/labs/barbara-wold/',\n",
       "  'award': 'UM1HG009443',\n",
       "  'encode_portal_host': 'www.encodeproject.org',\n",
       "  'star_container': 'https://woldlab.caltech.edu/~diane/containers/star-bash-2.7.10a.sif',\n",
       "  'mem_mb': 65536,\n",
       "  'disk_mb': 51200,\n",
       "  'automatic_submission': True,\n",
       "  'stranded': 'Forward',\n",
       "  'umi_length': 12,\n",
       "  'inclusion_accession': '737K-arc-v1(GEX)',\n",
       "  'library_accession': 'ENCLB549KAN',\n",
       "  'genome_accession': 'encode:starsolo-GRCh38-V29-male-index',\n",
       "  'include_intron': True,\n",
       "  'read1': ['ENCFF121UIQ', 'ENCFF663ZRK'],\n",
       "  'read2': ['ENCFF194KTM', 'ENCFF364QFJ']}]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(generate_config(\"Chromium Multiome ATAC + Gene Expression\", experiment_cache[\"ENCSR005TCL\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "married-bracelet",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Problems with ENCSR337HKS /libraries/ENCLB831IHU/\n",
      "Problems with ENCSR337HKS /libraries/ENCLB831IHU/\n",
      "Problems with ENCSR337HKS /libraries/ENCLB831IHU/\n",
      "Problems with ENCSR337HKS /libraries/ENCLB831IHU/\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[{'experiment_accession': 'ENCSR337HKS',\n",
       "  'description': '',\n",
       "  'lab': '/labs/barbara-wold/',\n",
       "  'award': 'UM1HG009443',\n",
       "  'encode_portal_host': 'www.encodeproject.org',\n",
       "  'star_container': 'https://woldlab.caltech.edu/~diane/containers/star-bash-2.7.10a.sif',\n",
       "  'mem_mb': 65536,\n",
       "  'disk_mb': 51200,\n",
       "  'automatic_submission': True,\n",
       "  'stranded': 'Forward',\n",
       "  'umi_length': 12,\n",
       "  'inclusion_accession': '737K-arc-v1(GEX)',\n",
       "  'library_accession': 'ENCLB831IHU',\n",
       "  'genome_accession': 'encode:starsolo-GRCh38-V29-male-index',\n",
       "  'include_intron': True,\n",
       "  'read1': ['ENCFF298GMK', 'ENCFF323AMX'],\n",
       "  'read2': ['ENCFF931UDZ', 'ENCFF878VQP']}]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(generate_config(\"Chromium Multiome ATAC + Gene Expression\", experiment_cache[\"ENCSR337HKS\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "freelance-interface",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Problems with ENCSR003FZY /libraries/ENCLB194KKC/\n",
      "Problems with ENCSR003FZY /libraries/ENCLB194KKC/\n",
      "Problems with ENCSR003FZY /libraries/ENCLB194KKC/\n",
      "Problems with ENCSR003FZY /libraries/ENCLB194KKC/\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[{'experiment_accession': 'ENCSR003FZY',\n",
       "  'description': 'Short-read Split-seq B6Cast F1 P18-20mo male hippocampus deep',\n",
       "  'lab': '/labs/barbara-wold/',\n",
       "  'award': 'UM1HG009443',\n",
       "  'encode_portal_host': 'www.encodeproject.org',\n",
       "  'star_container': 'https://woldlab.caltech.edu/~diane/containers/star-bash-2.7.10a.sif',\n",
       "  'mem_mb': 65536,\n",
       "  'disk_mb': 51200,\n",
       "  'automatic_submission': True,\n",
       "  'stranded': 'Forward',\n",
       "  'inclusion_list_url': 'https://woldlab.caltech.edu/~diane/genome/parse-splitseq-v1.tar.gz',\n",
       "  'library_accession': 'ENCLB946ISI',\n",
       "  'genome_accession': 'encode:starsolo-mm10-M21-male-index',\n",
       "  'include_intron': True,\n",
       "  'read1': ['ENCFF516EGU'],\n",
       "  'read2': ['ENCFF101CWQ']},\n",
       " {'experiment_accession': 'ENCSR003FZY',\n",
       "  'description': 'Short-read Split-seq B6Cast F1 P18-20mo male hippocampus deep',\n",
       "  'lab': '/labs/barbara-wold/',\n",
       "  'award': 'UM1HG009443',\n",
       "  'encode_portal_host': 'www.encodeproject.org',\n",
       "  'star_container': 'https://woldlab.caltech.edu/~diane/containers/star-bash-2.7.10a.sif',\n",
       "  'mem_mb': 65536,\n",
       "  'disk_mb': 51200,\n",
       "  'automatic_submission': True,\n",
       "  'stranded': 'Forward',\n",
       "  'inclusion_list_url': 'https://woldlab.caltech.edu/~diane/genome/parse-splitseq-v1.tar.gz',\n",
       "  'library_accession': 'ENCLB194KKC',\n",
       "  'genome_accession': 'encode:starsolo-mm10-M21-male-index',\n",
       "  'include_intron': True,\n",
       "  'read1': ['ENCFF329FOC'],\n",
       "  'read2': ['ENCFF387FAT']}]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    " list(generate_config(\"Split-seq\", experiment_cache['ENCSR003FZY']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "returning-angle",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_configs = {}\n",
    "for row in sorted(results, key=lambda x: (x[\"status\"], x[\"lab\"], x[\"experiment\"])):\n",
    "    if not row[\"has_processed\"]:\n",
    "        for config in generate_config(row[\"sc_type\"], experiment_cache[row['experiment']]):\n",
    "            library = config['library_accession']\n",
    "            if config is not None:\n",
    "                all_configs[library] = config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "ambient-maryland",
   "metadata": {},
   "outputs": [],
   "source": [
    "import yaml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "vanilla-growing",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'ENCLB002DZK'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-23-193ced05901b>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0myaml\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdump\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mall_configs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"ENCLB002DZK\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m: 'ENCLB002DZK'"
     ]
    }
   ],
   "source": [
    "print(yaml.dump(all_configs[\"ENCLB002DZK\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "deadly-saturn",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(yaml.dump(all_configs[\"ENCLB549KAN\"]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "billion-minimum",
   "metadata": {},
   "source": [
    "Ali said there's one multiome sample he'd like it processed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "heated-drilling",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(yaml.dump(all_configs[\"ENCLB297FDO\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "governmental-discount",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(yaml.dump(all_configs[\"ENCLB938HOJ\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "stopped-framing",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(yaml.dump(all_configs[\"ENCLB527WWJ\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "musical-genre",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(yaml.dump(all_configs[\"ENCLB062TYC\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "concerned-stereo",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(yaml.dump(all_configs[\"ENCLB349KYC\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "raising-valentine",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(yaml.dump(all_configs[\"ENCLB497BTJ\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "designing-replacement",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(yaml.dump(all_configs[\"ENCLB549SMV\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "southwest-column",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"# splitseq\")\n",
    "print(yaml.dump(all_configs[\"ENCLB946ISI\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "demonstrated-senior",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"# splitseq\")\n",
    "print(yaml.dump(all_configs[\"ENCLB194KKC\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "federal-singing",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"# splitseq\")\n",
    "print(yaml.dump(all_configs[\"ENCLB129LMS\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "medieval-entrance",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"# splitseq\")\n",
    "print(yaml.dump(all_configs[\"ENCLB952MZJ\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ready-satin",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"# splitseq\")\n",
    "print(yaml.dump(all_configs[\"ENCLB098ERF\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "minor-speaking",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"# splitseq\")\n",
    "print(yaml.dump(all_configs[\"ENCLB082IOS\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "formal-digest",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "instrumental-modification",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gene_model(config):\n",
    "    return \"GeneFull_Ex50pAS\" if config['include_intron'] else \"Gene\"\n",
    "\n",
    "\n",
    "def prepare_bulk_subworkflow(all_configs, libraries, submit=False):\n",
    "    batch_config = []\n",
    "    bulk_rule_input = []\n",
    "    \n",
    "    for library in libraries:\n",
    "        config = all_configs[library]\n",
    "        #config[\"automatic_submission\"] = False\n",
    "        submit_host = config[\"encode_portal_host\"]\n",
    "        library_dir = Path(library)\n",
    "        workflow_name = \"job_{}\".format(library)\n",
    "\n",
    "        current = \"\"\"subworkflow {workflow_name}:\n",
    "    workdir:\n",
    "        \"{workdir}\"\n",
    "    snakefile:\n",
    "        \"{snakefile}\"\n",
    "    configfile:\n",
    "        \"{configfile}\"\n",
    "\"\"\".format(\n",
    "        workflow_name=workflow_name,\n",
    "        workdir=str(library_dir),\n",
    "        snakefile=str(Path(\"/home/diane/woldlab-rna-seq/workflow/process-encode-10x.snakefile\")),\n",
    "        configfile=str(library_dir / \"config.yaml\"))\n",
    "        batch_config.append(current)\n",
    "        \n",
    "        posted_csv = \"posted.{}.csv\".format(submit_host)\n",
    "        metadata_csv = \"metadata.{}.csv\".format(submit_host)\n",
    "        target_rule = posted_csv if config[\"automatic_submission\"] else metadata_csv\n",
    "        bulk_rule_input.append(\"{}(\\\"{}\\\")\".format(workflow_name, target_rule))\n",
    "        final_out_rule = \"Log.final.out.{}.qc-upload\".format(submit_host)\n",
    "        bulk_rule_input.append(\"{}(\\\"{}\\\")\".format(workflow_name, final_out_rule))\n",
    "        summary_rule = \"Solo.out/{}/Summary.csv.{}.qc-upload\".format(get_gene_model(config), submit_host)\n",
    "        bulk_rule_input.append(\"{}(\\\"{}\\\")\".format(workflow_name, summary_rule))\n",
    "        pct_count_mt_rule = \"pct_count_mt.{}_EM_filtered.png.{}.qc-upload\".format(get_gene_model(config), submit_host)\n",
    "        bulk_rule_input.append(\"{}(\\\"{}\\\")\".format(workflow_name, pct_count_mt_rule))\n",
    "    \n",
    "    formatted_rules = []\n",
    "    for rule in bulk_rule_input:\n",
    "        formatted_rules.append(\"{prefix}{rule},\\n\".format(prefix=\" \" * 8, rule=rule))\n",
    "    batch_config.append(\"\"\"rule world:\n",
    "    input:\n",
    "{rules}\n",
    "\"\"\".format(rules=\"\".join(formatted_rules)))\n",
    "    return \"\\n\".join(batch_config)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "after-wiring",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(prepare_bulk_subworkflow(all_configs, [\"ENCLB002DZK\", \"ENCLB549KAN\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "grateful-influence",
   "metadata": {},
   "outputs": [],
   "source": [
    "def prepare_bulk_subworkflow_run(destination, all_configs, libraries):\n",
    "    with tarfile.open(destination, \"w:gz\") as archive:\n",
    "        for library in libraries:\n",
    "            library_dir = Path(library)\n",
    "            config_data = yaml.dump(all_configs[library]).encode(\"utf-8\")\n",
    "            config_stream = BytesIO(config_data)\n",
    "            config_filename = library_dir / \"config.yaml\"\n",
    "            config_info = tarfile.TarInfo(str(config_filename))\n",
    "            config_info.size = len(config_data)\n",
    "            \n",
    "            archive.addfile(config_info, fileobj=config_stream)\n",
    "            \n",
    "        world_data = prepare_bulk_subworkflow(all_configs, libraries).encode(\"utf-8\")\n",
    "        world_stream = BytesIO(world_data)\n",
    "        world_info = tarfile.TarInfo(\"Snakefile\")\n",
    "        world_info.size = len(world_data)\n",
    "        archive.addfile(world_info, world_stream)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dimensional-treaty",
   "metadata": {},
   "outputs": [],
   "source": [
    "prepare_bulk_subworkflow_run(\"/dev/shm/all.tar.gz\", all_configs, [\"ENCLB002DZK\", \"ENCLB549KAN\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "beautiful-accessory",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "lyric-juice",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "endangered-carolina",
   "metadata": {},
   "outputs": [],
   "source": [
    "def prepare_bulk_sbatch(all_configs, libraries, submit=False):\n",
    "    commands = [\"#!/bin/bash\", \"\"]\n",
    "    \n",
    "    for library in libraries:\n",
    "        config = all_configs[library]\n",
    "        library_dir = Path(library)\n",
    "\n",
    "        commands.append(\"pushd {}\".format(library_dir))\n",
    "        commands.append(\"sbatch --time 1-0 --mem 64GB --cpus-per-task 24 -- snakemake --use-singularity --snakefile ~/woldlab-rna-seq/workflow/process-encode-10x.snakefile --cluster \\\"sbatch --time 1-0 --mem {resources.mem_mb} --cpus-per-task {threads}\\\" --jobs 40 --cores 40\")\n",
    "        commands.append(\"popd\")\n",
    "    \n",
    "    return \"\\n\".join(commands)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "terminal-rating",
   "metadata": {},
   "outputs": [],
   "source": [
    "def prepare_bulk_sbatch_run(destination, all_configs, libraries):\n",
    "    with tarfile.open(destination, \"w:gz\") as archive:\n",
    "        for library in libraries:\n",
    "            library_dir = Path(library)\n",
    "            config_data = yaml.dump(all_configs[library]).encode(\"utf-8\")\n",
    "            config_stream = BytesIO(config_data)\n",
    "            config_filename = library_dir / \"config.yaml\"\n",
    "            config_info = tarfile.TarInfo(str(config_filename))\n",
    "            config_info.size = len(config_data)\n",
    "            \n",
    "            archive.addfile(config_info, fileobj=config_stream)\n",
    "            \n",
    "        world_data = prepare_bulk_sbatch(all_configs, libraries).encode(\"utf-8\")\n",
    "        world_stream = BytesIO(world_data)\n",
    "        world_info = tarfile.TarInfo(\"run.sh\")\n",
    "        world_info.size = len(world_data)\n",
    "        archive.addfile(world_info, world_stream)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "attended-brighton",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "atlantic-pride",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = pandas.DataFrame(results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "framed-silence",
   "metadata": {},
   "outputs": [],
   "source": [
    "experiments_to_run_2_6 = [\"ENCSR005TCL\",\"ENCSR011EDH\",\"ENCSR012APQ\",\"ENCSR021QIA\",\"ENCSR021YYL\"]\n",
    "\n",
    "experiments_to_run_7_21 = [\n",
    "\"ENCSR028MBY\",\n",
    "\"ENCSR035BDI\",\n",
    "\"ENCSR060SEA\",\n",
    "\"ENCSR060ZXZ\",\n",
    "\"ENCSR075HVN\",\n",
    "\"ENCSR079LVG\",\n",
    "\"ENCSR087YKG\",\n",
    "\"ENCSR132MEW\",\n",
    "\"ENCSR133XJX\",\n",
    "\"ENCSR142RRV\",\n",
    "\"ENCSR151ONW\",\n",
    "\"ENCSR157YJO\",\n",
    "\"ENCSR201WJO\",\n",
    "\"ENCSR203THX\",\n",
    "\"ENCSR224OUG\",\n",
    "]\n",
    "\n",
    "#experiments_to_run = set([\"ENCSR402PHR\",\"ENCSR157YJO\",\"ENCSR021YYL\",\"ENCSR988AMY\",\"ENCSR908CQZ\",\"ENCSR749GDE\",\"ENCSR224OUG\",\"ENCSR356VJZ\", \"ENCSR021YYL\"])\n",
    "experiments_to_run_22_51 = [\n",
    "\"ENCSR224RWH\",\n",
    "\"ENCSR225OII\",\n",
    "\"ENCSR258SZI\",\n",
    "\"ENCSR259VOY\",\n",
    "\"ENCSR261BXB\",\n",
    "\"ENCSR261LPM\",\n",
    "\"ENCSR281NBH\",\n",
    "\"ENCSR281RIH\",\n",
    "\"ENCSR289ANH\",\n",
    "\"ENCSR297ZTQ\",\n",
    "\"ENCSR298MPV\",\n",
    "\"ENCSR333SLJ\",\n",
    "\"ENCSR337HKS\",\n",
    "\"ENCSR352DXB\",\n",
    "\"ENCSR356VJZ\",\n",
    "\"ENCSR362HUP\",\n",
    "\"ENCSR362YDM\",\n",
    "\"ENCSR367YUX\",\n",
    "\"ENCSR369UFT\",\n",
    "\"ENCSR398PIW\",\n",
    "\"ENCSR402PHR\",\n",
    "\"ENCSR405YKM\",\n",
    "\"ENCSR415HEP\",\n",
    "\"ENCSR417FCM\",\n",
    "\"ENCSR421YXX\",\n",
    "\"ENCSR426PDU\",\n",
    "\"ENCSR439ZVQ\",\n",
    "\"ENCSR445LCA\",\n",
    "\"ENCSR460MGC\",\n",
    "\"ENCSR462HCD\",\n",
    "]\n",
    "\n",
    "experiments_to_run_52_81 = [\n",
    "\"ENCSR463EEP\",\n",
    "\"ENCSR467GGT\",\n",
    "\"ENCSR471YTA\",\n",
    "\"ENCSR472RRP\",\n",
    "\"ENCSR478FQR\",\n",
    "\"ENCSR481QQR\",\n",
    "\"ENCSR484YSK\",\n",
    "\"ENCSR522HNR\",\n",
    "\"ENCSR527MGL\",\n",
    "\"ENCSR533IGQ\",\n",
    "\"ENCSR539XHZ\",\n",
    "\"ENCSR548HED\",\n",
    "\"ENCSR568GGW\",\n",
    "\"ENCSR587SPO\",\n",
    "\"ENCSR602SGU\",\n",
    "\"ENCSR620AJZ\",\n",
    "\"ENCSR624WDN\",\n",
    "\"ENCSR632EIK\",\n",
    "\"ENCSR641MWO\",\n",
    "\"ENCSR642GFM\",\n",
    "\"ENCSR644VVE\",\n",
    "\"ENCSR656VWP\",\n",
    "\"ENCSR662VTU\",\n",
    "\"ENCSR665HPR\",\n",
    "\"ENCSR671FDY\",\n",
    "\"ENCSR683ONV\",\n",
    "\"ENCSR684KYI\",\n",
    "\"ENCSR700UVN\",\n",
    "\"ENCSR707ECV\",\n",
    "\"ENCSR707LLI\",\n",
    "]\n",
    "\n",
    "experiments_to_run_82_107 = [\"ENCSR716VLR\",\n",
    "\"ENCSR720TUP\",\n",
    "\"ENCSR724KET\",\n",
    "\"ENCSR726IPC\",\n",
    "\"ENCSR727OYO\",\n",
    "\"ENCSR742NWO\",\n",
    "\"ENCSR749GDE\",\n",
    "\"ENCSR751BHQ\",\n",
    "\"ENCSR753YPW\",\n",
    "\"ENCSR755CIF\",\n",
    "\"ENCSR762LZV\",\n",
    "\"ENCSR769KWM\",\n",
    "\"ENCSR789PLK\",\n",
    "\"ENCSR801DHT\",\n",
    "\"ENCSR813PWQ\",\n",
    "\"ENCSR835JFN\",\n",
    "\"ENCSR876EUH\",\n",
    "\"ENCSR908CQZ\",\n",
    "\"ENCSR938CZJ\",\n",
    "\"ENCSR943HBF\",\n",
    "\"ENCSR944XPC\",\n",
    "\"ENCSR964JZN\",\n",
    "\"ENCSR966DDY\",\n",
    "\"ENCSR968RPR\",\n",
    "\"ENCSR987SOK\",\n",
    "\"ENCSR988AMY\",\n",
    "\"ENCSR995ZKH\",\n",
    "]\n",
    "\n",
    "experiments_to_run = experiments_to_run_82_107\n",
    "print(\"Preparing {} jobs\".format(len(experiments_to_run)))\n",
    "ready_metadata = metadata[metadata[\"experiment\"].isin(experiments_to_run)]\n",
    "\n",
    "#prepare_bulk_sbatch_run(\"/dev/shm/test-run-82-108.tar.gz\", all_configs, ready_metadata[\"library\"].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "external-working",
   "metadata": {},
   "outputs": [],
   "source": [
    "#prepare_bulk_run(\"/dev/shm/adrenal_multiome.tar.gz\", all_configs, ready_metadata[\"library\"].values)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "mental-helicopter",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i, row in ready_metadata.iterrows():\n",
    "    config = all_configs[row.library]\n",
    "    print(\"https://{}/experiments/{}/\".format(config[\"encode_portal_host\"], row.experiment))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "independent-falls",
   "metadata": {},
   "outputs": [],
   "source": [
    "prepare_bulk_sbatch_run(\"/dev/shm/run-1-108.tar.gz\", all_configs, metadata[metadata[\"lab\"] == \"Stanford\"][\"library\"].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "quiet-boundary",
   "metadata": {},
   "outputs": [],
   "source": [
    "prepare_bulk_sbatch_run(\"/dev/shm/test-run-82-108.tar.gz\", all_configs, ready_metadata[\"library\"].values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "streaming-pierce",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_dcc_accession(server, accession):\n",
    "    metadata = None\n",
    "    try:\n",
    "        metadata = server.get_json(accession)\n",
    "    except HTTPError as e:\n",
    "        if e.response.status_code == 404:\n",
    "            print(\"Accession {} not found, discovery will fail\".format(accession))\n",
    "        else:\n",
    "            print(\"Other error {} for {}\".format(e.response.status_code, accession))\n",
    "            raise e\n",
    "\n",
    "    return metadata\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "arranged-accessory",
   "metadata": {},
   "outputs": [],
   "source": [
    "from io import StringIO\n",
    "submitted = pandas.read_csv(StringIO(\"\"\"uuid,accession,dataset,file_format,output_type,assembly,genome_annotation,derived_from:array,md5sum,file_size:integer,submitted_file_name,award,lab,step_run,aliases:array\n",
    ",,ENCSR224OUG,bam,alignments,mm10,M21,\"encode:starsolo-mm10-M21-male-index,/files/ENCFF494DXJ/,/files/ENCFF586ZOY/,/files/ENCFF201AFX/,/files/ENCFF468KNY/\",38bf053481a36b565a45e2e4fedb2bbf,14365291492,Aligned.sortedByCoord.out.bam,UM1HG009443,/labs/barbara-wold/,barbara-wold:starsolo-10x-alignment-step-run,barbara-wold:ENCLB607FDJ_alignment_2022-01-19\n",
    ",,ENCSR224OUG,tar,sparse gene count matrix of unique reads,mm10,M21,\"737K-arc-v1(GEX),barbara-wold:ENCLB607FDJ_alignment_2022-01-19\",d4214b016b19730e185ce11dd5396514,39391130,GeneFull_Ex50pAS_Unique_filtered.tar.gz,UM1HG009443,/labs/barbara-wold/,barbara-wold:starsolo-10x-quantification-step-run,\n",
    ",,ENCSR224OUG,tar,sparse gene count matrix of all reads,mm10,M21,\"737K-arc-v1(GEX),barbara-wold:ENCLB607FDJ_alignment_2022-01-19\",68d9aa21aae8e891f85f880cd5bd8cab,46821115,GeneFull_Ex50pAS_EM_filtered.tar.gz,UM1HG009443,/labs/barbara-wold/,barbara-wold:starsolo-10x-quantification-step-run,\n",
    ",,ENCSR224OUG,tar,unfiltered sparse gene count matrix of unique reads,mm10,M21,\"737K-arc-v1(GEX),barbara-wold:ENCLB607FDJ_alignment_2022-01-19\",920b78efae954ffb082faf2aaa921ba5,59389166,GeneFull_Ex50pAS_Unique_raw.tar.gz,UM1HG009443,/labs/barbara-wold/,barbara-wold:starsolo-10x-quantification-step-run,\n",
    ",,ENCSR224OUG,tar,unfiltered sparse gene count matrix of all reads,mm10,M21,\"737K-arc-v1(GEX),barbara-wold:ENCLB607FDJ_alignment_2022-01-19\",7d6019e2816f28f19eeecd47a3442608,76437934,GeneFull_Ex50pAS_EM_raw.tar.gz,UM1HG009443,/labs/barbara-wold/,barbara-wold:starsolo-10x-quantification-step-run,\n",
    ",,ENCSR224OUG,tar,unfiltered sparse splice junction count matrix of unique reads,mm10,M21,\"737K-arc-v1(GEX),barbara-wold:ENCLB607FDJ_alignment_2022-01-19\",2dd5935279637f61abf89e5778881a77,14019963,SJ_Unique_raw.tar.gz,UM1HG009443,/labs/barbara-wold/,barbara-wold:starsolo-10x-quantification-step-run,\n",
    "\"\"\"))\n",
    "submitted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "parliamentary-fossil",
   "metadata": {},
   "outputs": [],
   "source": [
    "problems = set(\"\"\"Problems with ENCSR203THX /libraries/ENCLB592LXS/\n",
    "Problems with ENCSR203THX /libraries/ENCLB592LXS/\n",
    "Problems with ENCSR203THX /libraries/ENCLB592LXS/\n",
    "Problems with ENCSR203THX /libraries/ENCLB592LXS/\n",
    "Problems with ENCSR298MPV /libraries/ENCLB814OYP/\n",
    "Problems with ENCSR298MPV /libraries/ENCLB814OYP/\n",
    "Problems with ENCSR298MPV /libraries/ENCLB814OYP/\n",
    "Problems with ENCSR298MPV /libraries/ENCLB814OYP/\n",
    "Problems with ENCSR337HKS /libraries/ENCLB831IHU/\n",
    "Problems with ENCSR337HKS /libraries/ENCLB831IHU/\n",
    "Problems with ENCSR337HKS /libraries/ENCLB831IHU/\n",
    "Problems with ENCSR337HKS /libraries/ENCLB831IHU/\n",
    "Problems with ENCSR405YKM /libraries/ENCLB181TPV/\n",
    "Problems with ENCSR405YKM /libraries/ENCLB181TPV/\n",
    "Problems with ENCSR405YKM /libraries/ENCLB181TPV/\n",
    "Problems with ENCSR405YKM /libraries/ENCLB181TPV/\n",
    "Problems with ENCSR463EEP /libraries/ENCLB619VVF/\n",
    "Problems with ENCSR463EEP /libraries/ENCLB619VVF/\n",
    "Problems with ENCSR463EEP /libraries/ENCLB619VVF/\n",
    "Problems with ENCSR463EEP /libraries/ENCLB619VVF/\n",
    "Problems with ENCSR527MGL /libraries/ENCLB909DQG/\n",
    "Problems with ENCSR527MGL /libraries/ENCLB909DQG/\n",
    "Problems with ENCSR527MGL /libraries/ENCLB909DQG/\n",
    "Problems with ENCSR527MGL /libraries/ENCLB909DQG/\n",
    "Problems with ENCSR662VTU /libraries/ENCLB615BOP/\n",
    "Problems with ENCSR662VTU /libraries/ENCLB615BOP/\n",
    "Problems with ENCSR662VTU /libraries/ENCLB615BOP/\n",
    "Problems with ENCSR662VTU /libraries/ENCLB615BOP/\n",
    "Problems with ENCSR700UVN /libraries/ENCLB527WWJ/\n",
    "Problems with ENCSR700UVN /libraries/ENCLB527WWJ/\n",
    "Problems with ENCSR700UVN /libraries/ENCLB527WWJ/\n",
    "Problems with ENCSR700UVN /libraries/ENCLB527WWJ/\n",
    "Problems with ENCSR769KWM /libraries/ENCLB812SMT/\n",
    "Problems with ENCSR769KWM /libraries/ENCLB812SMT/\n",
    "Problems with ENCSR769KWM /libraries/ENCLB812SMT/\n",
    "Problems with ENCSR769KWM /libraries/ENCLB812SMT/\n",
    "Problems with ENCSR003FZY /libraries/ENCLB194KKC/\n",
    "Problems with ENCSR003FZY /libraries/ENCLB194KKC/\n",
    "Problems with ENCSR003FZY /libraries/ENCLB194KKC/\n",
    "Problems with ENCSR003FZY /libraries/ENCLB194KKC/\n",
    "Problems with ENCSR024CHH /libraries/ENCLB491BDA/\n",
    "Problems with ENCSR024CHH /libraries/ENCLB491BDA/\n",
    "Problems with ENCSR029GVP /libraries/ENCLB631JOU/\n",
    "Problems with ENCSR029GVP /libraries/ENCLB631JOU/\n",
    "Problems with ENCSR029GVP /libraries/ENCLB631JOU/\n",
    "Problems with ENCSR029GVP /libraries/ENCLB631JOU/\n",
    "Problems with ENCSR029OIF /libraries/ENCLB048POP/\n",
    "Problems with ENCSR029OIF /libraries/ENCLB048POP/\n",
    "Problems with ENCSR029OIF /libraries/ENCLB048POP/\n",
    "Problems with ENCSR029OIF /libraries/ENCLB048POP/\n",
    "Problems with ENCSR035SWS /libraries/ENCLB315XSJ/\n",
    "Problems with ENCSR035SWS /libraries/ENCLB315XSJ/\n",
    "Problems with ENCSR035SWS /libraries/ENCLB315XSJ/\n",
    "Problems with ENCSR035SWS /libraries/ENCLB315XSJ/\n",
    "Problems with ENCSR043HKT /libraries/ENCLB771NKO/\n",
    "Problems with ENCSR043HKT /libraries/ENCLB771NKO/\n",
    "Problems with ENCSR043HKT /libraries/ENCLB771NKO/\n",
    "Problems with ENCSR043HKT /libraries/ENCLB771NKO/\n",
    "Problems with ENCSR044GSN /libraries/ENCLB102VRS/\n",
    "Problems with ENCSR044GSN /libraries/ENCLB102VRS/\n",
    "Problems with ENCSR044GSN /libraries/ENCLB102VRS/\n",
    "Problems with ENCSR044GSN /libraries/ENCLB102VRS/\n",
    "Problems with ENCSR062ICV /libraries/ENCLB440ABL/\n",
    "Problems with ENCSR062ICV /libraries/ENCLB440ABL/\n",
    "Problems with ENCSR062ICV /libraries/ENCLB440ABL/\n",
    "Problems with ENCSR062ICV /libraries/ENCLB440ABL/\n",
    "Problems with ENCSR062WSS /libraries/ENCLB046ZBE/\n",
    "Problems with ENCSR062WSS /libraries/ENCLB046ZBE/\n",
    "Problems with ENCSR062WSS /libraries/ENCLB046ZBE/\n",
    "Problems with ENCSR062WSS /libraries/ENCLB046ZBE/\n",
    "Problems with ENCSR080VLK /libraries/ENCLB400DWX/\n",
    "Problems with ENCSR080VLK /libraries/ENCLB400DWX/\n",
    "Problems with ENCSR081SRK /libraries/ENCLB798BJP/\n",
    "Problems with ENCSR081SRK /libraries/ENCLB798BJP/\n",
    "Problems with ENCSR081SRK /libraries/ENCLB798BJP/\n",
    "Problems with ENCSR081SRK /libraries/ENCLB798BJP/\n",
    "Problems with ENCSR087LWL /libraries/ENCLB892ZAW/\n",
    "Problems with ENCSR087LWL /libraries/ENCLB892ZAW/\n",
    "Problems with ENCSR087LWL /libraries/ENCLB892ZAW/\n",
    "Problems with ENCSR087LWL /libraries/ENCLB892ZAW/\n",
    "Problems with ENCSR090OFY /libraries/ENCLB061KPS/\n",
    "Problems with ENCSR090OFY /libraries/ENCLB061KPS/\n",
    "Problems with ENCSR090OFY /libraries/ENCLB061KPS/\n",
    "Problems with ENCSR090OFY /libraries/ENCLB061KPS/\n",
    "Problems with ENCSR093ZTB /libraries/ENCLB988YCJ/\n",
    "Problems with ENCSR093ZTB /libraries/ENCLB988YCJ/\n",
    "Problems with ENCSR093ZTB /libraries/ENCLB988YCJ/\n",
    "Problems with ENCSR093ZTB /libraries/ENCLB988YCJ/\n",
    "Problems with ENCSR096PUV /libraries/ENCLB169FGN/\n",
    "Problems with ENCSR096PUV /libraries/ENCLB169FGN/\n",
    "Problems with ENCSR096PUV /libraries/ENCLB169FGN/\n",
    "Problems with ENCSR096PUV /libraries/ENCLB169FGN/\n",
    "Problems with ENCSR105JHX /libraries/ENCLB792DPM/\n",
    "Problems with ENCSR105JHX /libraries/ENCLB792DPM/\n",
    "Problems with ENCSR105JHX /libraries/ENCLB792DPM/\n",
    "Problems with ENCSR105JHX /libraries/ENCLB792DPM/\n",
    "Problems with ENCSR115JCE /libraries/ENCLB156LUM/\n",
    "Problems with ENCSR115JCE /libraries/ENCLB156LUM/\n",
    "Problems with ENCSR115JCE /libraries/ENCLB156LUM/\n",
    "Problems with ENCSR115JCE /libraries/ENCLB156LUM/\n",
    "Problems with ENCSR119HVP /libraries/ENCLB372GFV/\n",
    "Problems with ENCSR119HVP /libraries/ENCLB372GFV/\n",
    "Problems with ENCSR119HVP /libraries/ENCLB372GFV/\n",
    "Problems with ENCSR119HVP /libraries/ENCLB372GFV/\n",
    "Problems with ENCSR120UHU /libraries/ENCLB343QQX/\n",
    "Problems with ENCSR120UHU /libraries/ENCLB343QQX/\n",
    "Problems with ENCSR120UHU /libraries/ENCLB343QQX/\n",
    "Problems with ENCSR120UHU /libraries/ENCLB343QQX/\n",
    "Problems with ENCSR126YAO /libraries/ENCLB848YMF/\n",
    "Problems with ENCSR126YAO /libraries/ENCLB848YMF/\n",
    "Problems with ENCSR126YAO /libraries/ENCLB848YMF/\n",
    "Problems with ENCSR126YAO /libraries/ENCLB848YMF/\n",
    "Problems with ENCSR135ELG /libraries/ENCLB722NFY/\n",
    "Problems with ENCSR135ELG /libraries/ENCLB722NFY/\n",
    "Problems with ENCSR145DBY /libraries/ENCLB230VGI/\n",
    "Problems with ENCSR145DBY /libraries/ENCLB230VGI/\n",
    "Problems with ENCSR146AFJ /libraries/ENCLB323HIM/\n",
    "Problems with ENCSR146AFJ /libraries/ENCLB323HIM/\n",
    "Problems with ENCSR146AFJ /libraries/ENCLB323HIM/\n",
    "Problems with ENCSR146AFJ /libraries/ENCLB323HIM/\n",
    "Problems with ENCSR147UYS /libraries/ENCLB564RVH/\n",
    "Problems with ENCSR147UYS /libraries/ENCLB564RVH/\n",
    "Problems with ENCSR147UYS /libraries/ENCLB564RVH/\n",
    "Problems with ENCSR147UYS /libraries/ENCLB564RVH/\n",
    "Problems with ENCSR156BQX /libraries/ENCLB894TFF/\n",
    "Problems with ENCSR156BQX /libraries/ENCLB894TFF/\n",
    "Problems with ENCSR156BQX /libraries/ENCLB894TFF/\n",
    "Problems with ENCSR156BQX /libraries/ENCLB894TFF/\n",
    "Problems with ENCSR162LWP /libraries/ENCLB672PSR/\n",
    "Problems with ENCSR162LWP /libraries/ENCLB672PSR/\n",
    "Problems with ENCSR162LWP /libraries/ENCLB672PSR/\n",
    "Problems with ENCSR162LWP /libraries/ENCLB672PSR/\n",
    "Problems with ENCSR164REI /libraries/ENCLB249TPV/\n",
    "Problems with ENCSR164REI /libraries/ENCLB249TPV/\n",
    "Problems with ENCSR164REI /libraries/ENCLB249TPV/\n",
    "Problems with ENCSR164REI /libraries/ENCLB249TPV/\n",
    "Problems with ENCSR165CAC /libraries/ENCLB323CCI/\n",
    "Problems with ENCSR165CAC /libraries/ENCLB323CCI/\n",
    "Problems with ENCSR169QXA /libraries/ENCLB213LKC/\n",
    "Problems with ENCSR169QXA /libraries/ENCLB213LKC/\n",
    "Problems with ENCSR169QXA /libraries/ENCLB213LKC/\n",
    "Problems with ENCSR169QXA /libraries/ENCLB213LKC/\n",
    "Problems with ENCSR171BEG /libraries/ENCLB143IPZ/\n",
    "Problems with ENCSR171BEG /libraries/ENCLB143IPZ/\n",
    "Problems with ENCSR171BEG /libraries/ENCLB143IPZ/\n",
    "Problems with ENCSR171BEG /libraries/ENCLB143IPZ/\n",
    "Problems with ENCSR172XWR /libraries/ENCLB820VCW/\n",
    "Problems with ENCSR172XWR /libraries/ENCLB820VCW/\n",
    "Problems with ENCSR172XWR /libraries/ENCLB820VCW/\n",
    "Problems with ENCSR172XWR /libraries/ENCLB820VCW/\n",
    "Problems with ENCSR178WRO /libraries/ENCLB829GDB/\n",
    "Problems with ENCSR178WRO /libraries/ENCLB829GDB/\n",
    "Problems with ENCSR186OUD /libraries/ENCLB072RWF/\n",
    "Problems with ENCSR186OUD /libraries/ENCLB072RWF/\n",
    "Problems with ENCSR186OUD /libraries/ENCLB072RWF/\n",
    "Problems with ENCSR186OUD /libraries/ENCLB072RWF/\n",
    "Problems with ENCSR186TDH /libraries/ENCLB716MVI/\n",
    "Problems with ENCSR186TDH /libraries/ENCLB716MVI/\n",
    "Problems with ENCSR186TDH /libraries/ENCLB716MVI/\n",
    "Problems with ENCSR186TDH /libraries/ENCLB716MVI/\n",
    "Problems with ENCSR188BZZ /libraries/ENCLB603YRZ/\n",
    "Problems with ENCSR188BZZ /libraries/ENCLB603YRZ/\n",
    "Problems with ENCSR188BZZ /libraries/ENCLB603YRZ/\n",
    "Problems with ENCSR188BZZ /libraries/ENCLB603YRZ/\n",
    "Problems with ENCSR191YNY /libraries/ENCLB266ZRP/\n",
    "Problems with ENCSR191YNY /libraries/ENCLB266ZRP/\n",
    "Problems with ENCSR192TYH /libraries/ENCLB925LWZ/\n",
    "Problems with ENCSR192TYH /libraries/ENCLB925LWZ/\n",
    "Problems with ENCSR192TYH /libraries/ENCLB925LWZ/\n",
    "Problems with ENCSR192TYH /libraries/ENCLB925LWZ/\n",
    "Problems with ENCSR193PLQ /libraries/ENCLB082IOS/\n",
    "Problems with ENCSR193PLQ /libraries/ENCLB082IOS/\n",
    "Problems with ENCSR193PLQ /libraries/ENCLB082IOS/\n",
    "Problems with ENCSR193PLQ /libraries/ENCLB082IOS/\n",
    "Problems with ENCSR204FUI /libraries/ENCLB072YJA/\n",
    "Problems with ENCSR204FUI /libraries/ENCLB072YJA/\n",
    "Problems with ENCSR204FUI /libraries/ENCLB072YJA/\n",
    "Problems with ENCSR204FUI /libraries/ENCLB072YJA/\n",
    "Problems with ENCSR226SBC /libraries/ENCLB889CHC/\n",
    "Problems with ENCSR226SBC /libraries/ENCLB889CHC/\n",
    "Problems with ENCSR226SBC /libraries/ENCLB889CHC/\n",
    "Problems with ENCSR226SBC /libraries/ENCLB889CHC/\n",
    "Problems with ENCSR227CPT /libraries/ENCLB793OLP/\n",
    "Problems with ENCSR227CPT /libraries/ENCLB793OLP/\n",
    "Problems with ENCSR227CPT /libraries/ENCLB793OLP/\n",
    "Problems with ENCSR227CPT /libraries/ENCLB793OLP/\n",
    "Problems with ENCSR236MGX /libraries/ENCLB692PHT/\n",
    "Problems with ENCSR236MGX /libraries/ENCLB692PHT/\n",
    "Problems with ENCSR236MGX /libraries/ENCLB692PHT/\n",
    "Problems with ENCSR236MGX /libraries/ENCLB692PHT/\n",
    "Problems with ENCSR237EZS /libraries/ENCLB010VHV/\n",
    "Problems with ENCSR237EZS /libraries/ENCLB010VHV/\n",
    "Problems with ENCSR237EZS /libraries/ENCLB010VHV/\n",
    "Problems with ENCSR237EZS /libraries/ENCLB010VHV/\n",
    "Problems with ENCSR240CYU /libraries/ENCLB601XFB/\n",
    "Problems with ENCSR240CYU /libraries/ENCLB601XFB/\n",
    "Problems with ENCSR240CYU /libraries/ENCLB601XFB/\n",
    "Problems with ENCSR240CYU /libraries/ENCLB601XFB/\n",
    "Problems with ENCSR244FME /libraries/ENCLB590EJB/\n",
    "Problems with ENCSR244FME /libraries/ENCLB590EJB/\n",
    "Problems with ENCSR244FME /libraries/ENCLB590EJB/\n",
    "Problems with ENCSR244FME /libraries/ENCLB590EJB/\n",
    "Problems with ENCSR248IOH /libraries/ENCLB570NOI/\n",
    "Problems with ENCSR248IOH /libraries/ENCLB570NOI/\n",
    "Problems with ENCSR248IOH /libraries/ENCLB570NOI/\n",
    "Problems with ENCSR248IOH /libraries/ENCLB570NOI/\n",
    "Problems with ENCSR250ZVV /libraries/ENCLB274ZBT/\n",
    "Problems with ENCSR250ZVV /libraries/ENCLB274ZBT/\n",
    "Problems with ENCSR250ZVV /libraries/ENCLB274ZBT/\n",
    "Problems with ENCSR250ZVV /libraries/ENCLB274ZBT/\n",
    "Problems with ENCSR266HBP /libraries/ENCLB931EHJ/\n",
    "Problems with ENCSR266HBP /libraries/ENCLB931EHJ/\n",
    "Problems with ENCSR266HBP /libraries/ENCLB931EHJ/\n",
    "Problems with ENCSR266HBP /libraries/ENCLB931EHJ/\n",
    "Problems with ENCSR268TST /libraries/ENCLB548SZL/\n",
    "Problems with ENCSR268TST /libraries/ENCLB548SZL/\n",
    "Problems with ENCSR284OLV /libraries/ENCLB164ZHY/\n",
    "Problems with ENCSR284OLV /libraries/ENCLB164ZHY/\n",
    "Problems with ENCSR284OLV /libraries/ENCLB164ZHY/\n",
    "Problems with ENCSR284OLV /libraries/ENCLB164ZHY/\n",
    "Problems with ENCSR287VRB /libraries/ENCLB544SEN/\n",
    "Problems with ENCSR287VRB /libraries/ENCLB544SEN/\n",
    "Problems with ENCSR287VRB /libraries/ENCLB544SEN/\n",
    "Problems with ENCSR287VRB /libraries/ENCLB544SEN/\n",
    "Problems with ENCSR299DGH /libraries/ENCLB773PPF/\n",
    "Problems with ENCSR299DGH /libraries/ENCLB773PPF/\n",
    "Problems with ENCSR299DGH /libraries/ENCLB773PPF/\n",
    "Problems with ENCSR299DGH /libraries/ENCLB773PPF/\n",
    "Problems with ENCSR306ZET /libraries/ENCLB018ZYA/\n",
    "Problems with ENCSR306ZET /libraries/ENCLB018ZYA/\n",
    "Problems with ENCSR306ZET /libraries/ENCLB018ZYA/\n",
    "Problems with ENCSR306ZET /libraries/ENCLB018ZYA/\n",
    "Problems with ENCSR311TRI /libraries/ENCLB321YXH/\n",
    "Problems with ENCSR311TRI /libraries/ENCLB321YXH/\n",
    "Problems with ENCSR311TRI /libraries/ENCLB321YXH/\n",
    "Problems with ENCSR311TRI /libraries/ENCLB321YXH/\n",
    "Problems with ENCSR321BZA /libraries/ENCLB715WPG/\n",
    "Problems with ENCSR321BZA /libraries/ENCLB715WPG/\n",
    "Problems with ENCSR322KUB /libraries/ENCLB908NNT/\n",
    "Problems with ENCSR322KUB /libraries/ENCLB908NNT/\n",
    "Problems with ENCSR322KUB /libraries/ENCLB908NNT/\n",
    "Problems with ENCSR322KUB /libraries/ENCLB908NNT/\n",
    "Problems with ENCSR327PSD /libraries/ENCLB919KOR/\n",
    "Problems with ENCSR327PSD /libraries/ENCLB919KOR/\n",
    "Problems with ENCSR331DYV /libraries/ENCLB905KRE/\n",
    "Problems with ENCSR331DYV /libraries/ENCLB905KRE/\n",
    "Problems with ENCSR331DYV /libraries/ENCLB905KRE/\n",
    "Problems with ENCSR331DYV /libraries/ENCLB905KRE/\n",
    "Problems with ENCSR331KXG /libraries/ENCLB661DWA/\n",
    "Problems with ENCSR331KXG /libraries/ENCLB661DWA/\n",
    "Problems with ENCSR334YXW /libraries/ENCLB457VVB/\n",
    "Problems with ENCSR334YXW /libraries/ENCLB457VVB/\n",
    "Problems with ENCSR345BZU /libraries/ENCLB211BVF/\n",
    "Problems with ENCSR345BZU /libraries/ENCLB211BVF/\n",
    "Problems with ENCSR356NFK /libraries/ENCLB797XAC/\n",
    "Problems with ENCSR356NFK /libraries/ENCLB797XAC/\n",
    "Problems with ENCSR356NFK /libraries/ENCLB797XAC/\n",
    "Problems with ENCSR356NFK /libraries/ENCLB797XAC/\n",
    "Problems with ENCSR356PVK /libraries/ENCLB604HIX/\n",
    "Problems with ENCSR356PVK /libraries/ENCLB604HIX/\n",
    "Problems with ENCSR356PVK /libraries/ENCLB604HIX/\n",
    "Problems with ENCSR356PVK /libraries/ENCLB604HIX/\n",
    "Problems with ENCSR358XDG /libraries/ENCLB239BXF/\n",
    "Problems with ENCSR358XDG /libraries/ENCLB239BXF/\n",
    "Problems with ENCSR358XDG /libraries/ENCLB239BXF/\n",
    "Problems with ENCSR358XDG /libraries/ENCLB239BXF/\n",
    "Problems with ENCSR366MPW /libraries/ENCLB129SKD/\n",
    "Problems with ENCSR366MPW /libraries/ENCLB129SKD/\n",
    "Problems with ENCSR366MPW /libraries/ENCLB129SKD/\n",
    "Problems with ENCSR366MPW /libraries/ENCLB129SKD/\n",
    "Problems with ENCSR382ZGH /libraries/ENCLB988KOO/\n",
    "Problems with ENCSR382ZGH /libraries/ENCLB988KOO/\n",
    "Problems with ENCSR382ZGH /libraries/ENCLB988KOO/\n",
    "Problems with ENCSR382ZGH /libraries/ENCLB988KOO/\n",
    "Problems with ENCSR388RND /libraries/ENCLB943OZA/\n",
    "Problems with ENCSR388RND /libraries/ENCLB943OZA/\n",
    "Problems with ENCSR388RND /libraries/ENCLB943OZA/\n",
    "Problems with ENCSR388RND /libraries/ENCLB943OZA/\n",
    "Problems with ENCSR389TZP /libraries/ENCLB306NVG/\n",
    "Problems with ENCSR389TZP /libraries/ENCLB306NVG/\n",
    "Problems with ENCSR389TZP /libraries/ENCLB306NVG/\n",
    "Problems with ENCSR389TZP /libraries/ENCLB306NVG/\n",
    "Problems with ENCSR390GEG /libraries/ENCLB690LBQ/\n",
    "Problems with ENCSR390GEG /libraries/ENCLB690LBQ/\n",
    "Problems with ENCSR390GEG /libraries/ENCLB690LBQ/\n",
    "Problems with ENCSR390GEG /libraries/ENCLB690LBQ/\n",
    "Problems with ENCSR400PBH /libraries/ENCLB010FKE/\n",
    "Problems with ENCSR400PBH /libraries/ENCLB010FKE/\n",
    "Problems with ENCSR400PBH /libraries/ENCLB010FKE/\n",
    "Problems with ENCSR400PBH /libraries/ENCLB010FKE/\n",
    "Problems with ENCSR405CPX /libraries/ENCLB614YSB/\n",
    "Problems with ENCSR405CPX /libraries/ENCLB614YSB/\n",
    "Problems with ENCSR405CPX /libraries/ENCLB614YSB/\n",
    "Problems with ENCSR405CPX /libraries/ENCLB614YSB/\n",
    "Problems with ENCSR405RFP /libraries/ENCLB261MTS/\n",
    "Problems with ENCSR405RFP /libraries/ENCLB261MTS/\n",
    "Problems with ENCSR405RFP /libraries/ENCLB261MTS/\n",
    "Problems with ENCSR405RFP /libraries/ENCLB261MTS/\n",
    "Problems with ENCSR408PZM /libraries/ENCLB336ZNO/\n",
    "Problems with ENCSR408PZM /libraries/ENCLB336ZNO/\n",
    "Problems with ENCSR408PZM /libraries/ENCLB336ZNO/\n",
    "Problems with ENCSR408PZM /libraries/ENCLB336ZNO/\n",
    "Problems with ENCSR408ZUR /libraries/ENCLB382NBT/\n",
    "Problems with ENCSR408ZUR /libraries/ENCLB382NBT/\n",
    "Problems with ENCSR408ZUR /libraries/ENCLB382NBT/\n",
    "Problems with ENCSR408ZUR /libraries/ENCLB382NBT/\n",
    "Problems with ENCSR415CYS /libraries/ENCLB867BPN/\n",
    "Problems with ENCSR415CYS /libraries/ENCLB867BPN/\n",
    "Problems with ENCSR415CYS /libraries/ENCLB867BPN/\n",
    "Problems with ENCSR415CYS /libraries/ENCLB867BPN/\n",
    "Problems with ENCSR421VUT /libraries/ENCLB673SUG/\n",
    "Problems with ENCSR421VUT /libraries/ENCLB673SUG/\n",
    "Problems with ENCSR421VUT /libraries/ENCLB673SUG/\n",
    "Problems with ENCSR421VUT /libraries/ENCLB673SUG/\n",
    "Problems with ENCSR423AXI /libraries/ENCLB489WXX/\n",
    "Problems with ENCSR423AXI /libraries/ENCLB489WXX/\n",
    "Problems with ENCSR423AXI /libraries/ENCLB489WXX/\n",
    "Problems with ENCSR423AXI /libraries/ENCLB489WXX/\n",
    "Problems with ENCSR423CNV /libraries/ENCLB622RLX/\n",
    "Problems with ENCSR423CNV /libraries/ENCLB622RLX/\n",
    "Problems with ENCSR423CNV /libraries/ENCLB622RLX/\n",
    "Problems with ENCSR423CNV /libraries/ENCLB622RLX/\n",
    "Problems with ENCSR423LFS /libraries/ENCLB516LRK/\n",
    "Problems with ENCSR423LFS /libraries/ENCLB516LRK/\n",
    "Problems with ENCSR423LFS /libraries/ENCLB516LRK/\n",
    "Problems with ENCSR423LFS /libraries/ENCLB516LRK/\n",
    "Problems with ENCSR424RYS /libraries/ENCLB228MHF/\n",
    "Problems with ENCSR424RYS /libraries/ENCLB228MHF/\n",
    "Problems with ENCSR424RYS /libraries/ENCLB228MHF/\n",
    "Problems with ENCSR424RYS /libraries/ENCLB228MHF/\n",
    "Problems with ENCSR426MLQ /libraries/ENCLB368QMP/\n",
    "Problems with ENCSR426MLQ /libraries/ENCLB368QMP/\n",
    "Problems with ENCSR427OIY /libraries/ENCLB319YUS/\n",
    "Problems with ENCSR427OIY /libraries/ENCLB319YUS/\n",
    "Problems with ENCSR427OIY /libraries/ENCLB319YUS/\n",
    "Problems with ENCSR427OIY /libraries/ENCLB319YUS/\n",
    "Problems with ENCSR429ZKT /libraries/ENCLB523JWN/\n",
    "Problems with ENCSR429ZKT /libraries/ENCLB523JWN/\n",
    "Problems with ENCSR429ZKT /libraries/ENCLB523JWN/\n",
    "Problems with ENCSR429ZKT /libraries/ENCLB523JWN/\n",
    "Problems with ENCSR441YBL /libraries/ENCLB903BTS/\n",
    "Problems with ENCSR441YBL /libraries/ENCLB903BTS/\n",
    "Problems with ENCSR441YBL /libraries/ENCLB903BTS/\n",
    "Problems with ENCSR441YBL /libraries/ENCLB903BTS/\n",
    "Problems with ENCSR462YAG /libraries/ENCLB954TLZ/\n",
    "Problems with ENCSR462YAG /libraries/ENCLB954TLZ/\n",
    "Problems with ENCSR462YAG /libraries/ENCLB954TLZ/\n",
    "Problems with ENCSR462YAG /libraries/ENCLB954TLZ/\n",
    "Problems with ENCSR470GMO /libraries/ENCLB592ZMI/\n",
    "Problems with ENCSR470GMO /libraries/ENCLB592ZMI/\n",
    "Problems with ENCSR470GMO /libraries/ENCLB592ZMI/\n",
    "Problems with ENCSR470GMO /libraries/ENCLB592ZMI/\n",
    "Problems with ENCSR477EIC /libraries/ENCLB983QCI/\n",
    "Problems with ENCSR477EIC /libraries/ENCLB983QCI/\n",
    "Problems with ENCSR488RZS /libraries/ENCLB989CJP/\n",
    "Problems with ENCSR488RZS /libraries/ENCLB989CJP/\n",
    "Problems with ENCSR488RZS /libraries/ENCLB989CJP/\n",
    "Problems with ENCSR488RZS /libraries/ENCLB989CJP/\n",
    "Problems with ENCSR489KPP /libraries/ENCLB399LDV/\n",
    "Problems with ENCSR489KPP /libraries/ENCLB399LDV/\n",
    "Problems with ENCSR489KPP /libraries/ENCLB399LDV/\n",
    "Problems with ENCSR489KPP /libraries/ENCLB399LDV/\n",
    "Problems with ENCSR493ZQU /libraries/ENCLB862YMG/\n",
    "Problems with ENCSR493ZQU /libraries/ENCLB862YMG/\n",
    "Problems with ENCSR493ZQU /libraries/ENCLB862YMG/\n",
    "Problems with ENCSR493ZQU /libraries/ENCLB862YMG/\n",
    "Problems with ENCSR498KBD /libraries/ENCLB188WUM/\n",
    "Problems with ENCSR498KBD /libraries/ENCLB188WUM/\n",
    "Problems with ENCSR510HAO /libraries/ENCLB805MAL/\n",
    "Problems with ENCSR510HAO /libraries/ENCLB805MAL/\n",
    "Problems with ENCSR510HAO /libraries/ENCLB805MAL/\n",
    "Problems with ENCSR510HAO /libraries/ENCLB805MAL/\n",
    "Problems with ENCSR511XOL /libraries/ENCLB460OEF/\n",
    "Problems with ENCSR511XOL /libraries/ENCLB460OEF/\n",
    "Problems with ENCSR511XOL /libraries/ENCLB460OEF/\n",
    "Problems with ENCSR511XOL /libraries/ENCLB460OEF/\n",
    "Problems with ENCSR519HYQ /libraries/ENCLB369KJQ/\n",
    "Problems with ENCSR519HYQ /libraries/ENCLB369KJQ/\n",
    "Problems with ENCSR519HYQ /libraries/ENCLB369KJQ/\n",
    "Problems with ENCSR519HYQ /libraries/ENCLB369KJQ/\n",
    "Problems with ENCSR526WJT /libraries/ENCLB656UZY/\n",
    "Problems with ENCSR526WJT /libraries/ENCLB656UZY/\n",
    "Problems with ENCSR526WJT /libraries/ENCLB656UZY/\n",
    "Problems with ENCSR526WJT /libraries/ENCLB656UZY/\n",
    "Problems with ENCSR536EAF /libraries/ENCLB891WFR/\n",
    "Problems with ENCSR536EAF /libraries/ENCLB891WFR/\n",
    "Problems with ENCSR536EAF /libraries/ENCLB891WFR/\n",
    "Problems with ENCSR536EAF /libraries/ENCLB891WFR/\n",
    "Problems with ENCSR542LRP /libraries/ENCLB700LWU/\n",
    "Problems with ENCSR542LRP /libraries/ENCLB700LWU/\n",
    "Problems with ENCSR555ATU /libraries/ENCLB908IUA/\n",
    "Problems with ENCSR555ATU /libraries/ENCLB908IUA/\n",
    "Problems with ENCSR560WNI /libraries/ENCLB354YXN/\n",
    "Problems with ENCSR560WNI /libraries/ENCLB354YXN/\n",
    "Problems with ENCSR560WNI /libraries/ENCLB354YXN/\n",
    "Problems with ENCSR560WNI /libraries/ENCLB354YXN/\n",
    "Problems with ENCSR561PWI /libraries/ENCLB785FXH/\n",
    "Problems with ENCSR561PWI /libraries/ENCLB785FXH/\n",
    "Problems with ENCSR563WPL /libraries/ENCLB782BWD/\n",
    "Problems with ENCSR563WPL /libraries/ENCLB782BWD/\n",
    "Problems with ENCSR563WPL /libraries/ENCLB782BWD/\n",
    "Problems with ENCSR563WPL /libraries/ENCLB782BWD/\n",
    "Problems with ENCSR566XHU /libraries/ENCLB416NMR/\n",
    "Problems with ENCSR566XHU /libraries/ENCLB416NMR/\n",
    "Problems with ENCSR576RBX /libraries/ENCLB902RKO/\n",
    "Problems with ENCSR576RBX /libraries/ENCLB902RKO/\n",
    "Problems with ENCSR586LMA /libraries/ENCLB332MGL/\n",
    "Problems with ENCSR586LMA /libraries/ENCLB332MGL/\n",
    "Problems with ENCSR586LMA /libraries/ENCLB332MGL/\n",
    "Problems with ENCSR586LMA /libraries/ENCLB332MGL/\n",
    "Problems with ENCSR595LVS /libraries/ENCLB678GYJ/\n",
    "Problems with ENCSR595LVS /libraries/ENCLB678GYJ/\n",
    "Problems with ENCSR595LVS /libraries/ENCLB678GYJ/\n",
    "Problems with ENCSR595LVS /libraries/ENCLB678GYJ/\n",
    "Problems with ENCSR601CQM /libraries/ENCLB524IGV/\n",
    "Problems with ENCSR601CQM /libraries/ENCLB524IGV/\n",
    "Problems with ENCSR601CQM /libraries/ENCLB524IGV/\n",
    "Problems with ENCSR601CQM /libraries/ENCLB524IGV/\n",
    "Problems with ENCSR635GFL /libraries/ENCLB589IWN/\n",
    "Problems with ENCSR635GFL /libraries/ENCLB589IWN/\n",
    "Problems with ENCSR635GFL /libraries/ENCLB589IWN/\n",
    "Problems with ENCSR635GFL /libraries/ENCLB589IWN/\n",
    "Problems with ENCSR635GRA /libraries/ENCLB250XVZ/\n",
    "Problems with ENCSR635GRA /libraries/ENCLB250XVZ/\n",
    "Problems with ENCSR635GRA /libraries/ENCLB250XVZ/\n",
    "Problems with ENCSR635GRA /libraries/ENCLB250XVZ/\n",
    "Problems with ENCSR637AZA /libraries/ENCLB581NBT/\n",
    "Problems with ENCSR637AZA /libraries/ENCLB581NBT/\n",
    "Problems with ENCSR637AZA /libraries/ENCLB581NBT/\n",
    "Problems with ENCSR637AZA /libraries/ENCLB581NBT/\n",
    "Problems with ENCSR642TKQ /libraries/ENCLB765ZDM/\n",
    "Problems with ENCSR642TKQ /libraries/ENCLB765ZDM/\n",
    "Problems with ENCSR644EUQ /libraries/ENCLB146JLP/\n",
    "Problems with ENCSR644EUQ /libraries/ENCLB146JLP/\n",
    "Problems with ENCSR644YWO /libraries/ENCLB517TXV/\n",
    "Problems with ENCSR644YWO /libraries/ENCLB517TXV/\n",
    "Problems with ENCSR644YWO /libraries/ENCLB517TXV/\n",
    "Problems with ENCSR644YWO /libraries/ENCLB517TXV/\n",
    "Problems with ENCSR651QGH /libraries/ENCLB308YIM/\n",
    "Problems with ENCSR651QGH /libraries/ENCLB308YIM/\n",
    "Problems with ENCSR651QGH /libraries/ENCLB308YIM/\n",
    "Problems with ENCSR651QGH /libraries/ENCLB308YIM/\n",
    "Problems with ENCSR671RJG /libraries/ENCLB885OKL/\n",
    "Problems with ENCSR671RJG /libraries/ENCLB885OKL/\n",
    "Problems with ENCSR671RJG /libraries/ENCLB885OKL/\n",
    "Problems with ENCSR671RJG /libraries/ENCLB885OKL/\n",
    "Problems with ENCSR692OZP /libraries/ENCLB194LYE/\n",
    "Problems with ENCSR692OZP /libraries/ENCLB194LYE/\n",
    "Problems with ENCSR692OZP /libraries/ENCLB194LYE/\n",
    "Problems with ENCSR692OZP /libraries/ENCLB194LYE/\n",
    "Problems with ENCSR692ZWB /libraries/ENCLB598TEK/\n",
    "Problems with ENCSR692ZWB /libraries/ENCLB598TEK/\n",
    "Problems with ENCSR692ZWB /libraries/ENCLB598TEK/\n",
    "Problems with ENCSR692ZWB /libraries/ENCLB598TEK/\n",
    "Problems with ENCSR694DQP /libraries/ENCLB170MZP/\n",
    "Problems with ENCSR694DQP /libraries/ENCLB170MZP/\n",
    "Problems with ENCSR694DQP /libraries/ENCLB170MZP/\n",
    "Problems with ENCSR694DQP /libraries/ENCLB170MZP/\n",
    "Problems with ENCSR732MDS /libraries/ENCLB701VGG/\n",
    "Problems with ENCSR732MDS /libraries/ENCLB701VGG/\n",
    "Problems with ENCSR732MDS /libraries/ENCLB701VGG/\n",
    "Problems with ENCSR732MDS /libraries/ENCLB701VGG/\n",
    "Problems with ENCSR745UMG /libraries/ENCLB709WAG/\n",
    "Problems with ENCSR745UMG /libraries/ENCLB709WAG/\n",
    "Problems with ENCSR773FPF /libraries/ENCLB536TLF/\n",
    "Problems with ENCSR773FPF /libraries/ENCLB536TLF/\n",
    "Problems with ENCSR773FPF /libraries/ENCLB536TLF/\n",
    "Problems with ENCSR773FPF /libraries/ENCLB536TLF/\n",
    "Problems with ENCSR774KTU /libraries/ENCLB727KBM/\n",
    "Problems with ENCSR774KTU /libraries/ENCLB727KBM/\n",
    "Problems with ENCSR774KTU /libraries/ENCLB727KBM/\n",
    "Problems with ENCSR774KTU /libraries/ENCLB727KBM/\n",
    "Problems with ENCSR799GLH /libraries/ENCLB899ORD/\n",
    "Problems with ENCSR799GLH /libraries/ENCLB899ORD/\n",
    "Problems with ENCSR799GLH /libraries/ENCLB899ORD/\n",
    "Problems with ENCSR799GLH /libraries/ENCLB899ORD/\n",
    "Problems with ENCSR801MJO /libraries/ENCLB449BXX/\n",
    "Problems with ENCSR801MJO /libraries/ENCLB449BXX/\n",
    "Problems with ENCSR801MJO /libraries/ENCLB449BXX/\n",
    "Problems with ENCSR801MJO /libraries/ENCLB449BXX/\n",
    "Problems with ENCSR813ULV /libraries/ENCLB963QFG/\n",
    "Problems with ENCSR813ULV /libraries/ENCLB963QFG/\n",
    "Problems with ENCSR813ULV /libraries/ENCLB963QFG/\n",
    "Problems with ENCSR813ULV /libraries/ENCLB963QFG/\n",
    "Problems with ENCSR827YHY /libraries/ENCLB389BSV/\n",
    "Problems with ENCSR827YHY /libraries/ENCLB389BSV/\n",
    "Problems with ENCSR827YHY /libraries/ENCLB389BSV/\n",
    "Problems with ENCSR827YHY /libraries/ENCLB389BSV/\n",
    "Problems with ENCSR828TLP /libraries/ENCLB611DGC/\n",
    "Problems with ENCSR828TLP /libraries/ENCLB611DGC/\n",
    "Problems with ENCSR828TLP /libraries/ENCLB611DGC/\n",
    "Problems with ENCSR828TLP /libraries/ENCLB611DGC/\n",
    "Problems with ENCSR830MSX /libraries/ENCLB231PCO/\n",
    "Problems with ENCSR830MSX /libraries/ENCLB231PCO/\n",
    "Problems with ENCSR830MSX /libraries/ENCLB231PCO/\n",
    "Problems with ENCSR830MSX /libraries/ENCLB231PCO/\n",
    "Problems with ENCSR834SCG /libraries/ENCLB849DVZ/\n",
    "Problems with ENCSR834SCG /libraries/ENCLB849DVZ/\n",
    "Problems with ENCSR834SCG /libraries/ENCLB849DVZ/\n",
    "Problems with ENCSR834SCG /libraries/ENCLB849DVZ/\n",
    "Problems with ENCSR847STS /libraries/ENCLB735WRJ/\n",
    "Problems with ENCSR847STS /libraries/ENCLB735WRJ/\n",
    "Problems with ENCSR847STS /libraries/ENCLB735WRJ/\n",
    "Problems with ENCSR847STS /libraries/ENCLB735WRJ/\n",
    "Problems with ENCSR851RNW /libraries/ENCLB349QMF/\n",
    "Problems with ENCSR851RNW /libraries/ENCLB349QMF/\n",
    "Problems with ENCSR851RNW /libraries/ENCLB349QMF/\n",
    "Problems with ENCSR851RNW /libraries/ENCLB349QMF/\n",
    "Problems with ENCSR852NQO /libraries/ENCLB299EAG/\n",
    "Problems with ENCSR852NQO /libraries/ENCLB299EAG/\n",
    "Problems with ENCSR852NQO /libraries/ENCLB299EAG/\n",
    "Problems with ENCSR852NQO /libraries/ENCLB299EAG/\n",
    "Problems with ENCSR855ZYP /libraries/ENCLB193HSH/\n",
    "Problems with ENCSR855ZYP /libraries/ENCLB193HSH/\n",
    "Problems with ENCSR855ZYP /libraries/ENCLB193HSH/\n",
    "Problems with ENCSR855ZYP /libraries/ENCLB193HSH/\n",
    "Problems with ENCSR877CFF /libraries/ENCLB227FXC/\n",
    "Problems with ENCSR877CFF /libraries/ENCLB227FXC/\n",
    "Problems with ENCSR877CFF /libraries/ENCLB227FXC/\n",
    "Problems with ENCSR877CFF /libraries/ENCLB227FXC/\n",
    "Problems with ENCSR877UAI /libraries/ENCLB747XAC/\n",
    "Problems with ENCSR877UAI /libraries/ENCLB747XAC/\n",
    "Problems with ENCSR877UAI /libraries/ENCLB747XAC/\n",
    "Problems with ENCSR877UAI /libraries/ENCLB747XAC/\n",
    "Problems with ENCSR880DYE /libraries/ENCLB977UQY/\n",
    "Problems with ENCSR880DYE /libraries/ENCLB977UQY/\n",
    "Problems with ENCSR880DYE /libraries/ENCLB977UQY/\n",
    "Problems with ENCSR880DYE /libraries/ENCLB977UQY/\n",
    "Problems with ENCSR880MVS /libraries/ENCLB586GGX/\n",
    "Problems with ENCSR880MVS /libraries/ENCLB586GGX/\n",
    "Problems with ENCSR880MVS /libraries/ENCLB586GGX/\n",
    "Problems with ENCSR880MVS /libraries/ENCLB586GGX/\n",
    "Problems with ENCSR888NNP /libraries/ENCLB639GHE/\n",
    "Problems with ENCSR888NNP /libraries/ENCLB639GHE/\n",
    "Problems with ENCSR888NNP /libraries/ENCLB639GHE/\n",
    "Problems with ENCSR888NNP /libraries/ENCLB639GHE/\n",
    "Problems with ENCSR894JVL /libraries/ENCLB868TVO/\n",
    "Problems with ENCSR894JVL /libraries/ENCLB868TVO/\n",
    "Problems with ENCSR894JVL /libraries/ENCLB868TVO/\n",
    "Problems with ENCSR894JVL /libraries/ENCLB868TVO/\n",
    "Problems with ENCSR905YQE /libraries/ENCLB822SVY/\n",
    "Problems with ENCSR905YQE /libraries/ENCLB822SVY/\n",
    "Problems with ENCSR905YQE /libraries/ENCLB822SVY/\n",
    "Problems with ENCSR905YQE /libraries/ENCLB822SVY/\n",
    "Problems with ENCSR912IHV /libraries/ENCLB849GNT/\n",
    "Problems with ENCSR912IHV /libraries/ENCLB849GNT/\n",
    "Problems with ENCSR912IHV /libraries/ENCLB849GNT/\n",
    "Problems with ENCSR912IHV /libraries/ENCLB849GNT/\n",
    "Problems with ENCSR915UNZ /libraries/ENCLB151ZAG/\n",
    "Problems with ENCSR915UNZ /libraries/ENCLB151ZAG/\n",
    "Problems with ENCSR915UNZ /libraries/ENCLB151ZAG/\n",
    "Problems with ENCSR915UNZ /libraries/ENCLB151ZAG/\n",
    "Problems with ENCSR923NSW /libraries/ENCLB143DFU/\n",
    "Problems with ENCSR923NSW /libraries/ENCLB143DFU/\n",
    "Problems with ENCSR923NSW /libraries/ENCLB143DFU/\n",
    "Problems with ENCSR923NSW /libraries/ENCLB143DFU/\n",
    "Problems with ENCSR929PEE /libraries/ENCLB959AIF/\n",
    "Problems with ENCSR929PEE /libraries/ENCLB959AIF/\n",
    "Problems with ENCSR932DTM /libraries/ENCLB905AUU/\n",
    "Problems with ENCSR932DTM /libraries/ENCLB905AUU/\n",
    "Problems with ENCSR932DTM /libraries/ENCLB905AUU/\n",
    "Problems with ENCSR932DTM /libraries/ENCLB905AUU/\n",
    "Problems with ENCSR948XEM /libraries/ENCLB315FQV/\n",
    "Problems with ENCSR948XEM /libraries/ENCLB315FQV/\n",
    "Problems with ENCSR962DRE /libraries/ENCLB697JVD/\n",
    "Problems with ENCSR962DRE /libraries/ENCLB697JVD/\n",
    "Problems with ENCSR962DRE /libraries/ENCLB697JVD/\n",
    "Problems with ENCSR962DRE /libraries/ENCLB697JVD/\n",
    "Problems with ENCSR969VQQ /libraries/ENCLB170FEN/\n",
    "Problems with ENCSR969VQQ /libraries/ENCLB170FEN/\n",
    "Problems with ENCSR972KJU /libraries/ENCLB762GOP/\n",
    "Problems with ENCSR972KJU /libraries/ENCLB762GOP/\n",
    "Problems with ENCSR972KJU /libraries/ENCLB762GOP/\n",
    "Problems with ENCSR972KJU /libraries/ENCLB762GOP/\n",
    "Problems with ENCSR973ZJO /libraries/ENCLB374JXM/\n",
    "Problems with ENCSR973ZJO /libraries/ENCLB374JXM/\n",
    "Problems with ENCSR973ZJO /libraries/ENCLB374JXM/\n",
    "Problems with ENCSR973ZJO /libraries/ENCLB374JXM/\n",
    "Problems with ENCSR974VTR /libraries/ENCLB415PDV/\n",
    "Problems with ENCSR974VTR /libraries/ENCLB415PDV/\n",
    "Problems with ENCSR974VTR /libraries/ENCLB415PDV/\n",
    "Problems with ENCSR974VTR /libraries/ENCLB415PDV/\n",
    "Problems with ENCSR976FIL /libraries/ENCLB124EIT/\n",
    "Problems with ENCSR976FIL /libraries/ENCLB124EIT/\n",
    "Problems with ENCSR976FIL /libraries/ENCLB124EIT/\n",
    "Problems with ENCSR976FIL /libraries/ENCLB124EIT/\n",
    "Problems with ENCSR983WDT /libraries/ENCLB708YOS/\n",
    "Problems with ENCSR983WDT /libraries/ENCLB708YOS/\n",
    "Problems with ENCSR983WDT /libraries/ENCLB708YOS/\n",
    "Problems with ENCSR983WDT /libraries/ENCLB708YOS/\n",
    "Problems with ENCSR985FZQ /libraries/ENCLB292IDD/\n",
    "Problems with ENCSR985FZQ /libraries/ENCLB292IDD/\n",
    "Problems with ENCSR985FZQ /libraries/ENCLB292IDD/\n",
    "Problems with ENCSR985FZQ /libraries/ENCLB292IDD/\n",
    "Problems with ENCSR986QTH /libraries/ENCLB571XKG/\n",
    "Problems with ENCSR986QTH /libraries/ENCLB571XKG/\n",
    "Problems with ENCSR109DAN /libraries/ENCLB527JDU/\n",
    "Problems with ENCSR109DAN /libraries/ENCLB527JDU/\n",
    "Problems with ENCSR388EIN /libraries/ENCLB952MZJ/\n",
    "Problems with ENCSR388EIN /libraries/ENCLB952MZJ/\n",
    "Problems with ENCSR452OYH /libraries/ENCLB294FBZ/\n",
    "Problems with ENCSR452OYH /libraries/ENCLB294FBZ/\n",
    "Problems with ENCSR503QZJ /libraries/ENCLB129LMS/\n",
    "Problems with ENCSR503QZJ /libraries/ENCLB129LMS/\n",
    "Problems with ENCSR572SMD /libraries/ENCLB273ZWE/\n",
    "Problems with ENCSR572SMD /libraries/ENCLB273ZWE/\n",
    "Problems with ENCSR877NAX /libraries/ENCLB514RKT/\n",
    "Problems with ENCSR877NAX /libraries/ENCLB514RKT/\"\"\".split(\"\\n\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "blind-subsection",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(problems)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "outer-tyler",
   "metadata": {},
   "outputs": [],
   "source": [
    "problem_experiments = []\n",
    "for row in problems:\n",
    "    problem_experiments.append(row.split()[2])\n",
    "problem_experiments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "direct-package",
   "metadata": {},
   "outputs": [],
   "source": [
    "processed = set()\n",
    "processed.update(experiments_to_run_2_6)\n",
    "processed.update(experiments_to_run_7_21)\n",
    "processed.update(experiments_to_run_22_51)\n",
    "processed.update(experiments_to_run_52_81)\n",
    "processed.update(experiments_to_run_82_107)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "rocky-decision",
   "metadata": {},
   "outputs": [],
   "source": [
    "processed.intersection(problem_experiments)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "manual-annex",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.1+"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
