{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "norwegian-racing",
   "metadata": {},
   "source": [
    "# Introduction"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "commercial-assurance",
   "metadata": {},
   "source": [
    "We discussed what to do witht the pipeline configuration files, and it was suggested I submit them as attachments.\n",
    "\n",
    "They attachments may also need to be added to Analysis objects that Jennifer added.\n",
    "\n",
    "The DCC doesn't support yaml, so I'll probably need to submit them as plain text."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "average-shift",
   "metadata": {},
   "outputs": [],
   "source": [
    "from datetime import datetime\n",
    "from pathlib import Path\n",
    "import pandas\n",
    "import os\n",
    "import sys\n",
    "from urllib.parse import quote_plus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "spiritual-utility",
   "metadata": {},
   "outputs": [],
   "source": [
    "EC = str(Path(\"~/proj/encoded_client\").expanduser())\n",
    "if EC not in sys.path:\n",
    "    sys.path.append(EC)\n",
    "\n",
    "from encoded_client.encoded import ENCODED, DCCValidator, Document, HTTPError"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "photographic-sellers",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total 616\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uuid</th>\n",
       "      <th>local_filename</th>\n",
       "      <th>remote_filename</th>\n",
       "      <th>mime_type</th>\n",
       "      <th>document type</th>\n",
       "      <th>md5sum</th>\n",
       "      <th>library_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>None</td>\n",
       "      <td>production/stanford/ENCLB527WWJ/config.yaml</td>\n",
       "      <td>production/stanford/ENCLB527WWJ/config.yaml.txt</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>None</td>\n",
       "      <td>ENCLB527WWJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>None</td>\n",
       "      <td>production/stanford/ENCLB619VVF/config.yaml</td>\n",
       "      <td>production/stanford/ENCLB619VVF/config.yaml.txt</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>None</td>\n",
       "      <td>ENCLB619VVF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>None</td>\n",
       "      <td>production/stanford/ENCLB002DZK/config.yaml</td>\n",
       "      <td>production/stanford/ENCLB002DZK/config.yaml.txt</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>None</td>\n",
       "      <td>ENCLB002DZK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>None</td>\n",
       "      <td>production/stanford/ENCLB814PUF/config.yaml</td>\n",
       "      <td>production/stanford/ENCLB814PUF/config.yaml.txt</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>None</td>\n",
       "      <td>ENCLB814PUF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>None</td>\n",
       "      <td>production/stanford/ENCLB280ZGL/config.yaml</td>\n",
       "      <td>production/stanford/ENCLB280ZGL/config.yaml.txt</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>None</td>\n",
       "      <td>ENCLB280ZGL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>611</th>\n",
       "      <td>None</td>\n",
       "      <td>production/uci/ENCLB595FHR/config.yaml</td>\n",
       "      <td>production/uci/ENCLB595FHR/config.yaml.txt</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>None</td>\n",
       "      <td>ENCLB595FHR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>612</th>\n",
       "      <td>None</td>\n",
       "      <td>production/uci/ENCLB947PWM/config.yaml</td>\n",
       "      <td>production/uci/ENCLB947PWM/config.yaml.txt</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>None</td>\n",
       "      <td>ENCLB947PWM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>613</th>\n",
       "      <td>None</td>\n",
       "      <td>production/uci/ENCLB106VMV/config.yaml</td>\n",
       "      <td>production/uci/ENCLB106VMV/config.yaml.txt</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>None</td>\n",
       "      <td>ENCLB106VMV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>614</th>\n",
       "      <td>None</td>\n",
       "      <td>production/uci/ENCLB327QWK/config.yaml</td>\n",
       "      <td>production/uci/ENCLB327QWK/config.yaml.txt</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>None</td>\n",
       "      <td>ENCLB327QWK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>615</th>\n",
       "      <td>None</td>\n",
       "      <td>production/uci/ENCLB217YGZ/config.yaml</td>\n",
       "      <td>production/uci/ENCLB217YGZ/config.yaml.txt</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>None</td>\n",
       "      <td>ENCLB217YGZ</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>616 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     uuid                               local_filename  \\\n",
       "0    None  production/stanford/ENCLB527WWJ/config.yaml   \n",
       "1    None  production/stanford/ENCLB619VVF/config.yaml   \n",
       "2    None  production/stanford/ENCLB002DZK/config.yaml   \n",
       "3    None  production/stanford/ENCLB814PUF/config.yaml   \n",
       "4    None  production/stanford/ENCLB280ZGL/config.yaml   \n",
       "..    ...                                          ...   \n",
       "611  None       production/uci/ENCLB595FHR/config.yaml   \n",
       "612  None       production/uci/ENCLB947PWM/config.yaml   \n",
       "613  None       production/uci/ENCLB106VMV/config.yaml   \n",
       "614  None       production/uci/ENCLB327QWK/config.yaml   \n",
       "615  None       production/uci/ENCLB217YGZ/config.yaml   \n",
       "\n",
       "                                     remote_filename   mime_type  \\\n",
       "0    production/stanford/ENCLB527WWJ/config.yaml.txt  text/plain   \n",
       "1    production/stanford/ENCLB619VVF/config.yaml.txt  text/plain   \n",
       "2    production/stanford/ENCLB002DZK/config.yaml.txt  text/plain   \n",
       "3    production/stanford/ENCLB814PUF/config.yaml.txt  text/plain   \n",
       "4    production/stanford/ENCLB280ZGL/config.yaml.txt  text/plain   \n",
       "..                                               ...         ...   \n",
       "611       production/uci/ENCLB595FHR/config.yaml.txt  text/plain   \n",
       "612       production/uci/ENCLB947PWM/config.yaml.txt  text/plain   \n",
       "613       production/uci/ENCLB106VMV/config.yaml.txt  text/plain   \n",
       "614       production/uci/ENCLB327QWK/config.yaml.txt  text/plain   \n",
       "615       production/uci/ENCLB217YGZ/config.yaml.txt  text/plain   \n",
       "\n",
       "         document type md5sum   library_id  \n",
       "0    workflow metadata   None  ENCLB527WWJ  \n",
       "1    workflow metadata   None  ENCLB619VVF  \n",
       "2    workflow metadata   None  ENCLB002DZK  \n",
       "3    workflow metadata   None  ENCLB814PUF  \n",
       "4    workflow metadata   None  ENCLB280ZGL  \n",
       "..                 ...    ...          ...  \n",
       "611  workflow metadata   None  ENCLB595FHR  \n",
       "612  workflow metadata   None  ENCLB947PWM  \n",
       "613  workflow metadata   None  ENCLB106VMV  \n",
       "614  workflow metadata   None  ENCLB327QWK  \n",
       "615  workflow metadata   None  ENCLB217YGZ  \n",
       "\n",
       "[616 rows x 7 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "library_configs = []\n",
    "production = Path(\"production\")\n",
    "for lab in production.iterdir():\n",
    "    for library in lab.iterdir():\n",
    "        if library.is_dir():\n",
    "            library_id = library.name\n",
    "            library_configs.append({\n",
    "                \"uuid\": None,\n",
    "                \"local_filename\": library / \"config.yaml\",\n",
    "                \"remote_filename\": str(library / \"config.yaml.txt\"),\n",
    "                \"mime_type\": \"text/plain\",\n",
    "                \"document type\": \"workflow metadata\",\n",
    "                \"md5sum\": None,\n",
    "                \"library_id\": library_id\n",
    "            })\n",
    "library_configs = pandas.DataFrame(library_configs)\n",
    "print(\"Total\", library_configs.shape[0])\n",
    "library_configs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "premium-stanford",
   "metadata": {},
   "outputs": [],
   "source": [
    "#server = ENCODED(\"test.encodedcc.org\")\n",
    "server = ENCODED(\"www.encodeproject.org\")\n",
    "validator = DCCValidator(server)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "stunning-vatican",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/analyses/ENCAN471PLI/'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_experiment_by_library_id(server, library_id):\n",
    "    result = server.search_jsonld(searchTerm=library_id)\n",
    "    for experiment_row in result[\"@graph\"]:\n",
    "        if 'Experiment' in experiment_row[\"@type\"]:\n",
    "            return server.get_json(experiment_row[\"@id\"])\n",
    "\n",
    "def get_analysis_by_library_id(server, library_id):\n",
    "    experiment = get_experiment_by_library_id(server, library_id)\n",
    "    if experiment is None:\n",
    "        return\n",
    "    \n",
    "    default_analysis = experiment.get(\"default_analysis\")\n",
    "    if default_analysis is None: \n",
    "        return\n",
    "    \n",
    "    for analysis in experiment[\"analyses\"]:\n",
    "        if analysis[\"@id\"] == default_analysis:\n",
    "            return analysis\n",
    "\n",
    "get_analysis_by_library_id(server, \"ENCLB280ZGL\")[\"@id\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "excess-japan",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>date_created</th>\n",
       "      <th>@id</th>\n",
       "      <th>description</th>\n",
       "      <th>document_type</th>\n",
       "      <th>mime_type</th>\n",
       "      <th>remote_filename</th>\n",
       "      <th>href</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2014-06-27 17:43:53.305108+00:00</td>\n",
       "      <td>/documents/e84de88e-4c31-4b2a-b6ac-96feb764b2d0/</td>\n",
       "      <td>mRNA Selection Protocol</td>\n",
       "      <td>extraction protocol</td>\n",
       "      <td>application/pdf</td>\n",
       "      <td>dynabeads_mrna_purification_man.pdf</td>\n",
       "      <td>@@download/attachment/dynabeads_mrna_purificat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2017-09-25 18:17:18.814210+00:00</td>\n",
       "      <td>/documents/92251b87-30e2-4a32-a3e3-c4b3c7856329/</td>\n",
       "      <td>This document describes the Fluidigm C1 cell i...</td>\n",
       "      <td>other</td>\n",
       "      <td>application/pdf</td>\n",
       "      <td>C1_mRNA-Seq_pr_100-7168L1 (2).pdf</td>\n",
       "      <td>@@download/attachment/C1_mRNA-Seq_pr_100-7168L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2022-06-13 22:51:25.495498+00:00</td>\n",
       "      <td>/documents/5441e3af-f996-4a51-b2a2-982e3427a6bc/</td>\n",
       "      <td>Configuration file for scRNA-seq pipeline</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>production%2Fuci%2FENCLB217YGZ%2Fconfig.yaml.txt</td>\n",
       "      <td>@@download/attachment/production%252Fuci%252FE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2022-06-13 22:51:34.446087+00:00</td>\n",
       "      <td>/documents/89831fba-ae30-4af5-a619-3eadaba73f72/</td>\n",
       "      <td>Configuration file for scRNA-seq pipeline</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>production%2Fstanford%2FENCLB123ZDT%2Fconfig.y...</td>\n",
       "      <td>@@download/attachment/production%252Fstanford%...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2022-06-13 22:52:43.578942+00:00</td>\n",
       "      <td>/documents/484d361b-8d11-433d-8398-44f58eb0f700/</td>\n",
       "      <td>Configuration file for scRNA-seq pipeline</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>production%2Fstanford%2FENCLB526RVS%2Fconfig.y...</td>\n",
       "      <td>@@download/attachment/production%252Fstanford%...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      date_created  \\\n",
       "0 2014-06-27 17:43:53.305108+00:00   \n",
       "1 2017-09-25 18:17:18.814210+00:00   \n",
       "2 2022-06-13 22:51:25.495498+00:00   \n",
       "3 2022-06-13 22:51:34.446087+00:00   \n",
       "4 2022-06-13 22:52:43.578942+00:00   \n",
       "\n",
       "                                                @id  \\\n",
       "0  /documents/e84de88e-4c31-4b2a-b6ac-96feb764b2d0/   \n",
       "1  /documents/92251b87-30e2-4a32-a3e3-c4b3c7856329/   \n",
       "2  /documents/5441e3af-f996-4a51-b2a2-982e3427a6bc/   \n",
       "3  /documents/89831fba-ae30-4af5-a619-3eadaba73f72/   \n",
       "4  /documents/484d361b-8d11-433d-8398-44f58eb0f700/   \n",
       "\n",
       "                                         description        document_type  \\\n",
       "0                            mRNA Selection Protocol  extraction protocol   \n",
       "1  This document describes the Fluidigm C1 cell i...                other   \n",
       "2          Configuration file for scRNA-seq pipeline    workflow metadata   \n",
       "3          Configuration file for scRNA-seq pipeline    workflow metadata   \n",
       "4          Configuration file for scRNA-seq pipeline    workflow metadata   \n",
       "\n",
       "         mime_type                                    remote_filename  \\\n",
       "0  application/pdf                dynabeads_mrna_purification_man.pdf   \n",
       "1  application/pdf                  C1_mRNA-Seq_pr_100-7168L1 (2).pdf   \n",
       "2       text/plain   production%2Fuci%2FENCLB217YGZ%2Fconfig.yaml.txt   \n",
       "3       text/plain  production%2Fstanford%2FENCLB123ZDT%2Fconfig.y...   \n",
       "4       text/plain  production%2Fstanford%2FENCLB526RVS%2Fconfig.y...   \n",
       "\n",
       "                                                href  \n",
       "0  @@download/attachment/dynabeads_mrna_purificat...  \n",
       "1  @@download/attachment/C1_mRNA-Seq_pr_100-7168L...  \n",
       "2  @@download/attachment/production%252Fuci%252FE...  \n",
       "3  @@download/attachment/production%252Fstanford%...  \n",
       "4  @@download/attachment/production%252Fstanford%...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_posted_documents(server):\n",
    "    posted_documents = []\n",
    "    query = \"/search/?searchTerm=@Documents&lab.title=Barbara%20Wold,%20Caltech\"\n",
    "    graph = server.get_json(query)\n",
    "    for row in graph[\"@graph\"]:\n",
    "        attachment = row[\"attachment\"]\n",
    "        posted_documents.append({\n",
    "            \"date_created\": datetime.strptime(row[\"date_created\"], \"%Y-%m-%dT%H:%M:%S.%f%z\"),\n",
    "            \"@id\": row[\"@id\"],\n",
    "            \"description\": row[\"description\"], \n",
    "            \"document_type\": row[\"document_type\"], \n",
    "            \"mime_type\": attachment[\"type\"], \n",
    "            \"remote_filename\": attachment[\"download\"],\n",
    "            \"href\": attachment[\"href\"]\n",
    "        })\n",
    "    posted_documents = pandas.DataFrame(posted_documents)\n",
    "    return posted_documents\n",
    "\n",
    "posted_documents = get_posted_documents(server)\n",
    "posted_documents.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "constant-vietnam",
   "metadata": {},
   "outputs": [],
   "source": [
    "#posted_documents.set_index(\"remote_filename\").loc['production/stanford/ENCLB527WWJ/config.yaml.txt']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "extraordinary-cornwall",
   "metadata": {},
   "outputs": [],
   "source": [
    "for name in posted_documents.set_index(\"remote_filename\").index:\n",
    "    if \"ENCLB527WWJ\" in name:\n",
    "        print(name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "stuck-wagon",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "searching production%2Fstanford%2FENCLB526RVS%2Fconfig.yaml.txt\n",
      "date_created                        2022-06-13 22:52:43.578942+00:00\n",
      "@id                 /documents/484d361b-8d11-433d-8398-44f58eb0f700/\n",
      "description                Configuration file for scRNA-seq pipeline\n",
      "document_type                                      workflow metadata\n",
      "mime_type                                                 text/plain\n",
      "remote_filename    production%2Fstanford%2FENCLB526RVS%2Fconfig.y...\n",
      "href               @@download/attachment/production%252Fstanford%...\n",
      "Name: 4, dtype: object\n",
      "searching hope_not_real.txt\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "def get_document_from_posted(posted, filename):\n",
    "    escaped = quote_plus(filename)\n",
    "    print(\"searching {}\".format(escaped))\n",
    "    candidates = posted[posted[\"remote_filename\"] == escaped]\n",
    "    if candidates.shape[0] == 0:\n",
    "        # no matches\n",
    "        return\n",
    "    elif isinstance(candidates, pandas.DataFrame):\n",
    "        return candidates.loc[candidates.first_valid_index()]\n",
    "    elif isinstance(candidates, pandas.Series):\n",
    "        return candidates\n",
    "    else:\n",
    "        raise RuntimeError(\"Unexpected internal type {}\".format(type(candidates)))\n",
    "    \n",
    "#print(get_document_from_posted(posted_documents, 'production/stanford/ENCLB527WWJ/config.yaml.txt'))\n",
    "print(get_document_from_posted(posted_documents, 'production/stanford/ENCLB526RVS/config.yaml.txt'))\n",
    "print(get_document_from_posted(posted_documents, \"hope_not_real.txt\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "artistic-penny",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "None\n"
     ]
    }
   ],
   "source": [
    "print(get_document_from_posted(posted_documents, 'production/stanford_heart_20220621/ENCLB138XBO/config.yaml'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "settled-positive",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>date_created</th>\n",
       "      <th>@id</th>\n",
       "      <th>description</th>\n",
       "      <th>document_type</th>\n",
       "      <th>mime_type</th>\n",
       "      <th>remote_filename</th>\n",
       "      <th>href</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2014-06-27 17:43:53.305108+00:00</td>\n",
       "      <td>/documents/e84de88e-4c31-4b2a-b6ac-96feb764b2d0/</td>\n",
       "      <td>mRNA Selection Protocol</td>\n",
       "      <td>extraction protocol</td>\n",
       "      <td>application/pdf</td>\n",
       "      <td>dynabeads_mrna_purification_man.pdf</td>\n",
       "      <td>@@download/attachment/dynabeads_mrna_purificat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2017-09-25 18:17:18.814210+00:00</td>\n",
       "      <td>/documents/92251b87-30e2-4a32-a3e3-c4b3c7856329/</td>\n",
       "      <td>This document describes the Fluidigm C1 cell i...</td>\n",
       "      <td>other</td>\n",
       "      <td>application/pdf</td>\n",
       "      <td>C1_mRNA-Seq_pr_100-7168L1 (2).pdf</td>\n",
       "      <td>@@download/attachment/C1_mRNA-Seq_pr_100-7168L...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2022-06-13 22:51:25.495498+00:00</td>\n",
       "      <td>/documents/5441e3af-f996-4a51-b2a2-982e3427a6bc/</td>\n",
       "      <td>Configuration file for scRNA-seq pipeline</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>production%2Fuci%2FENCLB217YGZ%2Fconfig.yaml.txt</td>\n",
       "      <td>@@download/attachment/production%252Fuci%252FE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2022-06-13 22:51:34.446087+00:00</td>\n",
       "      <td>/documents/89831fba-ae30-4af5-a619-3eadaba73f72/</td>\n",
       "      <td>Configuration file for scRNA-seq pipeline</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>production%2Fstanford%2FENCLB123ZDT%2Fconfig.y...</td>\n",
       "      <td>@@download/attachment/production%252Fstanford%...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2022-06-13 22:52:43.578942+00:00</td>\n",
       "      <td>/documents/484d361b-8d11-433d-8398-44f58eb0f700/</td>\n",
       "      <td>Configuration file for scRNA-seq pipeline</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>production%2Fstanford%2FENCLB526RVS%2Fconfig.y...</td>\n",
       "      <td>@@download/attachment/production%252Fstanford%...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2022-06-13 22:51:37.623874+00:00</td>\n",
       "      <td>/documents/8d6dd36c-210d-4d6d-ba26-c8582db6123e/</td>\n",
       "      <td>Configuration file for scRNA-seq pipeline</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>production%2Fstanford%2FENCLB063XDV%2Fconfig.y...</td>\n",
       "      <td>@@download/attachment/production%252Fstanford%...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2022-06-13 22:54:29.683734+00:00</td>\n",
       "      <td>/documents/f1971706-69cd-4e8a-8b3e-801aa6c73ebe/</td>\n",
       "      <td>Configuration file for scRNA-seq pipeline</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>production%2Fuci%2FENCLB870FRH%2Fconfig.yaml.txt</td>\n",
       "      <td>@@download/attachment/production%252Fuci%252FE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2022-06-13 22:54:12.522463+00:00</td>\n",
       "      <td>/documents/b56fe10c-42d3-48ba-b49a-936086b66350/</td>\n",
       "      <td>Configuration file for scRNA-seq pipeline</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>production%2Fuci%2FENCLB057CVT%2Fconfig.yaml.txt</td>\n",
       "      <td>@@download/attachment/production%252Fuci%252FE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2022-06-13 22:51:39.678880+00:00</td>\n",
       "      <td>/documents/279a8145-bf10-4e4a-b792-86905d1b20bf/</td>\n",
       "      <td>Configuration file for scRNA-seq pipeline</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>production%2Fstanford%2FENCLB025GOX%2Fconfig.y...</td>\n",
       "      <td>@@download/attachment/production%252Fstanford%...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2022-06-13 22:55:42.400302+00:00</td>\n",
       "      <td>/documents/fd8559fa-7535-4b44-ad3e-03f29d5ebc40/</td>\n",
       "      <td>Configuration file for scRNA-seq pipeline</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>production%2Fuci%2FENCLB215OBG%2Fconfig.yaml.txt</td>\n",
       "      <td>@@download/attachment/production%252Fuci%252FE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2022-06-13 22:55:40.327104+00:00</td>\n",
       "      <td>/documents/7fdad273-2fdc-46f9-a820-1b2dff8281fa/</td>\n",
       "      <td>Configuration file for scRNA-seq pipeline</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>production%2Fuci%2FENCLB070DDL%2Fconfig.yaml.txt</td>\n",
       "      <td>@@download/attachment/production%252Fuci%252FE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>2022-06-13 22:57:41.465593+00:00</td>\n",
       "      <td>/documents/95508fbf-0e26-46eb-bd13-7408c12cfc9c/</td>\n",
       "      <td>Configuration file for scRNA-seq pipeline</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>production%2Fuci%2FENCLB571LOU%2Fconfig.yaml.txt</td>\n",
       "      <td>@@download/attachment/production%252Fuci%252FE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>2022-06-13 22:54:47.611093+00:00</td>\n",
       "      <td>/documents/17efe021-4859-424a-b967-e2335e6af6a6/</td>\n",
       "      <td>Configuration file for scRNA-seq pipeline</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>production%2Fuci%2FENCLB613RTM%2Fconfig.yaml.txt</td>\n",
       "      <td>@@download/attachment/production%252Fuci%252FE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2022-06-13 22:57:46.674968+00:00</td>\n",
       "      <td>/documents/bfee1a9e-7f5f-4ffe-b7c3-56eb1fbaf4b7/</td>\n",
       "      <td>Configuration file for scRNA-seq pipeline</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>production%2Fuci%2FENCLB994FPH%2Fconfig.yaml.txt</td>\n",
       "      <td>@@download/attachment/production%252Fuci%252FE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>2022-06-13 22:57:39.426512+00:00</td>\n",
       "      <td>/documents/3c5a7074-b64d-4439-b671-4b44ee5619ed/</td>\n",
       "      <td>Configuration file for scRNA-seq pipeline</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>production%2Fuci%2FENCLB818NOG%2Fconfig.yaml.txt</td>\n",
       "      <td>@@download/attachment/production%252Fuci%252FE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>2022-06-13 22:57:40.464693+00:00</td>\n",
       "      <td>/documents/d7f61740-2f95-4ac7-a0f5-996360d1a7c8/</td>\n",
       "      <td>Configuration file for scRNA-seq pipeline</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>production%2Fuci%2FENCLB578XRO%2Fconfig.yaml.txt</td>\n",
       "      <td>@@download/attachment/production%252Fuci%252FE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>2022-06-13 22:55:48.863312+00:00</td>\n",
       "      <td>/documents/9e7604fa-6b51-41ca-be8d-07cdba4a426b/</td>\n",
       "      <td>Configuration file for scRNA-seq pipeline</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>production%2Fuci%2FENCLB861UVF%2Fconfig.yaml.txt</td>\n",
       "      <td>@@download/attachment/production%252Fuci%252FE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>2022-06-13 22:53:22.793171+00:00</td>\n",
       "      <td>/documents/7f011a09-8625-4fc6-b167-7aa9acc05eca/</td>\n",
       "      <td>Configuration file for scRNA-seq pipeline</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>production%2Fstanford%2FENCLB065ONF%2Fconfig.y...</td>\n",
       "      <td>@@download/attachment/production%252Fstanford%...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>2022-06-13 22:55:05.585719+00:00</td>\n",
       "      <td>/documents/d05f0992-18a7-4e83-ad36-1835bf4ba2b4/</td>\n",
       "      <td>Configuration file for scRNA-seq pipeline</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>production%2Fuci%2FENCLB858UWU%2Fconfig.yaml.txt</td>\n",
       "      <td>@@download/attachment/production%252Fuci%252FE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>2022-06-13 22:58:23.268411+00:00</td>\n",
       "      <td>/documents/3ae9bf2e-7c51-481c-a7b0-912e27a316a5/</td>\n",
       "      <td>Configuration file for scRNA-seq pipeline</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>production%2Fuci%2FENCLB998QFD%2Fconfig.yaml.txt</td>\n",
       "      <td>@@download/attachment/production%252Fuci%252FE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>2022-06-13 22:52:14.765160+00:00</td>\n",
       "      <td>/documents/3ef65bf2-792c-401d-af36-374400885039/</td>\n",
       "      <td>Configuration file for scRNA-seq pipeline</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>production%2Fstanford%2FENCLB442AMV%2Fconfig.y...</td>\n",
       "      <td>@@download/attachment/production%252Fstanford%...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>2022-06-13 22:57:59.709876+00:00</td>\n",
       "      <td>/documents/025834ee-0f3d-4eb2-ba66-c30511ea6808/</td>\n",
       "      <td>Configuration file for scRNA-seq pipeline</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>production%2Fuci%2FENCLB590ZKY%2Fconfig.yaml.txt</td>\n",
       "      <td>@@download/attachment/production%252Fuci%252FE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>2022-06-13 22:52:28.034901+00:00</td>\n",
       "      <td>/documents/638c27f9-99fc-4724-9cc2-c1a2204eb7b5/</td>\n",
       "      <td>Configuration file for scRNA-seq pipeline</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>production%2Fstanford%2FENCLB802MGH%2Fconfig.y...</td>\n",
       "      <td>@@download/attachment/production%252Fstanford%...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>2022-06-13 22:58:14.699134+00:00</td>\n",
       "      <td>/documents/13ec8e52-fc07-420f-9dc4-fe41e7f05718/</td>\n",
       "      <td>Configuration file for scRNA-seq pipeline</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>production%2Fuci%2FENCLB671FWJ%2Fconfig.yaml.txt</td>\n",
       "      <td>@@download/attachment/production%252Fuci%252FE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>2022-06-13 23:00:58.172719+00:00</td>\n",
       "      <td>/documents/0325a79f-fd7a-4144-b8f9-1c76f2f4e0a6/</td>\n",
       "      <td>Configuration file for scRNA-seq pipeline</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>production%2Fuci%2FENCLB415QIR%2Fconfig.yaml.txt</td>\n",
       "      <td>@@download/attachment/production%252Fuci%252FE...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       date_created  \\\n",
       "0  2014-06-27 17:43:53.305108+00:00   \n",
       "1  2017-09-25 18:17:18.814210+00:00   \n",
       "2  2022-06-13 22:51:25.495498+00:00   \n",
       "3  2022-06-13 22:51:34.446087+00:00   \n",
       "4  2022-06-13 22:52:43.578942+00:00   \n",
       "5  2022-06-13 22:51:37.623874+00:00   \n",
       "6  2022-06-13 22:54:29.683734+00:00   \n",
       "7  2022-06-13 22:54:12.522463+00:00   \n",
       "8  2022-06-13 22:51:39.678880+00:00   \n",
       "9  2022-06-13 22:55:42.400302+00:00   \n",
       "10 2022-06-13 22:55:40.327104+00:00   \n",
       "11 2022-06-13 22:57:41.465593+00:00   \n",
       "12 2022-06-13 22:54:47.611093+00:00   \n",
       "13 2022-06-13 22:57:46.674968+00:00   \n",
       "14 2022-06-13 22:57:39.426512+00:00   \n",
       "15 2022-06-13 22:57:40.464693+00:00   \n",
       "16 2022-06-13 22:55:48.863312+00:00   \n",
       "17 2022-06-13 22:53:22.793171+00:00   \n",
       "18 2022-06-13 22:55:05.585719+00:00   \n",
       "19 2022-06-13 22:58:23.268411+00:00   \n",
       "20 2022-06-13 22:52:14.765160+00:00   \n",
       "21 2022-06-13 22:57:59.709876+00:00   \n",
       "22 2022-06-13 22:52:28.034901+00:00   \n",
       "23 2022-06-13 22:58:14.699134+00:00   \n",
       "24 2022-06-13 23:00:58.172719+00:00   \n",
       "\n",
       "                                                 @id  \\\n",
       "0   /documents/e84de88e-4c31-4b2a-b6ac-96feb764b2d0/   \n",
       "1   /documents/92251b87-30e2-4a32-a3e3-c4b3c7856329/   \n",
       "2   /documents/5441e3af-f996-4a51-b2a2-982e3427a6bc/   \n",
       "3   /documents/89831fba-ae30-4af5-a619-3eadaba73f72/   \n",
       "4   /documents/484d361b-8d11-433d-8398-44f58eb0f700/   \n",
       "5   /documents/8d6dd36c-210d-4d6d-ba26-c8582db6123e/   \n",
       "6   /documents/f1971706-69cd-4e8a-8b3e-801aa6c73ebe/   \n",
       "7   /documents/b56fe10c-42d3-48ba-b49a-936086b66350/   \n",
       "8   /documents/279a8145-bf10-4e4a-b792-86905d1b20bf/   \n",
       "9   /documents/fd8559fa-7535-4b44-ad3e-03f29d5ebc40/   \n",
       "10  /documents/7fdad273-2fdc-46f9-a820-1b2dff8281fa/   \n",
       "11  /documents/95508fbf-0e26-46eb-bd13-7408c12cfc9c/   \n",
       "12  /documents/17efe021-4859-424a-b967-e2335e6af6a6/   \n",
       "13  /documents/bfee1a9e-7f5f-4ffe-b7c3-56eb1fbaf4b7/   \n",
       "14  /documents/3c5a7074-b64d-4439-b671-4b44ee5619ed/   \n",
       "15  /documents/d7f61740-2f95-4ac7-a0f5-996360d1a7c8/   \n",
       "16  /documents/9e7604fa-6b51-41ca-be8d-07cdba4a426b/   \n",
       "17  /documents/7f011a09-8625-4fc6-b167-7aa9acc05eca/   \n",
       "18  /documents/d05f0992-18a7-4e83-ad36-1835bf4ba2b4/   \n",
       "19  /documents/3ae9bf2e-7c51-481c-a7b0-912e27a316a5/   \n",
       "20  /documents/3ef65bf2-792c-401d-af36-374400885039/   \n",
       "21  /documents/025834ee-0f3d-4eb2-ba66-c30511ea6808/   \n",
       "22  /documents/638c27f9-99fc-4724-9cc2-c1a2204eb7b5/   \n",
       "23  /documents/13ec8e52-fc07-420f-9dc4-fe41e7f05718/   \n",
       "24  /documents/0325a79f-fd7a-4144-b8f9-1c76f2f4e0a6/   \n",
       "\n",
       "                                          description        document_type  \\\n",
       "0                             mRNA Selection Protocol  extraction protocol   \n",
       "1   This document describes the Fluidigm C1 cell i...                other   \n",
       "2           Configuration file for scRNA-seq pipeline    workflow metadata   \n",
       "3           Configuration file for scRNA-seq pipeline    workflow metadata   \n",
       "4           Configuration file for scRNA-seq pipeline    workflow metadata   \n",
       "5           Configuration file for scRNA-seq pipeline    workflow metadata   \n",
       "6           Configuration file for scRNA-seq pipeline    workflow metadata   \n",
       "7           Configuration file for scRNA-seq pipeline    workflow metadata   \n",
       "8           Configuration file for scRNA-seq pipeline    workflow metadata   \n",
       "9           Configuration file for scRNA-seq pipeline    workflow metadata   \n",
       "10          Configuration file for scRNA-seq pipeline    workflow metadata   \n",
       "11          Configuration file for scRNA-seq pipeline    workflow metadata   \n",
       "12          Configuration file for scRNA-seq pipeline    workflow metadata   \n",
       "13          Configuration file for scRNA-seq pipeline    workflow metadata   \n",
       "14          Configuration file for scRNA-seq pipeline    workflow metadata   \n",
       "15          Configuration file for scRNA-seq pipeline    workflow metadata   \n",
       "16          Configuration file for scRNA-seq pipeline    workflow metadata   \n",
       "17          Configuration file for scRNA-seq pipeline    workflow metadata   \n",
       "18          Configuration file for scRNA-seq pipeline    workflow metadata   \n",
       "19          Configuration file for scRNA-seq pipeline    workflow metadata   \n",
       "20          Configuration file for scRNA-seq pipeline    workflow metadata   \n",
       "21          Configuration file for scRNA-seq pipeline    workflow metadata   \n",
       "22          Configuration file for scRNA-seq pipeline    workflow metadata   \n",
       "23          Configuration file for scRNA-seq pipeline    workflow metadata   \n",
       "24          Configuration file for scRNA-seq pipeline    workflow metadata   \n",
       "\n",
       "          mime_type                                    remote_filename  \\\n",
       "0   application/pdf                dynabeads_mrna_purification_man.pdf   \n",
       "1   application/pdf                  C1_mRNA-Seq_pr_100-7168L1 (2).pdf   \n",
       "2        text/plain   production%2Fuci%2FENCLB217YGZ%2Fconfig.yaml.txt   \n",
       "3        text/plain  production%2Fstanford%2FENCLB123ZDT%2Fconfig.y...   \n",
       "4        text/plain  production%2Fstanford%2FENCLB526RVS%2Fconfig.y...   \n",
       "5        text/plain  production%2Fstanford%2FENCLB063XDV%2Fconfig.y...   \n",
       "6        text/plain   production%2Fuci%2FENCLB870FRH%2Fconfig.yaml.txt   \n",
       "7        text/plain   production%2Fuci%2FENCLB057CVT%2Fconfig.yaml.txt   \n",
       "8        text/plain  production%2Fstanford%2FENCLB025GOX%2Fconfig.y...   \n",
       "9        text/plain   production%2Fuci%2FENCLB215OBG%2Fconfig.yaml.txt   \n",
       "10       text/plain   production%2Fuci%2FENCLB070DDL%2Fconfig.yaml.txt   \n",
       "11       text/plain   production%2Fuci%2FENCLB571LOU%2Fconfig.yaml.txt   \n",
       "12       text/plain   production%2Fuci%2FENCLB613RTM%2Fconfig.yaml.txt   \n",
       "13       text/plain   production%2Fuci%2FENCLB994FPH%2Fconfig.yaml.txt   \n",
       "14       text/plain   production%2Fuci%2FENCLB818NOG%2Fconfig.yaml.txt   \n",
       "15       text/plain   production%2Fuci%2FENCLB578XRO%2Fconfig.yaml.txt   \n",
       "16       text/plain   production%2Fuci%2FENCLB861UVF%2Fconfig.yaml.txt   \n",
       "17       text/plain  production%2Fstanford%2FENCLB065ONF%2Fconfig.y...   \n",
       "18       text/plain   production%2Fuci%2FENCLB858UWU%2Fconfig.yaml.txt   \n",
       "19       text/plain   production%2Fuci%2FENCLB998QFD%2Fconfig.yaml.txt   \n",
       "20       text/plain  production%2Fstanford%2FENCLB442AMV%2Fconfig.y...   \n",
       "21       text/plain   production%2Fuci%2FENCLB590ZKY%2Fconfig.yaml.txt   \n",
       "22       text/plain  production%2Fstanford%2FENCLB802MGH%2Fconfig.y...   \n",
       "23       text/plain   production%2Fuci%2FENCLB671FWJ%2Fconfig.yaml.txt   \n",
       "24       text/plain   production%2Fuci%2FENCLB415QIR%2Fconfig.yaml.txt   \n",
       "\n",
       "                                                 href  \n",
       "0   @@download/attachment/dynabeads_mrna_purificat...  \n",
       "1   @@download/attachment/C1_mRNA-Seq_pr_100-7168L...  \n",
       "2   @@download/attachment/production%252Fuci%252FE...  \n",
       "3   @@download/attachment/production%252Fstanford%...  \n",
       "4   @@download/attachment/production%252Fstanford%...  \n",
       "5   @@download/attachment/production%252Fstanford%...  \n",
       "6   @@download/attachment/production%252Fuci%252FE...  \n",
       "7   @@download/attachment/production%252Fuci%252FE...  \n",
       "8   @@download/attachment/production%252Fstanford%...  \n",
       "9   @@download/attachment/production%252Fuci%252FE...  \n",
       "10  @@download/attachment/production%252Fuci%252FE...  \n",
       "11  @@download/attachment/production%252Fuci%252FE...  \n",
       "12  @@download/attachment/production%252Fuci%252FE...  \n",
       "13  @@download/attachment/production%252Fuci%252FE...  \n",
       "14  @@download/attachment/production%252Fuci%252FE...  \n",
       "15  @@download/attachment/production%252Fuci%252FE...  \n",
       "16  @@download/attachment/production%252Fuci%252FE...  \n",
       "17  @@download/attachment/production%252Fstanford%...  \n",
       "18  @@download/attachment/production%252Fuci%252FE...  \n",
       "19  @@download/attachment/production%252Fuci%252FE...  \n",
       "20  @@download/attachment/production%252Fstanford%...  \n",
       "21  @@download/attachment/production%252Fuci%252FE...  \n",
       "22  @@download/attachment/production%252Fstanford%...  \n",
       "23  @@download/attachment/production%252Fuci%252FE...  \n",
       "24  @@download/attachment/production%252Fuci%252FE...  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "posted_documents"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "mature-rotation",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "searching production%2Fuci%2FENCLB419YTE%2Fconfig.yaml.txt\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'create_document_log': None,\n",
       " 'library_id': 'ENCLB419YTE',\n",
       " 'document_id': 'would create',\n",
       " 'analysis_id': '/analyses/ENCAN597RYW/',\n",
       " 'filename': 'production/uci/ENCLB419YTE/config.yaml.txt'}"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def submit_document(server, row, library_configs, dry_run=True):\n",
    "    result = {\n",
    "        \"create_document_log\": None,\n",
    "        \"library_id\": None,\n",
    "        \"document_id\": None,\n",
    "        \"analysis_id\": None,\n",
    "        \"filename\": None\n",
    "    }\n",
    "\n",
    "    library_configs = library_configs.set_index(\"remote_filename\")\n",
    "    workflow_doc = Document(\n",
    "        row.local_filename, \n",
    "        \"workflow metadata\", \n",
    "        \"Configuration file for scRNA-seq pipeline\",\n",
    "        filename=row.remote_filename,\n",
    "        server=server,\n",
    "    )\n",
    "    result[\"filename\"] = workflow_doc.filename\n",
    "    \n",
    "    workflow_remote_filename = workflow_doc.filename\n",
    "    library_id = library_configs.loc[workflow_remote_filename][\"library_id\"]\n",
    "    \n",
    "    # try to block double escaping\n",
    "    assert \"%\" not in workflow_remote_filename\n",
    "    posted = get_posted_documents(server)\n",
    "    posted_document = get_document_from_posted(posted, workflow_remote_filename)\n",
    "    \n",
    "    if posted_document is None:\n",
    "        if not dry_run:\n",
    "            result[\"create_document_log\"] = workflow_doc.create_if_needed(server, workflow_doc.uuid, validator)\n",
    "            if result[\"create_document_log\"][\"status\"] == \"success\":\n",
    "                workflow_id = result[\"create_document_log\"][\"@graph\"][0][\"@id\"]\n",
    "            else:\n",
    "                workflow_id = None\n",
    "        else:\n",
    "            workflow_id = \"would create\"\n",
    "    else:\n",
    "        workflow_id = posted_document[\"@id\"]\n",
    "        \n",
    "    # 2) attach document to analysis object    \n",
    "    assert workflow_id is not None\n",
    "    assert library_id is not None\n",
    "    result[\"library_id\"] = library_id\n",
    "    result[\"document_id\"] = workflow_id\n",
    "    \n",
    "    analysis = get_analysis_by_library_id(server, library_id)\n",
    "    if workflow_id not in analysis[\"documents\"]:\n",
    "        #print(\"adding {} to {}\".format(workflow_id, analysis[\"@id\"]))\n",
    "        #documents = analysis[\"documents\"]\n",
    "        #documents.append(workflow_id)\n",
    "        #result = server.patch_json(analysis[\"@id\"], {\"documents\": documents})\n",
    "        #responses.append(result)\n",
    "        result[\"analysis_id\"] = analysis[\"@id\"]\n",
    "        \n",
    "    return result\n",
    "\n",
    "result = submit_document(server, library_configs.loc[556], library_configs, dry_run=True)\n",
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "metallic-tobacco",
   "metadata": {},
   "outputs": [],
   "source": [
    "results = []\n",
    "for i, row in library_configs.iterrows():\n",
    "    results.append(submit_document(server, row, library_configs, dry_run=True))\n",
    "#    if i > 5:\n",
    "#        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "coordinate-active",
   "metadata": {},
   "outputs": [],
   "source": [
    "posted_results = pandas.DataFrame(results)\n",
    "posted_results[[\"library_id\", \"document_id\", \"analysis_id\", \"filename\"]].to_csv(\n",
    "    \"document_analysis_{}.tsv\".format(server.server),\n",
    "    sep=\"\\t\",\n",
    "    index=False\n",
    ")\n",
    "posted_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "composite-pantyhose",
   "metadata": {},
   "outputs": [],
   "source": [
    "latest_configs = library_configs[library_configs[\"local_filename\"].apply(lambda lib: Path(lib).parts[1] in (\"stanford_heart_20220810\",))]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abandoned-trout",
   "metadata": {},
   "outputs": [],
   "source": [
    "if 0:\n",
    "    latest_results = []\n",
    "    for i, row in latest_configs.iterrows():\n",
    "        latest_results.append(submit_document(server, row, library_configs, dry_run=True))\n",
    "    #    if i > 5:\n",
    "    #        break\n",
    "\n",
    "    latest_results = pandas.DataFrame(latest_results)\n",
    "    target_name = Path(\"document_analysis_{}_20220810.tsv\".format(server.server))\n",
    "    if not target_name.exists():\n",
    "        latest_results[[\"library_id\", \"document_id\", \"analysis_id\", \"filename\"]].to_csv(\n",
    "            target_name,\n",
    "            sep=\"\\t\",\n",
    "            index=False\n",
    "        )\n",
    "    latest_results.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "accurate-vessel",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aquatic-mixture",
   "metadata": {},
   "outputs": [],
   "source": [
    "liz_configs = library_configs[library_configs[\"local_filename\"].apply(lambda lib: Path(lib).parts[1] in (\"liz_reprocess\",))]\n",
    "liz_configs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "logical-samoa",
   "metadata": {},
   "outputs": [],
   "source": [
    "if 0:\n",
    "    liz_results = []\n",
    "    for i, row in liz_configs.iterrows():\n",
    "        liz_results.append(submit_document(server, row, liz_configs, dry_run=True))\n",
    "\n",
    "    liz_results = pandas.DataFrame(liz_results)\n",
    "    target_name = Path(\"document_analysis_{}_liz_resubmit.tsv\".format(server.server))\n",
    "    if not target_name.exists():\n",
    "        liz_results[[\"library_id\", \"document_id\", \"analysis_id\", \"filename\"]].to_csv(\n",
    "            target_name,\n",
    "            sep=\"\\t\",\n",
    "            index=False\n",
    "        )\n",
    "    print(liz_results.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "swiss-enclosure",
   "metadata": {},
   "outputs": [],
   "source": [
    "analyses = pandas.DataFrame([[\"ENCSR398YBK\", \"ENCLB398IAZ\", \"barbara-wold:ENCSR398YBK_analysis\"],\n",
    "[\"ENCSR231FNL\", \"ENCLB398IAZ\", \"barbara-wold:ENCSR231FNL_analysis\"],\n",
    "[\"ENCSR176WWW\", \"ENCLB872TNB\", \"barbara-wold:ENCSR176WWW_analysis\"],\n",
    "[\"ENCSR980OCK\", \"ENCSR980OCK\", \"barbara-wold:ENCSR980OCK_analysis\"],\n",
    "[\"ENCSR814LMX\", \"ENCLB366ZFV\", \"barbara-wold:ENCSR814LMX_analysis\"],\n",
    "[\"ENCSR067BOK\", \"ENCLB849AUZ\", \"barbara-wold:ENCSR067BOK_analysis\"]], columns=[\"experiment_id\", \"library_id\", \"alias\"])\n",
    "\n",
    "analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "exposed-champagne",
   "metadata": {},
   "outputs": [],
   "source": [
    "latest_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "empty-desire",
   "metadata": {},
   "outputs": [],
   "source": [
    "library_configs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ultimate-spiritual",
   "metadata": {},
   "outputs": [],
   "source": [
    "missing_docs_query = \"https://www.encodeproject.org/report/?type=Analysis&files.assay_term_name=single-cell+RNA+sequencing+assay&documents!=*&pipelines=%2Fpipelines%2FENCPL257SYI%2F&field=%40id&field=status&field=pipeline_award_rfas&field=assembly&field=genome_annotation&field=datasets\"\n",
    "\n",
    "experiments_missing_docs = []\n",
    "for row in server.get_json(missing_docs_query)[\"@graph\"]:\n",
    "    experiment_id = row[\"datasets\"][0]\n",
    "    experiment = server.get_json(experiment_id)\n",
    "    for replicate in experiment[\"replicates\"]:\n",
    "        library = replicate[\"library\"]\n",
    "        experiments_missing_docs.append({\n",
    "            \"experiment\": experiment[\"accession\"],\n",
    "            \"library\": library[\"accession\"],\n",
    "            \"analyses\": row[\"@id\"],\n",
    "            \"date_created\": experiment[\"date_created\"],\n",
    "            \"summary\": experiment[\"simple_biosample_summary\"],\n",
    "            \n",
    "        })\n",
    "\n",
    "experiments_missing_docs = pandas.DataFrame(experiments_missing_docs)\n",
    "experiments_missing_docs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "hispanic-tackle",
   "metadata": {},
   "outputs": [],
   "source": [
    "libraries_missing_docs = set(experiments_missing_docs[\"library\"].to_list())\n",
    "library_configs[library_configs[\"library_id\"].isin(libraries_missing_docs)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "separate-auckland",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "incorporated-outreach",
   "metadata": {},
   "outputs": [],
   "source": [
    "ls production"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "modular-oxford",
   "metadata": {},
   "source": [
    "# Sept 22 heart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "indie-chocolate",
   "metadata": {},
   "outputs": [],
   "source": [
    "stanford_sept22_configs = library_configs[library_configs[\"local_filename\"].apply(lambda lib: Path(lib).parts[1] in (\"stanford_heart_20220822\",))]\n",
    "stanford_sept22_configs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "suited-academy",
   "metadata": {},
   "outputs": [],
   "source": [
    "if 0:\n",
    "    stanford_sept22_results = []\n",
    "    for i, row in stanford_sept22_configs.iterrows():\n",
    "        stanford_sept22_results.append(submit_document(server, row, stanford_sept22_configs, dry_run=True))\n",
    "\n",
    "    stanford_sept22_results = pandas.DataFrame(stanford_sept22_results)\n",
    "    target_name = Path(\"document_analysis_{}_stanford_20220822.tsv\".format(server.server))\n",
    "    if not target_name.exists():\n",
    "        stanford_sept22_results[[\"library_id\", \"document_id\", \"analysis_id\", \"filename\"]].to_csv(\n",
    "            target_name,\n",
    "            sep=\"\\t\",\n",
    "            index=False\n",
    "        )\n",
    "    print(stanford_sept22_results.head())\n",
    "    print(stanford_sept22_results.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "rolled-workshop",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "olive-minister",
   "metadata": {},
   "source": [
    "# Snyder Liver 2023 01 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "arctic-secretariat",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uuid</th>\n",
       "      <th>local_filename</th>\n",
       "      <th>remote_filename</th>\n",
       "      <th>mime_type</th>\n",
       "      <th>document type</th>\n",
       "      <th>md5sum</th>\n",
       "      <th>library_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>129</th>\n",
       "      <td>None</td>\n",
       "      <td>production/snyder_liver_20230110/ENCLB540OBF/c...</td>\n",
       "      <td>production/snyder_liver_20230110/ENCLB540OBF/c...</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>None</td>\n",
       "      <td>ENCLB540OBF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130</th>\n",
       "      <td>None</td>\n",
       "      <td>production/snyder_liver_20230110/ENCLB823WAF/c...</td>\n",
       "      <td>production/snyder_liver_20230110/ENCLB823WAF/c...</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>None</td>\n",
       "      <td>ENCLB823WAF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>None</td>\n",
       "      <td>production/snyder_liver_20230110/ENCLB566WWJ/c...</td>\n",
       "      <td>production/snyder_liver_20230110/ENCLB566WWJ/c...</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>None</td>\n",
       "      <td>ENCLB566WWJ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>None</td>\n",
       "      <td>production/snyder_liver_20230110/ENCLB054SXC/c...</td>\n",
       "      <td>production/snyder_liver_20230110/ENCLB054SXC/c...</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>None</td>\n",
       "      <td>ENCLB054SXC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>None</td>\n",
       "      <td>production/snyder_liver_20230110/ENCLB401DAZ/c...</td>\n",
       "      <td>production/snyder_liver_20230110/ENCLB401DAZ/c...</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>None</td>\n",
       "      <td>ENCLB401DAZ</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>None</td>\n",
       "      <td>production/snyder_liver_20230110/ENCLB851FFF/c...</td>\n",
       "      <td>production/snyder_liver_20230110/ENCLB851FFF/c...</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>None</td>\n",
       "      <td>ENCLB851FFF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>None</td>\n",
       "      <td>production/snyder_liver_20230110/ENCLB401UMP/c...</td>\n",
       "      <td>production/snyder_liver_20230110/ENCLB401UMP/c...</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>None</td>\n",
       "      <td>ENCLB401UMP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>None</td>\n",
       "      <td>production/snyder_liver_20230110/ENCLB713GHH/c...</td>\n",
       "      <td>production/snyder_liver_20230110/ENCLB713GHH/c...</td>\n",
       "      <td>text/plain</td>\n",
       "      <td>workflow metadata</td>\n",
       "      <td>None</td>\n",
       "      <td>ENCLB713GHH</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     uuid                                     local_filename  \\\n",
       "129  None  production/snyder_liver_20230110/ENCLB540OBF/c...   \n",
       "130  None  production/snyder_liver_20230110/ENCLB823WAF/c...   \n",
       "131  None  production/snyder_liver_20230110/ENCLB566WWJ/c...   \n",
       "132  None  production/snyder_liver_20230110/ENCLB054SXC/c...   \n",
       "133  None  production/snyder_liver_20230110/ENCLB401DAZ/c...   \n",
       "134  None  production/snyder_liver_20230110/ENCLB851FFF/c...   \n",
       "135  None  production/snyder_liver_20230110/ENCLB401UMP/c...   \n",
       "136  None  production/snyder_liver_20230110/ENCLB713GHH/c...   \n",
       "\n",
       "                                       remote_filename   mime_type  \\\n",
       "129  production/snyder_liver_20230110/ENCLB540OBF/c...  text/plain   \n",
       "130  production/snyder_liver_20230110/ENCLB823WAF/c...  text/plain   \n",
       "131  production/snyder_liver_20230110/ENCLB566WWJ/c...  text/plain   \n",
       "132  production/snyder_liver_20230110/ENCLB054SXC/c...  text/plain   \n",
       "133  production/snyder_liver_20230110/ENCLB401DAZ/c...  text/plain   \n",
       "134  production/snyder_liver_20230110/ENCLB851FFF/c...  text/plain   \n",
       "135  production/snyder_liver_20230110/ENCLB401UMP/c...  text/plain   \n",
       "136  production/snyder_liver_20230110/ENCLB713GHH/c...  text/plain   \n",
       "\n",
       "         document type md5sum   library_id  \n",
       "129  workflow metadata   None  ENCLB540OBF  \n",
       "130  workflow metadata   None  ENCLB823WAF  \n",
       "131  workflow metadata   None  ENCLB566WWJ  \n",
       "132  workflow metadata   None  ENCLB054SXC  \n",
       "133  workflow metadata   None  ENCLB401DAZ  \n",
       "134  workflow metadata   None  ENCLB851FFF  \n",
       "135  workflow metadata   None  ENCLB401UMP  \n",
       "136  workflow metadata   None  ENCLB713GHH  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "liver_jan10_configs = library_configs[library_configs[\"local_filename\"].apply(lambda lib: Path(lib).parts[1] in (\"snyder_liver_20230110\",))]\n",
    "liver_jan10_configs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "loved-digest",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "searching production%2Fsnyder_liver_20230110%2FENCLB540OBF%2Fconfig.yaml.txt\n",
      "searching production%2Fsnyder_liver_20230110%2FENCLB823WAF%2Fconfig.yaml.txt\n",
      "searching production%2Fsnyder_liver_20230110%2FENCLB566WWJ%2Fconfig.yaml.txt\n",
      "searching production%2Fsnyder_liver_20230110%2FENCLB054SXC%2Fconfig.yaml.txt\n",
      "searching production%2Fsnyder_liver_20230110%2FENCLB401DAZ%2Fconfig.yaml.txt\n",
      "searching production%2Fsnyder_liver_20230110%2FENCLB851FFF%2Fconfig.yaml.txt\n",
      "searching production%2Fsnyder_liver_20230110%2FENCLB401UMP%2Fconfig.yaml.txt\n",
      "searching production%2Fsnyder_liver_20230110%2FENCLB713GHH%2Fconfig.yaml.txt\n",
      "                                 create_document_log   library_id  \\\n",
      "0  {'status': 'success', '@type': ['result'], '@g...  ENCLB540OBF   \n",
      "1  {'status': 'success', '@type': ['result'], '@g...  ENCLB823WAF   \n",
      "2  {'status': 'success', '@type': ['result'], '@g...  ENCLB566WWJ   \n",
      "3  {'status': 'success', '@type': ['result'], '@g...  ENCLB054SXC   \n",
      "4  {'status': 'success', '@type': ['result'], '@g...  ENCLB401DAZ   \n",
      "\n",
      "                                        document_id             analysis_id  \\\n",
      "0  /documents/e6cd6c64-c378-4929-a8bb-d9431f24b5c2/  /analyses/ENCAN456IBQ/   \n",
      "1  /documents/767ecd3b-5c59-497f-b9b6-5a38199e2b6a/  /analyses/ENCAN592PVR/   \n",
      "2  /documents/9b00e956-1f39-4f03-b3a0-255b7491ff34/  /analyses/ENCAN809MWI/   \n",
      "3  /documents/9f02e2ac-418e-471e-b58e-a4586a8a15f4/  /analyses/ENCAN175UUQ/   \n",
      "4  /documents/b724970f-3cd6-4cd5-adfe-ec7585731b2d/  /analyses/ENCAN929BJL/   \n",
      "\n",
      "                                            filename  \n",
      "0  production/snyder_liver_20230110/ENCLB540OBF/c...  \n",
      "1  production/snyder_liver_20230110/ENCLB823WAF/c...  \n",
      "2  production/snyder_liver_20230110/ENCLB566WWJ/c...  \n",
      "3  production/snyder_liver_20230110/ENCLB054SXC/c...  \n",
      "4  production/snyder_liver_20230110/ENCLB401DAZ/c...  \n",
      "8\n"
     ]
    }
   ],
   "source": [
    "if 0:\n",
    "    liver_jan10_results = []\n",
    "    for i, row in liver_jan10_configs.iterrows():\n",
    "        liver_jan10_results.append(submit_document(server, row, liver_jan10_configs, dry_run=False))\n",
    "\n",
    "    liver_jan10_results = pandas.DataFrame(liver_jan10_results)\n",
    "    target_name = Path(\"document_analysis_{}_snyder_liver_20230110.tsv\".format(server.server))\n",
    "    if not target_name.exists():\n",
    "        liver_jan10_results[[\"library_id\", \"document_id\", \"analysis_id\", \"filename\"]].to_csv(\n",
    "            target_name,\n",
    "            sep=\"\\t\",\n",
    "            index=False\n",
    "        )\n",
    "    print(liver_jan10_results.head())\n",
    "    print(liver_jan10_results.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "homeless-contributor",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
